---
title: "Code to replicate the results of "Retrotransposon Expression Is Upregulated in Adulthood and Suppressed during Regeneration of the Limb in the Axolotl (Ambystoma mexicanum)" (Ruiz-Pérez et al., 2025)"
author: "Samuel Ruiz-Pérez"
date: "2025"
output:
  html_document:
    number_sections: true
    toc: true
    toc_depth: 6
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)

```

# Project directory structure

```{bash}
HOME_DIR=$(eval echo ~$USER)
PROJECT_DIR=$HOME_DIR/sruizperez/projects/axolotl
EXT_DATA_DIR=$HOME_DIR/sruizperez/projects/axolotl/ext_data/
SOFTWARE_DIR=$HOME_DIR/sruizperez/software
TMP_DIR=$HOME_DIR/sruizperez/tmp
# 32 CPUs maximum:
NTHREADS=8
NJOBS=4
RAM_LIMIT=130000000000

```

# Mamba environment installation

```{bash}
cd $SOFTWARE_DIR
curl -L -O "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh"
bash Miniforge3-$(uname)-$(uname -m).sh

```

```{bash}
mamba create \
  -n srp_axolote \
  -c conda-forge -c bioconda \
  pip pysam jq pandas seqkit \
  parallel pigz pbzip2 fastqc multiqc star salmon samtools sambamba bedtools bedops deeptools agat \
  uropa gffread eggnog-mapper cd-hit blast mafft muscle emboss pfam_scan gepard hmmer gffutils \
  r-base r-data.table r-stringr r-ggpubr r-ggplot2 r-svglite r-gridtext r-ashr r-ggh4x \
  r-dplyr r-tidyr r-readr r-seqinr r-foreach r-doparallel r-biocmanager r-devtools r-igraph=2.0.2 \
  r-concaveman r-cowplot r-ggrepel r-ggforce r-ggalt r-wgcna r-magrittr r-tidyverse r-rmariadb r-styler \
  bioconductor-summarizedexperiment bioconductor-deseq2 bioconductor-tximport bioconductor-tximeta \
  bioconductor-edger bioconductor-vsn bioconductor-rtracklayer bioconductor-biomart bioconductor-annotationdbi \
  bioconductor-enhancedvolcano bioconductor-complexheatmap bioconductor-apeglm \
  bioconductor-genomicfeatures bioconductor-clusterprofiler bioconductor-ihw \
  bioconductor-saturn bioconductor-enrichplot bioconductor-sva \
  bioconductor-karyoploter bioconductor-regioner bioconductor-plyranges \
  bioconductor-org.hs.eg.db bioconductor-org.mm.eg.db bioconductor-org.xl.eg.db \
  bioconductor-genomicranges bioconductor-iranges bioconductor-csaw

mamba activate srp_axolote

mamba list -n srp_axolote --export > $PROJECT_DIR/axoloTE/scripts/mamba_srp_axolote_package_list.txt
mamba list -n srp_axolote --json > $PROJECT_DIR/axoloTE/scripts/mamba_srp_axolote_package_list.json

```

# Preparing published RE libraries

## Downloading library and creating .out file (RepeatMasker file)

```{bash}
mkdir -p $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/
cd $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/

```

Downloaded libraries rmsk.outs (gzip compressed):
In https://www.axolotl-omics.org/assemblies look for hub: 
Annotated repeats (denovo predictions by RepeatModeler and LTR): 
https://www.axolotl-omics.org/trackhubs/AmexG_v6.0-DD/Repeats/hub.description 
load hub in UCSC genome browser (https://genome.axolotl-omics.org/) and download 
manually through table browser.
use filename format: ambMex60DD_ucsc_hub.RM.DNA.out.gz
save to $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/

```{bash}
cd $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/

gunzip *.out.gz

# remove "#" from "#chrom" in ambMex60DD_ucsc_hub.RM.DNA.out and save header in a new file
head -n 1 ambMex60DD_ucsc_hub.RM.DNA.out | awk 'NR==1 {sub(/^#/, "")} 1' > header.out

# Define an array of file names
files=("RM.DNA" "RM.LINE" "RM.LTR" "RM.Low_complexity" "RM.RC" "RM.SINE" "RM.Satellite" "RM.Simple_repeat" "RM.Unknown" "RM.rRNA" "LTR.LTR" "LTR.Low_complexity" "LTR.Satellite" "LTR.Simple_repeat")

# Loop over the array
for file in "${files[@]}"; do
    # Remove header
    awk 'NR > 1' FS='\t' "ambMex60DD_ucsc_hub.${file}.out" > "ambMex60DD_ucsc_hub.${file}.mod.out"
    
    # Count lines
    wc -l "ambMex60DD_ucsc_hub.${file}.mod.out"
done

# concat all files
cat header.out *mod.out > ambMex60DD_ucsc_hub.ALL.mod.out

rm header.out

# Replace any "/"  in column 13 with "_" to avoid the following error later:
# ExplorATE > RM2Bed.py > ValueError: too many values to unpack (expected 2)
awk -F'\t' 'NR==1 {print; next} {gsub(/\//, "_", $13)} 1' OFS='\t' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# change order of columns in ambMex60DD_ucsc_hub.ALL.mod.out to 7th, 8th, 9th, 10th, 1st, 2nd, 3rd, 11th, 6th, 4th, concat(12th,"/",13th), 14th, 15th, 16th
awk '{print $7"\t"$8"\t"$9"\t"$10"\t"$1"\t"$2"\t"$3"\t"$11"\t"$6"\t"$4"\t"$12"/"$13"\t"$14"\t"$15"\t"$16}' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# replace "-" in column 9 with "C" (skip first line in file)
awk -F'\t' 'NR==1 {print; next} {gsub(/-/, "C", $9)} 1' OFS='\t' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# Divide milliDev, milliDel and milliIns columns by 10 to match "perc div./perc.del/perc.ins" formats of rmsk.out files (skip first line in file)
awk 'BEGIN{FS=OFS="\t"} NR==1 { print; next } { $2 = sprintf("%.1f", $2 / 10); $3 = sprintf("%.1f", $3 / 10); $4 = sprintf("%.1f", $4 / 10);  print }' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# replace names
awk -F'\t' 'BEGIN {OFS="\t"} NR == 1 { $2="perc_div"; $3="perc_del"; $4="perc_ins" } { print }' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# in column 8th,12th,13th and 14th, if the number starts with a "-", remove the "-" and enclose the number in parenthesis (skip first line in file)
awk -F'\t' 'NR == 1 {print; next} {OFS=FS} NR > 1 && $8 ~ /^-/ {gsub(/^-/, "(", $8); gsub(/$/, ")", $8)} {print}' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out
awk -F'\t' 'NR == 1 {print; next} {OFS=FS} NR > 1 && $12 ~ /^-/ {gsub(/^-/, "(", $12); gsub(/$/, ")", $12)} {print}' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out
awk -F'\t' 'NR == 1 {print; next} {OFS=FS} NR > 1 && $13 ~ /^-/ {gsub(/^-/, "(", $13); gsub(/$/, ")", $13)} {print}' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out
awk -F'\t' 'NR == 1 {print; next} {OFS=FS} NR > 1 && $14 ~ /^-/ {gsub(/^-/, "(", $14); gsub(/$/, ")", $14)} {print}' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# Sort first by chromosome name and the by starting position
(head -n 1 ambMex60DD_ucsc_hub.ALL.mod.out && tail -n +2 ambMex60DD_ucsc_hub.ALL.mod.out | sort -t$'\t' -k5,5 -V -s -k6,6n --field-separator=$'\t') > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

# add an "ID" column to the right and fill it with numbers from 1 to n (skip first line in file)
awk -F'\t' 'NR==1 {print $0 "\tID"} NR>1 {print $0 "\t" (NR-1)}' ambMex60DD_ucsc_hub.ALL.mod.out > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.out

head ambMex60DD_ucsc_hub.ALL.mod.out

#cut -f 10 ambMex60DD_ucsc_hub.ALL.mod.out | sort | uniq | wc -l
#37989 families in total (maybe with duplicates, ExplorATE will discard them)

# wc -l ambMex60DD_ucsc_hub.ALL.mod.out
#40366448 lines in total

```

## Making RE GTF libraries for TETranscripts, TEcount, TElocal 

Reformatting RepeatMasker .out file

```{bash}
# following https://github.com/mhammell-laboratory/TEtranscripts/issues/83#issuecomment-734370631

cd $PROJECT_DIR/ext_data/axolotl_genome/repeats

# remove header from ambMex60DD_ucsc_hub.ALL.mod.out save to gtf file
tail -n +2 ambMex60DD_ucsc_hub.ALL.mod.out > ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp

# replace "C" in column 9 with "-"
awk 'BEGIN{FS=OFS="\t"} {if($9=="C") $9="-"; print}' ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp

# Split class and family columns
awk -F'\t' 'BEGIN{OFS="\t"} {split($11, arr, "/"); $16=arr[1]; $17=arr[2]; print}' ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp > tmp.out && mv tmp.out ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp

head ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp

#print column 16 of file
awk -F'\t' '{print $16}' ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp | sort | uniq

```

Downloading makeTEgtf (TEtranscripts) script and running it

```{bash}
cd $SOFTWARE_DIR
wget https://labshare.cshl.edu/shares/mhammelllab/www-data/TEtranscripts/TE_GTF/makeTEgtf.pl.gz
gunzip makeTEgtf.pl.gz

# see https://github.com/mhammell-laboratory/TEtranscripts/issues/63#issue-590217827
# see https://github.com/mhammell-laboratory/TEtranscripts/issues/83#issuecomment-734370631

chmod +x makeTEgtf.pl

```

```{bash}
perl ./makeTEgtf.pl \
  -c 5 -s 6 -e 7 -o 9 -t 10 -f 17 -C 16 -S 1 \
  -n ambMex60DD_ucsc_hub_563_RM_and_LTR_libs \
  $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp \
  > $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.gtf

# 33816789 lines in total, low complexity and simple repeats or small RNAs are removed (see source code of makeTEgtf.pl)

rm $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.gtf.tmp

```

Create TSV file

```{bash}
cd $PROJECT_DIR/ext_data/axolotl_genome/repeats

#gffutils
gtf_extract \
  -f exon \
  --fields chrom,source,feature,start,end,score,strand,gene_id,transcript_id,family_id,class_id \
  -o $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.tsv \
  $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.gtf

# add header 
sed -i '1s/^/chrom\tsource\tfeature\tstart\tend\tscore\tstrand\tgene_id\ttranscript_id\tfamily_id\tclass_id\n/' $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.tsv

```	

# Functional annotation of the axolotl proteome (eggNOG)

```{bash}
download_eggnog_data.py \
  -F \
  -P \
  --data_dir $PROJECT_DIR/protein_annotation/db_data

nohup emapper.py \
  --itype proteins \
  -m diamond \
  --pfam_realign realign \
  --sensmode ultra-sensitive \
  --dbmem \
  --cpu 20 \
  -i $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/AmexT_v47_protein.fa \
  --md5 \
  --report_orthologs \
  --excel \
  --output AmexT_v47_protein \
  --data_dir $PROJECT_DIR/protein_annotation/db_data/ \
  --temp_dir $PROJECT_DIR/protein_annotation/tmp/ \
  --output_dir $PROJECT_DIR/protein_annotation/output/ \
  > $PROJECT_DIR/protein_annotation/emapper.log 2>&1 &

# --temp_dir is necessary to avoid tempfile.py's 'PermissionError: [Errno 13] Permission denied'

```


## Gene/Transcript annotation

Generate a tx2gene table for tximport:

```{r}
library(dplyr) # To use %>% operator
library(plyranges) # To use read_gff

# Read annotation file and extract list of gene and transcript ids
tx2g <- read_gff(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/AmexT_v47-AmexG_v6.0-DD.gtf")) %>%
  filter(type == "transcript") %>%
  select(gene_id, transcript_id, type)
```

For salmon tx2gene table:
```{r}
tx2gene <- data.frame(transcript_id = tx2g$transcript_id, gene_id = tx2g$gene_id)

# Save list of gene and transcript ids
write.table(tx2gene, file.path(PROJECT_DIR, "/rnaseq/Salmon/index/tx2gene_fullid.tsv"), sep = "\t", row.names = FALSE)

# Remove everything before "AMEX" from transcript IDs
tx2gene$transcript_id <- gsub("^.*AMEX", "AMEX", tx2gene$transcript_id)

write.table(tx2gene, file.path(PROJECT_DIR, "/rnaseq/Salmon/index/tx2gene.tsv"), sep = "\t", row.names = FALSE)
```

```{r}
# in tx2g, copy transcript_id column to transcript_name_full
tx2g$transcript_name_full <- tx2g$transcript_id

# Remove everything before "AMEX" from transcript IDs
tx2g$transcript_id <- gsub("^.*AMEX", "AMEX", tx2g$transcript_id)

# Do the following to split the transcript_name_full column into multiple columns
tx2g$split_strings <- strsplit(tx2g$transcript_name_full, "\\|")

# Use sapply to extract the strings that match the patterns
tx2g$transcript_name_nr <- sapply(tx2g$split_strings, function(x) gsub(" \\[nr\\]", "", x[grepl("\\[nr\\]", x)]))
tx2g$transcript_name_hs <- sapply(tx2g$split_strings, function(x) gsub(" \\[hs\\]", "", x[grepl("\\[hs\\]", x)]))
tx2g$transcript_name_amex <- sapply(tx2g$split_strings, function(x) x[grepl("AMEX", x)])
tx2g$transcript_name_other <- sapply(tx2g$split_strings, function(x) x[!grepl("\\[nr\\]|\\[hs\\]|AMEX", x)])
tx2g$split_strings <- NULL

# change column type to character
tx2g$transcript_name_nr <- as.character(tx2g$transcript_name_nr)
tx2g$transcript_name_hs <- as.character(tx2g$transcript_name_hs)
tx2g$transcript_name_amex <- as.character(tx2g$transcript_name_amex)
tx2g$transcript_name_other <- as.character(tx2g$transcript_name_other)
tx2g$type <- as.character(tx2g$type)

# replace "N/A" or "UNKNOWN" with NA in GRanges object
tx2g$transcript_name_nr[tx2g$transcript_name_nr == "N/A" | tx2g$transcript_name_nr == "UNKNOWN" | tx2g$transcript_name_nr == "character(0)"] <- NA
tx2g$transcript_name_hs[tx2g$transcript_name_hs == "N/A" | tx2g$transcript_name_hs == "UNKNOWN" | tx2g$transcript_name_hs == "character(0)"] <- NA
tx2g$transcript_name_other[tx2g$transcript_name_other == "N/A" | tx2g$transcript_name_other == "UNKNOWN" | tx2g$transcript_name_other == "character(0)"] <- NA

```

Map transcript_name_nr to human genes to get entrez ID, GO terms and KEGG

```{r}
library(dplyr)

tx_names_nr <- data.frame(v = unlist(tx2g$transcript_name_nr), lv = tolower(unlist(tx2g$transcript_name_nr)), eid = gsub("^LOC", "", unlist(tx2g$transcript_name_nr)))

# sort tx_names_nr by v
tx_names_nr <- tx_names_nr[order(tx_names_nr$v), ]
# remove duplicate v
tx_names_nr <- tx_names_nr[!duplicated(tx_names_nr$v), ]

orgs <- c("Xl", "Hs", "Mm")
dbs <- c("org.Xl.eg.db", "org.Hs.eg.db", "org.Mm.eg.db")

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_go <- select(get(dbs[i]), keys = tx_names_nr$v, columns = c("SYMBOL", "GOALL"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      v_go <- v_go[, c(1, 2)]
      # Aggregate GOALL column based on SYMBOL column
      v_go <- aggregate(GOALL ~ SYMBOL, data = v_go, FUN = function(x) paste(x, collapse = ","))

      # append col v_go to tx_names_nr by matching SYMBOL column with the v column from tx_names_nr and grabing GOALL column
      tx_names_nr$v_go <- v_go$GOALL[match(tx_names_nr$v, v_go$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_go <- select(get(dbs[i]), keys = tx_names_nr$lv, columns = c("SYMBOL", "GOALL"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_go <- lv_go[, c(1, 2)]
      # Aggregate GOALL column based on SYMBOL column
      lv_go <- aggregate(GOALL ~ SYMBOL, data = lv_go, FUN = function(x) paste(x, collapse = ","))

      # append col lv_go to tx_names_nr by matching SYMBOL column with the lv column from tx_names_nr and grabing GOALL column
      tx_names_nr$lv_go <- lv_go$GOALL[match(tx_names_nr$lv, lv_go$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_go <- select(get(dbs[i]), keys = tx_names_nr$eid, columns = c("ENTREZID", "GOALL"), keytype = "ENTREZID")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      eid_go <- eid_go[, c(1, 2)]
      # Aggregate GOALL column based on ENTREZID column
      eid_go <- aggregate(GOALL ~ ENTREZID, data = eid_go, FUN = function(x) paste(x, collapse = ","))

      # append col eid_go to tx_names_nr by matching ENTREZID column with the eid column from tx_names_nr and grabing GOALL column
      tx_names_nr$eid_go <- eid_go$GOALL[match(tx_names_nr$eid, eid_go$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_nr$v_go, tx_names_nr$lv_go and tx_names_nr$eid_go IF THEY EXIST
  tx_names_nr$go <- paste(tx_names_nr$v_go, tx_names_nr$lv_go, tx_names_nr$eid_go, sep = ",")
  # Remove trailing commas
  tx_names_nr$go <- gsub(",$", "", tx_names_nr$go)

  # rename go column to org matching its name
  colnames(tx_names_nr)[colnames(tx_names_nr) == "go"] <- paste0("go_", orgs[i])

  # keep only columns
  tx_names_nr$v_go <- NULL
  tx_names_nr$lv_go <- NULL
  tx_names_nr$eid_go <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("go_", orgs[i]))
  tx_names_nr[, paste0("go_", orgs[i])] <- gsub(",NA", "", tx_names_nr[, paste0("go_", orgs[i])])

  # Remove duplicate GO terms
  tx_names_nr[, paste0("go_", orgs[i])] <- sapply(strsplit(tx_names_nr[, paste0("go_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

# KEGG

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_kegg <- select(get(dbs[i]), keys = tx_names_nr$v, columns = c("SYMBOL", "PATH"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      v_kegg <- v_kegg[, c(1, 2)]
      # Aggregate PATH column based on SYMBOL column
      v_kegg <- aggregate(PATH ~ SYMBOL, data = v_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col v_kegg to tx_names_nr by matching SYMBOL column with the v column from tx_names_nr and grabing PATH column
      tx_names_nr$v_kegg <- v_kegg$PATH[match(tx_names_nr$v, v_kegg$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_kegg <- select(get(dbs[i]), keys = tx_names_nr$lv, columns = c("SYMBOL", "PATH"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_kegg <- lv_kegg[, c(1, 2)]
      # Aggregate PATH column based on SYMBOL column
      lv_kegg <- aggregate(PATH ~ SYMBOL, data = lv_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col lv_kegg to tx_names_nr by matching SYMBOL column with the lv column from tx_names_nr and grabing PATH column
      tx_names_nr$lv_kegg <- lv_kegg$PATH[match(tx_names_nr$lv, lv_kegg$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_kegg <- select(get(dbs[i]), keys = tx_names_nr$eid, columns = c("ENTREZID", "PATH"), keytype = "ENTREZID")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      eid_kegg <- eid_kegg[, c(1, 2)]
      # Aggregate PATH column based on ENTREZID column
      eid_kegg <- aggregate(PATH ~ ENTREZID, data = eid_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col eid_kegg to tx_names_nr by matching ENTREZID column with the eid column from tx_names_nr and grabing PATH column
      tx_names_nr$eid_kegg <- eid_kegg$PATH[match(tx_names_nr$eid, eid_kegg$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_nr$v_kegg, tx_names_nr$lv_kegg and tx_names_nr$eid_kegg IF THEY EXIST
  tx_names_nr$kegg <- paste(tx_names_nr$v_kegg, tx_names_nr$lv_kegg, tx_names_nr$eid_kegg, sep = ",")
  # Remove trailing commas
  tx_names_nr$kegg <- gsub(",$", "", tx_names_nr$kegg)

  # rename kegg column to org matching its name
  colnames(tx_names_nr)[colnames(tx_names_nr) == "kegg"] <- paste0("kegg_", orgs[i])

  # keep only columns
  tx_names_nr$v_kegg <- NULL
  tx_names_nr$lv_kegg <- NULL
  tx_names_nr$eid_kegg <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("kegg_", orgs[i]))
  tx_names_nr[, paste0("kegg_", orgs[i])] <- gsub(",NA", "", tx_names_nr[, paste0("kegg_", orgs[i])])

  # Remove duplicate kegg terms
  tx_names_nr[, paste0("kegg_", orgs[i])] <- sapply(strsplit(tx_names_nr[, paste0("kegg_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

# ENTREZID

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_ezid <- select(get(dbs[i]), keys = tx_names_nr$v, columns = c("SYMBOL", "ENTREZID"), keytype = "SYMBOL")
      # remove extra cols
      v_ezid <- v_ezid[, c(1, 2)]
      # Aggregate ENTREZID column based on SYMBOL column
      v_ezid <- aggregate(ENTREZID ~ SYMBOL, data = v_ezid, FUN = function(x) paste(x, collapse = ","))

      # append col v_ezid to tx_names_nr by matching SYMBOL column with the v column from tx_names_nr and grabing ENTREZID column
      tx_names_nr$v_ezid <- v_ezid$ENTREZID[match(tx_names_nr$v, v_ezid$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_ezid <- select(get(dbs[i]), keys = tx_names_nr$lv, columns = c("SYMBOL", "ENTREZID"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_ezid <- lv_ezid[, c(1, 2)]
      # Aggregate ENTREZID column based on SYMBOL column
      lv_ezid <- aggregate(ENTREZID ~ SYMBOL, data = lv_ezid, FUN = function(x) paste(x, collapse = ","))

      # append col lv_ezid to tx_names_nr by matching SYMBOL column with the lv column from tx_names_nr and grabing ENTREZID column
      tx_names_nr$lv_ezid <- lv_ezid$ENTREZID[match(tx_names_nr$lv, lv_ezid$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_ezid <- select(get(dbs[i]), keys = tx_names_nr$eid, columns = c("SYMBOL", "ENTREZID"), keytype = "ENTREZID")

      # remove rows where SYMBOL is NA
      eid_ezid <- eid_ezid[!is.na(eid_ezid$SYMBOL), ]

      # append col eid_ezid to tx_names_nr by matching ENTREZID column with the eid column from tx_names_nr and grabing ENTREZID column
      tx_names_nr$eid_ezid <- eid_ezid$ENTREZID[match(tx_names_nr$eid, eid_ezid$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_nr$v_ezid, tx_names_nr$lv_ezid and tx_names_nr$eid_ezid IF THEY EXIST
  tx_names_nr$ezid <- paste(tx_names_nr$v_ezid, tx_names_nr$lv_ezid, tx_names_nr$eid_ezid, sep = ",")
  # Remove trailing commas
  tx_names_nr$ezid <- gsub(",$", "", tx_names_nr$ezid)

  # rename ezid column to org matching its name
  colnames(tx_names_nr)[colnames(tx_names_nr) == "ezid"] <- paste0("ezid_", orgs[i])

  # keep only columns
  tx_names_nr$v_ezid <- NULL
  tx_names_nr$lv_ezid <- NULL
  tx_names_nr$eid_ezid <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("ezid_", orgs[i]))
  tx_names_nr[, paste0("ezid_", orgs[i])] <- gsub(",NA", "", tx_names_nr[, paste0("ezid_", orgs[i])])

  # Remove duplicate ezid terms
  tx_names_nr[, paste0("ezid_", orgs[i])] <- sapply(strsplit(tx_names_nr[, paste0("ezid_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

tx_names_nr[tx_names_nr == "NA"] <- NA

# if v starts with "LOC", copy number part to NEW ezid_loc column (fill the rest with NA)
tx_names_nr$ezid_loc <- NA
tx_names_nr$ezid_loc[grepl("^LOC", tx_names_nr$v)] <- gsub("^LOC", "", tx_names_nr$v[grepl("^LOC", tx_names_nr$v)])


# remove lv and eid columns
tx_names_nr$lv <- NULL
tx_names_nr$eid <- NULL

# append "_nr" to every column name except the first one
colnames(tx_names_nr)[colnames(tx_names_nr) != "v"] <- paste0(colnames(tx_names_nr)[colnames(tx_names_nr) != "v"], "_nr")

rownames(tx_names_nr) <- NULL

head(tx_names_nr)


# join tx_names_nr$go_Xl_nr, tx_names_nr$go_Hs_nr and tx_names_nr$go_Mm_nr
tx_names_nr$go_all_nr <- paste(tx_names_nr$go_Xl_nr, tx_names_nr$go_Hs_nr, tx_names_nr$go_Mm_nr, sep = ",")
# Remove "NA," or ",NA" from tx_names_nr$go_all_nr
tx_names_nr$go_all_nr <- gsub("NA,", "", tx_names_nr$go_all_nr)
tx_names_nr$go_all_nr <- gsub(",NA", "", tx_names_nr$go_all_nr)
tx_names_nr[tx_names_nr == "NA"] <- NA

# join tx_names_nr$kegg_Xl_nr, tx_names_nr$kegg_Hs_nr and tx_names_nr$kegg_Mm_nr
tx_names_nr$kegg_all_nr <- paste(tx_names_nr$kegg_Xl_nr, tx_names_nr$kegg_Hs_nr, tx_names_nr$kegg_Mm_nr, sep = ",")
# Remove "NA," or ",NA" from tx_names_nr$kegg_all_nr
tx_names_nr$kegg_all_nr <- gsub("NA,", "", tx_names_nr$kegg_all_nr)
tx_names_nr$kegg_all_nr <- gsub(",NA", "", tx_names_nr$kegg_all_nr)
tx_names_nr[tx_names_nr == "NA"] <- NA

# join tx_names_nr$ezid_Xl_nr, tx_names_nr$ezid_Hs_nr and tx_names_nr$ezid_Mm_nr
tx_names_nr$ezid_all_nr <- paste(tx_names_nr$ezid_Xl_nr, tx_names_nr$ezid_Hs_nr, tx_names_nr$ezid_Mm_nr, tx_names_nr$ezid_loc_nr, sep = ",")
# Remove "NA," or ",NA" from tx_names_nr$ezid_all_nr
tx_names_nr$ezid_all_nr <- gsub("NA,", "", tx_names_nr$ezid_all_nr)
tx_names_nr$ezid_all_nr <- gsub(",NA", "", tx_names_nr$ezid_all_nr)
tx_names_nr[tx_names_nr == "NA"] <- NA

# add column transcript_name_nr_goall to tx2g by matching tx_names_nr$v with tx2g$transcript_name_nr and grabbing tx_names_nr$go_all_nr
tx2g$transcript_name_nr_goall <- tx_names_nr$go_all_nr[match(tx2g$transcript_name_nr, tx_names_nr$v)]

# add column transcript_name_nr_keggpath to tx2g by matching tx_names_nr$v with tx2g$transcript_name_nr and grabbing tx_names_nr$kegg_all_nr
tx2g$transcript_name_nr_keggpath <- tx_names_nr$kegg_all_nr[match(tx2g$transcript_name_nr, tx_names_nr$v)]

# add column transcript_name_nr_entrezid to tx2g by matching tx_names_nr$v with tx2g$transcript_name_nr and grabbing tx_names_nr$ezid_all_nr
tx2g$transcript_name_nr_entrezid <- tx_names_nr$ezid_all_nr[match(tx2g$transcript_name_nr, tx_names_nr$v)]
```

Map transcript_name_hs to human genes to get entrez ID, GO terms and KEGG

```{r}
library(dplyr)

tx_names_hs <- data.frame(v = unlist(tx2g$transcript_name_hs), lv = tolower(unlist(tx2g$transcript_name_hs)), eid = gsub("^LOC", "", unlist(tx2g$transcript_name_hs)))


# sort tx_names_hs by v
tx_names_hs <- tx_names_hs[order(tx_names_hs$v), ]
# remove duplicate v
tx_names_hs <- tx_names_hs[!duplicated(tx_names_hs$v), ]

orgs <- c("Hs")
dbs <- c("org.Hs.eg.db")

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_go <- select(get(dbs[i]), keys = tx_names_hs$v, columns = c("SYMBOL", "GOALL"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      v_go <- v_go[, c(1, 2)]
      # Aggregate GOALL column based on SYMBOL column
      v_go <- aggregate(GOALL ~ SYMBOL, data = v_go, FUN = function(x) paste(x, collapse = ","))

      # append col v_go to tx_names_hs by matching SYMBOL column with the v column from tx_names_hs and grabing GOALL column
      tx_names_hs$v_go <- v_go$GOALL[match(tx_names_hs$v, v_go$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_go <- select(get(dbs[i]), keys = tx_names_hs$lv, columns = c("SYMBOL", "GOALL"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_go <- lv_go[, c(1, 2)]
      # Aggregate GOALL column based on SYMBOL column
      lv_go <- aggregate(GOALL ~ SYMBOL, data = lv_go, FUN = function(x) paste(x, collapse = ","))

      # append col lv_go to tx_names_hs by matching SYMBOL column with the lv column from tx_names_hs and grabing GOALL column
      tx_names_hs$lv_go <- lv_go$GOALL[match(tx_names_hs$lv, lv_go$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_go <- select(get(dbs[i]), keys = tx_names_hs$eid, columns = c("ENTREZID", "GOALL"), keytype = "ENTREZID")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      eid_go <- eid_go[, c(1, 2)]
      # Aggregate GOALL column based on ENTREZID column
      eid_go <- aggregate(GOALL ~ ENTREZID, data = eid_go, FUN = function(x) paste(x, collapse = ","))

      # append col eid_go to tx_names_hs by matching ENTREZID column with the eid column from tx_names_hs and grabing GOALL column
      tx_names_hs$eid_go <- eid_go$GOALL[match(tx_names_hs$eid, eid_go$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_hs$v_go, tx_names_hs$lv_go and tx_names_hs$eid_go IF THEY EXIST
  tx_names_hs$go <- paste(tx_names_hs$v_go, tx_names_hs$lv_go, tx_names_hs$eid_go, sep = ",")
  # Remove trailing commas
  tx_names_hs$go <- gsub(",$", "", tx_names_hs$go)

  # rename go column to org matching its name
  colnames(tx_names_hs)[colnames(tx_names_hs) == "go"] <- paste0("go_", orgs[i])

  # keep only columns
  tx_names_hs$v_go <- NULL
  tx_names_hs$lv_go <- NULL
  tx_names_hs$eid_go <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("go_", orgs[i]))
  tx_names_hs[, paste0("go_", orgs[i])] <- gsub(",NA", "", tx_names_hs[, paste0("go_", orgs[i])])

  # Remove duplicate GO terms
  tx_names_hs[, paste0("go_", orgs[i])] <- sapply(strsplit(tx_names_hs[, paste0("go_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

# KEGG

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_kegg <- select(get(dbs[i]), keys = tx_names_hs$v, columns = c("SYMBOL", "PATH"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      v_kegg <- v_kegg[, c(1, 2)]
      # Aggregate PATH column based on SYMBOL column
      v_kegg <- aggregate(PATH ~ SYMBOL, data = v_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col v_kegg to tx_names_hs by matching SYMBOL column with the v column from tx_names_hs and grabing PATH column
      tx_names_hs$v_kegg <- v_kegg$PATH[match(tx_names_hs$v, v_kegg$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_kegg <- select(get(dbs[i]), keys = tx_names_hs$lv, columns = c("SYMBOL", "PATH"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_kegg <- lv_kegg[, c(1, 2)]
      # Aggregate PATH column based on SYMBOL column
      lv_kegg <- aggregate(PATH ~ SYMBOL, data = lv_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col lv_kegg to tx_names_hs by matching SYMBOL column with the lv column from tx_names_hs and grabing PATH column
      tx_names_hs$lv_kegg <- lv_kegg$PATH[match(tx_names_hs$lv, lv_kegg$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_kegg <- select(get(dbs[i]), keys = tx_names_hs$eid, columns = c("ENTREZID", "PATH"), keytype = "ENTREZID")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      eid_kegg <- eid_kegg[, c(1, 2)]
      # Aggregate PATH column based on ENTREZID column
      eid_kegg <- aggregate(PATH ~ ENTREZID, data = eid_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col eid_kegg to tx_names_hs by matching ENTREZID column with the eid column from tx_names_hs and grabing PATH column
      tx_names_hs$eid_kegg <- eid_kegg$PATH[match(tx_names_hs$eid, eid_kegg$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_hs$v_kegg, tx_names_hs$lv_kegg and tx_names_hs$eid_kegg IF THEY EXIST
  tx_names_hs$kegg <- paste(tx_names_hs$v_kegg, tx_names_hs$lv_kegg, tx_names_hs$eid_kegg, sep = ",")
  # Remove trailing commas
  tx_names_hs$kegg <- gsub(",$", "", tx_names_hs$kegg)

  # rename kegg column to org matching its name
  colnames(tx_names_hs)[colnames(tx_names_hs) == "kegg"] <- paste0("kegg_", orgs[i])

  # keep only columns
  tx_names_hs$v_kegg <- NULL
  tx_names_hs$lv_kegg <- NULL
  tx_names_hs$eid_kegg <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("kegg_", orgs[i]))
  tx_names_hs[, paste0("kegg_", orgs[i])] <- gsub(",NA", "", tx_names_hs[, paste0("kegg_", orgs[i])])

  # Remove duplicate kegg terms
  tx_names_hs[, paste0("kegg_", orgs[i])] <- sapply(strsplit(tx_names_hs[, paste0("kegg_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

# ENTREZID

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_ezid <- select(get(dbs[i]), keys = tx_names_hs$v, columns = c("SYMBOL", "ENTREZID"), keytype = "SYMBOL")
      # remove extra cols
      v_ezid <- v_ezid[, c(1, 2)]
      # Aggregate ENTREZID column based on SYMBOL column
      v_ezid <- aggregate(ENTREZID ~ SYMBOL, data = v_ezid, FUN = function(x) paste(x, collapse = ","))

      # append col v_ezid to tx_names_hs by matching SYMBOL column with the v column from tx_names_hs and grabing ENTREZID column
      tx_names_hs$v_ezid <- v_ezid$ENTREZID[match(tx_names_hs$v, v_ezid$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_ezid <- select(get(dbs[i]), keys = tx_names_hs$lv, columns = c("SYMBOL", "ENTREZID"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_ezid <- lv_ezid[, c(1, 2)]
      # Aggregate ENTREZID column based on SYMBOL column
      lv_ezid <- aggregate(ENTREZID ~ SYMBOL, data = lv_ezid, FUN = function(x) paste(x, collapse = ","))

      # append col lv_ezid to tx_names_hs by matching SYMBOL column with the lv column from tx_names_hs and grabing ENTREZID column
      tx_names_hs$lv_ezid <- lv_ezid$ENTREZID[match(tx_names_hs$lv, lv_ezid$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_ezid <- select(get(dbs[i]), keys = tx_names_hs$eid, columns = c("SYMBOL", "ENTREZID"), keytype = "ENTREZID")

      # remove rows where SYMBOL is NA
      eid_ezid <- eid_ezid[!is.na(eid_ezid$SYMBOL), ]

      # append col eid_ezid to tx_names_hs by matching ENTREZID column with the eid column from tx_names_hs and grabing ENTREZID column
      tx_names_hs$eid_ezid <- eid_ezid$ENTREZID[match(tx_names_hs$eid, eid_ezid$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_hs$v_ezid, tx_names_hs$lv_ezid and tx_names_hs$eid_ezid IF THEY EXIST
  tx_names_hs$ezid <- paste(tx_names_hs$v_ezid, tx_names_hs$lv_ezid, tx_names_hs$eid_ezid, sep = ",")
  # Remove trailing commas
  tx_names_hs$ezid <- gsub(",$", "", tx_names_hs$ezid)

  # rename ezid column to org matching its name
  colnames(tx_names_hs)[colnames(tx_names_hs) == "ezid"] <- paste0("ezid_", orgs[i])

  # keep only columns
  tx_names_hs$v_ezid <- NULL
  tx_names_hs$lv_ezid <- NULL
  tx_names_hs$eid_ezid <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("ezid_", orgs[i]))
  tx_names_hs[, paste0("ezid_", orgs[i])] <- gsub(",NA", "", tx_names_hs[, paste0("ezid_", orgs[i])])

  # Remove duplicate ezid terms
  tx_names_hs[, paste0("ezid_", orgs[i])] <- sapply(strsplit(tx_names_hs[, paste0("ezid_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

tx_names_hs[tx_names_hs == "NA"] <- NA

# if v starts with "LOC", copy number part to NEW ezid_loc column (fill the rest with NA)
tx_names_hs$ezid_loc <- NA
tx_names_hs$ezid_loc[grepl("^LOC", tx_names_hs$v)] <- gsub("^LOC", "", tx_names_hs$v[grepl("^LOC", tx_names_hs$v)])

# remove lv and eid columns
tx_names_hs$lv <- NULL
tx_names_hs$eid <- NULL

# append "_hs" to every column name except the first one
colnames(tx_names_hs)[colnames(tx_names_hs) != "v"] <- paste0(colnames(tx_names_hs)[colnames(tx_names_hs) != "v"], "_hs")

rownames(tx_names_hs) <- NULL

head(tx_names_hs)

# join tx_names_hs ezids
tx_names_hs$ezid_all_hs <- paste(tx_names_hs$ezid_Hs_hs, tx_names_hs$ezid_loc_hs, sep = ",")
# Remove "NA," or ",NA" from tx_names_hs$ezid_all_hs
tx_names_hs$ezid_all_hs <- gsub("NA,", "", tx_names_hs$ezid_all_hs)
tx_names_hs$ezid_all_hs <- gsub(",NA", "", tx_names_hs$ezid_all_hs)
tx_names_hs[tx_names_hs == "NA"] <- NA

tx2g$transcript_name_hs_goall <- tx_names_hs$go_Hs_hs[match(tx2g$transcript_name_hs, tx_names_hs$v)]

tx2g$transcript_name_hs_keggpath <- tx_names_hs$kegg_Hs_hs[match(tx2g$transcript_name_hs, tx_names_hs$v)]

tx2g$transcript_name_hs_entrezid <- tx_names_hs$ezid_all_hs[match(tx2g$transcript_name_hs, tx_names_hs$v)]
```

Map transcript_name_other to human genes to get entrez ID, GO terms and KEGG

```{r}
library(dplyr)

tx_names_other <- data.frame(v = unlist(tx2g$transcript_name_other), lv = tolower(unlist(tx2g$transcript_name_other)), eid = gsub("^LOC", "", unlist(tx2g$transcript_name_other)))

# sort tx_names_other by v
tx_names_other <- tx_names_other[order(tx_names_other$v), ]
# remove duplicate v
tx_names_other <- tx_names_other[!duplicated(tx_names_other$v), ]

orgs <- c("Xl", "Hs", "Mm")
dbs <- c("org.Xl.eg.db", "org.Hs.eg.db", "org.Mm.eg.db")

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_go <- select(get(dbs[i]), keys = tx_names_other$v, columns = c("SYMBOL", "GOALL"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      v_go <- v_go[, c(1, 2)]
      # Aggregate GOALL column based on SYMBOL column
      v_go <- aggregate(GOALL ~ SYMBOL, data = v_go, FUN = function(x) paste(x, collapse = ","))

      # append col v_go to tx_names_other by matching SYMBOL column with the v column from tx_names_other and grabing GOALL column
      tx_names_other$v_go <- v_go$GOALL[match(tx_names_other$v, v_go$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_go <- select(get(dbs[i]), keys = tx_names_other$lv, columns = c("SYMBOL", "GOALL"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_go <- lv_go[, c(1, 2)]
      # Aggregate GOALL column based on SYMBOL column
      lv_go <- aggregate(GOALL ~ SYMBOL, data = lv_go, FUN = function(x) paste(x, collapse = ","))

      # append col lv_go to tx_names_other by matching SYMBOL column with the lv column from tx_names_other and grabing GOALL column
      tx_names_other$lv_go <- lv_go$GOALL[match(tx_names_other$lv, lv_go$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_go <- select(get(dbs[i]), keys = tx_names_other$eid, columns = c("ENTREZID", "GOALL"), keytype = "ENTREZID")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      eid_go <- eid_go[, c(1, 2)]
      # Aggregate GOALL column based on ENTREZID column
      eid_go <- aggregate(GOALL ~ ENTREZID, data = eid_go, FUN = function(x) paste(x, collapse = ","))

      # append col eid_go to tx_names_other by matching ENTREZID column with the eid column from tx_names_other and grabing GOALL column
      tx_names_other$eid_go <- eid_go$GOALL[match(tx_names_other$eid, eid_go$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_other$v_go, tx_names_other$lv_go and tx_names_other$eid_go IF THEY EXIST
  tx_names_other$go <- paste(tx_names_other$v_go, tx_names_other$lv_go, tx_names_other$eid_go, sep = ",")
  # Remove trailing commas
  tx_names_other$go <- gsub(",$", "", tx_names_other$go)

  # rename go column to org matching its name
  colnames(tx_names_other)[colnames(tx_names_other) == "go"] <- paste0("go_", orgs[i])

  # keep only columns
  tx_names_other$v_go <- NULL
  tx_names_other$lv_go <- NULL
  tx_names_other$eid_go <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("go_", orgs[i]))
  tx_names_other[, paste0("go_", orgs[i])] <- gsub(",NA", "", tx_names_other[, paste0("go_", orgs[i])])

  # Remove duplicate GO terms
  tx_names_other[, paste0("go_", orgs[i])] <- sapply(strsplit(tx_names_other[, paste0("go_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

# KEGG

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_kegg <- select(get(dbs[i]), keys = tx_names_other$v, columns = c("SYMBOL", "PATH"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      v_kegg <- v_kegg[, c(1, 2)]
      # Aggregate PATH column based on SYMBOL column
      v_kegg <- aggregate(PATH ~ SYMBOL, data = v_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col v_kegg to tx_names_other by matching SYMBOL column with the v column from tx_names_other and grabing PATH column
      tx_names_other$v_kegg <- v_kegg$PATH[match(tx_names_other$v, v_kegg$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_kegg <- select(get(dbs[i]), keys = tx_names_other$lv, columns = c("SYMBOL", "PATH"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_kegg <- lv_kegg[, c(1, 2)]
      # Aggregate PATH column based on SYMBOL column
      lv_kegg <- aggregate(PATH ~ SYMBOL, data = lv_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col lv_kegg to tx_names_other by matching SYMBOL column with the lv column from tx_names_other and grabing PATH column
      tx_names_other$lv_kegg <- lv_kegg$PATH[match(tx_names_other$lv, lv_kegg$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_kegg <- select(get(dbs[i]), keys = tx_names_other$eid, columns = c("ENTREZID", "PATH"), keytype = "ENTREZID")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      eid_kegg <- eid_kegg[, c(1, 2)]
      # Aggregate PATH column based on ENTREZID column
      eid_kegg <- aggregate(PATH ~ ENTREZID, data = eid_kegg, FUN = function(x) paste(x, collapse = ","))

      # append col eid_kegg to tx_names_other by matching ENTREZID column with the eid column from tx_names_other and grabing PATH column
      tx_names_other$eid_kegg <- eid_kegg$PATH[match(tx_names_other$eid, eid_kegg$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_other$v_kegg, tx_names_other$lv_kegg and tx_names_other$eid_kegg IF THEY EXIST
  tx_names_other$kegg <- paste(tx_names_other$v_kegg, tx_names_other$lv_kegg, tx_names_other$eid_kegg, sep = ",")
  # Remove trailing commas
  tx_names_other$kegg <- gsub(",$", "", tx_names_other$kegg)

  # rename kegg column to org matching its name
  colnames(tx_names_other)[colnames(tx_names_other) == "kegg"] <- paste0("kegg_", orgs[i])

  # keep only columns
  tx_names_other$v_kegg <- NULL
  tx_names_other$lv_kegg <- NULL
  tx_names_other$eid_kegg <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("kegg_", orgs[i]))
  tx_names_other[, paste0("kegg_", orgs[i])] <- gsub(",NA", "", tx_names_other[, paste0("kegg_", orgs[i])])

  # Remove duplicate kegg terms
  tx_names_other[, paste0("kegg_", orgs[i])] <- sapply(strsplit(tx_names_other[, paste0("kegg_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

# ENTREZID

for (i in seq_along(orgs)) {
  library(dbs[i], character.only = TRUE)

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names..."))
      v_ezid <- select(get(dbs[i]), keys = tx_names_other$v, columns = c("SYMBOL", "ENTREZID"), keytype = "SYMBOL")
      # remove extra cols
      v_ezid <- v_ezid[, c(1, 2)]
      # Aggregate ENTREZID column based on SYMBOL column
      v_ezid <- aggregate(ENTREZID ~ SYMBOL, data = v_ezid, FUN = function(x) paste(x, collapse = ","))

      # append col v_ezid to tx_names_other by matching SYMBOL column with the v column from tx_names_other and grabing ENTREZID column
      tx_names_other$v_ezid <- v_ezid$ENTREZID[match(tx_names_other$v, v_ezid$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to lowercase transcript names..."))
      lv_ezid <- select(get(dbs[i]), keys = tx_names_other$lv, columns = c("SYMBOL", "ENTREZID"), keytype = "SYMBOL")
      # remove extra cols "EVIDENCEALL" and "ONTOLOGYALL"
      lv_ezid <- lv_ezid[, c(1, 2)]
      # Aggregate ENTREZID column based on SYMBOL column
      lv_ezid <- aggregate(ENTREZID ~ SYMBOL, data = lv_ezid, FUN = function(x) paste(x, collapse = ","))

      # append col lv_ezid to tx_names_other by matching SYMBOL column with the lv column from tx_names_other and grabing ENTREZID column
      tx_names_other$lv_ezid <- lv_ezid$ENTREZID[match(tx_names_other$lv, lv_ezid$SYMBOL)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )

  tryCatch(
    {
      message(paste("Mapping", orgs[i], "genes to transcript names with 'LOC' removed..."))
      eid_ezid <- select(get(dbs[i]), keys = tx_names_other$eid, columns = c("SYMBOL", "ENTREZID"), keytype = "ENTREZID")

      # remove rows where SYMBOL is NA
      eid_ezid <- eid_ezid[!is.na(eid_ezid$SYMBOL), ]

      # append col eid_ezid to tx_names_other by matching ENTREZID column with the eid column from tx_names_other and grabing ENTREZID column
      tx_names_other$eid_ezid <- eid_ezid$ENTREZID[match(tx_names_other$eid, eid_ezid$ENTREZID)]
    },
    error = function(e) {
      cat("ERROR :", conditionMessage(e), "\n")
    }
  )


  # join tx_names_other$v_ezid, tx_names_other$lv_ezid and tx_names_other$eid_ezid IF THEY EXIST
  tx_names_other$ezid <- paste(tx_names_other$v_ezid, tx_names_other$lv_ezid, tx_names_other$eid_ezid, sep = ",")
  # Remove trailing commas
  tx_names_other$ezid <- gsub(",$", "", tx_names_other$ezid)

  # rename ezid column to org matching its name
  colnames(tx_names_other)[colnames(tx_names_other) == "ezid"] <- paste0("ezid_", orgs[i])

  # keep only columns
  tx_names_other$v_ezid <- NULL
  tx_names_other$lv_ezid <- NULL
  tx_names_other$eid_ezid <- NULL

  # Remove NAs from the column whose name is the current organism (paste0("ezid_", orgs[i]))
  tx_names_other[, paste0("ezid_", orgs[i])] <- gsub(",NA", "", tx_names_other[, paste0("ezid_", orgs[i])])

  # Remove duplicate ezid terms
  tx_names_other[, paste0("ezid_", orgs[i])] <- sapply(strsplit(tx_names_other[, paste0("ezid_", orgs[i])], ","), function(x) paste(unique(x), collapse = ","))
}

tx_names_other[tx_names_other == "NA"] <- NA

# if v starts with "LOC", copy number part to NEW ezid_loc column (fill the rest with NA)
tx_names_other$ezid_loc <- NA
tx_names_other$ezid_loc[grepl("^LOC", tx_names_other$v)] <- gsub("^LOC", "", tx_names_other$v[grepl("^LOC", tx_names_other$v)])


# remove lv and eid columns
tx_names_other$lv <- NULL
tx_names_other$eid <- NULL

# append "_other" to every column name except the first one
colnames(tx_names_other)[colnames(tx_names_other) != "v"] <- paste0(colnames(tx_names_other)[colnames(tx_names_other) != "v"], "_other")



rownames(tx_names_other) <- NULL

head(tx_names_other)



# join tx_names_other$go_Xl_other, tx_names_other$go_Hs_other and tx_names_other$go_Mm_other
tx_names_other$go_all_other <- paste(tx_names_other$go_Xl_other, tx_names_other$go_Hs_other, tx_names_other$go_Mm_other, sep = ",")
# Remove "NA," or ",NA" from tx_names_other$go_all_other
tx_names_other$go_all_other <- gsub("NA,", "", tx_names_other$go_all_other)
tx_names_other$go_all_other <- gsub(",NA", "", tx_names_other$go_all_other)
tx_names_other[tx_names_other == "NA"] <- NA

# join tx_names_other$kegg_Xl_other, tx_names_other$kegg_Hs_other and tx_names_other$kegg_Mm_other
tx_names_other$kegg_all_other <- paste(tx_names_other$kegg_Xl_other, tx_names_other$kegg_Hs_other, tx_names_other$kegg_Mm_other, sep = ",")
# Remove "NA," or ",NA" from tx_names_other$kegg_all_other
tx_names_other$kegg_all_other <- gsub("NA,", "", tx_names_other$kegg_all_other)
tx_names_other$kegg_all_other <- gsub(",NA", "", tx_names_other$kegg_all_other)
tx_names_other[tx_names_other == "NA"] <- NA

# join tx_names_other$ezid_Xl_other, tx_names_other$ezid_Hs_other and tx_names_other$ezid_Mm_other
tx_names_other$ezid_all_other <- paste(tx_names_other$ezid_Xl_other, tx_names_other$ezid_Hs_other, tx_names_other$ezid_Mm_other, tx_names_other$ezid_loc_other, sep = ",")
# Remove "NA," or ",NA" from tx_names_other$ezid_all_other
tx_names_other$ezid_all_other <- gsub("NA,", "", tx_names_other$ezid_all_other)
tx_names_other$ezid_all_other <- gsub(",NA", "", tx_names_other$ezid_all_other)
tx_names_other[tx_names_other == "NA"] <- NA

# add column transcript_name_other_goall to tx2g by matching tx_names_other$v with tx2g$transcript_name_other and grabbing tx_names_other$go_all_other
tx2g$transcript_name_other_goall <- tx_names_other$go_all_other[match(tx2g$transcript_name_other, tx_names_other$v)]

# add column transcript_name_other_keggpath to tx2g by matching tx_names_other$v with tx2g$transcript_name_other and grabbing tx_names_other$kegg_all_other
tx2g$transcript_name_other_keggpath <- tx_names_other$kegg_all_other[match(tx2g$transcript_name_other, tx_names_other$v)]

# add column transcript_name_other_entrezid to tx2g by matching tx_names_other$v with tx2g$transcript_name_other and grabbing tx_names_other$ezid_all_other
tx2g$transcript_name_other_entrezid <- tx_names_other$ezid_all_other[match(tx2g$transcript_name_other, tx_names_other$v)]
```

Adding eggnog mapper results

```{bash}
# remove first 4 lines of eggnog annotations
sed '1,4d' $PROJECT_DIR/protein_annotation/output/AmexT_v47_protein.emapper.annotations > $PROJECT_DIR/protein_annotation/output/AmexT_v47_protein.emapper.annotations.nocomm.txt


```

```{r}
# read the necessary information from the eggNOG-mapper output file
eggnog_data <- read.csv(file.path(PROJECT_DIR, "/protein_annotation/output/AmexT_v47_protein.emapper.annotations.nocomm.txt"), header = TRUE, sep = "\t")

# Replace any cells with "-" with NA
eggnog_data[eggnog_data == "-"] <- NA
# Replace any cells with "NA" with NA
eggnog_data[eggnog_data == "NA"] <- NA

# remove X. from first column name
colnames(eggnog_data)[1] <- "query"

# Remove semicolon from transcript ids
eggnog_data$query <- gsub(";", "", eggnog_data$query)


# transcript_name_eggnog is eggnog_data$Prefered_name
tx2g$transcript_name_eggnog <- eggnog_data$Preferred_name[match(tx2g$transcript_id, eggnog_data$query)]


# transcript_name_eggnog_goall is eggnog_data$GOs
tx2g$transcript_name_eggnog_goall <- eggnog_data$GOs[match(tx2g$transcript_id, eggnog_data$query)]

# for transcript_name_eggnog_keggpath if there are duplicates of the form "ko000001" and "map000001" keep only "ko000001"
kegg_data <- eggnog_data[c("KEGG_Pathway", "query")]

library(data.table)
# expand, since some genes/proteins will have multiple assigned KO terms
kegg_data <- data.table(kegg_data)
kegg_data <- kegg_data[, list(KEGG_Pathway = unlist(strsplit(KEGG_Pathway, ","))), by = query]

# select the needed columns
kegg_data <- kegg_data[, c(2, 1)]

# Create temp column by removing map and ko from KEGG_Pathway IDs
kegg_data$temp <- gsub("[^[:digit:]]", "", kegg_data$KEGG_Pathway)

library(dplyr)
# Keep only one instance of quey and temp combination (to remove redundant map IDs)
kegg_data <- distinct(kegg_data, temp, query, .keep_all = TRUE)

# Remove temp column
kegg_data$temp <- NULL

# aggregate KEGG_Pathway column based on query column
kegg_data <- aggregate(KEGG_Pathway ~ query, data = kegg_data, FUN = function(x) paste(x, collapse = ","))

# add column transcript_name_eggnog_keggpath to tx2g
tx2g$transcript_name_eggnog_keggpath <- kegg_data$KEGG_Pathway[match(tx2g$transcript_id, kegg_data$query)]
```

```{r}
# add transcript_name_all by pasting transcript_name_nr, transcript_name_hs, transcript_name_other and transcript_name_eggnog

tx2g$transcript_name_all <- paste(tx2g$transcript_name_nr, tx2g$transcript_name_hs, tx2g$transcript_name_other, tx2g$transcript_name_eggnog, sep = ",")

# remove any ",NA" from transcript_name_all
tx2g$transcript_name_all <- gsub(",NA", "", tx2g$transcript_name_all)

# remove any "NA," from transcript_name_all
tx2g$transcript_name_all <- gsub("NA,", "", tx2g$transcript_name_all)

# remove duplicates ignoring case
tx2g$transcript_name_all <- sapply(strsplit(tx2g$transcript_name_all, ","), function(x) paste(unique(toupper(x)), collapse = ","))

# replace "NA" with NA
tx2g$transcript_name_all[tx2g$transcript_name_all == "NA"] <- NA


# select one transcript_name_selected by the following priority: transcript_name_hs, transcript_name_eggnog, transcript_name_nr, transcript_name_other

tx2g <- tx2g %>%
  mutate(transcript_name_selected = coalesce(transcript_name_hs, transcript_name_eggnog, transcript_name_nr, transcript_name_other, transcript_name_amex))
```

```{r}
# add transcript_goall_all by pasting transcript_name_nr_goall, transcript_name_hs_goall, transcript_name_other_goall and transcript_name_eggnog_goall

tx2g$transcript_goall_all <- paste(tx2g$transcript_name_nr_goall, tx2g$transcript_name_hs_goall, tx2g$transcript_name_other_goall, tx2g$transcript_name_eggnog_goall, sep = ",")

# remove any ",NA" from transcript_goall_all
tx2g$transcript_goall_all <- gsub(",NA", "", tx2g$transcript_goall_all)

# remove any "NA," from transcript_goall_all
tx2g$transcript_goall_all <- gsub("NA,", "", tx2g$transcript_goall_all)

# remove duplicates from transcript_goall_all
tx2g$transcript_goall_all <- sapply(strsplit(tx2g$transcript_goall_all, ","), function(x) paste(unique(x), collapse = ","))

# replace "NA" with NA
tx2g$transcript_goall_all[tx2g$transcript_goall_all == "NA"] <- NA
```

```{r}	
# add transcript_keggpath_all by pasting transcript_name_nr_keggpath, transcript_name_hs_keggpath, transcript_name_other_keggpath and transcript_name_eggnog_keggpath

tx2g$transcript_keggpath_all <- paste(tx2g$transcript_name_nr_keggpath, tx2g$transcript_name_hs_keggpath, tx2g$transcript_name_other_keggpath, tx2g$transcript_name_eggnog_keggpath, sep = ",")

# remove any ",NA" from transcript_keggpath_all
tx2g$transcript_keggpath_all <- gsub(",NA", "", tx2g$transcript_keggpath_all)

# remove any "NA," from transcript_keggpath_all
tx2g$transcript_keggpath_all <- gsub("NA,", "", tx2g$transcript_keggpath_all)

# split, and if transcript_keggpath_all is a number and does not start with, "ko", prepend it
tx2g$transcript_keggpath_all <- sapply(strsplit(tx2g$transcript_keggpath_all, ","), function(x) paste(ifelse(grepl("^[0-9]", x) & !grepl("^ko", x), paste0("ko", x), x), collapse = ","))

# remove duplicates from transcript_keggpath_all

tx2g$transcript_keggpath_all <- sapply(strsplit(tx2g$transcript_keggpath_all, ","), function(x) paste(unique(x), collapse = ","))

# replace "NA" with NA
tx2g$transcript_keggpath_all[tx2g$transcript_keggpath_all == "NA"] <- NA

# save
saveRDS(tx2g, file = file.path(PROJECT_DIR, "/metadata/transcript_metadata.granges.rds"))


transcript_metadata <- as(tx2g, "data.frame")

write.table(transcript_metadata, file = file.path(PROJECT_DIR, "/metadata/transcript_metadata.txt", sep = "\t"))

# transcript_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/transcript_metadata.txt"), header = TRUE, sep = "\t", stringsAsFactors = FALSE)

# gene level

gene_metadata <- transcript_metadata

# remove first 5 columns
gene_metadata <- gene_metadata[, 6:ncol(gene_metadata)]

# remove type column, transcript_name_full column
gene_metadata$type <- NULL
gene_metadata$transcript_name_full <- NULL

# aggregate all columns based on gene_id

gene_metadata <- gene_metadata %>%
  group_by(gene_id) %>%
  summarize_all(~ paste(unique(na.omit(.)), collapse = ","))

# remove duplicates from transcript_goall_all
gene_metadata$gene_goall <- sapply(strsplit(gene_metadata$transcript_goall_all, ","), function(x) paste(unique(x), collapse = ","))

# remove duplicates from transcript_keggpath_all
gene_metadata$gene_keggpath <- sapply(strsplit(gene_metadata$transcript_keggpath_all, ","), function(x) paste(unique(x), collapse = ","))

# gene_symbol is the first transcript_name_selected that does not start with "AMEX"
gene_metadata$gene_symbol <- sapply(strsplit(gene_metadata$transcript_name_selected, ","), function(x) x[which(!grepl("^AMEX", x))[1]])

# if gene_symbol is NA, the copy gene_id to gene_symbol
gene_metadata$gene_symbol[is.na(gene_metadata$gene_symbol)] <- gene_metadata$gene_id[is.na(gene_metadata$gene_symbol)]

# convert all gene_symbol to uppercase
gene_metadata$gene_symbol <- toupper(gene_metadata$gene_symbol)

make.unique.2 <- function(x, sep = ".") {
  ave(x, x, FUN = function(a) {
    if (length(a) > 1) {
      paste(a, seq_along(a), sep = sep)
    } else {
      a
    }
  })
}

# make.unique.2(c("a","a","a","b","b","c","d","d"))

# make gene_symbol unique
gene_metadata$gene_symbol <- make.unique.2(gene_metadata$gene_symbol)

# convert all empty to NA
gene_metadata[gene_metadata == ""] <- NA

gene_metadata <- gene_metadata[, c("gene_id", "transcript_id", "gene_symbol", "gene_goall", "gene_keggpath")]

write.table(gene_metadata, file = file.path(PROJECT_DIR, "/metadata/gene_metadata.txt", sep = "\t"))
```

```{r}	
# KEGG pathway tables

# from https://gist.github.com/ATpoint/c6d8a4e0d7665bca538d35a50daa4331
```

```{bash}
GENOME_DIR=$PROJECT_DIR/ext_data/axolotl_genome

# replace "gene_name" with "transcript_name" in transcripts
while IFS= read -r line; do
  if [[ "$line" =~ [[:space:]]transcript[[:space:]] ]]; then
    # Replace "gene_name" with "transcript_name" in the line
    line=$(echo "$line" | sed 's/gene_name/transcript_name/')
  fi
  echo "$line" >> $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test
done < $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf

# keep only id in transcript id
sed 's/transcript_id "[^"]*AMEX/transcript_id "AMEX/g' \
  $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test \
  > $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test2

# remove everything after CDS from transcripts
while IFS=$'\t' read -r col1 col2 col3 rest_of_line; do
    if [[ $col3 == "transcript" ]]; then
        modified_line=$(echo "$rest_of_line" | sed 's/; CDS.*;/;/')
        echo -e "$col1\t$col2\t$col3\t$modified_line"
    else
        echo -e "$col1\t$col2\t$col3\t$rest_of_line"
    fi
done < $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test2 > $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test3

# remove trailing spaces from all lines
sed -e 's/[[:space:]]*$//' $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test3 > $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test4

agat_sp_add_introns.pl \
    -f $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test4 \
    --out $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.gff3

agat_sp_add_intergenic_regions.pl \
    -f $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.gff3 \
    --out $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gff3


rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test
rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test2
rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test3
rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.test4
rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.gff3

agat_convert_sp_gff2gtf.pl \
    --gff $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gff3 \
    --gtf_version relax \
    -o $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gtf

cat $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.agat.log $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.agat.log $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.agat.log > $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.agat.final.log

rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.agat.log
rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.gtf.agat.log
rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.agat.log

# fix so that uropa can index the gtf more easily

# prepend "chr" to all lines that do not start with "chr" and "#"
awk '!/^chr/ && !/^#/ { $0 = "chr" $0 } 1' $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gtf > $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gtf.test

grep -v "^#" $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gtf.test | sort -k1,1 -k4,4n > $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.sorted.gtf

rm $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.wIntrons.wIntergenic.gtf.test

```

# RNA-seq data analysis

## Metadata tables (coldata)

```{r}
glob_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/sample_metadata.txt"),
                            header = TRUE,
                            sep = "\t")

glob_metadata$Run <- gsub("\\-", "_", glob_metadata$Run)

glob_metadata$Batch <- gsub("\\s", "_", glob_metadata$Batch)
glob_metadata$Batch <- gsub("\\(|\\)", "", glob_metadata$Batch)
glob_metadata$Batch <- gsub("\\-", "_", glob_metadata$Batch)

glob_metadata$Age_group <- gsub("\\s", "_", glob_metadata$Age_group)
glob_metadata$Age_group <- gsub("\\(|\\)", "", glob_metadata$Age_group)
glob_metadata$Age_group <- gsub("\\-", "_", glob_metadata$Age_group)

glob_metadata$Tissue <- gsub("\\s", "_", glob_metadata$Tissue)
glob_metadata$Tissue <- gsub("\\(|\\)", "", glob_metadata$Tissue)
glob_metadata$Tissue <- gsub("\\-", "_", glob_metadata$Tissue)

glob_metadata$Tissue_group <- gsub("\\s", "_", glob_metadata$Tissue_group)
glob_metadata$Tissue_group <- gsub("\\(|\\)", "", glob_metadata$Tissue_group)
glob_metadata$Tissue_group <- gsub("\\-", "_", glob_metadata$Tissue_group)

rownames(glob_metadata) <- glob_metadata$Run

glob_metadata <- glob_metadata[!(glob_metadata$Age_group == "Embryo" | glob_metadata$Age_group == "Not_applicable"), ]

write.table(glob_metadata,
  file = file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), sep = "\t",
  quote = TRUE
)

nat_metadata <- glob_metadata[glob_metadata$Batch == "Ruiz_Perez_et_al._n.d.", ]

write.table(nat_metadata,
  file = file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"), sep = "\t",
  quote = TRUE
)

# nat_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"), header = TRUE, sep = "\t")
```

## RNA-seq samples download

```{bash}
sra_samples=$(awk -F'\t' 'NR==1 { for (i=1; i<=NF; i++) { f[$i] = i } } { print $(f["Run"])}' $PROJECT_DIR/rnaseq/raw/fastq/sample_metadata.txt | grep SRR | uniq) #https://unix.stackexchange.com/a/359699

echo "Downloading the following files:"
echo $sra_samples

# download fastqc files
cd $PROJECT_DIR/rnaseq/raw/fastq
for i in $sra_samples
do
  $SRATOOLKIT_PATH/prefetch --max-size 50G --verbose $i
  $SRATOOLKIT_PATH/fasterq-dump --verbose $i --outdir ./
done

# Delete all temp folders (not files) from working directory
find -mindepth 1 -maxdepth 1 -type d -exec rm -r {} \;

gzip -v ./*.fastq
 
```

## Quality assessment

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/raw/fastqc
mkdir -p $PROJECT_DIR/rnaseq/raw/multiqc

fastqc \
  $PROJECT_DIR/rnaseq/raw/fastq/*.fastq.gz \
  -t $NTHREADS \
  -o $PROJECT_DIR/rnaseq/raw/fastqc/

multiqc \
  --interactive \
  $PROJECT_DIR/rnaseq/raw/fastqc/ \
  --filename raw_multiqc \
  --outdir $PROJECT_DIR/rnaseq/raw/multiqc

```

## Salmon and ExplorATE pipeline

### Salmon: Genes and transcripts

```{bash}
# Salmon indexing needs the transcriptome.fa.
# The transcriptome file in https://www.axolotl-omics.org/assemblies is corrupted,
# So we generate a new one based on the gtf and the genome.fa

cd $PROJECT_DIR/ext_data/axolotl_genome

gffread \
  --w-nocds \
  -w AmexT_v47-AmexG_v6.0-DD.fa \
  -g AmexG_v6.0-DD.fa \
  AmexT_v47-AmexG_v6.0-DD.gtf

# remove everything before "AMEX" from sequence IDs
sed -i.bak -e 's/^>.*AMEX/>AMEX/g' AmexT_v47-AmexG_v6.0-DD.fa

chmod -w AmexT_v47-AmexG_v6.0-DD.fa

```

#### Quantification

```{bash salmon_quant.sh}

GENOME_DIR=$PROJECT_DIR/ext_data/axolotl_genome/

SALMON_OUT_DIR=$PROJECT_DIR/rnaseq/Salmon/

cd $PROJECT_DIR/rnaseq/Salmon/

grep "^>" $GENOME_DIR/AmexG_v6.0-DD.fa | cut -d " " -f 1 > $PROJECT_DIR/rnaseq/Salmon/index/decoys.txt
sed -i.bak -e 's/>//g' $PROJECT_DIR/rnaseq/Salmon/index/decoys.txt
rm $PROJECT_DIR/rnaseq/Salmon/index/decoys.txt.bak

echo "Concatenating genome and transcriptome"
nohup cat $GENOME_DIR/AmexT_v47-AmexG_v6.0-DD.fa $GENOME_DIR/AmexG_v6.0-DD.fa > $PROJECT_DIR/rnaseq/Salmon/index/gentrome.fa && gzip $PROJECT_DIR/rnaseq/Salmon/index/gentrome.fa > $PROJECT_DIR/rnaseq/Salmon/index/cat_gentrome.log 2>&1 &

cd $PROJECT_DIR/rnaseq/Salmon/index/
echo "Salmon indexing"
nohup salmon index \
  --transcripts $PROJECT_DIR/rnaseq/Salmon/index/gentrome.fa.gz \
  --decoys $PROJECT_DIR/rnaseq/Salmon/index/decoys.txt \
  --threads 20 \
  --index axolotl_index \
  > $PROJECT_DIR/rnaseq/Salmon/index/salmon_index.log 2>&1 &


# Create new list with sample names 
list_files=$(ls $PROJECT_DIR/rnaseq/raw/fastq | egrep '\.fastq.gz$' | awk '$1 ~ /_R/' | awk -F'_R' '{print $1}' | sort -u)
#printf '%s\n' "${list_files[@]}" | wc -w

cd $PROJECT_DIR/rnaseq/Salmon/quant

for i in $list_files
do
  echo "Initiating quantification for sample $i"
  read1=$PROJECT_DIR/rnaseq/raw/fastq/$i\_R1.fastq.gz; echo "Read 1: $read1"
  read2=$PROJECT_DIR/rnaseq/raw/fastq/$i\_R2.fastq.gz; echo "Read 2: $read2"
  salmon quant \
    --index $PROJECT_DIR/rnaseq/Salmon/index/axolotl_index \
    --libType A \
    --gcBias \
    --mates1 $read1 \
    --mates2 $read2 \
    --threads 20 \
    --output $PROJECT_DIR/rnaseq/Salmon/quant/$i/
done

mkdir -p $PROJECT_DIR/rnaseq/Salmon/multiqc/
cd $PROJECT_DIR/rnaseq/Salmon/multiqc/

multiqc \
  --interactive \
  $PROJECT_DIR/rnaseq/Salmon/quant/ \
  --filename salmon_quant_multiqc \
  --outdir $PROJECT_DIR/rnaseq/Salmon/multiqc/

```

```{bash}
nohup \
  $PROJECT_DIR/main/scripts/salmon_quant.sh \
  > $PROJECT_DIR/rnaseq/Salmon/quant/salmon_quant.log \
  2>&1 &

```

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/Salmon/multiqc

multiqc \
  --interactive \
  $PROJECT_DIR/rnaseq/Salmon/quant \
  --filename salmon_quant_multiqc \
  --outdir $PROJECT_DIR/rnaseq/Salmon/multiqc

```

#### Importing quantification data

```{r}
library(tximport)

path.sal <- file.path(PROJECT_DIR, "/rnaseq/Salmon/quant/quant")

dir <- list.dirs(path.sal,
  recursive = F,
  full.names = F
)

files <- file.path(path.sal, dir, "quant.sf")

names(files) <- dir

tx2gene <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/quant/index/tx2gene.tsv"))

transcript_salmon_quant <- tximport(files,
  type = "salmon",
  txOut = TRUE
)

saveRDS(transcript_salmon_quant,
        file = file.path(PROJECT_DIR, "/rnaseq/Salmon/import/salmon_transcript_quant_all_samples.rds"))

# save each list element as a separate .tsv
for (i in seq_along(transcript_salmon_quant)) {
  write.table(transcript_salmon_quant[[i]],
              file = paste0(file.path(PROJECT_DIR,"/rnaseq/Salmon/import/salmon_transcript_", names(transcript_salmon_quant)[i], "_all_samples.tsv"), sep = "\t"))
}

gene_salmon_quant <- summarizeToGene(transcript_salmon_quant, tx2gene = tx2gene)

save(gene_salmon_quant,
     file = file.path(PROJECT_DIR, "/rnaseq/quantification/salmon/import/gene_salmon_quant.RData"))
load(file.path(PROJECT_DIR, "/rnaseq/Salmon/import/gene_salmon_quant.RData"))


saveRDS(gene_salmon_quant,
        file = file.path(PROJECT_DIR, "/rnaseq/Salmon/import/salmon_gene_quant_all_samples.rds"))
# gene_salmon_quant <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/import/salmon_gene_quant_all_samples.rds"))

# save each list element as a separate .tsv
for (i in seq_along(gene_salmon_quant)) {
  write.table(gene_salmon_quant[[i]], file = paste0(file.path(PROJECT_DIR, "/rnaseq/Salmon/import/salmon_gene_", names(gene_salmon_quant)[i], "_all_samples.tsv"), sep = "\t"))
}
```

#### Differential expression analysis

##### Native samples

###### Setting DESeq2 model

```{r}

# use lapply to replace "-" in colnames with "_"
nat_gene_salmon_quant <- lapply(gene_salmon_quant, function(x) {
  if (is.matrix(x)) {
    colnames(x) <- gsub("-", "_", colnames(x))
    x
  } else {
    return(x)
  }
})


nat_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"),
  header = TRUE,
  sep = "\t"
)

# replace "-" with "_" in nat_metadata$new_name_run
nat_metadata$new_name_run <- gsub("-", "_", nat_metadata$new_name_run)
rownames(nat_metadata) <- nat_metadata$new_name_run


# use lapply to subset each list element in gene_salmon_quant to only include the samples in nat_metadata
nat_gene_salmon_quant <- lapply(nat_gene_salmon_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% nat_metadata$new_name_run])
} else {
  return(x)
})


library(DESeq2)

nat_salmon_gene_dds_raw <- DESeqDataSetFromTximport(
  txi = nat_gene_salmon_quant,
  colData = nat_metadata,
  design = ~ Age_group + Tissue_group
)

# Explicitly setting the factors levels
nat_salmon_gene_dds_raw$Age_group <- relevel(nat_salmon_gene_dds_raw$Age_group, ref = "Sub_adult")
nat_salmon_gene_dds_raw$Tissue_group <- relevel(nat_salmon_gene_dds_raw$Tissue_group, ref = "Limb")

saveRDS(nat_salmon_gene_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_salmon_gene_dds_raw.rds"))
# nat_salmon_gene_dds_raw <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_salmon_gene_dds_raw.rds"))
```

###### Pre-filtering

```{r}
summary(rowSums(counts(nat_salmon_gene_dds_raw) >= 10) >= 3)

#   Mode   FALSE    TRUE
# logical   74414   22542

# keep <- rowSums(counts(nat_salmon_gene_dds_raw) >= 10 ) >= 3
# nat_salmon_gene_dds_filt <- nat_salmon_gene_dds_raw[keep,]
```

###### Pre-visualization

####### Variance Stabilizing Transformation (VST)

```{r}
library(vsn)
library(ggplot2)
library(scales)
library(DESeq2)

nat_salmon_gene_dds_vst <- vst(nat_salmon_gene_dds_raw, blind = FALSE)

saveRDS(nat_salmon_gene_dds_vst, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_salmon_gene_dds_vst.rds"))
# nat_salmon_gene_dds_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_salmon_gene_dds_vst.rds"))
```

######## cPCA (vst)

```{r}
library(cowplot)
library(ggpubr)
library(ggrepel)
library(ggh4x)
```

Getting underlying data:
```{r}
nat_salmon_gene_pcaData <- plotPCA(nat_salmon_gene_dds_vst, intgroup = c("Age_group", "Tissue_group"), returnData = TRUE)
nat_salmon_gene_percentVar <- round(100 * attr(nat_salmon_gene_pcaData, "percentVar"))

merged <- merge(nat_salmon_gene_pcaData, nat_metadata, by.x = "name", by.y = "new_name_run", all.x = TRUE)
merged <- merged[order(merged$name), ]

nat_salmon_gene_pcaData <- nat_salmon_gene_pcaData[order(row.names(nat_salmon_gene_pcaData)), ]

library(dplyr)
library(readr)

nat_salmon_gene_pcaData <- as_tibble(nat_salmon_gene_pcaData)

write_tsv(nat_salmon_gene_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/previsualization/nat_salmon_gene_pcaData.tsv"))

```

Plotting:
```{r}

nat_salmon_gene_pca_main <- ggplot(nat_salmon_gene_pcaData, aes(PC1, PC2, color = Tissue_group, fill = Tissue_group, shape = Age_group, label = name)) +
  geom_point(size = 3.5) +
  scale_shape_manual(
    name = "Age group", values = c(24, 22),
    breaks = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"),
    guide = guide_legend(order = 1, label.theme = element_text(size = 9))
  ) +
  scale_color_manual(
    name = "Tissue", values = c("#4dac26", "#d01c8b"),
    breaks = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema")
  ) +
  scale_fill_manual(
    name = "Tissue", values = c("#4dac26", "#d01c8b"),
    breaks = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema")
  ) +
  xlab(paste0("PC1: ", nat_salmon_gene_percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", nat_salmon_gene_percentVar[2], "% variance")) +
  theme_bw() +
  theme(
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.4),
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  guides(fill = guide_legend(order = 2), color = guide_legend(order = 2)) +
  geom_text_repel(
    show.legend = FALSE,
    color = "black", size = 3,
    label.padding = 0.5,
    box.padding = 0.3
  ) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  )

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/previsualization/nat_salmon_gene_cPCA.png"),
  plot = nat_salmon_gene_pca_main,
  device = "png", path = NULL, scale = 1,
  width = 150, height = 90, units = "mm",
  bg = "white", dpi = 1200
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/previsualization/nat_salmon_gene_cPCA.svg"),
  plot = nat_salmon_gene_pca_main,
  device = "svg", path = NULL, scale = 1,
  width = 150, height = 90, units = "mm",
  bg = "white"
)
```

######## Heatmap of sample clustering

```{r}
library(ComplexHeatmap)
library(viridisLite)
library(RColorBrewer)
library(DESeq2)

# Extract distance matrix
nat_salmon_gene_cor_dist_mat <- (1 - cor(assay(nat_salmon_gene_dds_vst), method = "pearson"))

# To replace sample names
match_idx <- match(rownames(nat_salmon_gene_cor_dist_mat), nat_metadata$new_name_run)
replacement_vals <- nat_metadata$new_name_run[match_idx]
rownames(nat_salmon_gene_cor_dist_mat) <- replacement_vals
colnames(nat_salmon_gene_cor_dist_mat) <- replacement_vals

write.table(nat_salmon_gene_cor_dist_mat,
  sep = "\t",
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/previsualization/nat_salmon_gene_cor_dist_mat.tsv")
)

# Prepare heatmap annotations
annotation <- as.data.frame(colData(nat_salmon_gene_dds_vst))
annotation <- annotation[c("Age_group", "Tissue_group")]
rownames(annotation) <- replacement_vals

# This changes some of the global options of ComplexHeatmap to be able to add a whitespace to the left of a merged legend.
# Should be reset after generating all heatmaps
# ht_opt(HEATMAP_LEGEND_PADDING = unit(c(6,0,0,0), "mm"), ANNOTATION_LEGEND_PADDING = unit(c(6,0,0,0), "mm"))
# ht_opt(RESET = TRUE)


column_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b"
    )
  ),
  annotation_label = gt_render(c("Age group", "Tissue")),
  annotation_legend_param = list(
    Age_group = list(
      title = "Age group",
      at = c("Adult", "Sub_adult"),
      labels = c("Adult (8 years)", "Sub-adult (8 months)"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    ),
    Tissue_group = list(
      title = "Tissue",
      at = c("Limb", "Limb_blastema"),
      labels = c("Limb", "Limb blastema"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    )
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

# Create ComplexHeatmap
# Color palette copied from: https://www.rdocumentation.org/packages/pheatmap/versions/1.0.12/topics/pheatmap
nat_salmon_gene_sample_dist_heatmap <- Heatmap(nat_salmon_gene_cor_dist_mat,
  name = "Distance",
  clustering_distance_rows = function(m) as.dist(m),
  clustering_distance_columns = function(m) as.dist(m),
  row_dend_side = "right",
  row_names_side = "left",
  top_annotation = column_ha,
  row_title = "RNA-seq run", row_title_rot = 90, row_title_gp = gpar(fontface = "bold", fontsize = 11),
  column_title = "RNA-seq run", column_title_side = "bottom", column_title_gp = gpar(fontface = "bold", fontsize = 11),
  row_names_gp = gpar(fontsize = 8.5),
  column_names_gp = gpar(fontsize = 8.5),
  row_dend_width = unit(0.6, "cm"),
  column_dend_height = unit(0.6, "cm"),
  col = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
  width = unit(3, "in"), height = unit(3, "in"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.5),
  heatmap_legend_param = list(
    legend_height = unit(3, "cm"),
    labels_gp = gpar(fontsize = 9),
    title_gp = gpar(
      fontsize = 11,
      fontface = "bold"
    )
  ),
  cell_fun = function(j, i, x, y, width, height, fill) {
    if (nat_salmon_gene_cor_dist_mat[i, j] != 0) {
      ifelse(nat_salmon_gene_cor_dist_mat[i, j] < 0.22, grid.text(sprintf("%.2f", nat_salmon_gene_cor_dist_mat[i, j]), x, y, gp = gpar(fontsize = 7, col = "gray55")), grid.text(sprintf("%.2f", nat_salmon_gene_cor_dist_mat[i, j]), x, y, gp = gpar(fontsize = 8.5, col = "gray85")))
    }
  },
)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/previsualization/nat_salmon_gene_sample_dist_heatmap.png"),
  width = 18, height = 13, units = "cm", res = 1200
)

draw(nat_salmon_gene_sample_dist_heatmap, merge_legends = TRUE)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/previsualization/nat_salmon_gene_sample_dist_heatmap.svg"),
  width = 18, height = 5.11811
)

draw(nat_salmon_gene_sample_dist_heatmap, merge_legends = TRUE)

dev.off()
```

######## Differential expression analysis (DEA)

######### Adult limbs vs Sub-adult limbs

```{r}
library(DESeq2)

# Only keep rows where Tissue_group is Limb
nat_age_metadata <- nat_metadata[nat_metadata$Tissue_group == "Limb", ]

# use lapply to subset each list element in nat_gene_salmon_quant to only include the samples in nat_age_metadata
nat_age_gene_salmon_quant <- lapply(nat_gene_salmon_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% nat_age_metadata$new_name_run])
} else {
  return(x)
})

all(rownames(nat_age_metadata) %in% colnames(nat_age_gene_salmon_quant$abundance))
all(rownames(nat_age_metadata) == colnames(nat_age_gene_salmon_quant$abundance))

nat_age_salmon_gene_dds_raw <- DESeqDataSetFromTximport(
  txi = nat_age_gene_salmon_quant,
  colData = nat_age_metadata,
  design = ~ Age_group
)

# Explicitly setting the factors levels
nat_age_salmon_gene_dds_raw$Age_group <- relevel(nat_age_salmon_gene_dds_raw$Age_group, ref = "Sub_adult")

saveRDS(nat_age_salmon_gene_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_age_salmon_gene_dds_raw.rds"))
# nat_age_salmon_gene_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_age_salmon_gene_dds_raw.rds"))
```

####### Running DESeq

```{r}
nat_age_salmon_gene_dds_de <- DESeq(nat_age_salmon_gene_dds_raw)

saveRDS(nat_age_salmon_gene_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_age_salmon_gene_dds_de.rds"))
```

####### Extracting contrast tables

```{r}
library("IHW")
library(ashr)
library(dplyr)
library(readr)

resultsNames(nat_age_salmon_gene_dds_de)
```

```{r}
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs <- results(nat_age_salmon_gene_dds_de,
  filterFun = ihw,
  contrast = c("Age_group", "Adult", "Sub_adult"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs)
# out of 50335 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 16, 0.032%
# LFC < -0.58 (down) : 84, 0.17%
# outliers [1]       : 373, 0.74%

nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink <- lfcShrink(nat_age_salmon_gene_dds_de,
  res = nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink)
# out of 50335 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 15, 0.03%
# LFC < -0.58 (down) : 79, 0.16%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs) <- paste0(colnames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs), "_MLE")
colnames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink) <- paste0(colnames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink), "_MMSE")
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE <- nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink$log2FoldChange_MMSE
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$lfcSE_MMSE <- nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink$lfcSE_MMSE
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$svalue_MMSE <- nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$gene_symbol[is.na(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$gene_symbol)] <- rownames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs)[is.na(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$gene_symbol)]
# add metadata columns
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$quant_tool <- "Salmon"
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$diffexp_contrast <- "adult_limbs_vs_subadult_limbs"
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$sample_group <- "native"
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) | abs(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$padj_MLE) | nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs <- as_tibble(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs, rownames = "gene_id")

write_tsv(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs.tsv")
)

# nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs.tsv"))

```


####### Sub-adult blastemas vs Sub-adult limbs

```{r}
library(DESeq2)

# Only keep rows where Age_group is Limb
nat_reg_metadata <- nat_metadata[nat_metadata$Age_group == "Sub_adult", ]

# use lapply to subset each list element in nat_gene_salmon_quant to only include the samples in nat_reg_metadata
nat_reg_gene_salmon_quant <- lapply(nat_gene_salmon_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% nat_reg_metadata$new_name_run])
} else {
  return(x)
})

all(rownames(nat_reg_metadata) %in% colnames(nat_reg_gene_salmon_quant$abundance))
all(rownames(nat_reg_metadata) == colnames(nat_reg_gene_salmon_quant$abundance))

nat_reg_salmon_gene_dds_raw <- DESeqDataSetFromTximport(
  txi = nat_reg_gene_salmon_quant,
  colData = nat_reg_metadata,
  design = ~ Tissue_group
)

# Explicitly setting the factors levels
nat_reg_salmon_gene_dds_raw$Age_group <- relevel(nat_reg_salmon_gene_dds_raw$Tissue_group, ref = "Limb")

saveRDS(nat_reg_salmon_gene_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_reg_salmon_gene_dds_raw.rds"))
# nat_reg_salmon_gene_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_reg_salmon_gene_dds_raw.rds"))
```

####### Running DESeq

```{r}
nat_reg_salmon_gene_dds_de <- DESeq(nat_reg_salmon_gene_dds_raw)

saveRDS(nat_reg_salmon_gene_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/nat_reg_salmon_gene_dds_de.rds"))
```

####### Extracting contrast tables

```{r}
library("IHW")
library(ashr)
library(dplyr)
library(readr)

resultsNames(nat_reg_salmon_gene_dds_de)
```

```{r}
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs <- results(nat_reg_salmon_gene_dds_de,
  filterFun = ihw,
  contrast = c("Tissue_group", "Limb_blastema", "Limb"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs)
  # out of 52033 with nonzero total read count
  # adjusted p-value < 0.05
  # LFC > 0.58 (up)    : 137, 0.26%
  # LFC < -0.58 (down) : 943, 1.8%
  # outliers [1]       : 743, 1.4%

nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink <- lfcShrink(nat_reg_salmon_gene_dds_de,
  res = nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink)
# out of 52033 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 135, 0.26%
# LFC < -0.58 (down) : 846, 1.6%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs) <- paste0(colnames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs), "_MLE")
colnames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink) <- paste0(colnames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink), "_MMSE")
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE <- nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink$log2FoldChange_MMSE
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$lfcSE_MMSE <- nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink$lfcSE_MMSE
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$svalue_MMSE <- nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$gene_symbol[is.na(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$gene_symbol)] <- rownames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs)[is.na(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$gene_symbol)]
# add metadata columns
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$quant_tool <- "Salmon"
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_limbs"
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$sample_group <- "native"
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$diffexp_res_lfc <- ifelse(is.na(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE) | abs(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$diffexp_res_padj <- ifelse(is.na(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$padj_MLE) | nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$padj_MLE > 0.05, "not_significant", "significant")

nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs <- as_tibble(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs, rownames = "gene_id")

write_tsv(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs.tsv")
)

# nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs.tsv"))

```

####### Volcano plots

```{r}
dir.create(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/volcanoes"))
```


```{r}

nat_salmon_gene_res_all <- bind_rows(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs, nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs)

# remove rows ehere padj_MLE is NA or log2FoldChange_MMSE is NA
nat_salmon_gene_res_all <- nat_salmon_gene_res_all %>%
  filter(!is.na(padj_MLE) & !is.na(log2FoldChange_MMSE))
```
 
```{r}
library(ggh4x)
library(EnhancedVolcano)

# Associating colors to each type of value
keyvals <- ifelse(nat_salmon_gene_res_all$padj_MLE > 0.05 & nat_salmon_gene_res_all$log2FoldChange_MMSE > log2(1.5),
  "#c5c5c5",
  ifelse(nat_salmon_gene_res_all$padj_MLE > 0.05 & nat_salmon_gene_res_all$log2FoldChange_MMSE < -log2(1.5),
    "#c5c5c5",
    ifelse(nat_salmon_gene_res_all$padj_MLE > 0.05 & abs(nat_salmon_gene_res_all$log2FoldChange_MMSE) < log2(1.5),
      "#c5c5c5",
      ifelse(nat_salmon_gene_res_all$padj_MLE < 0.05 & abs(nat_salmon_gene_res_all$log2FoldChange_MMSE) < log2(1.5),
        "#c5c5c5",
        ifelse(nat_salmon_gene_res_all$padj_MLE < 0.05 & nat_salmon_gene_res_all$log2FoldChange_MMSE < -log2(1.5),
          "#0072B2",
          "#D55E00"
        )
      )
    )
  )
)

keyvals[is.na(keyvals)] <- "#c5c5c5"

names(keyvals)[keyvals == "#c5c5c5"] <- "up_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "mid_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "down_NS"
names(keyvals)[keyvals == "#0072B2"] <- "down_S"
names(keyvals)[keyvals == "#D55E00"] <- "up_S"
names(keyvals)[keyvals == "#c5c5c5"] <- "middle_S"

shapevals <- keyvals
names(shapevals) <- keyvals

shapevals <- ifelse(keyvals == "#c5c5c5", 21,
      ifelse(keyvals == "#0072B2", 25,
        ifelse(keyvals == "#D55E00", 24, 21)
      )
  )

# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- nat_salmon_gene_res_all %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE < -log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(gene_symbol)

selectGenes1_2 <- nat_salmon_gene_res_all %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE > log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(gene_symbol)

selectGenes2 <- nat_salmon_gene_res_all %>%
                  group_by(diffexp_contrast) %>%
                  slice_min(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(gene_symbol)

selectGenes3 <- nat_salmon_gene_res_all %>%
                  group_by(diffexp_contrast) %>%
                  slice_max(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(gene_symbol)

selectGenes <- unique(c(selectGenes1, selectGenes1_2, selectGenes2, selectGenes3))

#if element starts with "AMEX", remove from selectGenes
selectGenes <- selectGenes[!grepl("^AMEX", selectGenes)]
selectGenes <- selectGenes[!grepl("^LOC", selectGenes)]


# Creating volcano plot
volcanoes_nat_salmon_gene <- EnhancedVolcano(

  as.data.frame(nat_salmon_gene_res_all),
  lab = as.data.frame(nat_salmon_gene_res_all)$gene_symbol,
  selectLab = selectGenes,
  drawConnectors = TRUE,
  widthConnectors = 0.3,
  arrowheads = FALSE,
  maxoverlapsConnectors = Inf,
  labSize = 2,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +


  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +


  # change y scale of first plot, keep the second plot as it is
  
  facetted_pos_scales(
  y = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_salmon_gene_res_all)$padj_MLE)[as.data.frame(nat_salmon_gene_res_all)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_salmon_gene_res_all)$padj_MLE)[as.data.frame(nat_salmon_gene_res_all)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(nat_salmon_gene_res_all)$log2FoldChange_MMSE[as.data.frame(nat_salmon_gene_res_all)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(nat_salmon_gene_res_all)$log2FoldChange_MMSE[as.data.frame(nat_salmon_gene_res_all)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/volcanoes/volcanoes_nat_salmon_gene.png"),
  plot = volcanoes_nat_salmon_gene,
  scale = 1, width = 200, height = 72, units = "mm", bg = "white", dpi = 1200
)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/volcanoes/volcanoes_nat_salmon_gene.svg"),
  width = 7.87402, height = 2.83465
)

volcanoes_nat_salmon_gene

dev.off()
```


##### Gene set enrichment analysis (GSEA)

```{r}
library(clusterProfiler)
library(enrichplot)
library(data.table)
library(DESeq2)
library(ggh4x)
```

###### Gene Ontology (GO)

####### Formatting gene metadata (GO information/TERM2GENE)

```{r}
# load gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# get columns gene_id and GO IDs
go_TERM2GENE <- gene_metadata[c("gene_id", "gene_goall")]
# remove rows where gene_goall is NA
go_TERM2GENE <- go_TERM2GENE[!is.na(go_TERM2GENE$gene_goall), ]

# expand, since some genes/proteins will have multiple assigned GO terms
go_TERM2GENE <- data.table(go_TERM2GENE)
go_TERM2GENE <- go_TERM2GENE[, list(gene_goall = unlist(strsplit(gene_goall, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_goall combination (to remove redundant pairs)
go_TERM2GENE <- distinct(go_TERM2GENE, gene_id, gene_goall, .keep_all = TRUE)

go_TERM2GENE <- go_TERM2GENE[, c(2, 1)]
```

####### Running FGSEA (clusterProfiler::compareCluster)

```{r}
library(readr)

nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs.tsv"))
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs.tsv"))

nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList <- nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE
names(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList) <- nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs$gene_id
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList <- na.omit(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList)
nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList <- sort(nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList, decreasing = TRUE)

nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList <- nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE
names(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList) <- nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs$gene_id
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList <- na.omit(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList)
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList <- sort(nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList, decreasing = TRUE)

nat_salmon_gene_gsea_List <- list(
  "Adult limbs vs Sub-adult limbs" = nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs_geneList,
  "Sub-adult blastemas vs Sub-adult lims" = nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs_geneList
)

gsea_nat_salmon_gene <- compareCluster(
  geneClusters = nat_salmon_gene_gsea_List,
  fun = "GSEA",
  TERM2GENE = go_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_nat_salmon_gene
```
  
```{r}
library(GO.db)
library(AnnotationDbi)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_nat_salmon_gene@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_nat_salmon_gene@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_nat_salmon_gene@compareClusterResult$ID[i]))
        gsea_nat_salmon_gene@compareClusterResult$Description[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_nat_salmon_gene@compareClusterResult$ID[i], column = c("TERM"), keytype = "GOID")
        gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_nat_salmon_gene@compareClusterResult$ID[i], column = c("ONTOLOGY"), keytype = "GOID")
      }
    },
    error = function(e) {
      cat("ERROR while mapping ", gsea_nat_salmon_gene@compareClusterResult$ID[i], ":", conditionMessage(e), "\n")
    }
  )
}

# In gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY replace BP with Biological Process, MF with Molecular Function, CC with Cellular Component
gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY_full <- gsub("BP", "Biological process", gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY)
gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY_full <- gsub("MF", "Molecular function", gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY_full)
gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY_full <- gsub("CC", "Cellular component", gsea_nat_salmon_gene@compareClusterResult$ONTOLOGY_full)

# To remove obsolete or removed GO terms
# Remove from gsea_nat_salmon_gene@compareClusterResult all rows for which Description starts with GO:
gsea_nat_salmon_gene@compareClusterResult <- gsea_nat_salmon_gene@compareClusterResult[!grepl("^GO:", gsea_nat_salmon_gene@compareClusterResult$Description), ]

# Add type column to gsea_nat_salmon_gene@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_nat_salmon_gene@compareClusterResult$type <- ifelse(gsea_nat_salmon_gene@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_nat_salmon_gene@compareClusterResult$type <- factor(gsea_nat_salmon_gene@compareClusterResult$type, levels = c("Suppressed", "Activated"))

dir.create(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich"))

saveRDS(gsea_nat_salmon_gene, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/gsea_go_nat_salmon_gene.rds"))
# gsea_nat_salmon_gene <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/gsea_go_nat_salmon_gene.rds"))

# get summary of type per cluster
table(gsea_nat_salmon_gene@compareClusterResult$Cluster, gsea_nat_salmon_gene@compareClusterResult$type)


gsea_nat_salmon_gene_tb <- as_tibble(gsea_nat_salmon_gene@compareClusterResult)
write_tsv(gsea_nat_salmon_gene_tb, file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/gsea_go_nat_salmon_gene.tsv"))


```

######## Dotplot

```{r}
library(stringr)
dp_gsea_nat_salmon_gene <- dotplot(gsea_nat_salmon_gene,
  showCategory = 28,
  size = "Count",
  x = "GeneRatio",
  label_format = 50
) +

  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 30)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )
# switch = "y"

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/dp_gsea_go_nat_salmon_gene.png"),
  plot = dp_gsea_nat_salmon_gene, width = 20, height = 28, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/dp_gsea_go_nat_salmon_gene.svg"),
  plot = dp_gsea_nat_salmon_gene, width = 20, height = 28, units = "cm"
)
```

###### KEGG

```{r}
# get columns gene_id and KEGG IDs
kegg_TERM2GENE <- gene_metadata[c("gene_id", "gene_keggpath")]
# remove rows where gene_keggpath is NA
kegg_TERM2GENE <- kegg_TERM2GENE[!is.na(kegg_TERM2GENE$gene_keggpath), ]

# expand, since some genes/proteins will have multiple assigned GO terms
kegg_TERM2GENE <- data.table(kegg_TERM2GENE)
kegg_TERM2GENE <- kegg_TERM2GENE[, list(gene_keggpath = unlist(strsplit(gene_keggpath, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_keggpath combination (to remove redundant pairs)
kegg_TERM2GENE <- distinct(kegg_TERM2GENE, gene_id, gene_keggpath, .keep_all = TRUE)

kegg_TERM2GENE <- kegg_TERM2GENE[, c(2, 1)]
```

####### compareCluster

```{r}

gsea_kegg_nat_salmon_gene <- compareCluster(
  geneClusters = nat_salmon_gene_gsea_List,
  fun = "GSEA",
  TERM2GENE = kegg_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_kegg_nat_salmon_gene
```
  
```{r}
# To map KEGG IDs to Pathway names (there's a 10-element limitation)
library(KEGGREST)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_kegg_nat_salmon_gene@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_kegg_nat_salmon_gene@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_kegg_nat_salmon_gene@compareClusterResult$ID[i]))
        gsea_kegg_nat_salmon_gene@compareClusterResult$Description[i] <- keggGet(gsea_kegg_nat_salmon_gene@compareClusterResult$ID[i])[[1]]$NAME
      }
    },
    error = function(e) {
      cat("ERROR: ", gsea_kegg_nat_salmon_gene@compareClusterResult$ID[i], conditionMessage(e), "\n")
    }
  )
}


# To remove obsolete or removed KEGG terms
# Remove from gsea_kegg_nat_salmon_gene@compareClusterResult all rows for which Description starts with GO:
gsea_kegg_nat_salmon_gene@compareClusterResult <- gsea_kegg_nat_salmon_gene@compareClusterResult[!grepl("^ko", gsea_kegg_nat_salmon_gene@compareClusterResult$Description), ]

# Add type column to gsea_kegg_nat_salmon_gene@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_kegg_nat_salmon_gene@compareClusterResult$type <- ifelse(gsea_kegg_nat_salmon_gene@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_kegg_nat_salmon_gene@compareClusterResult$type <- factor(gsea_kegg_nat_salmon_gene@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_kegg_nat_salmon_gene, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/gsea_kegg_nat_salmon_gene.rds"))
# gsea_kegg_nat_salmon_gene <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/gsea_kegg_nat_salmon_gene.rds"))


# get summary of type per cluster
table(gsea_kegg_nat_salmon_gene@compareClusterResult$Cluster, gsea_kegg_nat_salmon_gene@compareClusterResult$type)


gsea_kegg_nat_salmon_gene_tb <- as_tibble(gsea_kegg_nat_salmon_gene@compareClusterResult)
write_tsv(gsea_kegg_nat_salmon_gene_tb, file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/gsea_kegg_nat_salmon_gene.tsv"))
```

######## Dotplot
```{r}
library(stringr)
library(ggrepel)

# add dummy column to gsea_kegg_nat_salmon_gene@compareClusterResult
gsea_kegg_nat_salmon_gene@compareClusterResult$dummy <- "same"

dp_gsea_kegg_nat_salmon_gene <- dotplot(gsea_kegg_nat_salmon_gene,
  showCategory = 16,
  size = "Count",
  x = "GeneRatio",
  label_format = 50
) +

  # add text labels to each dot
  # geom_text_repel(aes(label = Count), size = 3, direction = "x", nudge_x = ifelse(Count < 0.5, 0.5, -0.5)) +


  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 30)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +


  facet_nested(dummy ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )


ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/dp_gsea_kegg_nat_salmon_gene.png"),
  plot = dp_gsea_kegg_nat_salmon_gene, width = 20, height = 16, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/dp_gsea_kegg_nat_salmon_gene.svg"),
  plot = dp_gsea_kegg_nat_salmon_gene, width = 20, height = 16, units = "cm"
)

```

```{r}
# make ggarrange of dp_gsea_nat_salmon_gene and dp_gsea_kegg_nat_salmon_gene
library(ggpubr)

# remove strip titles
dp_gsea_kegg_nat_salmon_gene_tmp <- dp_gsea_kegg_nat_salmon_gene +
  theme(strip.text.x = element_blank())

# remove x axis text, title and ticks
dp_gsea_nat_salmon_gene_tmp <- dp_gsea_nat_salmon_gene +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

dp_gsea_nat_salmon_gene_arrange <- ggarrange(dp_gsea_nat_salmon_gene_tmp, dp_gsea_kegg_nat_salmon_gene_tmp,
  ncol = 1, nrow = 2,
  common.legend = TRUE,
  align = "v",
  legend = "bottom",
  heights = c(65, 25)
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/dp_gsea_nat_salmon_gene_arrange.png"),
  plot = dp_gsea_nat_salmon_gene_arrange, width = 22, height = 46, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/enrich/dp_gsea_nat_salmon_gene_arrange.svg"),
  plot = dp_gsea_nat_salmon_gene_arrange, width = 22, height = 46, units = "cm"
)
```

###### Global axolotl sample analysis

####### Setting DESeq2 model

```{r}
glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"),
  sep = "\t",
  header = TRUE
)


# use lapply to replace "-" in colnames with "_"
glob_gene_salmon_quant <- lapply(gene_salmon_quant, function(x) {
  if (is.matrix(x)) {
    colnames(x) <- gsub("-", "_", colnames(x))
    x
  } else {
    return(x)
  }
})

# use lapply to subset each list element in gene_salmon_quant to only include the samples in glob_metadata
glob_gene_salmon_quant <- lapply(glob_gene_salmon_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% glob_metadata$Run])
} else {
  return(x)
})

library(DESeq2)

glob_salmon_gene_dds <- DESeqDataSetFromTximport(
  txi = glob_gene_salmon_quant,
  colData = glob_metadata,
  design = ~ Batch + Age_group + Tissue_group + Age_group:Tissue_group
)

# Explicitly setting the factors levels
glob_salmon_gene_dds$Age_group <- relevel(glob_salmon_gene_dds$Age_group, ref = "Sub_adult")
glob_salmon_gene_dds$Batch <- relevel(glob_salmon_gene_dds$Batch, ref = "Ruiz_Perez_et_al._n.d.")
glob_salmon_gene_dds$Tissue_group <- relevel(glob_salmon_gene_dds$Tissue_group, ref = "Non_limb")

save(glob_salmon_gene_dds, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/glob_salmon_gene_dds.RData"))
```

####### Pre-filtering

```{r}
summary(rowSums(counts(glob_salmon_gene_dds)) >= 10)

#   Mode   FALSE    TRUE
# logical   18575   78381

# No filtering was applied
```

####### Pre-visualization

```{r}
glob_salmon_gene_dds_vst <- vst(glob_salmon_gene_dds, blind = FALSE)

saveRDS(glob_salmon_gene_dds_vst, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/glob_salmon_gene_dds_vst.rds"))
# glob_salmon_gene_dds_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/glob_salmon_gene_dds_vst.rds"))

```

####### cPCA (vst)

```{r}
library(DESeq2)
library(cowplot)
library(ggpubr)
library(ggrepel)
library(ggforce)
library(ggh4x)
library(ggalt)


glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")
```

Personalized cPCA

```{r}
glob_salmon_pcaData <- plotPCA(glob_salmon_gene_dds_vst, intgroup = c("Age_group", "Batch", "Tissue_group"), returnData = TRUE)
percentVar <- round(100 * attr(glob_salmon_pcaData, "percentVar"))

merged <- merge(glob_salmon_pcaData, glob_metadata, by.x = "name", by.y = "Run", all.x = TRUE)
merged <- merged[order(merged$name), ]

glob_salmon_pcaData <- glob_salmon_pcaData[order(row.names(glob_salmon_pcaData)), ]
glob_salmon_pcaData$new_name <- merged$new_name_run

saveRDS(glob_salmon_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/previsualization/glob_salmon_pcaData.rds"))

write.csv(glob_salmon_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/previsualization/glob_salmon_pcaData.csv"))
```

Plotting:
```{r}
glob_salmon_pca_main <- ggplot(glob_salmon_pcaData, aes(PC1, PC2, fill = Tissue_group, shape = Age_group)) +
  geom_encircle(aes(group = Batch, colour = Batch), alpha = 0.4) +
  geom_point(size = 3.2, stroke = 0.2) +
  scale_shape_manual(
    name = "Age group", values = c(24, 22),
    breaks = c("Adult", "Sub_adult"), labels = c("Adult", "Sub-adult"),
    guide = guide_legend(order = 1, label.theme = element_text(size = 9, margin = margin(l = 25)))
  ) +
  scale_fill_manual(
    name = "Tissue_group",
    breaks = c("Non_limb", "Limb", "Limb_blastema"),
    values = c("#57452dff", "#4dac26", "#d01c8b")
  ) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  ) +
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  # guides(color = guide_legend(override.aes = list(size = 3))) +

  coord_fixed() +
  guides(
    fill = guide_legend(order = 2),
    color = guide_legend(order = 2)
  )


ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/previsualization/pca_tecount_vst.png"),
  plot = glob_salmon_pca_main,
  device = "png",
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white",
  dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/previsualization/pca_tecount_vst.svg"),
  plot = glob_salmon_pca_main,
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white"
)
```

Temporal plot with different points to overlay

```{r}
glob_salmon_pca_main <- ggplot(glob_salmon_pcaData, aes(PC1, PC2, color = Batch)) +
  geom_point(size = 12, alpha = 0.5, stroke = 0) +
  scale_color_manual(
    name = "Batch",
    values = c("#d36027ff", "#0773b3ff", "#f2e646ff", "#6b6a66ff", "#e79f26ff"),
    breaks = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"),
    labels = c("This study", bquote("Ye" ~ italic("et al.") ~ "(2022)"), bquote("Nowoshilow" ~ italic("et al.") ~ "(2018)"), bquote("Caballero-Pérez" ~ italic("et al.") ~ "(2018)"), bquote("Bryant" ~ italic("et al.") ~ "(2017)"))
  ) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  ) +
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  # guides(color = guide_legend(override.aes = list(size = 3))) +

  coord_fixed() +
  guides(
    fill = guide_legend(order = 2),
    color = guide_legend(order = 2)
  )

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/previsualization/pca_tecount_vst2.png"),
  plot = glob_salmon_pca_main,
  device = "png",
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white",
  dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/previsualization/pca_tecount_vst2.svg"),
  plot = glob_salmon_pca_main,
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white"
)
```


######## Differential expression (DE) analysis

######### Running DESeq


```{r}
glob_salmon_dds_de <- DESeq(glob_salmon_gene_dds)

saveRDS(glob_salmon_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/glob_salmon_dds_de.rds"))
```

```{r}
# Remove not converging genes (https://support.bioconductor.org/p/65091/#65114)

glob_salmon_ddsClean_de <- glob_salmon_dds_de[which(mcols(glob_salmon_dds_de)$betaConv), ]

saveRDS(glob_salmon_ddsClean_de, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/glob_salmon_ddsClean_de.rds"))
```

####### Summarizing DE results

```{r}
library("IHW")
```

```{r}
resultsNames(glob_salmon_ddsClean_de)

# [1] "Intercept"
# [2] "Batch_Bryant_et_al._2017_vs_Ruiz_Perez_et_al._n.d."
# [3] "Batch_Caballero_Perez_et_al._2018_vs_Ruiz_Perez_et_al._n.d."
# [4] "Batch_Nowoshilow_et_al._2018_vs_Ruiz_Perez_et_al._n.d."
# [5] "Batch_Ye_et_al._2022_vs_Ruiz_Perez_et_al._n.d."
# [6] "Age_group_Adult_vs_Sub_adult"
# [7] "Tissue_group_Limb_vs_Non_limb"
# [8] "Tissue_group_Limb_blastema_vs_Non_limb"
# [9] "Age_groupAdult.Tissue_groupLimb"
# [10] "Age_groupAdult.Tissue_groupLimb_blastema"
```

######### Limb vs Non-limb (for subadults)

```{r}
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 1, 0, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1250, 1.4%
# LFC < -0.58 (down) : 979, 1.1%
# outliers [1]       : 398, 0.43%

## LFC Shrinkage

library(ashr)

glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 826, 0.9%
# LFC < -0.58 (down) : 575, 0.62%


# Saving all columns (MLE and posterior LFC) in the same dataframe.

colnames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs) <- paste0(colnames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs), "_MLE")
colnames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink) <- paste0(colnames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink), "_MMSE")
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE <- glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink$log2FoldChange_MMSE
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$lfcSE_MMSE <- glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink$lfcSE_MMSE
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$svalue_MMSE <- glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol[is.na(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol)] <- rownames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs)[is.na(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol)]
# add metadata columns
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$quant_tool <- "Salmon"
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$diffexp_contrast <- "subadult_limbs_vs_subadult_nonlimbs"
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$sample_group <- "global"
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) | abs(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$padj_MLE) | glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs <- as_tibble(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs, rownames = "gene_id")

print(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs, width = Inf)

readr::write_tsv(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs.csv")
)
```

######### Limb vs Non-limb (for adults)

```{r}
glob_salmon_res_adult_limbs_vs_adult_nonlimbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 1, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_adult_limbs_vs_adult_nonlimbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 218, 0.24%
# LFC < -0.58 (down) : 847, 0.92%
# outliers [1]       : 398, 0.43%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_adult_limbs_vs_adult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 168, 0.18%
# LFC < -0.58 (down) : 583, 0.63%


colnames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs) <- paste0(colnames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs), "_MLE")
colnames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink) <- paste0(colnames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink), "_MMSE")
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE <- glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink$log2FoldChange_MMSE
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$lfcSE_MMSE <- glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink$lfcSE_MMSE
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$svalue_MMSE <- glob_salmon_res_adult_limbs_vs_adult_nonlimbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$gene_symbol[is.na(glob_salmon_res_adult_limbs_vs_adult_nonlimbs$gene_symbol)] <- rownames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs)[is.na(glob_salmon_res_adult_limbs_vs_adult_nonlimbs$gene_symbol)]
# add metadata columns
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$quant_tool <- "Salmon"
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$diffexp_contrast <- "adult_limbs_vs_adult_nonlimbs"
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$sample_group <- "global"
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_adult_limbs_vs_adult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE) | abs(glob_salmon_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_adult_limbs_vs_adult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_adult_limbs_vs_adult_nonlimbs$padj_MLE) | glob_salmon_res_adult_limbs_vs_adult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_adult_limbs_vs_adult_nonlimbs <- as_tibble(glob_salmon_res_adult_limbs_vs_adult_nonlimbs, rownames = "gene_id")

print(glob_salmon_res_adult_limbs_vs_adult_nonlimbs, width = Inf)

readr::write_tsv(glob_salmon_res_adult_limbs_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_limbs_vs_adult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_adult_limbs_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_limbs_vs_adult_nonlimbs.csv")
)
```

######### Blastema vs Non-limb (for subadults)

```{r}
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 1, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 281, 0.31%
# LFC < -0.58 (down) : 1878, 2%
# outliers [1]       : 398, 0.43%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 161, 0.17%
# LFC < -0.58 (down) : 1219, 1.3%


colnames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs) <- paste0(colnames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs), "_MLE")
colnames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink) <- paste0(colnames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink), "_MMSE")
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE <- glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink$log2FoldChange_MMSE
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$lfcSE_MMSE <- glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink$lfcSE_MMSE
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$svalue_MMSE <- glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol[is.na(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol)] <- rownames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs)[is.na(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol)]
# add metadata columns
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$quant_tool <- "Salmon"
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_nonlimbs"
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$sample_group <- "global"
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE) | abs(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$padj_MLE) | glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs <- as_tibble(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs, rownames = "gene_id")

print(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs, width = Inf)

readr::write_tsv(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs.csv")
)
```

######### Blastema vs Non-limb (for adults)

```{r}
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 1, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1253, 1.4%
# LFC < -0.58 (down) : 7883, 8.6%
# outliers [1]       : 398, 0.43%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_adult_blastemas_vs_adult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 1261, 1.4%
# LFC < -0.58 (down) : 7300, 7.9%


colnames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs) <- paste0(colnames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs), "_MLE")
colnames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink) <- paste0(colnames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink), "_MMSE")

glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE <- glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink$log2FoldChange_MMSE
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$lfcSE_MMSE <- glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink$lfcSE_MMSE
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$svalue_MMSE <- glob_salmon_res_adult_blastemas_vs_adult_nonlimbs_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol[is.na(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol)] <- rownames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs)[is.na(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol)]
# add metadata columns
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$quant_tool <- "Salmon"
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$diffexp_contrast <- "adult_blastemas_vs_adult_nonlimbs"
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$sample_group <- "global"
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE) | abs(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$padj_MLE) | glob_salmon_res_adult_blastemas_vs_adult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_adult_blastemas_vs_adult_nonlimbs <- as_tibble(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs, rownames = "gene_id")

print(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs, width = Inf)

readr::write_tsv(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_blastemas_vs_adult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_adult_blastemas_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_blastemas_vs_adult_nonlimbs.csv")
)
```

######### Blastema vs Limb (for subadults)

```{r}
glob_salmon_res_subadult_blastemas_vs_subadult_limbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, -1, 1, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_subadult_blastemas_vs_subadult_limbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 259, 0.28%
# LFC < -0.58 (down) : 3003, 3.3%
# outliers [1]       : 398, 0.43%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_subadult_blastemas_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 183, 0.2%
# LFC < -0.58 (down) : 2214, 2.4%


colnames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs) <- paste0(colnames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs), "_MLE")
colnames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink) <- paste0(colnames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink), "_MMSE")

glob_salmon_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE <- glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink$log2FoldChange_MMSE
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$lfcSE_MMSE <- glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink$lfcSE_MMSE
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$svalue_MMSE <- glob_salmon_res_subadult_blastemas_vs_subadult_limbs_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$gene_symbol[is.na(glob_salmon_res_subadult_blastemas_vs_subadult_limbs$gene_symbol)] <- rownames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs)[is.na(glob_salmon_res_subadult_blastemas_vs_subadult_limbs$gene_symbol)]
# add metadata columns
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$quant_tool <- "Salmon"
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_limbs"
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$sample_group <- "global"
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_subadult_blastemas_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE) | abs(glob_salmon_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_subadult_blastemas_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_subadult_blastemas_vs_subadult_limbs$padj_MLE) | glob_salmon_res_subadult_blastemas_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_subadult_blastemas_vs_subadult_limbs <- as_tibble(glob_salmon_res_subadult_blastemas_vs_subadult_limbs, rownames = "gene_id")

print(glob_salmon_res_subadult_blastemas_vs_subadult_limbs, width = Inf)

readr::write_tsv(glob_salmon_res_subadult_blastemas_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_blastemas_vs_subadult_limbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_subadult_blastemas_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_blastemas_vs_subadult_limbs.csv")
)
```

######### Blastema vs Limb (for adults)

```{r}
glob_salmon_res_adult_blastemas_vs_adult_limbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, -1, 1, -1, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_adult_blastemas_vs_adult_limbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 932, 1%
# LFC < -0.58 (down) : 4289, 4.7%
# outliers [1]       : 398, 0.43%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_adult_blastemas_vs_adult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 765, 0.83%
# LFC < -0.58 (down) : 3421, 3.7%


colnames(glob_salmon_res_adult_blastemas_vs_adult_limbs) <- paste0(colnames(glob_salmon_res_adult_blastemas_vs_adult_limbs), "_MLE")
colnames(glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink) <- paste0(colnames(glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink), "_MMSE")

glob_salmon_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE <- glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink$log2FoldChange_MMSE
glob_salmon_res_adult_blastemas_vs_adult_limbs$lfcSE_MMSE <- glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink$lfcSE_MMSE
glob_salmon_res_adult_blastemas_vs_adult_limbs$svalue_MMSE <- glob_salmon_res_adult_blastemas_vs_adult_limbs_shrink$svalue_MMSE



# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_adult_blastemas_vs_adult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_adult_blastemas_vs_adult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_adult_blastemas_vs_adult_limbs$gene_symbol[is.na(glob_salmon_res_adult_blastemas_vs_adult_limbs$gene_symbol)] <- rownames(glob_salmon_res_adult_blastemas_vs_adult_limbs)[is.na(glob_salmon_res_adult_blastemas_vs_adult_limbs$gene_symbol)]
# add metadata columns
glob_salmon_res_adult_blastemas_vs_adult_limbs$quant_tool <- "Salmon"
glob_salmon_res_adult_blastemas_vs_adult_limbs$diffexp_contrast <- "adult_blastemas_vs_adult_limbs"
glob_salmon_res_adult_blastemas_vs_adult_limbs$sample_group <- "global"
glob_salmon_res_adult_blastemas_vs_adult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_adult_blastemas_vs_adult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_adult_blastemas_vs_adult_limbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE) | abs(glob_salmon_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_adult_blastemas_vs_adult_limbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_adult_blastemas_vs_adult_limbs$padj_MLE) | glob_salmon_res_adult_blastemas_vs_adult_limbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_adult_blastemas_vs_adult_limbs <- as_tibble(glob_salmon_res_adult_blastemas_vs_adult_limbs, rownames = "gene_id")

print(glob_salmon_res_adult_blastemas_vs_adult_limbs, width = Inf)

readr::write_tsv(glob_salmon_res_adult_blastemas_vs_adult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_blastemas_vs_adult_limbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_adult_blastemas_vs_adult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_blastemas_vs_adult_limbs.csv")
)
```

######### Adults vs Sub-adults (for non-limb)


```{r}
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 4281, 4.7%
# LFC < -0.58 (down) : 1190, 1.3%
# outliers [1]       : 398, 0.43%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 3199, 3.5%
# LFC < -0.58 (down) : 939, 1%


colnames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs) <- paste0(colnames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs), "_MLE")
colnames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink) <- paste0(colnames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink), "_MMSE")

glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE <- glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink$log2FoldChange_MMSE
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$lfcSE_MMSE <- glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink$lfcSE_MMSE
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$svalue_MMSE <- glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol[is.na(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol)] <- rownames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs)[is.na(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol)]
# add metadata columns
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$quant_tool <- "Salmon"
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$diffexp_contrast <- "adult_nonlimbs_vs_subadult_nonlimbs"
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$sample_group <- "global"
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) | abs(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$padj_MLE) | glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs <- as_tibble(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs, rownames = "gene_id")

print(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs, width = Inf)

readr::write_tsv(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs.csv")
)
```

######### Adults vs Sub-adults (for limb)

```{r}
glob_salmon_res_adult_limbs_vs_subadult_limbs <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_adult_limbs_vs_subadult_limbs)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 4763, 5.2%
# LFC < -0.58 (down) : 2490, 2.7%
# outliers [1]       : 398, 0.43%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_adult_limbs_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 4058, 4.4%
# LFC < -0.58 (down) : 1986, 2.2%


colnames(glob_salmon_res_adult_limbs_vs_subadult_limbs) <- paste0(colnames(glob_salmon_res_adult_limbs_vs_subadult_limbs), "_MLE")
colnames(glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink) <- paste0(colnames(glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink), "_MMSE")

glob_salmon_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE <- glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink$log2FoldChange_MMSE
glob_salmon_res_adult_limbs_vs_subadult_limbs$lfcSE_MMSE <- glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink$lfcSE_MMSE
glob_salmon_res_adult_limbs_vs_subadult_limbs$svalue_MMSE <- glob_salmon_res_adult_limbs_vs_subadult_limbs_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_adult_limbs_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_adult_limbs_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_adult_limbs_vs_subadult_limbs$gene_symbol[is.na(glob_salmon_res_adult_limbs_vs_subadult_limbs$gene_symbol)] <- rownames(glob_salmon_res_adult_limbs_vs_subadult_limbs)[is.na(glob_salmon_res_adult_limbs_vs_subadult_limbs$gene_symbol)]
# add metadata columns
glob_salmon_res_adult_limbs_vs_subadult_limbs$quant_tool <- "Salmon"
glob_salmon_res_adult_limbs_vs_subadult_limbs$diffexp_contrast <- "adult_limbs_vs_subadult_limbs"
glob_salmon_res_adult_limbs_vs_subadult_limbs$sample_group <- "global"
glob_salmon_res_adult_limbs_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_adult_limbs_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_adult_limbs_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) | abs(glob_salmon_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_adult_limbs_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_adult_limbs_vs_subadult_limbs$padj_MLE) | glob_salmon_res_adult_limbs_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_adult_limbs_vs_subadult_limbs <- as_tibble(glob_salmon_res_adult_limbs_vs_subadult_limbs, rownames = "gene_id")

print(glob_salmon_res_adult_limbs_vs_subadult_limbs, width = Inf)

readr::write_tsv(glob_salmon_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_limbs_vs_subadult_limbs.tsv")
)
readr::write_excel_csv(glob_salmon_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_limbs_vs_subadult_limbs.csv")
)
```

######### Adults vs Sub-adults (for blastema)

```{r}
glob_salmon_res_adult_vs_subadult_blastema <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_adult_vs_subadult_blastema)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 748, 0.81%
# LFC < -0.58 (down) : 421, 0.46%
# outliers [1]       : 398, 0.43%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_adult_vs_subadult_blastema_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_adult_vs_subadult_blastema,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_adult_vs_subadult_blastema_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 543, 0.59%
# LFC < -0.58 (down) : 261, 0.28%


colnames(glob_salmon_res_adult_vs_subadult_blastema) <- paste0(colnames(glob_salmon_res_adult_vs_subadult_blastema), "_MLE")
colnames(glob_salmon_res_adult_vs_subadult_blastema_shrink) <- paste0(colnames(glob_salmon_res_adult_vs_subadult_blastema_shrink), "_MMSE")

glob_salmon_res_adult_vs_subadult_blastema$log2FoldChange_MMSE <- glob_salmon_res_adult_vs_subadult_blastema_shrink$log2FoldChange_MMSE
glob_salmon_res_adult_vs_subadult_blastema$lfcSE_MMSE <- glob_salmon_res_adult_vs_subadult_blastema_shrink$lfcSE_MMSE
glob_salmon_res_adult_vs_subadult_blastema$svalue_MMSE <- glob_salmon_res_adult_vs_subadult_blastema_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_adult_vs_subadult_blastema$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_adult_vs_subadult_blastema), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_adult_vs_subadult_blastema$gene_symbol[is.na(glob_salmon_res_adult_vs_subadult_blastema$gene_symbol)] <- rownames(glob_salmon_res_adult_vs_subadult_blastema)[is.na(glob_salmon_res_adult_vs_subadult_blastema$gene_symbol)]
# add metadata columns
glob_salmon_res_adult_vs_subadult_blastema$quant_tool <- "Salmon"
glob_salmon_res_adult_vs_subadult_blastema$diffexp_contrast <- "adults_blastemas_vs_subadults_blastemas"
glob_salmon_res_adult_vs_subadult_blastema$sample_group <- "global"
glob_salmon_res_adult_vs_subadult_blastema$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_adult_vs_subadult_blastema)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_adult_vs_subadult_blastema$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_adult_vs_subadult_blastema$log2FoldChange_MMSE) | abs(glob_salmon_res_adult_vs_subadult_blastema$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_adult_vs_subadult_blastema$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_adult_vs_subadult_blastema$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_adult_vs_subadult_blastema$padj_MLE) | glob_salmon_res_adult_vs_subadult_blastema$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_adult_vs_subadult_blastema <- as_tibble(glob_salmon_res_adult_vs_subadult_blastema, rownames = "gene_id")

print(glob_salmon_res_adult_vs_subadult_blastema, width = Inf)

readr::write_tsv(glob_salmon_res_adult_vs_subadult_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_vs_subadult_blastema.tsv")
)
readr::write_excel_csv(glob_salmon_res_adult_vs_subadult_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_vs_subadult_blastema.csv")
)
```

######### Interaction term: Age_group effect in Limb vs Non-limb (Age_groupAdult.Tissue_groupLimb)

 this tests if the Age_group effect is different in Limb compared to Non-limb

```{r}
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 362, 0.39%
# LFC < -0.58 (down) : 1002, 1.1%
# outliers [1]       : 398, 0.43%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 223, 0.24%
# LFC < -0.58 (down) : 482, 0.52%

colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb) <- paste0(colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb), "_MLE")
colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink) <- paste0(colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink), "_MMSE")

glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE <- glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink$log2FoldChange_MMSE
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$lfcSE_MMSE <- glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink$lfcSE_MMSE
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$svalue_MMSE <- glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol[is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol)] <- rownames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb)[is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol)]
# add metadata columns
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$quant_tool <- "Salmon"
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_contrast <- "age_effect_limbs_vs_nonlimbs"
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$sample_group <- "global"
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE) | abs(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$padj_MLE) | glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb <- as_tibble(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb, rownames = "gene_id")

print(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb, width = Inf)

readr::write_tsv(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb.tsv")
)
readr::write_excel_csv(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb.csv")
)
```

######### Interaction term: Age_group effect in Blastema vs Non-limb (Age_groupAdult.Tissue_groupLimb_blastema)
 
 this tests if the Age_group effect is different in Blastema compared to Non-limb

```{r}
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 746, 0.81%
# LFC < -0.58 (down) : 1853, 2%
# outliers [1]       : 398, 0.43%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 488, 0.53%
# LFC < -0.58 (down) : 1162, 1.3%

colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema) <- paste0(colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema), "_MLE")
colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink) <- paste0(colnames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink), "_MMSE")

glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE <- glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$log2FoldChange_MMSE
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$lfcSE_MMSE <- glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$lfcSE_MMSE
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$svalue_MMSE <- glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$svalue_MMSE


# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol[is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol)] <- rownames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema)[is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol)]
# add metadata columns
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$quant_tool <- "Salmon"
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_contrast <- "age_effect_blastemas_vs_nonlimbs"
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$sample_group <- "global"
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE) | abs(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$padj_MLE) | glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- as_tibble(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema, rownames = "gene_id")

print(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema, width = Inf)

readr::write_tsv(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema.tsv")
)
readr::write_excel_csv(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema.csv")
)
```

######### Interaction term: Age_group effect in Blastema vs Limb (Age_groupAdult.Tissue_groupLimb_blastema vs Age_groupAdult.Tissue_groupLimb)
 
 this tests if the Age_group effect is different in Blastema compared to Limb

```{r}
glob_salmon_res_int_Age_group_effect_blastema_vs_limb <- results(glob_salmon_ddsClean_de,
  filterFun = ihw,
  contrast = list("Age_groupAdult.Tissue_groupLimb_blastema", "Age_groupAdult.Tissue_groupLimb"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_salmon_res_int_Age_group_effect_blastema_vs_limb)
# out of 92003 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1145, 1.2%
# LFC < -0.58 (down) : 1588, 1.7%
# outliers [1]       : 398, 0.43%



## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink <- lfcShrink(glob_salmon_ddsClean_de,
  res = glob_salmon_res_int_Age_group_effect_blastema_vs_limb,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink)
# out of 92003 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 682, 0.74%
# LFC < -0.58 (down) : 844, 0.92%


colnames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb) <- paste0(colnames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb), "_MLE")
colnames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink) <- paste0(colnames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink), "_MMSE")

glob_salmon_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE <- glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink$log2FoldChange_MMSE
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$lfcSE_MMSE <- glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink$lfcSE_MMSE
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$svalue_MMSE <- glob_salmon_res_int_Age_group_effect_blastema_vs_limb_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$gene_symbol[is.na(glob_salmon_res_int_Age_group_effect_blastema_vs_limb$gene_symbol)] <- rownames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb)[is.na(glob_salmon_res_int_Age_group_effect_blastema_vs_limb$gene_symbol)]
# add metadata columns
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$quant_tool <- "Salmon"
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$diffexp_contrast <- "age_effect_blastemas_vs_limbs"
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$sample_group <- "global"
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_salmon_res_int_Age_group_effect_blastema_vs_limb)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$diffexp_res_lfc <- ifelse(is.na(glob_salmon_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE) | abs(glob_salmon_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_salmon_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_salmon_res_int_Age_group_effect_blastema_vs_limb$diffexp_res_padj <- ifelse(is.na(glob_salmon_res_int_Age_group_effect_blastema_vs_limb$padj_MLE) | glob_salmon_res_int_Age_group_effect_blastema_vs_limb$padj_MLE > 0.05, "not_significant", "significant")

glob_salmon_res_int_Age_group_effect_blastema_vs_limb <- as_tibble(glob_salmon_res_int_Age_group_effect_blastema_vs_limb, rownames = "gene_id")

print(glob_salmon_res_int_Age_group_effect_blastema_vs_limb, width = Inf)

readr::write_tsv(glob_salmon_res_int_Age_group_effect_blastema_vs_limb,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_group_effect_blastema_vs_limb.tsv")
)
readr::write_excel_csv(glob_salmon_res_int_Age_group_effect_blastema_vs_limb,
  file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_group_effect_blastema_vs_limb.csv")
)
```


Volcano plots 

```{r}

glob_salmon_res_1 <- all_diffexp_results %>%
    filter(quant_tool == "Salmon",
           feature_biotype == "gene",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs",
                                   "adult_blastemas_vs_adult_limbs",
                                   "adult_limbs_vs_subadult_limbs",
                                   "adults_blastemas_vs_subadults_blastemas",
                                   "age_effect_blastemas_vs_limbs"))
           

library(ggh4x)
library(EnhancedVolcano)


# Associating colors to each type of value
keyvals <- ifelse(glob_salmon_res_1$padj_MLE > 0.05 & glob_salmon_res_1$log2FoldChange_MMSE > log2(1.5),
  "#c5c5c5",
  ifelse(glob_salmon_res_1$padj_MLE > 0.05 & glob_salmon_res_1$log2FoldChange_MMSE < -log2(1.5),
    "#c5c5c5",
    ifelse(glob_salmon_res_1$padj_MLE > 0.05 & abs(glob_salmon_res_1$log2FoldChange_MMSE) < log2(1.5),
      "#c5c5c5",
      ifelse(glob_salmon_res_1$padj_MLE < 0.05 & abs(glob_salmon_res_1$log2FoldChange_MMSE) < log2(1.5),
        "#c5c5c5",
        ifelse(glob_salmon_res_1$padj_MLE < 0.05 & glob_salmon_res_1$log2FoldChange_MMSE < -log2(1.5),
          "#0072B2",
          "#D55E00"
        )
      )
    )
  )
)

keyvals[is.na(keyvals)] <- "#c5c5c5"

names(keyvals)[keyvals == "#c5c5c5"] <- "up_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "mid_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "down_NS"
names(keyvals)[keyvals == "#0072B2"] <- "down_S"
names(keyvals)[keyvals == "#D55E00"] <- "up_S"
names(keyvals)[keyvals == "#c5c5c5"] <- "middle_S"

shapevals <- keyvals
names(shapevals) <- keyvals

shapevals <- ifelse(keyvals == "#c5c5c5", 21,
      ifelse(keyvals == "#0072B2", 25,
        ifelse(keyvals == "#D55E00", 24, 21)
      )
  )


# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- glob_salmon_res_1 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE < -log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes1_2 <- glob_salmon_res_1 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE > log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes2 <- glob_salmon_res_1 %>%
                  group_by(diffexp_contrast) %>%
                  slice_min(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes3 <- glob_salmon_res_1 %>%
                  group_by(diffexp_contrast) %>%
                  slice_max(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes <- unique(c(selectGenes1, selectGenes1_2, selectGenes2, selectGenes3))

#if element starts with "AMEX", remove from selectGenes
selectGenes <- selectGenes[!grepl("^AMEX", selectGenes)]
selectGenes <- selectGenes[!grepl("^LOC", selectGenes)]


# Creating volcano plot
volcanoes_glob_salmon_1 <- EnhancedVolcano(

  as.data.frame(glob_salmon_res_1),
  lab = as.data.frame(glob_salmon_res_1)$SYMBOL,
  selectLab = selectGenes,
  drawConnectors = TRUE,
  widthConnectors = 0.3,
  arrowheads = FALSE,
  maxoverlapsConnectors = Inf,
  labSize = 2,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +


  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +

  facetted_pos_scales(
  y = list(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_1)$padj_MLE)[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
          diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_1)$padj_MLE)[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
          diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_1)$padj_MLE)[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
          diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_1)$padj_MLE)[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
          diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_1)$padj_MLE)[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
          ),

  x = list(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_1)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
          diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_1)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
          diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_1)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
          diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_1)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
          diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_1)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))  
    )
  ) +


  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/volcanoes/volcanoes_glob_salmon_1.png"),
  plot = volcanoes_glob_salmon_1,
  scale = 1, width = 40, height = 9, units = "cm", bg = "white", dpi = 600
)

# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/volcanoes/volcanoes_glob_salmon_1.svg"),
  plot = volcanoes_glob_salmon_1,
  scale = 1, width = 40, height = 9, units = "cm", bg = "white", fix_text_size = FALSE
)

```

glob_salmon_res_2

```{r}

glob_salmon_res_2 <- all_diffexp_results %>%
    filter(quant_tool == "Salmon",
           feature_biotype == "gene",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                                   "adult_blastemas_vs_adult_nonlimbs",
                                   "subadult_blastemas_vs_subadult_nonlimbs",
                                   "adult_limbs_vs_adult_nonlimbs",
                                   "subadult_limbs_vs_subadult_nonlimbs",
                                   "age_effect_blastemas_vs_nonlimbs",
                                   "age_effect_limbs_vs_nonlimbs"))

library(ggh4x)
library(EnhancedVolcano)


# Associating colors to each type of value
keyvals <- ifelse(glob_salmon_res_2$padj_MLE > 0.05 & glob_salmon_res_2$log2FoldChange_MMSE > log2(1.5),
  "#c5c5c5",
  ifelse(glob_salmon_res_2$padj_MLE > 0.05 & glob_salmon_res_2$log2FoldChange_MMSE < -log2(1.5),
    "#c5c5c5",
    ifelse(glob_salmon_res_2$padj_MLE > 0.05 & abs(glob_salmon_res_2$log2FoldChange_MMSE) < log2(1.5),
      "#c5c5c5",
      ifelse(glob_salmon_res_2$padj_MLE < 0.05 & abs(glob_salmon_res_2$log2FoldChange_MMSE) < log2(1.5),
        "#c5c5c5",
        ifelse(glob_salmon_res_2$padj_MLE < 0.05 & glob_salmon_res_2$log2FoldChange_MMSE < -log2(1.5),
          "#0072B2",
          "#D55E00"
        )
      )
    )
  )
)

keyvals[is.na(keyvals)] <- "#c5c5c5"

names(keyvals)[keyvals == "#c5c5c5"] <- "up_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "mid_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "down_NS"
names(keyvals)[keyvals == "#0072B2"] <- "down_S"
names(keyvals)[keyvals == "#D55E00"] <- "up_S"
names(keyvals)[keyvals == "#c5c5c5"] <- "middle_S"

shapevals <- keyvals
names(shapevals) <- keyvals

shapevals <- ifelse(keyvals == "#c5c5c5", 21,
      ifelse(keyvals == "#0072B2", 25,
        ifelse(keyvals == "#D55E00", 24, 21)
      )
  )


# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- glob_salmon_res_2 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE < -log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes1_2 <- glob_salmon_res_2 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE > log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes2 <- glob_salmon_res_2 %>%
                  group_by(diffexp_contrast) %>%
                  slice_min(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes3 <- glob_salmon_res_2 %>%
                  group_by(diffexp_contrast) %>%
                  slice_max(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes <- unique(c(selectGenes1, selectGenes1_2, selectGenes2, selectGenes3))


#if element starts with "AMEX", remove from selectGenes
selectGenes <- selectGenes[!grepl("^AMEX", selectGenes)]
selectGenes <- selectGenes[!grepl("^LOC", selectGenes)]


# Creating volcano plot
volcanoes_glob_salmon_2 <- EnhancedVolcano(

  as.data.frame(glob_salmon_res_2),
  lab = as.data.frame(glob_salmon_res_2)$SYMBOL,
  selectLab = selectGenes,
  drawConnectors = TRUE,
  widthConnectors = 0.3,
  arrowheads = FALSE,
  maxoverlapsConnectors = Inf,
  labSize = 2,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +


  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +

  facetted_pos_scales(
  y = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_salmon_res_2)$padj_MLE)[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))),

  x = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_salmon_res_2)$log2FoldChange_MMSE[as.data.frame(glob_salmon_res_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
          
    )) +


  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/volcanoes/volcanoes_glob_salmon_2.png"),
  plot = volcanoes_glob_salmon_2,
  scale = 1, width = 50, height = 9, units = "cm", bg = "white", dpi = 600
)

# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/volcanoes/volcanoes_glob_salmon_2.svg"),
  plot = volcanoes_glob_salmon_2,
  scale = 1, width = 50, height = 9, units = "cm", bg = "white", fix_text_size = FALSE
)

```


###### Gene set enrichment analysis

```{r}
library(clusterProfiler)
library(enrichplot)
library(data.table)
library(DESeq2)
library(ggh4x)
library(stringr)
library(ggrepel)
```

######## Generating gene metadata

######## (GO information/TERM2GENE)
```{r}
# load gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# get columns gene_id and GO IDs
go_TERM2GENE <- gene_metadata[c("gene_id", "gene_goall")]
# remove rows where gene_goall is NA
go_TERM2GENE <- go_TERM2GENE[!is.na(go_TERM2GENE$gene_goall), ]

# expand, since some genes/proteins will have multiple assigned GO terms
go_TERM2GENE <- data.table(go_TERM2GENE)
go_TERM2GENE <- go_TERM2GENE[, list(gene_goall = unlist(strsplit(gene_goall, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_goall combination (to remove redundant pairs)
go_TERM2GENE <- distinct(go_TERM2GENE, gene_id, gene_goall, .keep_all = TRUE)

go_TERM2GENE <- go_TERM2GENE[, c(2, 1)]
```

######## KEGG metadata

```{r}
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# get columns gene_id and KEGG IDs
kegg_TERM2GENE <- gene_metadata[c("gene_id", "gene_keggpath")]
# remove rows where gene_keggpath is NA
kegg_TERM2GENE <- kegg_TERM2GENE[!is.na(kegg_TERM2GENE$gene_keggpath), ]

# expand, since some genes/proteins will have multiple assigned GO terms
kegg_TERM2GENE <- data.table(kegg_TERM2GENE)
kegg_TERM2GENE <- kegg_TERM2GENE[, list(gene_keggpath = unlist(strsplit(gene_keggpath, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_keggpath combination (to remove redundant pairs)
kegg_TERM2GENE <- distinct(kegg_TERM2GENE, gene_id, gene_keggpath, .keep_all = TRUE)

kegg_TERM2GENE <- kegg_TERM2GENE[, c(2, 1)]
```

Group 1 of contrasts
```{r}

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

df <- all_diffexp_results %>%
  filter(quant_tool == "Salmon",
         sample_group == "global",
         feature_biotype == "gene",
         !is.na(padj_MLE),
         !is.na(log2FoldChange_MMSE),
         diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs",
                                 "age_effect_blastemas_vs_limbs",
                                 "subadult_blastemas_vs_subadult_limbs",
                                 "adults_blastemas_vs_subadults_blastemas",
                                 "adult_blastemas_vs_adult_limbs")) %>%
  group_by(diffexp_contrast) %>%
  # sort each group first by log2FoldChange_MMSE and then by padj_MLE
  arrange(desc(log2FoldChange_MMSE), padj_MLE, .by_group = TRUE) %>%
  # get gene vectors where name is ID and value is log2FoldChange_MMSE
  summarise(geneList = list(ID),
            log2FoldChange_MMSE = list(log2FoldChange_MMSE),
            .groups = "keep")
  
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = diffexp_contrast) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            # lapply so that geneList are names and log2FoldChange_MMSE the values
            lapply(function(x) {
              unlist(x$log2FoldChange_MMSE) %>% setNames(unlist(x$geneList))
            })


gsea_glob_salmon_1 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = go_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_glob_salmon_1

```
  
```{r}
library(GO.db)
library(AnnotationDbi)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_glob_salmon_1@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_glob_salmon_1@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_glob_salmon_1@compareClusterResult$ID[i]))
        gsea_glob_salmon_1@compareClusterResult$Description[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_salmon_1@compareClusterResult$ID[i], column = c("TERM"), keytype = "GOID")
        gsea_glob_salmon_1@compareClusterResult$ONTOLOGY[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_salmon_1@compareClusterResult$ID[i], column = c("ONTOLOGY"), keytype = "GOID")
      }
    },
    error = function(e) {
      cat("ERROR while mapping ", gsea_glob_salmon_1@compareClusterResult$ID[i], ":", conditionMessage(e), "\n")
    }
  )
}

# In gsea_glob_salmon_1@compareClusterResult$ONTOLOGY replace BP with Biological Process, MF with Molecular Function, CC with Cellular Component
gsea_glob_salmon_1@compareClusterResult$ONTOLOGY_full <- gsub("BP", "Biological process", gsea_glob_salmon_1@compareClusterResult$ONTOLOGY)
gsea_glob_salmon_1@compareClusterResult$ONTOLOGY_full <- gsub("MF", "Molecular function", gsea_glob_salmon_1@compareClusterResult$ONTOLOGY_full)
gsea_glob_salmon_1@compareClusterResult$ONTOLOGY_full <- gsub("CC", "Cellular component", gsea_glob_salmon_1@compareClusterResult$ONTOLOGY_full)

# To remove obsolete or removed GO terms
# Remove from gsea_glob_salmon_1@compareClusterResult all rows for which Description starts with GO:
gsea_glob_salmon_1@compareClusterResult <- gsea_glob_salmon_1@compareClusterResult[!grepl("^GO:", gsea_glob_salmon_1@compareClusterResult$Description), ]

# Add type column to gsea_glob_salmon_1@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_glob_salmon_1@compareClusterResult$type <- ifelse(gsea_glob_salmon_1@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_glob_salmon_1@compareClusterResult$type <- factor(gsea_glob_salmon_1@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_glob_salmon_1, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_go_glob_tecount_1.rds"))
# gsea_glob_salmon_1 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_go_glob_tecount_1.rds")

gsea_glob_salmon_1_tb <- as_tibble(gsea_glob_salmon_1@compareClusterResult)
write_tsv(gsea_glob_salmon_1_tb, file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_go_glob_tecount_1.tsv"))
```

Dotplot
```{r}
dp_gsea_glob_salmon_1 <- dotplot(gsea_glob_salmon_1,
  showCategory = 12,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
# switch = "y"
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 6),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_1.png"),
  plot = dp_gsea_glob_salmon_1, width = 40, height = 28, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_1.svg"),
  plot = dp_gsea_glob_salmon_1, width = 40, height = 28, units = "cm", fix_text_size = FALSE
)
```

KEGG

```{r}
gsea_kegg_glob_salmon_1 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = kegg_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_kegg_glob_salmon_1
```
  
```{r}
# To map KEGG IDs to Pathway names (there's a 10-element limitation)
library(KEGGREST)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_kegg_glob_salmon_1@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_kegg_glob_salmon_1@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_kegg_glob_salmon_1@compareClusterResult$ID[i]))
        gsea_kegg_glob_salmon_1@compareClusterResult$Description[i] <- keggGet(gsea_kegg_glob_salmon_1@compareClusterResult$ID[i])[[1]]$NAME
      }
    },
    error = function(e) {
      cat("ERROR: ", gsea_kegg_glob_salmon_1@compareClusterResult$ID[i], conditionMessage(e), "\n")
    }
  )
}


# To remove obsolete or removed KEGG terms
# Remove from gsea_kegg_glob_salmon_1@compareClusterResult all rows for which Description starts with GO:
gsea_kegg_glob_salmon_1@compareClusterResult <- gsea_kegg_glob_salmon_1@compareClusterResult[!grepl("^ko", gsea_kegg_glob_salmon_1@compareClusterResult$Description), ]

# Add type column to gsea_kegg_glob_salmon_1@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_kegg_glob_salmon_1@compareClusterResult$type <- ifelse(gsea_kegg_glob_salmon_1@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_kegg_glob_salmon_1@compareClusterResult$type <- factor(gsea_kegg_glob_salmon_1@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_kegg_glob_salmon_1, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_kegg_glob_salmon_1.rds"))
# gsea_kegg_glob_salmon_1 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_kegg_glob_salmon_1.rds")

gsea_kegg_glob_salmon_1_tb <- as_tibble(gsea_kegg_glob_salmon_1@compareClusterResult)
write_tsv(gsea_kegg_glob_salmon_1_tb, file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_kegg_glob_salmon_1.tsv"))
```

######## Dotplot
```{r}
# add dummy column to gsea_kegg_glob_salmon_1@compareClusterResult
gsea_kegg_glob_salmon_1@compareClusterResult$dummy <- "same"

dp_gsea_kegg_glob_salmon_1 <- dotplot(gsea_kegg_glob_salmon_1,
  showCategory = 11,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  # add text labels to each dot
  # geom_text_repel(aes(label = Count), size = 3, direction = "x", nudge_x = ifelse(Count < 0.5, 0.5, -0.5)) +


  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(dummy ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 6),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )


ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_kegg_glob_salmon_1.png"),
  plot = dp_gsea_kegg_glob_salmon_1, width = 40, height = 20, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_kegg_glob_salmon_1.svg"),
  plot = dp_gsea_kegg_glob_salmon_1, width = 40, height = 20, units = "cm", dpi = 600
)


# make ggarrange of dp_gsea_nat_tecount and dp_gsea_kegg_glob_salmon_1
library(ggpubr)

# remove strip titles
dp_gsea_kegg_glob_salmon_1_tmp <- dp_gsea_kegg_glob_salmon_1 +
  theme(strip.text.x = element_blank())

# remove x axis text, title and ticks
dp_gsea_glob_salmon_1_tmp <- dp_gsea_glob_salmon_1 +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

dp_gsea_glob_salmon_arrange_1 <- ggarrange(dp_gsea_glob_salmon_1_tmp, dp_gsea_kegg_glob_salmon_1_tmp,
  ncol = 1, nrow = 2,
  common.legend = TRUE,
  align = "v",
  legend = "bottom",
  heights = c(56, 28)
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_arrange_1.png"),
  plot = dp_gsea_glob_salmon_arrange_1, width = 44, height = 32, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_arrange_1.svg"),
  plot = dp_gsea_glob_salmon_arrange_1, width = 44, height = 32, units = "cm"
)
```

Group 2 of contrasts

```{r}

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

df <- all_diffexp_results %>%
  filter(quant_tool == "Salmon",
         sample_group == "global",
         feature_biotype == "gene",
         !is.na(padj_MLE),
         !is.na(log2FoldChange_MMSE),
         diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                          "adult_blastemas_vs_adult_nonlimbs",
                          "subadult_blastemas_vs_subadult_nonlimbs",
                          "adult_limbs_vs_adult_nonlimbs",
                          "subadult_limbs_vs_subadult_nonlimbs",
                          "age_effect_blastemas_vs_nonlimbs",
                          "age_effect_limbs_vs_nonlimbs")) %>%

  group_by(diffexp_contrast) %>%
  # sort each group first by log2FoldChange_MMSE and then by padj_MLE
  arrange(desc(log2FoldChange_MMSE), padj_MLE, .by_group = TRUE) %>%
  # get gene vectors where name is ID and value is log2FoldChange_MMSE
  summarise(geneList = list(ID),
            log2FoldChange_MMSE = list(log2FoldChange_MMSE),
            .groups = "keep")
  
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = diffexp_contrast) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            # lapply so that geneList are names and log2FoldChange_MMSE the values
            lapply(function(x) {
              unlist(x$log2FoldChange_MMSE) %>% setNames(unlist(x$geneList))
            })


gsea_glob_salmon_2 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = go_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_glob_salmon_2

```
  
```{r}
library(GO.db)
library(AnnotationDbi)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_glob_salmon_2@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_glob_salmon_2@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_glob_salmon_2@compareClusterResult$ID[i]))
        gsea_glob_salmon_2@compareClusterResult$Description[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_salmon_2@compareClusterResult$ID[i], column = c("TERM"), keytype = "GOID")
        gsea_glob_salmon_2@compareClusterResult$ONTOLOGY[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_salmon_2@compareClusterResult$ID[i], column = c("ONTOLOGY"), keytype = "GOID")
      }
    },
    error = function(e) {
      cat("ERROR while mapping ", gsea_glob_salmon_2@compareClusterResult$ID[i], ":", conditionMessage(e), "\n")
    }
  )
}

# In gsea_glob_salmon_2@compareClusterResult$ONTOLOGY replace BP with Biological Process, MF with Molecular Function, CC with Cellular Component
gsea_glob_salmon_2@compareClusterResult$ONTOLOGY_full <- gsub("BP", "Biological process", gsea_glob_salmon_2@compareClusterResult$ONTOLOGY)
gsea_glob_salmon_2@compareClusterResult$ONTOLOGY_full <- gsub("MF", "Molecular function", gsea_glob_salmon_2@compareClusterResult$ONTOLOGY_full)
gsea_glob_salmon_2@compareClusterResult$ONTOLOGY_full <- gsub("CC", "Cellular component", gsea_glob_salmon_2@compareClusterResult$ONTOLOGY_full)

# To remove obsolete or removed GO terms
# Remove from gsea_glob_salmon_2@compareClusterResult all rows for which Description starts with GO:
gsea_glob_salmon_2@compareClusterResult <- gsea_glob_salmon_2@compareClusterResult[!grepl("^GO:", gsea_glob_salmon_2@compareClusterResult$Description), ]

# Add type column to gsea_glob_salmon_2@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_glob_salmon_2@compareClusterResult$type <- ifelse(gsea_glob_salmon_2@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_glob_salmon_2@compareClusterResult$type <- factor(gsea_glob_salmon_2@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_glob_salmon_2, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_go_glob_tecount_2.rds"))
# gsea_glob_salmon_2 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_go_glob_tecount_2.rds")

gsea_glob_salmon_2_tb <- as_tibble(gsea_glob_salmon_2@compareClusterResult)
write_tsv(gsea_glob_salmon_2_tb, file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_go_glob_tecount_2.tsv"))
```

Dotplot
```{r}
dp_gsea_glob_salmon_2 <- dotplot(gsea_glob_salmon_2,
  showCategory = 12,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
# switch = "y"
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 6),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 8, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank(),
    legend.position = "bottom"
  )

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_2.png"),
  plot = dp_gsea_glob_salmon_2, width = 44, height = 38.1, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_2.svg"),
  plot = dp_gsea_glob_salmon_2, width = 44, height = 38.1, units = "cm"
)
```

KEGG

```{r}
gsea_kegg_glob_salmon_2 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = kegg_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_kegg_glob_salmon_2
```
  
```{r}
# To map KEGG IDs to Pathway names (there's a 10-element limitation)
library(KEGGREST)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_kegg_glob_salmon_2@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_kegg_glob_salmon_2@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_kegg_glob_salmon_2@compareClusterResult$ID[i]))
        gsea_kegg_glob_salmon_2@compareClusterResult$Description[i] <- keggGet(gsea_kegg_glob_salmon_2@compareClusterResult$ID[i])[[1]]$NAME
      }
    },
    error = function(e) {
      cat("ERROR: ", gsea_kegg_glob_salmon_2@compareClusterResult$ID[i], conditionMessage(e), "\n")
    }
  )
}


# To remove obsolete or removed KEGG terms
# Remove from gsea_kegg_glob_salmon_2@compareClusterResult all rows for which Description starts with GO:
gsea_kegg_glob_salmon_2@compareClusterResult <- gsea_kegg_glob_salmon_2@compareClusterResult[!grepl("^ko", gsea_kegg_glob_salmon_2@compareClusterResult$Description), ]

# Add type column to gsea_kegg_glob_salmon_2@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_kegg_glob_salmon_2@compareClusterResult$type <- ifelse(gsea_kegg_glob_salmon_2@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_kegg_glob_salmon_2@compareClusterResult$type <- factor(gsea_kegg_glob_salmon_2@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_kegg_glob_salmon_2, file = file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_kegg_glob_salmon_2.rds"))
# gsea_kegg_glob_salmon_2 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_kegg_glob_salmon_2.rds")

gsea_kegg_glob_salmon_2_tb <- as_tibble(gsea_kegg_glob_salmon_2@compareClusterResult)
write_tsv(gsea_kegg_glob_salmon_2_tb, file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/gsea_kegg_glob_salmon_2.tsv"))
```

######## Dotplot
```{r}
# add dummy column to gsea_kegg_glob_salmon_2@compareClusterResult
gsea_kegg_glob_salmon_2@compareClusterResult$dummy <- "same"

dp_gsea_kegg_glob_salmon_2 <- dotplot(gsea_kegg_glob_salmon_2,
  showCategory = 11,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  # add text labels to each dot
  # geom_text_repel(aes(label = Count), size = 3, direction = "x", nudge_x = ifelse(Count < 0.5, 0.5, -0.5)) +


  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(dummy ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 6),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 8, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank(),
    legend.position = "bottom"
  )


ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_kegg_glob_salmon_2.png"),
  plot = dp_gsea_kegg_glob_salmon_2, width = 44, height = 38.1, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_kegg_glob_salmon_2.svg"),
  plot = dp_gsea_kegg_glob_salmon_2, width = 44, height = 38.1, units = "cm", dpi = 600
)


# make ggarrange of dp_gsea_nat_tecount and dp_gsea_kegg_glob_salmon_2
library(ggpubr)

# remove strip titles
dp_gsea_kegg_glob_salmon_2_tmp <- dp_gsea_kegg_glob_salmon_2 +
  theme(strip.text.x = element_blank())

# remove x axis text, title and ticks
dp_gsea_glob_salmon_2_tmp <- dp_gsea_glob_salmon_2 +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

dp_gsea_glob_salmon_arrange_2 <- ggarrange(dp_gsea_glob_salmon_2_tmp, dp_gsea_kegg_glob_salmon_2_tmp,
  ncol = 1, nrow = 2,
  common.legend = TRUE,
  align = "v",
  legend = "bottom",
  heights = c(56, 28)
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_arrange_2.png"),
  plot = dp_gsea_glob_salmon_arrange_2, width = 44, height = 32, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/enrich/dp_gsea_glob_salmon_arrange_2.svg"),
  plot = dp_gsea_glob_salmon_arrange_2, width = 44, height = 32, units = "cm"
)

```





### ExplorATE: RE families and loci

```{bash}
cd $SOFTWARE_DIR
git clone https://github.com/FemeniasM/ExplorATE_shell_script

```

```{bash}
cd $SOFTWARE_DIR/ExplorATE_shell_script
mkdir -p $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/

chmod +x ExplorATE

#modified explorate mo script to keep satellites

nohup bash ./ExplorATE mo \
  -p 20 \
  -f $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.fa \
  -g $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/AmexT_v47-AmexG_v6.0-DD.gtf \
  -r $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.out \
  -e pe \
  -l $PROJECT_DIR/raw_fastq/ \
  -o $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/ \
  -v higher_score \
  > $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/ambMex60DD_ucsc_hub.ALL.explorate.log 2>&1 &

nohup bash $SOFTWARE_DIR/ExplorATE_shell_script/bin/ExplorATE_mo_edit.sh \
  -p 20 \
  -f $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.fa \
  -g $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/AmexT_v47-AmexG_v6.0-DD.gtf \
  -r $PROJECT_DIR/ext_data/axolotl_genome/repeats/ambMex60DD_ucsc_hub.ALL.mod.out \
  -e pe \
  -l $PROJECT_DIR/raw_fastq/temp_fastq \
  -o $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/ \
  -v higher_score \
  > $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/ambMex60DD_ucsc_hub.ALL.explorate.edit.log 2>&1 &

#error incomplete sample

mamba activate srp_explorate

nohup salmon quant \
  -i $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/temp/SALMON_INDEX \
  -l A \
  --gcBias \
  --useVBOpt \
  -1 $PROJECT_DIR/raw_fastq/SRR2885268_R1.fastq.gz \
  -2 $PROJECT_DIR/raw_fastq/SRR2885268_R2.fastq.gz \
  -p 20 \
  --validateMappings \
  -o $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/quant_out/SRR2885268/ \
  > $PROJECT_DIR/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/missing.sample.log 2>&1 &

2752109

```

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/ExplorATE/multiqc

multiqc --interactive $PROJECT_DIR/rnaseq/ExplorATE/quant_out --filename explorate_quant_multiqc --outdir $PROJECT_DIR/rnaseq/ExplorATE/multiqc


```

```{r}
devtools::install_github("FemeniasM/ExplorATEproject")


path.sal <- file.path(PROJECT_DIR, "/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/quant_out")

dir <- list.dirs(path.sal,
  recursive = F,
  full.names = F
)

files <- file.path(path.sal, dir, "quant.sf")

names(files) <- dir

re.references <- read.csv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/ambMex60DD_ucsc_hub.ALL/references.csv"),
  sep = ";",
  header = FALSE
)


locus_explorate_quant <- tximport::tximport(files,
  type = "salmon",
  txOut = TRUE
)


saveRDS(locus_explorate_quant, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/import/explorate_re_locus_quant_all_samples.rds"))
# locus_explorate_quant <- readRDS(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/import/explorate_re_locus_quant_all_samples.rds"))

# save each list element as a separate .tsv
for (i in seq_along(locus_explorate_quant)) {
  write.table(locus_explorate_quant[[i]], file = paste0(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/import/explorate_re_locus_", names(locus_explorate_quant)[i], "_all_samples.tsv"), sep = "\t"))
}




family_explorate_quant <- summarizeToGene(locus_explorate_quant,
  tx2gene = re.references
)

# reorder re classification to match TEcount results
rownames(family_explorate_quant$abundance) <- sapply(strsplit(rownames(family_explorate_quant$abundance), "/"), function(x) paste(x[1], x[3], x[2], sep = ":"))
rownames(family_explorate_quant$counts) <- sapply(strsplit(rownames(family_explorate_quant$counts), "/"), function(x) paste(x[1], x[3], x[2], sep = ":"))
rownames(family_explorate_quant$length) <- sapply(strsplit(rownames(family_explorate_quant$length), "/"), function(x) paste(x[1], x[3], x[2], sep = ":"))

saveRDS(family_explorate_quant, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/import/explorate_re_subfamily_quant_all_samples.rds"))
# family_explorate_quant <- readRDS(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/import/explorate_re_subfamily_quant_all_samples.rds"))

for (i in seq_along(family_explorate_quant)) {
  write.table(family_explorate_quant[[i]],
              file = file.path(PROJECT_DIR, paste0("/rnaseq/ExplorATE/import/explorate_re_subfamily_", names(family_explorate_quant)[i], "_all_samples.tsv")),
  sep = "\t",
  quote = FALSE)
}
```

##### Differential expression analysis

######  Native samples

####### Setting DESeq2 model

```{r}

# use lapply to replace "-" in colnames with "_"
nat_family_explorate_quant <- lapply(family_explorate_quant, function(x) {
  if (is.matrix(x)) {
    colnames(x) <- gsub("-", "_", colnames(x))
    x
  } else {
    return(x)
  }
})


nat_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"),
  header = TRUE,
  sep = "\t"
)


# replace "-" with "_" in nat_metadata$new_name_run
nat_metadata$new_name_run <- gsub("-", "_", nat_metadata$new_name_run)
rownames(nat_metadata) <- nat_metadata$new_name_run


# use lapply to subset each list element in family_explorate_quant to only include the samples in nat_metadata
nat_family_explorate_quant <- lapply(nat_family_explorate_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% nat_metadata$new_name_run])
} else {
  return(x)
})

library(DESeq2)

nat_explorate_fam_dds_raw <- DESeqDataSetFromTximport(
  txi = nat_family_explorate_quant,
  colData = nat_metadata,
  design = ~ Age_group + Tissue_group
)

# Explicitly setting the factors levels
nat_explorate_fam_dds_raw$Age_group <- relevel(nat_explorate_fam_dds_raw$Age_group, ref = "Sub_adult")
nat_explorate_fam_dds_raw$Tissue_group <- relevel(nat_explorate_fam_dds_raw$Tissue_group, ref = "Limb")

saveRDS(nat_explorate_fam_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_explorate_fam_dds_raw.rds"))
# nat_explorate_fam_dds_raw <- readRDS(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_explorate_fam_dds_raw.rds"))

```

####### Pre-filtering

```{r}
summary(rowSums(counts(nat_explorate_fam_dds_raw)) >= 10)

#   Mode   FALSE    TRUE
# logical       3    2427

# No filtering was applied
```

####### Pre-visualization

######### Variance Stabilizing Transformation (VST)

```{r}
library(vsn)
library(ggplot2)
library(scales)
library(DESeq2)

nat_explorate_fam_dds_vst <- vst(nat_explorate_fam_dds_raw, blind = FALSE)

saveRDS(nat_explorate_fam_dds_vst, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_explorate_fam_dds_vst.rds"))
```

######## cPCA (vst)

```{r}
library(cowplot)
library(ggpubr)
library(ggrepel)
library(ggh4x)
```

Get underlying data:
```{r}
nat_explorate_fam_pcaData <- plotPCA(nat_explorate_fam_dds_vst, intgroup = c("Age_group", "Tissue_group"), returnData = TRUE)
nat_explorate_fam_percentVar <- round(100 * attr(nat_explorate_fam_pcaData, "percentVar"))

merged <- merge(nat_explorate_fam_pcaData, nat_metadata, by.x = "name", by.y = "new_name_run", all.x = TRUE)
merged <- merged[order(merged$name), ]

nat_explorate_fam_pcaData <- nat_explorate_fam_pcaData[order(row.names(nat_explorate_fam_pcaData)), ]


library(dplyr)
library(readr)

nat_explorate_fam_pcaData <- as_tibble(nat_explorate_fam_pcaData)

write_tsv(nat_explorate_fam_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/previsualization/nat_explorate_fam_pcaData.tsv"))

```

Plotting:
```{r}
nat_explorate_fam_pca_main <- ggplot(nat_explorate_fam_pcaData, aes(PC1, PC2, color = Tissue_group, fill = Tissue_group, shape = Age_group, label = name)) +
  geom_point(size = 3.5) +
  scale_shape_manual(
    name = "Age group", values = c(24, 22),
    breaks = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"),
    guide = guide_legend(order = 1, label.theme = element_text(size = 9))
  ) +
  scale_color_manual(
    name = "Tissue", values = c("#4dac26", "#d01c8b"),
    breaks = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema")
  ) +
  scale_fill_manual(
    name = "Tissue", values = c("#4dac26", "#d01c8b"),
    breaks = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema")
  ) +
  xlab(paste0("PC1: ", nat_explorate_fam_percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", nat_explorate_fam_percentVar[2], "% variance")) +
  theme_bw() +
  theme(
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.4),
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  guides(fill = guide_legend(order = 2), color = guide_legend(order = 2)) +
  geom_text_repel(
    show.legend = FALSE,
    color = "black", size = 3,
    #max.overlaps = Inf,
    box.padding = 0.3
  ) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/previsualization/nat_explorate_fam_cPCA.png"),
  plot = nat_explorate_fam_pca_main,
  device = "png", path = NULL, scale = 1,
  width = 150, height = 90, units = "mm",
  bg = "white", dpi = 1200
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/previsualization/nat_explorate_fam_cPCA.svg"),
  plot = nat_explorate_fam_pca_main,
  device = "svg", path = NULL, scale = 1,
  width = 150, height = 90, units = "mm",
  bg = "white"
)
```

######## Heatmaps of sample clustering

```{r}
library(ComplexHeatmap)
library(viridisLite)
library(RColorBrewer)
library(DESeq2)

# Extract distance matrix
nat_explorate_fam_cor_dist_mat <- (1 - cor(assay(nat_explorate_fam_dds_vst), method = "pearson"))

# To replace sample names
match_idx <- match(rownames(nat_explorate_fam_cor_dist_mat), nat_metadata$new_name_run)
replacement_vals <- nat_metadata$new_name_run[match_idx]
rownames(nat_explorate_fam_cor_dist_mat) <- replacement_vals
colnames(nat_explorate_fam_cor_dist_mat) <- replacement_vals


write.table(nat_explorate_fam_cor_dist_mat,
  sep = "\t",
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/previsualization/nat_explorate_fam_cor_dist_mat.tsv")
)

# Prepare heatmap annotations
annotation <- as.data.frame(colData(nat_explorate_fam_dds_vst))
annotation <- annotation[c("Age_group", "Tissue_group")]
rownames(annotation) <- replacement_vals


# This changes some of the global options of ComplexHeatmap to be able to add a whitespace to the left of a merged legend.
# Should be reset after generating all heatmaps
# ht_opt(HEATMAP_LEGEND_PADDING = unit(c(6,0,0,0), "mm"), ANNOTATION_LEGEND_PADDING = unit(c(6,0,0,0), "mm"))
# ht_opt(RESET = TRUE)

column_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b"
    )
  ),
  annotation_label = gt_render(c("Age group", "Tissue")),
  annotation_legend_param = list(
    Age_group = list(
      title = "Age group",
      at = c("Adult", "Sub_adult"),
      labels = c("Adult (8 years)", "Sub-adult (8 months)"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    ),
    Tissue_group = list(
      title = "Tissue",
      at = c("Limb", "Limb_blastema"),
      labels = c("Limb", "Limb blastema"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    )
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

# Create ComplexHeatmap
# Color palette copied from: https://www.rdocumentation.org/packages/pheatmap/versions/1.0.12/topics/pheatmap
nat_explorate_fam_sample_dist_heatmap <- Heatmap(nat_explorate_fam_cor_dist_mat,
  name = "Distance",
  clustering_distance_rows = function(m) as.dist(m),
  clustering_distance_columns = function(m) as.dist(m),
  row_dend_side = "right",
  row_names_side = "left",
  top_annotation = column_ha,
  row_title = "RNA-seq run", row_title_rot = 90, row_title_gp = gpar(fontface = "bold", fontsize = 11),
  column_title = "RNA-seq run", column_title_side = "bottom", column_title_gp = gpar(fontface = "bold", fontsize = 11),
  row_names_gp = gpar(fontsize = 8.5),
  column_names_gp = gpar(fontsize = 8.5),
  row_dend_width = unit(0.6, "cm"),
  column_dend_height = unit(0.6, "cm"),
  col = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
   width = unit(3, "in"), height = unit(3, "in"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.5),
  heatmap_legend_param = list(
    legend_height = unit(3, "cm"),
    labels_gp = gpar(fontsize = 9),
    title_gp = gpar(
      fontsize = 11,
      fontface = "bold"
    )
  ),
  cell_fun = function(j, i, x, y, width, height, fill) {
    if (nat_explorate_fam_cor_dist_mat[i, j] != 0) {
      ifelse(nat_explorate_fam_cor_dist_mat[i, j] < 0.22, grid.text(sprintf("%.2f", nat_explorate_fam_cor_dist_mat[i, j]), x, y, gp = gpar(fontsize = 7, col = "gray55")), grid.text(sprintf("%.2f", nat_explorate_fam_cor_dist_mat[i, j]), x, y, gp = gpar(fontsize = 8.5, col = "gray85")))
    }
  },
)


# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/previsualization/nat_explorate_fam_sample_dist_heatmap.png"),
  width = 18, height = 13, units = "cm", res = 1200
)

draw(nat_explorate_fam_sample_dist_heatmap, merge_legends = TRUE)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/previsualization/nat_explorate_fam_sample_dist_heatmap.svg"),
  width = 18, height = 5.11811
)

draw(nat_explorate_fam_sample_dist_heatmap, merge_legends = TRUE)

dev.off()
```

######## Differential expression analysis (DEA)

######### Adult limbs vs Sub-adult limbs

```{r}
library(DESeq2)

# Only keep rows where Tissue_group is Limb
nat_age_metadata <- nat_metadata[nat_metadata$Tissue_group == "Limb", ]

# use lapply to subset each list element in nat_family_explorate_quant to only include the samples in nat_age_metadata
nat_age_family_explorate_quant <- lapply(nat_family_explorate_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% nat_age_metadata$new_name_run])
} else {
  return(x)
})

all(rownames(nat_age_metadata) %in% colnames(nat_age_family_explorate_quant$abundance))
all(rownames(nat_age_metadata) == colnames(nat_age_family_explorate_quant$abundance))

nat_age_explorate_fam_dds_raw <- DESeqDataSetFromTximport(
  txi = nat_age_family_explorate_quant,
  colData = nat_age_metadata,
  design = ~ Age_group
)

# Explicitly setting the factors levels
nat_age_explorate_fam_dds_raw$Age_group <- relevel(nat_age_explorate_fam_dds_raw$Age_group, ref = "Sub_adult")

saveRDS(nat_age_explorate_fam_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_age_explorate_fam_dds_raw.rds"))
# nat_age_explorate_fam_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_age_explorate_fam_dds_raw.rds"))
```

####### Running DESeq

```{r}
nat_age_explorate_fam_dds_de <- DESeq(nat_age_explorate_fam_dds_raw)

saveRDS(nat_age_explorate_fam_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_age_explorate_fam_dds_de.rds"))
```

####### Extracting contrast tables

```{r}
library("IHW")
library(ashr)
library(dplyr)
library(readr)

resultsNames(nat_age_explorate_fam_dds_de)
```

```{r}
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs <- results(nat_age_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c("Age_group", "Adult", "Sub_adult"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
#Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs)
# out of 2428 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 13, 0.54%
# LFC < -0.58 (down) : 12, 0.49%
# outliers [1]       : 51, 2.1%

nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink <- lfcShrink(nat_age_explorate_fam_dds_de,
  res = nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink)
# out of 2428 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 12, 0.49%
# LFC < -0.58 (down) : 6, 0.25%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs) <- paste0(colnames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs), "_MLE")
colnames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink) <- paste0(colnames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink), "_MMSE")
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE <- nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink$log2FoldChange_MMSE
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$lfcSE_MMSE <- nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink$lfcSE_MMSE
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$svalue_MMSE <- nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$gene_symbol[is.na(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$gene_symbol)] <- rownames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs)[is.na(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$gene_symbol)]
# add metadata columns
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$quant_tool <- "ExplorATE"
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$diffexp_contrast <- "adult_limbs_vs_subadult_limbs"
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$sample_group <- "native"
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) | abs(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$padj_MLE) | nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs <- as_tibble(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs, rownames = "gene_id")

write_tsv(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs.tsv")
)

# nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs.tsv"))

```

####### Sub-adult blastemas vs Sub-adult limbs

```{r}
library(DESeq2)

# Only keep rows where Age_group is Limb
nat_reg_metadata <- nat_metadata[nat_metadata$Age_group == "Sub_adult", ]

# use lapply to subset each list element in nat_family_explorate_quant to only include the samples in nat_reg_metadata
nat_reg_explorate_family_quant <- lapply(nat_family_explorate_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% nat_reg_metadata$new_name_run])
} else {
  return(x)
})

all(rownames(nat_reg_metadata) %in% colnames(nat_reg_explorate_family_quant$abundance))
all(rownames(nat_reg_metadata) == colnames(nat_reg_explorate_family_quant$abundance))

nat_reg_explorate_fam_dds_raw <- DESeqDataSetFromTximport(
  txi = nat_reg_explorate_family_quant,
  colData = nat_reg_metadata,
  design = ~ Tissue_group
)

# Explicitly setting the factors levels
nat_reg_explorate_fam_dds_raw$Age_group <- relevel(nat_reg_explorate_fam_dds_raw$Tissue_group, ref = "Limb")

saveRDS(nat_reg_explorate_fam_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_reg_explorate_fam_dds_raw.rds"))
# nat_reg_explorate_fam_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_reg_explorate_fam_dds_raw.rds"))
```

####### Running DESeq

```{r}
nat_reg_explorate_fam_dds_de <- DESeq(nat_reg_explorate_fam_dds_raw)

saveRDS(nat_reg_explorate_fam_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/nat_reg_explorate_fam_dds_de.rds"))
```

####### Extracting contrast tables

```{r}
library("IHW")
library(ashr)
library(dplyr)
library(readr)

resultsNames(nat_reg_explorate_fam_dds_de)
```

```{r}
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs <- results(nat_reg_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c("Tissue_group", "Limb_blastema", "Limb"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs)
# out of 2428 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 9, 0.37%
# LFC < -0.58 (down) : 3, 0.12%
# outliers [1]       : 66, 2.7%

nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink <- lfcShrink(nat_reg_explorate_fam_dds_de,
  res = nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink)
# out of 2428 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 3, 0.12%
# LFC < -0.58 (down) : 1, 0.041%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs) <- paste0(colnames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs), "_MLE")
colnames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink) <- paste0(colnames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink), "_MMSE")
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE <- nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink$log2FoldChange_MMSE
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$lfcSE_MMSE <- nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink$lfcSE_MMSE
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$svalue_MMSE <- nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$gene_symbol[is.na(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$gene_symbol)] <- rownames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs)[is.na(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$gene_symbol)]
# add metadata columns
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$quant_tool <- "ExplorATE"
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_limbs"
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$sample_group <- "native"
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$diffexp_res_lfc <- ifelse(is.na(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE) | abs(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$diffexp_res_padj <- ifelse(is.na(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$padj_MLE) | nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs$padj_MLE > 0.05, "not_significant", "significant")

nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs <- as_tibble(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs, rownames = "gene_id")

write_tsv(nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs.tsv")
)

# nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs.tsv"))

```


####### Volcano plots

```{r}

nat_explorate_fam_all <- bind_rows(nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs, nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs)

# remove rows ehere padj_MLE is NA or log2FoldChange_MMSE is NA
nat_explorate_fam_all <- nat_explorate_fam_all %>%
  filter(!is.na(padj_MLE) & !is.na(log2FoldChange_MMSE))

```

```{r}
library(ggh4x)
library(EnhancedVolcano)
library(tidyr)
library(RColorBrewer)


# split gene_id column into three columns, based on the ":" separator, rename: c("RE_subfam", "RE_superfam", "RE_class")
nat_explorate_fam_all <- nat_explorate_fam_all %>%
  tidyr::separate(gene_id, c("RE_subfam", "RE_class", "RE_superfam"), sep = "/", remove = FALSE) %>%
  # if RE_subfam starts with "rnd-", replace with "Unknown"
  mutate(RE_superfam = ifelse(grepl("^rnd-", RE_superfam), "Unknown", RE_superfam)) %>%
  # replace Gypsy with Ty3, not necessarilly the full word
  mutate(RE_superfam = gsub("Gypsy", "Ty3", RE_superfam)) %>%
  # remove "?" in RE_class
  mutate(RE_class = gsub("\\?", "", RE_class))


# add shape, color and fill columns
nat_explorate_fam_all <- nat_explorate_fam_all %>%
  mutate(shape = ifelse(log2FoldChange_MMSE < -log2(1.5), 25, ifelse(log2FoldChange_MMSE > log2(1.5), 24, 21)),
  # for each unique value in RE_class, assign a color from a palette
        color = factor(RE_class, levels = sort(unique(RE_class)), labels = brewer.pal(n = length(unique(RE_class)), name = "Dark2")))

# Associating colors to each type of value
# convert to named vector
keyvals <- setNames(as.character(nat_explorate_fam_all$color), as.character(nat_explorate_fam_all$RE_class))
shapevals <- setNames(nat_explorate_fam_all$shape, nat_explorate_fam_all$shape)


# Creating volcano plot
volcanoes_nat_explorate_fam <- EnhancedVolcano(

  as.data.frame(nat_explorate_fam_all),
  lab = NA,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +

  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +


  # change y scale of first plot, keep the second plot as it is
  
  facetted_pos_scales(
  y = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_explorate_fam_all)$padj_MLE)[as.data.frame(nat_explorate_fam_all)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_explorate_fam_all)$padj_MLE)[as.data.frame(nat_explorate_fam_all)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(nat_explorate_fam_all)$log2FoldChange_MMSE[as.data.frame(nat_explorate_fam_all)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(nat_explorate_fam_all)$log2FoldChange_MMSE[as.data.frame(nat_explorate_fam_all)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )

# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/volcanoes/volcanoes_nat_explorate_fam.png"),
  plot = volcanoes_nat_explorate_fam,
  scale = 1, width = 200, height = 90, units = "mm", bg = "white", dpi = 1200
)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/volcanoes/volcanoes_nat_tecount_re.svg"),
  width = 6.29921, height = 2.83465
)

volcanoes_nat_explorate_fam

dev.off()

```

###### Global samples

####### Setting DESeq2 model

```{r}
glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"),
  sep = "\t",
  header = TRUE
)


# use lapply to replace "-" in colnames with "_"
glob_family_explorate_quant <- lapply(family_explorate_quant, function(x) {
  if (is.matrix(x)) {
    colnames(x) <- gsub("-", "_", colnames(x))
    x
  } else {
    return(x)
  }
})

# use lapply to subset each list element in family_explorate_quant to only include the samples in glob_metadata
glob_family_explorate_quant <- lapply(glob_family_explorate_quant, function(x) if (is.matrix(x)) {
  return(x[, colnames(x) %in% glob_metadata$Run])
} else {
  return(x)
})

library(DESeq2)

glob_explorate_fam_dds <- DESeqDataSetFromTximport(
  txi = glob_family_explorate_quant,
  colData = glob_metadata,
  design = ~ Batch + Age_group + Tissue_group + Age_group:Tissue_group
)

# Explicitly setting the factors levels
glob_explorate_fam_dds$Age_group <- relevel(glob_explorate_fam_dds$Age_group, ref = "Sub_adult")
glob_explorate_fam_dds$Batch <- relevel(glob_explorate_fam_dds$Batch, ref = "Ruiz_Perez_et_al._n.d.")
glob_explorate_fam_dds$Tissue_group <- relevel(glob_explorate_fam_dds$Tissue_group, ref = "Non_limb")

save(glob_explorate_fam_dds, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/glob_explorate_fam_dds.RData"))
```

####### Pre-filtering

```{r}
summary(rowSums(counts(glob_explorate_fam_dds)) >= 10)

#   Mode    TRUE
# logical    2430

# No filtering was applied
```

```{r}
glob_explorate_fam_dds_vst <- vst(glob_explorate_fam_dds, blind = FALSE)

saveRDS(glob_explorate_fam_dds_vst, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/glob_explorate_fam_dds_vst.rds"))


```

####### cPCA (vst)

```{r}
library(DESeq2)
library(cowplot)
library(ggpubr)
library(ggrepel)
library(ggforce)
library(ggh4x)
library(ggalt)


glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")
```

Personalized cPCA

```{r}
glob_explorate_pcaData <- plotPCA(glob_explorate_fam_dds_vst, intgroup = c("Age_group", "Batch", "Tissue_group"), returnData = TRUE)
percentVar <- round(100 * attr(glob_explorate_pcaData, "percentVar"))

merged <- merge(glob_explorate_pcaData, glob_metadata, by.x = "name", by.y = "Run", all.x = TRUE)
merged <- merged[order(merged$name), ]

glob_explorate_pcaData <- glob_explorate_pcaData[order(row.names(glob_explorate_pcaData)), ]
glob_explorate_pcaData$new_name <- merged$new_name_run

saveRDS(glob_explorate_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/previsualization/glob_explorate_pcaData.rds"))

write.csv(glob_explorate_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/previsualization/glob_explorate_pcaData.csv"))
```

Plotting:
```{r}
glob_explorate_pca_main <- ggplot(glob_explorate_pcaData, aes(PC1, PC2, fill = Tissue_group, shape = Age_group)) +
  geom_encircle(aes(group = Batch, colour = Batch), alpha = 0.4) +
  geom_point(size = 3.2, stroke = 0.2) +
  scale_shape_manual(
    name = "Age group", values = c(24, 22),
    breaks = c("Adult", "Sub_adult"), labels = c("Adult", "Sub-adult"),
    guide = guide_legend(order = 1, label.theme = element_text(size = 9, margin = margin(l = 25)))
  ) +
  scale_fill_manual(
    name = "Tissue_group",
    breaks = c("Non_limb", "Limb", "Limb_blastema"),
    values = c("#57452dff", "#4dac26", "#d01c8b")
  ) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  ) +
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  # guides(color = guide_legend(override.aes = list(size = 3))) +

  coord_fixed() +
  guides(
    fill = guide_legend(order = 2),
    color = guide_legend(order = 2)
  )


ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/previsualization/pca_explorate_vst.png"),
  plot = glob_explorate_pca_main,
  device = "png",
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white",
  dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/previsualization/pca_explorate_vst.svg"),
  plot = glob_explorate_pca_main,
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white"
)
```

Temporal plot with different points to overlay

```{r}
glob_explorate_pca_main <- ggplot(glob_explorate_pcaData, aes(PC1, PC2, color = Batch)) +
  geom_point(size = 12, alpha = 0.5, stroke = 0) +
  scale_color_manual(
    name = "Batch",
    values = c("#d36027ff", "#0773b3ff", "#f2e646ff", "#6b6a66ff", "#e79f26ff"),
    breaks = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"),
    labels = c("This study", bquote("Ye" ~ italic("et al.") ~ "(2022)"), bquote("Nowoshilow" ~ italic("et al.") ~ "(2018)"), bquote("Caballero-Pérez" ~ italic("et al.") ~ "(2018)"), bquote("Bryant" ~ italic("et al.") ~ "(2017)"))
  ) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  ) +
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  # guides(color = guide_legend(override.aes = list(size = 3))) +

  coord_fixed() +
  guides(
    fill = guide_legend(order = 2),
    color = guide_legend(order = 2)
  )

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/previsualization/pca_explorate_vst2.png"),
  plot = glob_explorate_pca_main,
  device = "png",
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white",
  dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/previsualization/pca_explorate_vst2.svg"),
  plot = glob_explorate_pca_main,
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white"
)
```



######## Differential expression (DE) analysis

######### Running DESeq

```{r}
glob_explorate_fam_dds_de <- DESeq(glob_explorate_fam_dds)

save(glob_explorate_fam_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/glob_explorate_fam_dds_de.RData"))
```

######### Summarizing DE results

```{r}
library("IHW")
library(ashr)
```

```{r}
resultsNames(glob_explorate_fam_dds_de)

# [1] "Intercept"
# [2] "Batch_Bryant_et_al._2017_vs_Ruiz_Perez_et_al._n.d."
# [3] "Batch_Caballero_Perez_et_al._2018_vs_Ruiz_Perez_et_al._n.d."
# [4] "Batch_Nowoshilow_et_al._2018_vs_Ruiz_Perez_et_al._n.d."
# [5] "Batch_Ye_et_al._2022_vs_Ruiz_Perez_et_al._n.d."
# [6] "Age_group_Adult_vs_Sub_adult"
# [7] "Tissue_group_Limb_vs_Non_limb"
# [8] "Tissue_group_Limb_blastema_vs_Non_limb"
# [9] "Age_groupAdult.Tissue_groupLimb"
# [10] "Age_groupAdult.Tissue_groupLimb_blastema"
```

########## Limb vs Non-limb (for subadults)

```{r}
glob_explorate_fam_res_l_vs_nl_sa <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 1, 0, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_l_vs_nl_sa)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 23, 0.95%
# LFC < -0.58 (down) : 43, 1.8%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_l_vs_nl_sa_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_l_vs_nl_sa, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_l_vs_nl_sa_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 20, 0.82%
# LFC < -0.58 (down) : 46, 1.9%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_l_vs_nl_sa) <- paste0(colnames(glob_explorate_fam_res_l_vs_nl_sa), "_MLE")
colnames(glob_explorate_fam_res_l_vs_nl_sa_shrink) <- paste0(colnames(glob_explorate_fam_res_l_vs_nl_sa_shrink), "_MMSE")
glob_explorate_fam_res_l_vs_nl_sa$log2FoldChange_MMSE <- glob_explorate_fam_res_l_vs_nl_sa_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_l_vs_nl_sa$lfcSE_MMSE <- glob_explorate_fam_res_l_vs_nl_sa_shrink$lfcSE_MMSE
glob_explorate_fam_res_l_vs_nl_sa$svalue_MMSE <- glob_explorate_fam_res_l_vs_nl_sa_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_l_vs_nl_sa$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_l_vs_nl_sa), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_l_vs_nl_sa$gene_symbol[is.na(glob_explorate_fam_res_l_vs_nl_sa$gene_symbol)] <- rownames(glob_explorate_fam_res_l_vs_nl_sa)[is.na(glob_explorate_fam_res_l_vs_nl_sa$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_l_vs_nl_sa$quant_tool <- "ExplorATE"
glob_explorate_fam_res_l_vs_nl_sa$diffexp_contrast <- "subadult_limbs_vs_subadult_nonlimbs"
glob_explorate_fam_res_l_vs_nl_sa$sample_group <- "global"
glob_explorate_fam_res_l_vs_nl_sa$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_l_vs_nl_sa)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_l_vs_nl_sa$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_l_vs_nl_sa$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_l_vs_nl_sa$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_l_vs_nl_sa$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_l_vs_nl_sa$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_l_vs_nl_sa$padj_MLE) | glob_explorate_fam_res_l_vs_nl_sa$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_l_vs_nl_sa <- as_tibble(glob_explorate_fam_res_l_vs_nl_sa, rownames = "gene_id")


print(glob_explorate_fam_res_l_vs_nl_sa, width = Inf)

readr::write_tsv(glob_explorate_fam_res_l_vs_nl_sa,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_l_vs_nl_sa.tsv")
)

```

########## Limb vs Non-limb (for adults)

```{r}
glob_explorate_fam_res_l_vs_nl_a <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 1, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_l_vs_nl_a)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 0, 0%
# LFC < -0.58 (down) : 0, 0%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_l_vs_nl_a_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_l_vs_nl_a, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_l_vs_nl_a_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 0, 0%
# LFC < -0.58 (down) : 1, 0.041%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_l_vs_nl_a) <- paste0(colnames(glob_explorate_fam_res_l_vs_nl_a), "_MLE")
colnames(glob_explorate_fam_res_l_vs_nl_a_shrink) <- paste0(colnames(glob_explorate_fam_res_l_vs_nl_a_shrink), "_MMSE")
glob_explorate_fam_res_l_vs_nl_a$log2FoldChange_MMSE <- glob_explorate_fam_res_l_vs_nl_a_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_l_vs_nl_a$lfcSE_MMSE <- glob_explorate_fam_res_l_vs_nl_a_shrink$lfcSE_MMSE
glob_explorate_fam_res_l_vs_nl_a$svalue_MMSE <- glob_explorate_fam_res_l_vs_nl_a_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_l_vs_nl_a$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_l_vs_nl_a), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_l_vs_nl_a$gene_symbol[is.na(glob_explorate_fam_res_l_vs_nl_a$gene_symbol)] <- rownames(glob_explorate_fam_res_l_vs_nl_a)[is.na(glob_explorate_fam_res_l_vs_nl_a$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_l_vs_nl_a$quant_tool <- "ExplorATE"
glob_explorate_fam_res_l_vs_nl_a$diffexp_contrast <- "adult_limbs_vs_adult_nonlimbs"
glob_explorate_fam_res_l_vs_nl_a$sample_group <- "global"
glob_explorate_fam_res_l_vs_nl_a$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_l_vs_nl_a)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_l_vs_nl_a$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_l_vs_nl_a$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_l_vs_nl_a$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_l_vs_nl_a$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_l_vs_nl_a$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_l_vs_nl_a$padj_MLE) | glob_explorate_fam_res_l_vs_nl_a$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_l_vs_nl_a <- as_tibble(glob_explorate_fam_res_l_vs_nl_a, rownames = "gene_id")


print(glob_explorate_fam_res_l_vs_nl_a, width = Inf)

readr::write_tsv(glob_explorate_fam_res_l_vs_nl_a,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_l_vs_nl_a.tsv")
)

```

########## Blastema vs Non-limb (for subadults)

```{r}
glob_explorate_fam_res_b_vs_nl_sa <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 1, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_b_vs_nl_sa)
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 19, 0.78%
# LFC < -0.58 (down) : 27, 1.1%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_b_vs_nl_sa_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_b_vs_nl_sa, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_b_vs_nl_sa_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 14, 0.58%
# LFC < -0.58 (down) : 21, 0.86%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_b_vs_nl_sa) <- paste0(colnames(glob_explorate_fam_res_b_vs_nl_sa), "_MLE")
colnames(glob_explorate_fam_res_b_vs_nl_sa_shrink) <- paste0(colnames(glob_explorate_fam_res_b_vs_nl_sa_shrink), "_MMSE")
glob_explorate_fam_res_b_vs_nl_sa$log2FoldChange_MMSE <- glob_explorate_fam_res_b_vs_nl_sa_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_b_vs_nl_sa$lfcSE_MMSE <- glob_explorate_fam_res_b_vs_nl_sa_shrink$lfcSE_MMSE
glob_explorate_fam_res_b_vs_nl_sa$svalue_MMSE <- glob_explorate_fam_res_b_vs_nl_sa_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_b_vs_nl_sa$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_b_vs_nl_sa), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_b_vs_nl_sa$gene_symbol[is.na(glob_explorate_fam_res_b_vs_nl_sa$gene_symbol)] <- rownames(glob_explorate_fam_res_b_vs_nl_sa)[is.na(glob_explorate_fam_res_b_vs_nl_sa$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_b_vs_nl_sa$quant_tool <- "ExplorATE"
glob_explorate_fam_res_b_vs_nl_sa$diffexp_contrast <- "subadult_blastemas_vs_subadult_nonlimbs"
glob_explorate_fam_res_b_vs_nl_sa$sample_group <- "global"
glob_explorate_fam_res_b_vs_nl_sa$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_b_vs_nl_sa)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_b_vs_nl_sa$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_b_vs_nl_sa$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_b_vs_nl_sa$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_b_vs_nl_sa$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_b_vs_nl_sa$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_b_vs_nl_sa$padj_MLE) | glob_explorate_fam_res_b_vs_nl_sa$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_b_vs_nl_sa <- as_tibble(glob_explorate_fam_res_b_vs_nl_sa, rownames = "gene_id")


print(glob_explorate_fam_res_b_vs_nl_sa, width = Inf)

readr::write_tsv(glob_explorate_fam_res_b_vs_nl_sa,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_nl_sa.tsv")
)

```

########## Blastema vs Non-limb (for adults)

```{r}
glob_explorate_fam_res_b_vs_nl_a <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 1, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_b_vs_nl_a)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 24, 0.99%
# LFC < -0.58 (down) : 110, 4.5%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_b_vs_nl_a_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_b_vs_nl_a, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_b_vs_nl_a_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 29, 1.2%
# LFC < -0.58 (down) : 108, 4.4%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_b_vs_nl_a) <- paste0(colnames(glob_explorate_fam_res_b_vs_nl_a), "_MLE")
colnames(glob_explorate_fam_res_b_vs_nl_a_shrink) <- paste0(colnames(glob_explorate_fam_res_b_vs_nl_a_shrink), "_MMSE")
glob_explorate_fam_res_b_vs_nl_a$log2FoldChange_MMSE <- glob_explorate_fam_res_b_vs_nl_a_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_b_vs_nl_a$lfcSE_MMSE <- glob_explorate_fam_res_b_vs_nl_a_shrink$lfcSE_MMSE
glob_explorate_fam_res_b_vs_nl_a$svalue_MMSE <- glob_explorate_fam_res_b_vs_nl_a_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_b_vs_nl_a$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_b_vs_nl_a), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_b_vs_nl_a$gene_symbol[is.na(glob_explorate_fam_res_b_vs_nl_a$gene_symbol)] <- rownames(glob_explorate_fam_res_b_vs_nl_a)[is.na(glob_explorate_fam_res_b_vs_nl_a$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_b_vs_nl_a$quant_tool <- "ExplorATE"
glob_explorate_fam_res_b_vs_nl_a$diffexp_contrast <- "adult_blastemas_vs_adult_nonlimbs"
glob_explorate_fam_res_b_vs_nl_a$sample_group <- "global"
glob_explorate_fam_res_b_vs_nl_a$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_b_vs_nl_a)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_b_vs_nl_a$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_b_vs_nl_a$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_b_vs_nl_a$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_b_vs_nl_a$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_b_vs_nl_a$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_b_vs_nl_a$padj_MLE) | glob_explorate_fam_res_b_vs_nl_a$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_b_vs_nl_a <- as_tibble(glob_explorate_fam_res_b_vs_nl_a, rownames = "gene_id")


print(glob_explorate_fam_res_b_vs_nl_a, width = Inf)

readr::write_tsv(glob_explorate_fam_res_b_vs_nl_a,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_nl_a.tsv")
)

```

########## Blastema vs Limb (for subadults)

```{r}
glob_explorate_fam_res_b_vs_l_sa <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, -1, 1, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_b_vs_l_sa)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 23, 0.95%
# LFC < -0.58 (down) : 15, 0.62%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_b_vs_l_sa_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_b_vs_l_sa, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_b_vs_l_sa_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 24, 0.99%
# LFC < -0.58 (down) : 14, 0.58%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_b_vs_l_sa) <- paste0(colnames(glob_explorate_fam_res_b_vs_l_sa), "_MLE")
colnames(glob_explorate_fam_res_b_vs_l_sa_shrink) <- paste0(colnames(glob_explorate_fam_res_b_vs_l_sa_shrink), "_MMSE")
glob_explorate_fam_res_b_vs_l_sa$log2FoldChange_MMSE <- glob_explorate_fam_res_b_vs_l_sa_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_b_vs_l_sa$lfcSE_MMSE <- glob_explorate_fam_res_b_vs_l_sa_shrink$lfcSE_MMSE
glob_explorate_fam_res_b_vs_l_sa$svalue_MMSE <- glob_explorate_fam_res_b_vs_l_sa_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_b_vs_l_sa$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_b_vs_l_sa), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_b_vs_l_sa$gene_symbol[is.na(glob_explorate_fam_res_b_vs_l_sa$gene_symbol)] <- rownames(glob_explorate_fam_res_b_vs_l_sa)[is.na(glob_explorate_fam_res_b_vs_l_sa$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_b_vs_l_sa$quant_tool <- "ExplorATE"
glob_explorate_fam_res_b_vs_l_sa$diffexp_contrast <- "subadult_blastemas_vs_subadult_limbs"
glob_explorate_fam_res_b_vs_l_sa$sample_group <- "global"
glob_explorate_fam_res_b_vs_l_sa$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_b_vs_l_sa)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_b_vs_l_sa$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_b_vs_l_sa$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_b_vs_l_sa$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_b_vs_l_sa$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_b_vs_l_sa$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_b_vs_l_sa$padj_MLE) | glob_explorate_fam_res_b_vs_l_sa$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_b_vs_l_sa <- as_tibble(glob_explorate_fam_res_b_vs_l_sa, rownames = "gene_id")


print(glob_explorate_fam_res_b_vs_l_sa, width = Inf)

readr::write_tsv(glob_explorate_fam_res_b_vs_l_sa,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_l_sa.tsv")
)

```

########## Blastema vs Limb (for adults)

```{r}
glob_explorate_fam_res_b_vs_l_a <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, -1, 1, -1, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_b_vs_l_a)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 12, 0.49%
# LFC < -0.58 (down) : 56, 2.3%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_b_vs_l_a_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_b_vs_l_a, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_b_vs_l_a_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 13, 0.53%
# LFC < -0.58 (down) : 55, 2.3%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_b_vs_l_a) <- paste0(colnames(glob_explorate_fam_res_b_vs_l_a), "_MLE")
colnames(glob_explorate_fam_res_b_vs_l_a_shrink) <- paste0(colnames(glob_explorate_fam_res_b_vs_l_a_shrink), "_MMSE")
glob_explorate_fam_res_b_vs_l_a$log2FoldChange_MMSE <- glob_explorate_fam_res_b_vs_l_a_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_b_vs_l_a$lfcSE_MMSE <- glob_explorate_fam_res_b_vs_l_a_shrink$lfcSE_MMSE
glob_explorate_fam_res_b_vs_l_a$svalue_MMSE <- glob_explorate_fam_res_b_vs_l_a_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_b_vs_l_a$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_b_vs_l_a), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_b_vs_l_a$gene_symbol[is.na(glob_explorate_fam_res_b_vs_l_a$gene_symbol)] <- rownames(glob_explorate_fam_res_b_vs_l_a)[is.na(glob_explorate_fam_res_b_vs_l_a$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_b_vs_l_a$quant_tool <- "ExplorATE"
glob_explorate_fam_res_b_vs_l_a$diffexp_contrast <- "adult_blastemas_vs_adult_limbs"
glob_explorate_fam_res_b_vs_l_a$sample_group <- "global"
glob_explorate_fam_res_b_vs_l_a$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_b_vs_l_a)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_b_vs_l_a$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_b_vs_l_a$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_b_vs_l_a$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_b_vs_l_a$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_b_vs_l_a$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_b_vs_l_a$padj_MLE) | glob_explorate_fam_res_b_vs_l_a$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_b_vs_l_a <- as_tibble(glob_explorate_fam_res_b_vs_l_a, rownames = "gene_id")


print(glob_explorate_fam_res_b_vs_l_a, width = Inf)

readr::write_tsv(glob_explorate_fam_res_b_vs_l_a,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_l_a.tsv")
)

```

########## Adults vs Sub-adults (for non-limb)

```{r}
glob_explorate_fam_res_a_vs_sa_nl <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_a_vs_sa_nl)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 134, 5.5%
# LFC < -0.58 (down) : 31, 1.3%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_a_vs_sa_nl_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_a_vs_sa_nl, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_a_vs_sa_nl_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 116, 4.8%
# LFC < -0.58 (down) : 34, 1.4%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_a_vs_sa_nl) <- paste0(colnames(glob_explorate_fam_res_a_vs_sa_nl), "_MLE")
colnames(glob_explorate_fam_res_a_vs_sa_nl_shrink) <- paste0(colnames(glob_explorate_fam_res_a_vs_sa_nl_shrink), "_MMSE")
glob_explorate_fam_res_a_vs_sa_nl$log2FoldChange_MMSE <- glob_explorate_fam_res_a_vs_sa_nl_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_a_vs_sa_nl$lfcSE_MMSE <- glob_explorate_fam_res_a_vs_sa_nl_shrink$lfcSE_MMSE
glob_explorate_fam_res_a_vs_sa_nl$svalue_MMSE <- glob_explorate_fam_res_a_vs_sa_nl_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_a_vs_sa_nl$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_a_vs_sa_nl), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_a_vs_sa_nl$gene_symbol[is.na(glob_explorate_fam_res_a_vs_sa_nl$gene_symbol)] <- rownames(glob_explorate_fam_res_a_vs_sa_nl)[is.na(glob_explorate_fam_res_a_vs_sa_nl$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_a_vs_sa_nl$quant_tool <- "ExplorATE"
glob_explorate_fam_res_a_vs_sa_nl$diffexp_contrast <- "adult_nonlimbs_vs_subadult_nonlimbs"
glob_explorate_fam_res_a_vs_sa_nl$sample_group <- "global"
glob_explorate_fam_res_a_vs_sa_nl$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_a_vs_sa_nl)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_a_vs_sa_nl$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_a_vs_sa_nl$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_a_vs_sa_nl$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_a_vs_sa_nl$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_a_vs_sa_nl$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_a_vs_sa_nl$padj_MLE) | glob_explorate_fam_res_a_vs_sa_nl$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_a_vs_sa_nl <- as_tibble(glob_explorate_fam_res_a_vs_sa_nl, rownames = "gene_id")


print(glob_explorate_fam_res_a_vs_sa_nl, width = Inf)

readr::write_tsv(glob_explorate_fam_res_a_vs_sa_nl,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_a_vs_sa_nl.tsv")
)

```

########## Adults vs Sub-adults (for limb)

```{r}
glob_explorate_fam_res_a_vs_sa_l <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_a_vs_sa_l)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 223, 9.2%
# LFC < -0.58 (down) : 24, 0.99%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_a_vs_sa_l_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_a_vs_sa_l, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_a_vs_sa_l_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 230, 9.5%
# LFC < -0.58 (down) : 28, 1.2%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_a_vs_sa_l) <- paste0(colnames(glob_explorate_fam_res_a_vs_sa_l), "_MLE")
colnames(glob_explorate_fam_res_a_vs_sa_l_shrink) <- paste0(colnames(glob_explorate_fam_res_a_vs_sa_l_shrink), "_MMSE")
glob_explorate_fam_res_a_vs_sa_l$log2FoldChange_MMSE <- glob_explorate_fam_res_a_vs_sa_l_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_a_vs_sa_l$lfcSE_MMSE <- glob_explorate_fam_res_a_vs_sa_l_shrink$lfcSE_MMSE
glob_explorate_fam_res_a_vs_sa_l$svalue_MMSE <- glob_explorate_fam_res_a_vs_sa_l_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_a_vs_sa_l$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_a_vs_sa_l), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_a_vs_sa_l$gene_symbol[is.na(glob_explorate_fam_res_a_vs_sa_l$gene_symbol)] <- rownames(glob_explorate_fam_res_a_vs_sa_l)[is.na(glob_explorate_fam_res_a_vs_sa_l$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_a_vs_sa_l$quant_tool <- "ExplorATE"
glob_explorate_fam_res_a_vs_sa_l$diffexp_contrast <- "adult_limbs_vs_subadult_limbs"
glob_explorate_fam_res_a_vs_sa_l$sample_group <- "global"
glob_explorate_fam_res_a_vs_sa_l$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_a_vs_sa_l)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_a_vs_sa_l$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_a_vs_sa_l$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_a_vs_sa_l$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_a_vs_sa_l$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_a_vs_sa_l$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_a_vs_sa_l$padj_MLE) | glob_explorate_fam_res_a_vs_sa_l$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_a_vs_sa_l <- as_tibble(glob_explorate_fam_res_a_vs_sa_l, rownames = "gene_id")


print(glob_explorate_fam_res_a_vs_sa_l, width = Inf)

readr::write_tsv(glob_explorate_fam_res_a_vs_sa_l,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_a_vs_sa_l.tsv")
)

```

########## Adults vs Sub-adults (for blastema)

```{r}
glob_explorate_fam_res_a_vs_sa_b <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

summary(glob_explorate_fam_res_a_vs_sa_b)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 29, 1.2%
# LFC < -0.58 (down) : 20, 0.82%
# outliers [1]       : 19, 0.78%

# LFC Shrinkage
glob_explorate_fam_res_a_vs_sa_b_shrink <- lfcShrink(glob_explorate_fam_dds_de, res = glob_explorate_fam_res_a_vs_sa_b, lfcThreshold = log2(1.5), alpha = 0.05, type = "ashr")

summary(glob_explorate_fam_res_a_vs_sa_b_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 21, 0.86%
# LFC < -0.58 (down) : 15, 0.62%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(glob_explorate_fam_res_a_vs_sa_b) <- paste0(colnames(glob_explorate_fam_res_a_vs_sa_b), "_MLE")
colnames(glob_explorate_fam_res_a_vs_sa_b_shrink) <- paste0(colnames(glob_explorate_fam_res_a_vs_sa_b_shrink), "_MMSE")
glob_explorate_fam_res_a_vs_sa_b$log2FoldChange_MMSE <- glob_explorate_fam_res_a_vs_sa_b_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_a_vs_sa_b$lfcSE_MMSE <- glob_explorate_fam_res_a_vs_sa_b_shrink$lfcSE_MMSE
glob_explorate_fam_res_a_vs_sa_b$svalue_MMSE <- glob_explorate_fam_res_a_vs_sa_b_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_a_vs_sa_b$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_a_vs_sa_b), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_a_vs_sa_b$gene_symbol[is.na(glob_explorate_fam_res_a_vs_sa_b$gene_symbol)] <- rownames(glob_explorate_fam_res_a_vs_sa_b)[is.na(glob_explorate_fam_res_a_vs_sa_b$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_a_vs_sa_b$quant_tool <- "ExplorATE"
glob_explorate_fam_res_a_vs_sa_b$diffexp_contrast <- "adults_blastemas_vs_subadults_blastemas"
glob_explorate_fam_res_a_vs_sa_b$sample_group <- "global"
glob_explorate_fam_res_a_vs_sa_b$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_a_vs_sa_b)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_a_vs_sa_b$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_a_vs_sa_b$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_a_vs_sa_b$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_a_vs_sa_b$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_a_vs_sa_b$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_a_vs_sa_b$padj_MLE) | glob_explorate_fam_res_a_vs_sa_b$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_a_vs_sa_b <- as_tibble(glob_explorate_fam_res_a_vs_sa_b, rownames = "gene_id")

print(glob_explorate_fam_res_a_vs_sa_b, width = Inf)

readr::write_tsv(glob_explorate_fam_res_a_vs_sa_b,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_a_vs_sa_b.tsv")
)

```

######### Interaction term: Age_group effect in Limb vs Non-limb (Age_groupAdult.Tissue_groupLimb)

this tests if the Age_group effect is different in Limb compared to Non-limb

```{r}
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 27, 1.1%
# LFC < -0.58 (down) : 7, 0.29%
# outliers [1]       : 19, 0.78%

## LFC Shrinkage
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink <- lfcShrink(glob_explorate_fam_dds_de,
  res = glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 24, 0.99%
# LFC < -0.58 (down) : 4, 0.16%

colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb) <- paste0(colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb), "_MLE")
colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink) <- paste0(colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink), "_MMSE")
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE <- glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$lfcSE_MMSE <- glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink$lfcSE_MMSE
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$svalue_MMSE <- glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol[is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol)] <- rownames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb)[is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$quant_tool <- "ExplorATE"
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_contrast <- "age_effect_limbs_vs_nonlimbs"
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$sample_group <- "global"
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$padj_MLE) | glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb <- as_tibble(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb, rownames = "gene_id")

print(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb, width = Inf)

readr::write_tsv(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb.tsv")
)

```

######### Interaction term: Age_group effect in Blastema vs Non-limb (Age_groupAdult.Tissue_groupLimb_blastema)
 
this tests if the Age_group effect is different in Blastema compared to Non-limb

```{r}
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema)
# Only 1 bin; IHW reduces to Benjamini Hochberg (uniform weights)

# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 31, 1.3%
# LFC < -0.58 (down) : 44, 1.8%
# outliers [1]       : 19, 0.78%

## LFC Shrinkage

glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink <- lfcShrink(glob_explorate_fam_dds_de,
  res = glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 26, 1.1%
# LFC < -0.58 (down) : 35, 1.4%

colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema) <- paste0(colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema), "_MLE")
colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink) <- paste0(colnames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink), "_MMSE")
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE <- glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$lfcSE_MMSE <- glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$lfcSE_MMSE
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$svalue_MMSE <- glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol[is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol)] <- rownames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema)[is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$quant_tool <- "ExplorATE"
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_contrast <- "age_effect_blastemas_vs_nonlimbs"
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$sample_group <- "global"
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$padj_MLE) | glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- as_tibble(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema, rownames = "gene_id")

print(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema, width = Inf)

readr::write_tsv(glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema.tsv")
)

```

######### Interaction term: Age_group effect in Blastema vs Limb (Age_groupAdult.Tissue_groupLimb_blastema vs Age_groupAdult.Tissue_groupLimb)

this tests if the Age_group effect is different in Blastema compared to Limb

```{r}
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb <- results(glob_explorate_fam_dds_de,
  filterFun = ihw,
  contrast = list("Age_groupAdult.Tissue_groupLimb_blastema", "Age_groupAdult.Tissue_groupLimb"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb)
# out of 2430 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 11, 0.45%
# LFC < -0.58 (down) : 33, 1.4%
# outliers [1]       : 19, 0.78%

## LFC Shrinkage
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink <- lfcShrink(glob_explorate_fam_dds_de,
  res = glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink)
# out of 2430 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 12, 0.49%
# LFC < -0.58 (down) : 31, 1.3%

colnames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb) <- paste0(colnames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb), "_MLE")
colnames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink) <- paste0(colnames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink), "_MMSE")
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE <- glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink$log2FoldChange_MMSE
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$lfcSE_MMSE <- glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink$lfcSE_MMSE
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$svalue_MMSE <- glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$gene_symbol[is.na(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$gene_symbol)] <- rownames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb)[is.na(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$gene_symbol)]
# add metadata columns
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$quant_tool <- "ExplorATE"
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$diffexp_contrast <- "age_effect_blastemas_vs_limbs"
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$sample_group <- "global"
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$diffexp_res_lfc <- ifelse(is.na(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE) | abs(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$diffexp_res_padj <- ifelse(is.na(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$padj_MLE) | glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb$padj_MLE > 0.05, "not_significant", "significant")

glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb <- as_tibble(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb, rownames = "gene_id")

print(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb, width = Inf)

readr::write_tsv(glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb,
  file = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb.tsv")
)

```


Volcano plots

```{r}

library(ggh4x)
library(EnhancedVolcano)
library(tidyr)
library(RColorBrewer)

```

Set 1

```{r}

glob_explorate_res_re_1 <- all_diffexp_results %>%
    filter(quant_tool == "ExplorATE",
           feature_biotype == "repetitive_element",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c( "subadult_blastemas_vs_subadult_limbs",
                                    "adult_blastemas_vs_adult_limbs",
                                    "adult_limbs_vs_subadult_limbs",
                                    "adults_blastemas_vs_subadults_blastemas",
                                    "age_effect_blastemas_vs_limbs")
)

# split gene_id column into three columns, based on the ":" separator, rename: c("RE_subfam", "RE_superfam", "RE_class")
glob_explorate_res_re_1 <- glob_explorate_res_re_1 %>%
  tidyr::separate(ID, c("RE_subfam", "RE_superfam", "RE_class"), sep = ":", remove = FALSE) %>%
  # if RE_subfam starts with "rnd-", replace with "Unknown"
  mutate(RE_superfam = ifelse(grepl("^rnd-", RE_superfam), "Unknown", RE_superfam)) %>%
  # replace Gypsy with Ty3, not necessarilly the full word
  mutate(RE_superfam = gsub("Gypsy", "Ty3", RE_superfam)) %>%
  # remove "?" in RE_class
  mutate(RE_class = gsub("\\?", "", RE_class))

# add shape, color and fill columns
glob_explorate_res_re_1 <- glob_explorate_res_re_1 %>%
  mutate(shape = ifelse(log2FoldChange_MMSE < -log2(1.5), 25, ifelse(log2FoldChange_MMSE > log2(1.5), 24, 21)),
  # for each unique value in RE_class, assign a color from a palette
        color = factor(RE_class, levels = sort(unique(RE_class)), labels = brewer.pal(n = length(unique(RE_class)), name = "Dark2")))

# Associating colors to each type of value
# convert to named vector
keyvals <- setNames(as.character(glob_explorate_res_re_1$color), as.character(glob_explorate_res_re_1$RE_class))
shapevals <- setNames(glob_explorate_res_re_1$shape, glob_explorate_res_re_1$shape)

# Creating volcano plot
volcanoes_glob_explorate_re_1 <- EnhancedVolcano(

  as.data.frame(glob_explorate_res_re_1),
  lab = NA,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +

  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +


  # change y scale of first plot, keep the second plot as it is
  
  facetted_pos_scales(
  y = list(
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_1)$padj_MLE)[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_1)$padj_MLE)[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_1)$padj_MLE)[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_1)$padj_MLE)[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_1)$padj_MLE)[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )

# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/volcanoes/volcanoes_glob_explorate_re_1.png"),
  plot = volcanoes_glob_explorate_re_1,
  scale = 1, width = 400, height = 90, units = "mm", bg = "white", dpi = 1200
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/volcanoes/volcanoes_glob_explorate_re_1.svg"),
  plot = volcanoes_glob_explorate_re_1,
  scale = 1, width = 400, height = 90, units = "mm", bg = "white", fix_text_size = FALSE
)

```

Set 2

```{r}

glob_explorate_res_re_2 <- all_diffexp_results %>%
    filter(quant_tool == "ExplorATE",
           feature_biotype == "repetitive_element",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                          "adult_blastemas_vs_adult_nonlimbs",
                          "subadult_blastemas_vs_subadult_nonlimbs",
                          "adult_limbs_vs_adult_nonlimbs",
                          "subadult_limbs_vs_subadult_nonlimbs",
                          "age_effect_blastemas_vs_nonlimbs",
                          "age_effect_limbs_vs_nonlimbs")
)

# split gene_id column into three columns, based on the ":" separator, rename: c("RE_subfam", "RE_superfam", "RE_class")
glob_explorate_res_re_2 <- glob_explorate_res_re_2 %>%
  tidyr::separate(ID, c("RE_subfam", "RE_superfam", "RE_class"), sep = ":", remove = FALSE) %>%
  # if RE_subfam starts with "rnd-", replace with "Unknown"
  mutate(RE_superfam = ifelse(grepl("^rnd-", RE_superfam), "Unknown", RE_superfam)) %>%
  # replace Gypsy with Ty3, not necessarilly the full word
  mutate(RE_superfam = gsub("Gypsy", "Ty3", RE_superfam)) %>%
  # remove "?" in RE_class
  mutate(RE_class = gsub("\\?", "", RE_class))

# add shape, color and fill columns
glob_explorate_res_re_2 <- glob_explorate_res_re_2 %>%
  mutate(shape = ifelse(log2FoldChange_MMSE < -log2(1.5), 25, ifelse(log2FoldChange_MMSE > log2(1.5), 24, 21)),
  # for each unique value in RE_class, assign a color from a palette
        color = factor(RE_class, levels = sort(unique(RE_class)), labels = brewer.pal(n = length(unique(RE_class)), name = "Dark2")))

# Associating colors to each type of value
# convert to named vector
keyvals <- setNames(as.character(glob_explorate_res_re_2$color), as.character(glob_explorate_res_re_2$RE_class))
shapevals <- setNames(glob_explorate_res_re_2$shape, glob_explorate_res_re_2$shape)

# Creating volcano plot
volcanoes_glob_explorate_re_2 <- EnhancedVolcano(

  as.data.frame(glob_explorate_res_re_2),
  lab = NA,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +

  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +
 
  facetted_pos_scales(
  y = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 2.6))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_explorate_res_re_2)$padj_MLE)[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))),
  x = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_explorate_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_explorate_res_re_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))))) +

  force_panelsizes(rows = unit(2, "in"), cols = unit(2, "in"))


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/volcanoes/volcanoes_glob_explorate_re_2.png"),
  plot = volcanoes_glob_explorate_re_2,
  scale = 1, width = 55, height = 9, units = "cm", bg = "white", dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/volcanoes/volcanoes_glob_explorate_re_2.svg"),
  plot = volcanoes_glob_explorate_re_2,
  scale = 1, width = 55, height = 9, units = "cm", bg = "white", fix_text_size = FALSE
)

```

## STAR and TEcount pipeline

### Mapping reads to the genome

#### Generating genome index

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/STAR/genome_index/
mkdir -p $PROJECT_DIR/rnaseq/STAR/alignments/
mkdir -p $PROJECT_DIR/rnaseq/STAR/multiqc/

# Generating genome index with STAR
nohup STAR \
      --runThreadN $NTHREADS \
      --limitGenomeGenerateRAM=$RAM_LIMIT \
      --runMode genomeGenerate \
      --genomeDir $PROJECT_DIR/rnaseq/STAR/genome_index/ \
      --genomeFastaFiles $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.fa \
      --sjdbGTFfile $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/AmexT_v47-AmexG_v6.0-DD.gtf \
> $PROJECT_DIR/rnaseq/STAR/genome_index/genomeGenerate_star.log 2>&1 &

``` 
  
#### Aligning reads to the genome

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/STAR/genome_index/tmp
cd $PROJECT_DIR/rnaseq/STAR/genome_index/tmp

# Loading STAR index into memory for the first time
STAR \
  --genomeLoad LoadAndExit \
  --genomeDir $PROJECT_DIR/rnaseq/STAR/genome_index/

# get list of fastq files
#list_fastqs=($(ls $PROJECT_DIR/rnaseq/raw/fastq/*.fastq.gz | sed 's/_1.fastq.gz//g' | sed 's/_2.fastq.gz//g' | sort -u))
list_fastqs=($(ls $PROJECT_DIR/rnaseq/raw/fastq/*.fastq.gz | sed 's/_R1.fastq.gz//g' | sed 's/_R2.fastq.gz//g' | sort -u))

# run alignment script in parallel
parallel \
  --keep-order --line-buffer --jobs $NJOBS \
  $PROJECT_DIR/axoloTE/scripts/star_align_rnaseq.sh {} \
  $PROJECT_DIR/rnaseq/STAR/genome_index/ \
  $NTHREADS \
  $RAM_LIMIT \
  $PROJECT_DIR/rnaseq/STAR/alignments/ \
  ::: ${list_fastqs[@]}

# Removing STAR index from memory
cd $PROJECT_DIR/rnaseq/STAR/genome_index/tmp
STAR \
  --genomeLoad Remove \
  --genomeDir $PROJECT_DIR/rnaseq/STAR/genome_index/

# Generating MultiQC report
multiqc --interactive \
  $PROJECT_DIR/rnaseq/STAR/alignments/ \
  --filename aligned_multiqc \
  --outdir $PROJECT_DIR/rnaseq/STAR/multiqc/

```

### TEcount: Gene and RE subfamily quantification 

Downloading and installing TEtranscripts (TEcount)

```{bash}
cd $SOFTWARE_DIR
wget https://github.com/mhammell-laboratory/TEtranscripts/archive/refs/tags/2.2.3.tar.gz
tar -xvf 2.2.3.tar.gz
rm 2.2.3.tar.gz
cd TEtranscripts-2.2.3/

pip install -e $SOFTWARE_DIR/TEtranscripts-2.2.3

python setup.py install --prefix $HOME_DIR
  
```
  
#### Prebuilding TEcount gene and RE indexes

```{bash}
cd $SOFTWARE_DIR
wget https://github.com/mhammell-laboratory/TEtranscripts/files/6118032/TEtranscripts_indexer.gz
#https://github.com/mhammell-laboratory/TEtranscripts/issues/87#issuecomment-795863539
gunzip TEtranscripts_indexer.gz
chmod +x TEtranscripts_indexer

```

```{bash}
mkdir -p $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/TEcount_GENE_index
ln -s $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/AmexT_v47-AmexG_v6.0-DD.gtf $PROJECT_DIR/rnaseq/STAR/TEcount/indexes/TEcount_GENE_index/

cd $PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD/TEcount_GENE_index
nohup $SOFTWARE_DIR/TEtranscripts_indexer \
  --afile $PROJECT_DIR/rnaseq/STAR/TEcount/indexes/TEcount_GENE_index/AmexT_v47-AmexG_v6.0-DD.gtf \
  --itype gene \
  --verbose 3 \
  > $PROJECT_DIR/rnaseq/STAR/TEcount/indexes/TEcount_GENE_index/tecount_gene_index.log 2>&1 &

mkdir -p $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/TEcount_TE_index
ln -s $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/ambMex60DD_ucsc_hub.ALL.mod.gtf $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/TEcount_TE_index/

cd $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/TEcount_TE_index
nohup $SOFTWARE_DIR/TEtranscripts_indexer \
  --afile $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/TEcount_TE_index/ambMex60DD_ucsc_hub.ALL.mod.gtf \
  --itype TE \
  --verbose 3 \
  > $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/TEcount_TE_index/tecount_te_index.log 2>&1 &

```

To determine the expected format (strandedness) of the libraries, we will quickly read the JSON files generated during the Salmon quantification step.
The expected format will be used as an input for the TEcount run.

```{bash}
# Iterate through subdirectories
for subdirectory in $PROJECT_DIR/rnaseq/Salmon/quant/*; do
    if [ -d "$subdirectory" ]; then
        json_file="$subdirectory/lib_format_counts.json"
        
        # Check if the JSON file exists
        if [ -f "$json_file" ]; then
            # Extract "expected_format" value from JSON using jq (make sure jq is installed)
            expected_format=$(jq -r '.expected_format' "$json_file")
            
            # Print the value
            echo "Subdirectory: $subdirectory"
            echo "Expected Format: $expected_format"
            echo
        fi
    fi
done

```

All libraries were unstranded (an unstranded paired-end library where the reads face each other)

#### Running TEcount

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/STAR/TEcount/quant
cd $PROJECT_DIR/rnaseq/STAR/TEcount/quant

# get list of bam files
list_rnabams=($(ls $PROJECT_DIR/rnaseq/STAR/alignments/*.nsort.bam))

chmod +x $PROJECT_DIR/axoloTE/scripts/tecount_quant.sh

parallel \
  --keep-order --line-buffer --jobs $NTHREADS \
  $PROJECT_DIR/axoloTE/scripts/tecount_quant.sh {} \
  $PROJECT_DIR/rnaseq/STAR/TEcount/indexes/TEcount_GENE_index/AmexT_v47-AmexG_v6.0-DD.gtf.ind \
  $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/TEcount_TE_index/ambMex60DD_ucsc_hub.ALL.mod.gtf.ind \
  no \
  multi \
  $PROJECT_DIR/rnaseq/STAR/TEcount/quant/ \
  ::: ${list_rnabams[@]}

```

#### Importing the count matrices into R

```{r}
PROJECT_DIR <- "../sruizperez/projects/axolotl/"

```

```{r}
library(readr)
library(plyr)
library(dplyr)

counts_list <- list.files(path = file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/quant/"), pattern = ".cntTable$", full.names = TRUE)

sample_names <- gsub(".cntTable", "", counts_list) # remove the extension
sample_names <- gsub(".*/", "", sample_names) # remove the path

list.data <- list()

for (i in seq_along(counts_list)) {
  table <- read_tsv(counts_list[i])
  colnames(table) <- c("feature_id", sample_names[i])
  list.data[[i]] <- table
}

names(list.data) <- sample_names

count_matrix <- join_all(list.data, by = "feature_id", type = "full")

rownames(count_matrix) <- count_matrix$feature_id
count_matrix$feature_id <- NULL

write.table(count_matrix,
  sep = "\t",
  file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/quant/cntTable_all.tsv")
)

```

#### Generating bigwig files

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/STAR/TEcount/bigwigs

```

```{r}
count_matrix <- read.table(file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/quant/cntTable_all.tsv"),
  header = TRUE,
  sep = "\t",
  row.names = 1,
  check.names = FALSE # some sample names may start with a number
)

# Calculate scaling factors with edgeR following:
# https://www.biostars.org/p/413626/#414440
# https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html

library(edgeR)

NormFactor <- calcNormFactors(
  object = count_matrix,
  method = "TMM"
) # for DESeq2 strategy use method="RLE" instead

## raw library size:
LibSize <- colSums(count_matrix)

## calculate size factors:
SizeFactors <- NormFactor * LibSize / 1000000

## Reciprocal, please read section below:
SizeFactors.Reciprocal <- 1 / SizeFactors

## save the size factors to a file:
write.table(SizeFactors.Reciprocal,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/bigwigs/scaling_factors.tsv"),
  sep = "\t",
  quote = FALSE,
  col.names = FALSE
)

```

```{bash}
# run bamCoverage using the first column of scaling_factors.txt as input and the second as --scaleFactor

parallel \
  --keep-order --line-buffer --jobs $NJOBS \
  --link bamCoverage \
            --numberOfProcessors 8 \
            --verbose \
            -b $PROJECT_DIR/rnaseq/STAR/alignments/{}.Aligned.sortedByCoord.out.bam \
            --scaleFactor {2} \
            -o $PROJECT_DIR/rnaseq/STAR/TEcount/bigwigs/{1}.bw \
            ::: $(cut -f 1,2 $PROJECT_DIR/rnaseq/STAR/TEcount/bigwigs/scaling_factors.tsv)

```

#### Native axolotl samples analysis

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/STAR/TEcount/native/DEA

```

##### Setting DESeq2 model for previsualization

```{r} 
library(DESeq2)

nat_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"),
  header = TRUE,
  sep = "\t"
)

count_matrix <- read.table(file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/quant/cntTable_all.tsv"),
  header = TRUE,
  sep = "\t",
  row.names = 1,
  check.names = FALSE # some sample names may start with a number
)

# filtering to remove other datasets and keep only native
nat_tecount_counts <- count_matrix[, rownames(nat_metadata)]

all(rownames(nat_metadata) %in% colnames(nat_tecount_counts))
all(rownames(nat_metadata) == colnames(nat_tecount_counts))

nat_tecount_dds_raw <- DESeqDataSetFromMatrix(
  countData = nat_tecount_counts,
  colData = nat_metadata,
  design = ~ Age_group + Tissue_group
)

# Explicitly setting the factors levels
nat_tecount_dds_raw$Age_group <- relevel(nat_tecount_dds_raw$Age_group, ref = "Sub_adult")
nat_tecount_dds_raw$Tissue_group <- relevel(nat_tecount_dds_raw$Tissue_group, ref = "Limb")

saveRDS(nat_tecount_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_tecount_dds_raw.rds"))
# nat_tecount_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_tecount_dds_raw.rds"))
```

##### Prefiltering

```{r}
summary(rowSums(counts(nat_tecount_dds_raw) >= 10) >= 3)
#   Mode   FALSE    TRUE
# logical   76588   25456
```

##### Pre-visualization

###### Variance Stabilizing Transformation (VST)

```{r}
library(vsn)
library(ggplot2)
library(scales)
library(DESeq2)

nat_tecount_dds_vst <- vst(nat_tecount_dds_raw, blind = FALSE)

saveRDS(nat_tecount_dds_vst, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_tecount_dds_vst.rds"))
# nat_tecount_dds_vst <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_tecount_dds_vst.rds"))
```

###### cPCA

```{r}
library(cowplot)
library(ggpubr)
library(ggrepel)
library(ggh4x)
```

Getting underlying data:
```{r}
nat_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"),
  header = TRUE,
  sep = "\t"
)

nat_tecount_pcaData <- plotPCA(nat_tecount_dds_vst, intgroup = c("Age_group", "Tissue_group"), returnData = TRUE)
nat_tecount_percentVar <- round(100 * attr(nat_tecount_pcaData, "percentVar"))

merged <- merge(nat_tecount_pcaData, nat_metadata, by.x = "name", by.y = "Run", all.x = TRUE)
merged <- merged[order(merged$name), ]

nat_tecount_pcaData <- nat_tecount_pcaData[order(row.names(nat_tecount_pcaData)), ]
nat_tecount_pcaData$new_name <- merged$new_name_run

library(dplyr)
library(readr)

nat_tecount_pcaData <- as_tibble(nat_tecount_pcaData)

write_tsv(nat_tecount_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/previsualization/nat_tecount_pcaData.tsv"))
```

Plotting:
```{r}
nat_tecount_pca_main <- ggplot(nat_tecount_pcaData, aes(PC1, PC2, color = Tissue_group, fill = Tissue_group, shape = Age_group, label = new_name)) +
  geom_point(size = 3.5) +
  scale_shape_manual(
    name = "Age group", values = c(24, 22),
    breaks = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"),
    guide = guide_legend(order = 1, label.theme = element_text(size = 9))
  ) +
  scale_color_manual(
    name = "Tissue", values = c("#4dac26", "#d01c8b"),
    breaks = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema")
  ) +
  scale_fill_manual(
    name = "Tissue", values = c("#4dac26", "#d01c8b"),
    breaks = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema")
  ) +
  xlab(paste0("PC1: ", nat_tecount_percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", nat_tecount_percentVar[2], "% variance")) +
  theme_bw() +
  theme(
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.4),
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  guides(fill = guide_legend(order = 2), color = guide_legend(order = 2)) +
  geom_text_repel(
    show.legend = FALSE,
    color = "black", size = 3,
    label.padding = 0.5,
    box.padding = 0.3
  ) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  )

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/previsualization/nat_tecount_cPCA.png"),
  plot = nat_tecount_pca_main,
  device = "png", path = NULL, scale = 1,
  width = 150, height = 90, units = "mm",
  bg = "white", dpi = 1200
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/previsualization/nat_tecount_cPCA.svg"),
  plot = nat_tecount_pca_main,
  device = "svg", path = NULL, scale = 1,
  width = 150, height = 90, units = "mm",
  bg = "white"
)
```

###### Heatmap of sample clustering

```{r}
library(ComplexHeatmap)
library(viridisLite)
library(RColorBrewer)
library(DESeq2)

# Extract distance matrix
nat_tecount_cor_dist_mat <- (1 - cor(assay(nat_tecount_dds_vst), method = "pearson"))

write.table(nat_tecount_cor_dist_mat,
  sep = "\t",
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/previsualization/nat_tecount_cor_dist_mat.tsv")
)

# Prepare heatmap annotations
annotation <- as.data.frame(colData(nat_tecount_dds_vst))
annotation <- annotation[c("Age_group", "Tissue_group")]

# This changes some of the global options of ComplexHeatmap to be able to add a whitespace to the left of a merged legend.
# Should be reset after generating all heatmaps
# ht_opt(HEATMAP_LEGEND_PADDING = unit(c(6,0,0,0), "mm"), ANNOTATION_LEGEND_PADDING = unit(c(6,0,0,0), "mm"))
# ht_opt(RESET = TRUE)

column_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b"
    )
  ),
  annotation_label = gt_render(c("Age group", "Tissue")),
  annotation_legend_param = list(
    Age_group = list(
      title = "Age group",
      at = c("Adult", "Sub_adult"),
      labels = c("Adult (8 years)", "Sub-adult (8 months)"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    ),
    Tissue_group = list(
      title = "Tissue",
      at = c("Limb", "Limb_blastema"),
      labels = c("Limb", "Limb blastema"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    )
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

# Create ComplexHeatmap
# Color palette copied from: https://www.rdocumentation.org/packages/pheatmap/versions/1.0.12/topics/pheatmap
nat_tecount_sample_dist_heatmap <- Heatmap(nat_tecount_cor_dist_mat,
  name = "Distance",
  clustering_distance_rows = function(m) as.dist(m),
  clustering_distance_columns = function(m) as.dist(m),
  row_dend_side = "right",
  row_names_side = "left",
  top_annotation = column_ha,
  row_title = "RNA-seq run", row_title_rot = 90, row_title_gp = gpar(fontface = "bold", fontsize = 11),
  column_title = "RNA-seq run", column_title_side = "bottom", column_title_gp = gpar(fontface = "bold", fontsize = 11),
  row_names_gp = gpar(fontsize = 8.5),
  column_names_gp = gpar(fontsize = 8.5),
  row_dend_width = unit(0.6, "cm"),
  column_dend_height = unit(0.6, "cm"),
  col = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
  width = unit(3, "in"), height = unit(2.6, "in"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.5),
  heatmap_legend_param = list(
    legend_height = unit(3, "cm"),
    labels_gp = gpar(fontsize = 9),
    title_gp = gpar(
      fontsize = 11,
      fontface = "bold"
    )
  ),
  cell_fun = function(j, i, x, y, width, height, fill) {
    if (nat_tecount_cor_dist_mat[i, j] != 0) {
      ifelse(nat_tecount_cor_dist_mat[i, j] < 0.22, grid.text(sprintf("%.2f", nat_tecount_cor_dist_mat[i, j]), x, y, gp = gpar(fontsize = 7, col = "gray55")), grid.text(sprintf("%.2f", nat_tecount_cor_dist_mat[i, j]), x, y, gp = gpar(fontsize = 8.5, col = "gray85")))
    }
  },
)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/previsualization/nat_tecount_sample_dist_heatmap.png"),
  width = 18, height = 12.7, units = "cm", res = 1200
)

draw(nat_tecount_sample_dist_heatmap, merge_legends = TRUE)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/previsualization/nat_tecount_sample_dist_heatmap.svg"),
  width = 7.08661, height = 5
)

draw(nat_tecount_sample_dist_heatmap, merge_legends = TRUE)

dev.off()
```

##### Differential expression analysis (DEA)

###### Adult limbs vs Sub-adult limbs

```{r}
library(DESeq2)

nat_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"),
  header = TRUE,
  sep = "\t"
)

count_matrix <- read.table(file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/quant/cntTable_all.tsv"),
  header = TRUE,
  sep = "\t",
  row.names = 1,
  check.names = FALSE # some sample names may start with a number
)

# Only keep rows where Tissue_group is Limb
nat_age_metadata <- nat_metadata[nat_metadata$Tissue_group == "Limb", ]

# replace "-" in new_name_run with "_"
nat_age_metadata$new_name_run <- gsub("-", "_", nat_age_metadata$new_name_run)

rownames(nat_age_metadata) <- nat_age_metadata$new_name_run

# filtering to remove other datasets and keep only native
nat_age_tecount_counts <- count_matrix[, rownames(nat_age_metadata)]

all(rownames(nat_age_metadata) %in% colnames(nat_age_tecount_counts))
all(rownames(nat_age_metadata) == colnames(nat_age_tecount_counts))

nat_age_tecount_dds_raw <- DESeqDataSetFromMatrix(
  countData = nat_age_tecount_counts,
  colData = nat_age_metadata,
  design = ~ Age_group
)

# Explicitly setting the factors levels
nat_age_tecount_dds_raw$Age_group <- relevel(nat_age_tecount_dds_raw$Age_group, ref = "Sub_adult")

saveRDS(nat_age_tecount_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_age_tecount_dds_raw.rds"))
# nat_age_tecount_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_age_tecount_dds_raw.rds"))
```

####### Running DESeq

```{r}
nat_age_tecount_dds_de <- DESeq(nat_age_tecount_dds_raw)

saveRDS(nat_age_tecount_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_age_tecount_dds_de.rds"))
```

####### Extracting contrast tables

```{r}
library("IHW")
library(ashr)
library(dplyr)
library(readr)

resultsNames(nat_age_tecount_dds_de)
```

```{r}
nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- results(nat_age_tecount_dds_de,
  filterFun = ihw,
  contrast = c("Age_group", "Adult", "Sub_adult"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(nat_age_tecount_res_adult_limbs_vs_subadult_limbs)
# out of 53526 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 23, 0.043%
# LFC < -0.58 (down) : 52, 0.097%
# outliers [1]       : 530, 0.99%

nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink <- lfcShrink(nat_age_tecount_dds_de,
  res = nat_age_tecount_res_adult_limbs_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink)
# out of 53526 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 25, 0.047%
# LFC < -0.58 (down) : 68, 0.13%

# Saving all columns (MLE and posterior LFC) in the same dataframe
colnames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs) <- paste0(colnames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs), "_MLE")
colnames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink) <- paste0(colnames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink), "_MMSE")
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink$log2FoldChange_MMSE
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$lfcSE_MMSE <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink$lfcSE_MMSE
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$svalue_MMSE <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol[is.na(nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol)] <- rownames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs)[is.na(nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol)]
# add metadata columns
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$quant_tool <- "TEcount"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$diffexp_contrast <- "adult_limbs_vs_subadult_limbs"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$sample_group <- "native"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(nat_age_tecount_res_adult_limbs_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(nat_age_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) | abs(nat_age_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(nat_age_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
nat_age_tecount_res_adult_limbs_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(nat_age_tecount_res_adult_limbs_vs_subadult_limbs$padj_MLE) | nat_age_tecount_res_adult_limbs_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- as_tibble(nat_age_tecount_res_adult_limbs_vs_subadult_limbs, rownames = "gene_id")

write_tsv(nat_age_tecount_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_age_tecount_res_adult_limbs_vs_subadult_limbs.tsv")
)

# nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_age_tecount_res_adult_limbs_vs_subadult_limbs.tsv"))

```

####### Sub-adult blastemas vs Sub-adult limbs

```{r}
library(DESeq2)

nat_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"),
  header = TRUE,
  sep = "\t"
)

count_matrix <- read.table(file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/quant/cntTable_all.tsv"),
  header = TRUE,
  sep = "\t",
  row.names = 1,
  check.names = FALSE # some sample names may start with a number
)

# Only keep rows where Tissue_group is Limb
nat_reg_metadata <- nat_metadata[nat_metadata$Age_group == "Sub_adult", ]

# replace "-" in new_name_run with "_"
nat_reg_metadata$new_name_run <- gsub("-", "_", nat_reg_metadata$new_name_run)

rownames(nat_reg_metadata) <- nat_reg_metadata$new_name_run

# filtering to remove other datasets and keep only native
nat_reg_tecount_counts <- count_matrix[, rownames(nat_reg_metadata)]

all(rownames(nat_reg_metadata) %in% colnames(nat_reg_tecount_counts))
all(rownames(nat_reg_metadata) == colnames(nat_reg_tecount_counts))

nat_reg_tecount_dds_raw <- DESeqDataSetFromMatrix(
  countData = nat_reg_tecount_counts,
  colData = nat_reg_metadata,
  design = ~ Tissue_group
)

# Explicitly setting the factors levels
nat_reg_tecount_dds_raw$Tissue_group <- relevel(nat_reg_tecount_dds_raw$Tissue_group, ref = "Limb")

saveRDS(nat_reg_tecount_dds_raw, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_reg_tecount_dds_raw.rds"))
# nat_reg_tecount_dds_raw <- readRDS(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_reg_tecount_dds_raw.rds"))
```

####### Running DESeq

```{r}
nat_reg_tecount_dds_de <- DESeq(nat_reg_tecount_dds_raw)

saveRDS(nat_reg_tecount_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_reg_tecount_dds_de.rds"))
```

####### Extracting contrast tables

```{r}
library("IHW")
library(ashr)
library(dplyr)
library(readr)

resultsNames(nat_reg_tecount_dds_de)
```

```{r}
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- results(nat_reg_tecount_dds_de,
  filterFun = ihw,
  contrast = c("Tissue_group", "Limb_blastema", "Limb"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs)
# out of 54660 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 195, 0.36%
# LFC < -0.58 (down) : 1089, 2%
# outliers [1]       : 657, 1.2%

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink <- lfcShrink(nat_reg_tecount_dds_de,
  res = nat_reg_tecount_res_sa_blastemas_vs_sa_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink)
# out of 54660 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 206, 0.38%
# LFC < -0.58 (down) : 958, 1.8%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs) <- paste0(colnames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs), "_MLE")
colnames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink) <- paste0(colnames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink), "_MMSE")
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink$log2FoldChange_MMSE
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$lfcSE_MMSE <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink$lfcSE_MMSE
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$svalue_MMSE <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_symbol[is.na(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_symbol)] <- rownames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs)[is.na(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_symbol)]
# add metadata columns
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$quant_tool <- "TEcount"
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_limbs"
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$sample_group <- "native"
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$diffexp_res_lfc <- ifelse(is.na(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE) | abs(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$diffexp_res_padj <- ifelse(is.na(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$padj_MLE) | nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$padj_MLE > 0.05, "not_significant", "significant")

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- as_tibble(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs, rownames = "gene_id")

print(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs, width = Inf)

readr::write_tsv(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_reg_tecount_res_sa_blastemas_vs_sa_limbs.tsv")
)

# nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_reg_tecount_res_sa_blastemas_vs_sa_limbs.tsv")

```

###### Volcano plots

```{r}
dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/volcanoes"))
```

####### Genes

```{r}

#nat_reg_tecount_res_sa_blastemas_vs_sa_limbs
#nat_age_tecount_res_adult_limbs_vs_subadult_limbs

# keep only rows where rowname starts with "AMEX"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_genes <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs %>%
  filter(grepl("^AMEX", gene_id))

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_genes <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs %>%
  filter(grepl("^AMEX", gene_id))

nat_tecount_res_all_genes <- bind_rows(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_genes, nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_genes)

# remove rows ehere padj_MLE is NA or log2FoldChange_MMSE is NA
nat_tecount_res_all_genes <- nat_tecount_res_all_genes %>%
  filter(!is.na(padj_MLE) & !is.na(log2FoldChange_MMSE))

```

```{r}
library(ggh4x)
library(EnhancedVolcano)

# Associating colors to each type of value
keyvals <- ifelse(nat_tecount_res_all_genes$padj_MLE > 0.05 & nat_tecount_res_all_genes$log2FoldChange_MMSE > log2(1.5),
  "#c5c5c5",
  ifelse(nat_tecount_res_all_genes$padj_MLE > 0.05 & nat_tecount_res_all_genes$log2FoldChange_MMSE < -log2(1.5),
    "#c5c5c5",
    ifelse(nat_tecount_res_all_genes$padj_MLE > 0.05 & abs(nat_tecount_res_all_genes$log2FoldChange_MMSE) < log2(1.5),
      "#c5c5c5",
      ifelse(nat_tecount_res_all_genes$padj_MLE < 0.05 & abs(nat_tecount_res_all_genes$log2FoldChange_MMSE) < log2(1.5),
        "#c5c5c5",
        ifelse(nat_tecount_res_all_genes$padj_MLE < 0.05 & nat_tecount_res_all_genes$log2FoldChange_MMSE < -log2(1.5),
          "#0072B2",
          "#D55E00"
        )
      )
    )
  )
)

keyvals[is.na(keyvals)] <- "#c5c5c5"

names(keyvals)[keyvals == "#c5c5c5"] <- "up_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "mid_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "down_NS"
names(keyvals)[keyvals == "#0072B2"] <- "down_S"
names(keyvals)[keyvals == "#D55E00"] <- "up_S"
names(keyvals)[keyvals == "#c5c5c5"] <- "middle_S"

shapevals <- keyvals
names(shapevals) <- keyvals

shapevals <- ifelse(keyvals == "#c5c5c5", 21,
      ifelse(keyvals == "#0072B2", 25,
        ifelse(keyvals == "#D55E00", 24, 21)
      )
  )

# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- nat_tecount_res_all_genes %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE < -log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(gene_symbol)

selectGenes1_2 <- nat_tecount_res_all_genes %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE > log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(gene_symbol)

selectGenes2 <- nat_tecount_res_all_genes %>%
                  group_by(diffexp_contrast) %>%
                  slice_min(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(gene_symbol)

selectGenes3 <- nat_tecount_res_all_genes %>%
                  group_by(diffexp_contrast) %>%
                  slice_max(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(gene_symbol)

selectGenes <- unique(c(selectGenes1, selectGenes1_2, selectGenes2, selectGenes3))


#if element starts with "AMEX", remove from selectGenes
selectGenes <- selectGenes[!grepl("^AMEX", selectGenes)]
selectGenes <- selectGenes[!grepl("^LOC", selectGenes)]


# Creating volcano plot
volcanoes_nat_tecount_gene <- EnhancedVolcano(

  as.data.frame(nat_tecount_res_all_genes),
  lab = as.data.frame(nat_tecount_res_all_genes)$gene_symbol,
  selectLab = selectGenes,
  drawConnectors = TRUE,
  widthConnectors = 0.3,
  arrowheads = FALSE,
  maxoverlapsConnectors = Inf,
  labSize = 2,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +


  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +


  # change y scale of first plot, keep the second plot as it is
  
  facetted_pos_scales(
  y = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_tecount_res_all_genes)$padj_MLE)[as.data.frame(nat_tecount_res_all_genes)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_tecount_res_all_genes)$padj_MLE)[as.data.frame(nat_tecount_res_all_genes)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(nat_tecount_res_all_genes)$log2FoldChange_MMSE[as.data.frame(nat_tecount_res_all_genes)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(nat_tecount_res_all_genes)$log2FoldChange_MMSE[as.data.frame(nat_tecount_res_all_genes)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.11, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/volcanoes/volcanoes_nat_tecount_gene.png"),
  plot = volcanoes_nat_tecount_gene,
  scale = 1, width = 200, height = 72, units = "mm", bg = "white", dpi = 1200
)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/volcanoes/volcanoes_nat_tecount_gene.svg"),
  width = 7.87402, height = 2.83465
)

volcanoes_nat_tecount_gene

dev.off()
```

####### Repetitive elements (REs)

```{r}
# keep only rows where rowname starts with "AMEX"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs %>%
  filter(!grepl("^AMEX", gene_id))

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs %>%
  filter(!grepl("^AMEX", gene_id))

nat_tecount_res_all_re <- bind_rows(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re, nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re)

# remove rows ehere padj_MLE is NA or log2FoldChange_MMSE is NA
nat_tecount_res_all_re <- nat_tecount_res_all_re %>%
  filter(!is.na(padj_MLE) & !is.na(log2FoldChange_MMSE))

```

```{r}
library(ggh4x)
library(EnhancedVolcano)
library(tidyr)
library(RColorBrewer)

# split gene_id column into three columns, based on the ":" separator, rename: c("RE_subfam", "RE_superfam", "RE_class")
nat_tecount_res_all_re <- nat_tecount_res_all_re %>%
  tidyr::separate(ID, c("RE_subfam", "RE_superfam", "RE_class"), sep = ":", remove = FALSE) %>%
  # if RE_subfam starts with "rnd-", replace with "Unknown"
  mutate(RE_superfam = ifelse(grepl("^rnd-", RE_superfam), "Unknown", RE_superfam)) %>%
  # replace Gypsy with Ty3, not necessarilly the full word
  mutate(RE_superfam = gsub("Gypsy", "Ty3", RE_superfam)) %>%
  # remove "?" in RE_class
  mutate(RE_class = gsub("\\?", "", RE_class))


# get unique values of RE_superfam where abs(log2FoldChange_MMSE) > log2(1.5) and padj_MLE < 0.05
unique(nat_tecount_res_all_re$RE_superfam[abs(nat_tecount_res_all_re$log2FoldChange_MMSE) > log2(1.5)])

# get unique values of RE_superfam where RE_class is "LTR"
unique(nat_tecount_res_all_re$RE_superfam[nat_tecount_res_all_re$RE_class == "LTR"])


# add shape, color and fill columns
nat_tecount_res_all_re <- nat_tecount_res_all_re %>%
  mutate(shape = ifelse(log2FoldChange_MMSE < -log2(1.5), 25, ifelse(log2FoldChange_MMSE > log2(1.5), 24, 21)),
  # for each unique value in RE_class, assign a color from a palette
        color = factor(RE_class, levels = sort(unique(RE_class)), labels = brewer.pal(n = length(unique(RE_class)), name = "Dark2")))

# Associating colors to each type of value
# convert to named vector
keyvals <- setNames(as.character(nat_tecount_res_all_re$color), as.character(nat_tecount_res_all_re$RE_class))
shapevals <- setNames(nat_tecount_res_all_re$shape, nat_tecount_res_all_re$shape)


# Creating volcano plot
volcanoes_nat_tecount_re <- EnhancedVolcano(

  as.data.frame(nat_tecount_res_all_re),
  lab = NA,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +

  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +


  # change y scale of first plot, keep the second plot as it is
  
  facetted_pos_scales(
  y = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_tecount_res_all_re)$padj_MLE)[as.data.frame(nat_tecount_res_all_re)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15), add = c(0, 0.5))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(nat_tecount_res_all_re)$padj_MLE)[as.data.frame(nat_tecount_res_all_re)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "adult_limbs_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(nat_tecount_res_all_re)$log2FoldChange_MMSE[as.data.frame(nat_tecount_res_all_re)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1), add = c(0, 1))),
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(nat_tecount_res_all_re)$log2FoldChange_MMSE[as.data.frame(nat_tecount_res_all_re)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )

# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/volcanoes/volcanoes_nat_tecount_re.png"),
  plot = volcanoes_nat_tecount_re,
  scale = 1, width = 200, height = 90, units = "mm", bg = "white", dpi = 1200
)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/volcanoes/volcanoes_nat_tecount_re.svg"),
  width = 6.29921, height = 2.83465
)

volcanoes_nat_tecount_re

dev.off()
```

###### Heatmap of gene/RE clustering

```{r}
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
library(gridtext)
library(dplyr)
library(readr)
library(DESeq2)

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/heatmap/"))

nat_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/nat_metadata.txt"), header = TRUE, sep = "\t")

nat_tecount_dds_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/nat_tecount_dds_vst.rds"))


nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_age_tecount_res_adult_limbs_vs_subadult_limbs.tsv"))
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_reg_tecount_res_sa_blastemas_vs_sa_limbs.tsv"))

nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs %>%
  filter(!is.na(log2FoldChange_MMSE) & !is.na(padj_MLE))

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs %>%
  filter(!is.na(log2FoldChange_MMSE) & !is.na(padj_MLE))

# remove genes in nat_age_tecount_res_adult_limbs_vs_subadult_limbs not present in nat_reg_tecount_res_sa_blastemas_vs_sa_limbs
nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs %>%
  filter(gene_id %in% nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_id)

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs %>%
  filter(gene_id %in% nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_id)

# keep only rows where rowname starts with "AMEX"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs %>%
  filter(!grepl("^AMEX", gene_id))

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs %>%
  filter(!grepl("^AMEX", gene_id))

nat_tecount_res_all_re <- bind_rows(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re, nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re)

# remove rows ehere padj_MLE is NA or log2FoldChange_MMSE is NA
nat_tecount_res_all_re <- nat_tecount_res_all_re %>%
  filter(!is.na(padj_MLE) & !is.na(log2FoldChange_MMSE))

```

```{r}
# This changes some of the global options of ComplexHeatmap to be able to add a whitespace to the left of a merged legend.
# Should be reset after generating all heatmaps
ht_opt(HEATMAP_LEGEND_PADDING = unit(c(7.5, 0, 0, 0), "cm"), ANNOTATION_LEGEND_PADDING = unit(c(6, 0, 0, 0), "mm"))

#nat_reg_tecount_res_sa_blastemas_vs_sa_limbs
#nat_age_tecount_res_adult_limbs_vs_subadult_limbs

# Filter by DE and significance and reorder
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_top <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs %>%
  filter(padj_MLE < 0.05 & abs(log2FoldChange_MMSE) > log2(1.5)) %>%
  arrange(desc(log2FoldChange_MMSE))

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_top <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs %>%
  filter(padj_MLE < 0.05 & abs(log2FoldChange_MMSE) > log2(1.5)) %>%
  arrange(desc(log2FoldChange_MMSE))

# Keep top and bottom genes from both contrasts
num_keep <- 30

genes_keep_1 <- c(seq(1:num_keep), seq((nrow(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_top) - num_keep + 1), nrow(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_top)))
genes_ids_1 <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs_top$gene_id[genes_keep_1]

genes_keep_2 <- c(seq(1:num_keep), seq((nrow(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_top) - num_keep + 1), nrow(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_top)))
genes_ids_2 <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_top$gene_id[genes_keep_2]

# Adding differentially expressed REs even if not among top DE
# remove rows where padj_MLE or log2FoldChange_MMSE is NA
re_keep_1 <- nat_tecount_res_all_re[!is.na(nat_tecount_res_all_re$padj_MLE) & !is.na(nat_tecount_res_all_re$log2FoldChange_MMSE), ]

# re_keep_1 are rows of significant REs in adults vs subadults
re_keep_1 <- unique(re_keep_1$gene_id[re_keep_1$padj_MLE < 0.05 & abs(re_keep_1$log2FoldChange_MMSE) > log2(1.5)])

# Merging all gene/RE ids
ids <- unique(c(genes_ids_1, genes_ids_2, re_keep_1))

# Get vst values
mat <- assay(nat_tecount_dds_vst)[ids, ]
mat <- mat - rowMeans(mat)

# replace colnames(mat) by matching with rownames(nat_metadata) and replacing with nat_metadata$new_name_run
colnames(mat) <- nat_metadata$new_name_run[match(colnames(mat), rownames(nat_metadata))]

# Get lfc and basemean values
lfc_vals_a_vs_sa <- as.matrix(nat_age_tecount_res_adult_limbs_vs_subadult_limbs[match(ids, nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_id), ]$log2FoldChange_MMSE)
colnames(lfc_vals_a_vs_sa) <- "A vs SA l"

lfc_vals_b_vs_l <- as.matrix(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs[match(ids, nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_id), ]$log2FoldChange_MMSE)
colnames(lfc_vals_b_vs_l) <- "sa B vs sa L"

#basemean <- as.matrix(nat_age_tecount_res_adult_limbs_vs_subadult_limbs[match(ids, nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_id), ]$baseMean_MLE) # baseMean is the same for both contrasts
#colnames(basemean) <- "baseMean"
```

```{r}
# Prepare heatmap annotations
annotation <- as.data.frame(colData(nat_tecount_dds_vst))
annotation <- annotation[c("Age_group", "Tissue_group")]
rownames(annotation) <- colnames(mat)

main_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b"
    )
  ),
  annotation_label = gt_render(c("Age group", "Tissue")),
  annotation_legend_param = list(
    Age_group = list(
      title = "Age group",
      at = c("Adult", "Sub_adult"),
      labels = c("Adult (8 years)", "Sub-adult (8 months)"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(
        fontsize = 11,
        fontface = "bold"
      )
    ),
    Tissue_group = list(
      title = "Tissue",
      at = c("Limb", "Limb_blastema"),
      labels = c("Limb", "Limb blastema"),
      labels_gp = gpar(fontsize = 9),
      title_gp = gpar(fontsize = 11, fontface = "bold")
    )
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

boxplotCol <- HeatmapAnnotation(
  boxplot = anno_boxplot(
    mat,
    border = TRUE,
    gp = gpar(fill = "#CCCCCC"),
    pch = ".",
    size = unit(2, "mm"),
    axis = TRUE,
    axis_param = list(
      gp = gpar(fontsize = 12),
      side = "left"
    )
  ),
  annotation_width = unit(c(2.0), "cm"),
  which = "col"
)

ha <- HeatmapAnnotation(summary = anno_summary(
  gp = gpar(fill = 2),
  height = unit(2, "cm")
))

# maps values between b/w/r for min and max lfc values
lfc_col <- colorRamp2(c(min(c(lfc_vals_a_vs_sa, lfc_vals_b_vs_l)), 0, max(c(lfc_vals_a_vs_sa, lfc_vals_b_vs_l))), c("#0072B2", "white", "#D55E00"))

# maps between 0% quantile, and 75% quantile of basemean values
#basemean_col <- colorRamp2(c(quantile(basemean)[1], quantile(basemean)[4], quantile(basemean)[5]), c("white", "red", "darkred"))

# Get gene names
symbols <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol[match(ids, nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_id)]

symbols <- gsub("Gypsy_Chromoviridae_Vclade", "Chromoviridae:Ty3", symbols)
symbols <- gsub("Gypsy_Mag", "Mag:Ty3", symbols)

# Generate main heatmap
heatmap_main <- Heatmap(mat[ids, ],
  cluster_rows = T,
  show_row_names = F,
  column_labels = colnames(mat),
  name = "Deviation from the mean (vst)",
  cluster_columns = T,
  top_annotation = main_ha,
  bottom_annotation = boxplotCol,
  width = unit(4.5, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9),
  # row_km = 3,
  # row_km_repeats = 100,
  split = 3,
  row_title = "Cluster %s",
  row_gap = unit(1.5, "mm"),
  row_title_gp = gpar(fontface = "bold")

  # split = pamClusters$clustering,
  # cluster_row_slices = FALSE
)

##### TO SAVE CLUSTERS

r.dend <- row_dend(heatmap_main) # If needed, extract row dendrogram
rcl.list <- row_order(heatmap_main) # Extract clusters (output is a list)


lapply(rcl.list, function(x) length(x)) # check/confirm size gene clusters

clu_df <- lapply(seq_along((rcl.list)), function(i) {
  out <- data.frame(
    ID = rownames(mat[rcl.list[[i]], ]),
    nat_heatmap_cluster = paste0("Cluster ", i),
    stringsAsFactors = FALSE
  )
  return(out)
}) %>% # pipe (forward) the output 'out' to the function rbind to create 'clu_df'
  do.call(rbind, .)

clu_df <- as_tibble(clu_df)

write_tsv(clu_df, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/heatmap/nat_heatmap_clusters.tsv"))

# Generate lfc (a_vs_sa) column
heatmap_lfc_a_vs_sa <- Heatmap(lfc_vals_a_vs_sa, # row_labels = ids,
  cluster_rows = F, name = "ash-log2(FC)", bottom_annotation = ha, col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
    grid.text(sprintf("%.2f", round(lfc_vals_a_vs_sa[i, j], 2)), x, y,
      gp = gpar(fontsize = 8.5)
    )
  },
  width = unit(1.2, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Generate lfc (b_vs_l) column
heatmap_lfc_b_vs_l <- Heatmap(lfc_vals_b_vs_l, # row_labels = ids,
  cluster_rows = F, name = "ash-log2(FC)", bottom_annotation = ha, col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
    grid.text(sprintf("%.2f", round(lfc_vals_b_vs_l[i, j], 2)), x, y,
      gp = gpar(fontsize = 8.5)
    )
  },
  width = unit(1.2, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Generate baseMean column
# heatmap_basemean <- Heatmap(basemean, # row_labels = ids,
#   cluster_rows = F, name = "baseMean", col = basemean_col,
#   cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
#     grid.text(round(basemean[i, j], 2), x, y,
#       gp = gpar(fontsize = 9.5)
#     )
#   },
#   width = unit(1.6, "cm"),
#   border_gp = gpar(col = "black", lwd = 0.7),
#   rect_gp = gpar(col = "white", lwd = 0.9)
# )

# Gene symbol annotation
gene_symb_ann <- rowAnnotation(gene_symbol = anno_text(symbols, gp = gpar(fontsize = 8.5)))


# Merge all heatmaps
#final_heatmap <- heatmap_main + heatmap_lfc_a_vs_sa + heatmap_lfc_b_vs_l + heatmap_basemean + gene_symb_ann
final_heatmap <- heatmap_main + heatmap_lfc_a_vs_sa + heatmap_lfc_b_vs_l + gene_symb_ann

png(filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/heatmap/heatmap_all.png"), width = 40.4, height = 40.4, units = "cm", res = 600)

draw(final_heatmap, merge_legends = TRUE)

dev.off()

svg(filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/heatmap/heatmap_all.svg"), width = 15.748, height = 15.748)

draw(final_heatmap, merge_legends = TRUE)

dev.off()

ht_opt(RESET = TRUE)
```

##### Gene set enrichment analysis (GSEA)

```{r}
library(clusterProfiler)
library(enrichplot)
library(data.table)
library(DESeq2)
library(ggh4x)
```

###### Gene Ontology (GO)

####### Formatting gene metadata (GO information/TERM2GENE)

```{r}
# load gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# get columns gene_id and GO IDs
go_TERM2GENE <- gene_metadata[c("gene_id", "gene_goall")]
# remove rows where gene_goall is NA
go_TERM2GENE <- go_TERM2GENE[!is.na(go_TERM2GENE$gene_goall), ]

# expand, since some genes/proteins will have multiple assigned GO terms
go_TERM2GENE <- data.table(go_TERM2GENE)
go_TERM2GENE <- go_TERM2GENE[, list(gene_goall = unlist(strsplit(gene_goall, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_goall combination (to remove redundant pairs)
go_TERM2GENE <- distinct(go_TERM2GENE, gene_id, gene_goall, .keep_all = TRUE)

go_TERM2GENE <- go_TERM2GENE[, c(2, 1)]
```

####### Running FGSEA (clusterProfiler::compareCluster)

```{r}
library(readr)

nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_age_tecount_res_adult_limbs_vs_subadult_limbs.tsv"))
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- read_tsv(file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_reg_tecount_res_sa_blastemas_vs_sa_limbs.tsv"))

nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE
names(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList) <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_id
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList <- na.omit(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList)
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList <- sort(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList, decreasing = TRUE)

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$log2FoldChange_MMSE
names(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList) <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_id
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList <- na.omit(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList)
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList <- sort(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList, decreasing = TRUE)

nat_tecount_gsea_List <- list(
  "Adult limbs vs Sub-adult limbs" = nat_age_tecount_res_adult_limbs_vs_subadult_limbs_geneList,
  "Sub-adult blastemas vs Sub-adult limbs" = nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_geneList
)

gsea_nat_tecount <- compareCluster(
  geneClusters = nat_tecount_gsea_List,
  fun = "GSEA",
  TERM2GENE = go_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_nat_tecount
```
  
```{r}
library(GO.db)
library(AnnotationDbi)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_nat_tecount@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_nat_tecount@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_nat_tecount@compareClusterResult$ID[i]))
        gsea_nat_tecount@compareClusterResult$Description[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_nat_tecount@compareClusterResult$ID[i], column = c("TERM"), keytype = "GOID")
        gsea_nat_tecount@compareClusterResult$ONTOLOGY[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_nat_tecount@compareClusterResult$ID[i], column = c("ONTOLOGY"), keytype = "GOID")
      }
    },
    error = function(e) {
      cat("ERROR while mapping ", gsea_nat_tecount@compareClusterResult$ID[i], ":", conditionMessage(e), "\n")
    }
  )
}

# In gsea_nat_tecount@compareClusterResult$ONTOLOGY replace BP with Biological Process, MF with Molecular Function, CC with Cellular Component
gsea_nat_tecount@compareClusterResult$ONTOLOGY_full <- gsub("BP", "Biological process", gsea_nat_tecount@compareClusterResult$ONTOLOGY)
gsea_nat_tecount@compareClusterResult$ONTOLOGY_full <- gsub("MF", "Molecular function", gsea_nat_tecount@compareClusterResult$ONTOLOGY_full)
gsea_nat_tecount@compareClusterResult$ONTOLOGY_full <- gsub("CC", "Cellular component", gsea_nat_tecount@compareClusterResult$ONTOLOGY_full)

# To remove obsolete or removed GO terms
# Remove from gsea_nat_tecount@compareClusterResult all rows for which Description starts with GO:
gsea_nat_tecount@compareClusterResult <- gsea_nat_tecount@compareClusterResult[!grepl("^GO:", gsea_nat_tecount@compareClusterResult$Description), ]

# Add type column to gsea_nat_tecount@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_nat_tecount@compareClusterResult$type <- ifelse(gsea_nat_tecount@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_nat_tecount@compareClusterResult$type <- factor(gsea_nat_tecount@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_nat_tecount, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/gsea_go_nat_tecount.rds"))
# gsea_nat_tecount <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/gsea_go_nat_tecount.rds"))

# how many type per Cluster
#table(gsea_nat_tecount@compareClusterResult$Cluster, gsea_nat_tecount@compareClusterResult$type)

gsea_nat_tecount_tb <- as_tibble(gsea_nat_tecount@compareClusterResult)
write_tsv(gsea_nat_tecount_tb, file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/gsea_go_nat_tecount.tsv"))



```

######## Dotplot

```{r}
library(stringr)
dp_gsea_nat_tecount <- dotplot(gsea_nat_tecount,
  showCategory = 28,
  size = "Count",
  x = "GeneRatio",
  label_format = 50
) +

  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 30)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster + type,
    scales = "free_y",
    space = "free"
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )
# switch = "y"

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/dp_gsea_nat_tecount.png"),
  plot = dp_gsea_nat_tecount, width = 20, height = 28, units = "cm", dpi = 600
)


# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/dp_gsea_nat_tecount.svg"),
  plot = dp_gsea_nat_tecount, width = 20, height = 28, units = "cm"
)
```

###### KEGG

```{r}
# get columns gene_id and KEGG IDs
kegg_TERM2GENE <- gene_metadata[c("gene_id", "gene_keggpath")]
# remove rows where gene_keggpath is NA
kegg_TERM2GENE <- kegg_TERM2GENE[!is.na(kegg_TERM2GENE$gene_keggpath), ]

# expand, since some genes/proteins will have multiple assigned GO terms
kegg_TERM2GENE <- data.table(kegg_TERM2GENE)
kegg_TERM2GENE <- kegg_TERM2GENE[, list(gene_keggpath = unlist(strsplit(gene_keggpath, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_keggpath combination (to remove redundant pairs)
kegg_TERM2GENE <- distinct(kegg_TERM2GENE, gene_id, gene_keggpath, .keep_all = TRUE)

kegg_TERM2GENE <- kegg_TERM2GENE[, c(2, 1)]
```

####### compareCluster

```{r}

gsea_kegg_nat_tecount <- compareCluster(
  geneClusters = nat_tecount_gsea_List,
  fun = "GSEA",
  TERM2GENE = kegg_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_kegg_nat_tecount
```
  
```{r}
# To map KEGG IDs to Pathway names (there's a 10-element limitation)
library(KEGGREST)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_kegg_nat_tecount@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_kegg_nat_tecount@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_kegg_nat_tecount@compareClusterResult$ID[i]))
        gsea_kegg_nat_tecount@compareClusterResult$Description[i] <- keggGet(gsea_kegg_nat_tecount@compareClusterResult$ID[i])[[1]]$NAME
      }
    },
    error = function(e) {
      cat("ERROR: ", gsea_kegg_nat_tecount@compareClusterResult$ID[i], conditionMessage(e), "\n")
    }
  )
}


# To remove obsolete or removed KEGG terms
# Remove from gsea_kegg_nat_tecount@compareClusterResult all rows for which Description starts with GO:
gsea_kegg_nat_tecount@compareClusterResult <- gsea_kegg_nat_tecount@compareClusterResult[!grepl("^ko", gsea_kegg_nat_tecount@compareClusterResult$Description), ]

# Add type column to gsea_kegg_nat_tecount@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_kegg_nat_tecount@compareClusterResult$type <- ifelse(gsea_kegg_nat_tecount@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_kegg_nat_tecount@compareClusterResult$type <- factor(gsea_kegg_nat_tecount@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_kegg_nat_tecount, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/gsea_kegg_nat_tecount.rds"))
# gsea_kegg_nat_tecount <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/gsea_kegg_nat_tecount.rds"))


# how many type per Cluster
#table(gsea_kegg_nat_tecount@compareClusterResult$Cluster, gsea_kegg_nat_tecount@compareClusterResult$type)


gsea_kegg_nat_tecount_tb <- as_tibble(gsea_kegg_nat_tecount@compareClusterResult)
write_tsv(gsea_kegg_nat_tecount_tb, file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/gsea_kegg_nat_tecount.tsv"))
```

######## Dotplot
```{r}
library(stringr)
library(ggrepel)

# add dummy column to gsea_kegg_nat_tecount@compareClusterResult
gsea_kegg_nat_tecount@compareClusterResult$dummy <- "same"

dp_gsea_kegg_nat_tecount <- dotplot(gsea_kegg_nat_tecount,
  showCategory = 16,
  size = "Count",
  x = "GeneRatio",
  label_format = 50
) +

  # add text labels to each dot
  # geom_text_repel(aes(label = Count), size = 3, direction = "x", nudge_x = ifelse(Count < 0.5, 0.5, -0.5)) +


  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 30)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(dummy ~ Cluster + type,
    scales = "free_y",
    space = "free"
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )


ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/dp_gsea_kegg_nat_tecount.png"),
  plot = dp_gsea_kegg_nat_tecount, width = 20, height = 16, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/dp_gsea_kegg_nat_tecount.svg"),
  plot = dp_gsea_kegg_nat_tecount, width = 20, height = 16, units = "cm"
)

```

```{r}
# make ggarrange of dp_gsea_nat_tecount and dp_gsea_kegg_nat_tecount
library(ggpubr)

# remove strip titles
dp_gsea_kegg_nat_tecount_tmp <- dp_gsea_kegg_nat_tecount +
  theme(strip.text.x = element_blank())

# remove x axis text, title and ticks
dp_gsea_nat_tecount_tmp <- dp_gsea_nat_tecount +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

dp_gsea_nat_tecount_arrange <- ggarrange(dp_gsea_nat_tecount_tmp, dp_gsea_kegg_nat_tecount_tmp,
  ncol = 1, nrow = 2,
  common.legend = TRUE,
  align = "v",
  legend = "bottom",
  heights = c(56, 30)
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/dp_gsea_nat_tecount_arrange.png"),
  plot = dp_gsea_nat_tecount_arrange, width = 20, height = 46, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/enrich/dp_gsea_nat_tecount_arrange.svg"),
  plot = dp_gsea_nat_tecount_arrange, width = 20, height = 46, units = "cm"
)
```

#### Global axolotl sample analysis

##### Setting DESeq2 model

```{r}
library(DESeq2)

glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")

counts <- as.matrix(read.csv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/import/cntTable_all.csv"),
  header = TRUE, sep = ",", row.names = 1
))
colnames(counts) <- sub("^X", "", colnames(counts))
colnames(counts) <- gsub("\\.", "_", colnames(counts))


# filtering to remove other datasets and keep only native
glob_tecount_quant <- counts[, rownames(glob_metadata)]

all(rownames(glob_metadata) %in% colnames(glob_tecount_quant))

all(rownames(glob_metadata) == colnames(glob_tecount_quant))

glob_tecount_dds <- DESeqDataSetFromMatrix(
  countData = glob_tecount_quant,
  colData = glob_metadata,
  design = ~ Batch + Age_group + Tissue_group + Age_group:Tissue_group
)


# Explicitly setting the factors levels
glob_tecount_dds$Age_group <- relevel(glob_tecount_dds$Age_group, ref = "Sub_adult")
glob_tecount_dds$Batch <- relevel(glob_tecount_dds$Batch, ref = "Ruiz_Perez_et_al._n.d.")
glob_tecount_dds$Tissue_group <- relevel(glob_tecount_dds$Tissue_group, ref = "Non_limb")

saveRDS(glob_tecount_dds, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/glob_tecount_dds.rds"))
```

###### Pre-filtering

```{r}
summary(rowSums(counts(glob_tecount_dds) >= 10) >= 3)

#   Mode   FALSE    TRUE
# logical   55554   46490
```

###### Pre-visualization

```{r}
glob_tecount_dds_vst <- vst(glob_tecount_dds, blind = FALSE)

saveRDS(glob_tecount_dds_vst,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/glob_tecount_dds_vst.rds")
)
```

####### cPCA (vst)

```{r}
library(DESeq2)
library(cowplot)
library(ggpubr)
library(ggrepel)
library(ggforce)
library(ggh4x)
library(ggalt)

glob_tecount_dds_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/glob_tecount_dds_vst.rds"))

glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")
```

Personalized cPCA

```{r}
glob_tecount_pcaData <- plotPCA(glob_tecount_dds_vst, intgroup = c("Age_group", "Batch", "Tissue_group"), returnData = TRUE)
percentVar <- round(100 * attr(glob_tecount_pcaData, "percentVar"))

merged <- merge(glob_tecount_pcaData, glob_metadata, by.x = "name", by.y = "Run", all.x = TRUE)
merged <- merged[order(merged$name), ]

glob_tecount_pcaData <- glob_tecount_pcaData[order(row.names(glob_tecount_pcaData)), ]
glob_tecount_pcaData$new_name <- merged$new_name_run

saveRDS(glob_tecount_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/previsualization/glob_tecount_pcaData.rds"))

write.csv(glob_tecount_pcaData, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/previsualization/glob_tecount_pcaData.csv"))
```

Plotting:
```{r}
glob_tecount_pca_main <- ggplot(glob_tecount_pcaData, aes(PC1, PC2, fill = Tissue_group, shape = Age_group)) +
  geom_encircle(aes(group = Batch, colour = Batch), alpha = 0.4) +
  geom_point(size = 3.2, stroke = 0.2) +
  scale_shape_manual(
    name = "Age group", values = c(24, 22),
    breaks = c("Adult", "Sub_adult"), labels = c("Adult", "Sub-adult"),
    guide = guide_legend(order = 1, label.theme = element_text(size = 9, margin = margin(l = 25)))
  ) +
  scale_fill_manual(
    name = "Tissue_group",
    breaks = c("Non_limb", "Limb", "Limb_blastema"),
    values = c("#57452dff", "#4dac26", "#d01c8b")
  ) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  ) +
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  # guides(color = guide_legend(override.aes = list(size = 3))) +

  coord_fixed() +
  guides(
    fill = guide_legend(order = 2),
    color = guide_legend(order = 2)
  )


ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/previsualization/pca_tecount_vst.png"),
  plot = glob_tecount_pca_main,
  device = "png",
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white",
  dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/previsualization/pca_tecount_vst.svg"),
  plot = glob_tecount_pca_main,
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white"
)
```

Temporal plot with different points to overlay

```{r}
glob_tecount_pca_main <- ggplot(glob_tecount_pcaData, aes(PC1, PC2, color = Batch)) +
  geom_point(size = 12, alpha = 0.5, stroke = 0) +
  scale_color_manual(
    name = "Batch",
    values = c("#d36027ff", "#0773b3ff", "#f2e646ff", "#6b6a66ff", "#e79f26ff"),
    breaks = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"),
    labels = c("This study", bquote("Ye" ~ italic("et al.") ~ "(2022)"), bquote("Nowoshilow" ~ italic("et al.") ~ "(2018)"), bquote("Caballero-Pérez" ~ italic("et al.") ~ "(2018)"), bquote("Bryant" ~ italic("et al.") ~ "(2017)"))
  ) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  force_panelsizes(
    rows = unit(3, "in"),
    cols = unit(3, "in")
  ) +
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold"),
    axis.title.x = element_text(face = "bold"),
    axis.title.y = element_text(face = "bold")
  ) +
  # guides(color = guide_legend(override.aes = list(size = 3))) +

  coord_fixed() +
  guides(
    fill = guide_legend(order = 2),
    color = guide_legend(order = 2)
  )

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/previsualization/pca_tecount_vst2.png"),
  plot = glob_tecount_pca_main,
  device = "png",
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white",
  dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/previsualization/pca_tecount_vst2.svg"),
  plot = glob_tecount_pca_main,
  path = NULL,
  scale = 1,
  width = 165,
  height = 115,
  units = "mm",
  bg = "white"
)
```

###### Differential expression (DE) analysis

####### Running DESeq

```{r}
glob_tecount_dds_de <- DESeq(glob_tecount_dds)

# estimating size factors
# estimating dispersions
# gene-wise dispersion estimates
# mean-dispersion relationship
# final dispersion estimates
# fitting model and testing
# 865 rows did not converge in beta, labelled in #mcols(object)$betaConv. Use larger maxit #argument with nbinomWaldTest
#-- replacing outliers and refitting for 2267 genes
#-- DESeq argument 'minReplicatesForReplace' = 7
#-- original counts are preserved in counts(dds)
# estimating dispersions
# fitting model and testing
# 121 rows did not converge in beta, labelled in #mcols(object)$betaConv. Use larger maxit #argument with nbinomWaldTest

saveRDS(glob_tecount_dds_de, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/glob_tecount_dds_de.rds"))
```

```{r}
# Remove not converging genes (https://support.bioconductor.org/p/65091/#65114)

glob_tecount_ddsClean_de <- glob_tecount_dds_de[which(mcols(glob_tecount_dds_de)$betaConv), ]

saveRDS(glob_tecount_ddsClean_de, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/glob_tecount_ddsClean_de.rds"))
```

```{r}
png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/DispEsts_glob_tecount_dds_de.png"))

plotDispEsts(glob_tecount_dds_de)

dev.off()

png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/DispEsts_glob_tecount_ddsClean_de.png"))

plotDispEsts(glob_tecount_ddsClean_de)

dev.off()
```

```{r}
summary(rowSums(counts(glob_tecount_dds) >= 10) >= 3)

#   Mode   FALSE    TRUE
# logical   55554   46490

keep <- rowSums(counts(glob_tecount_dds) >= 10) >= 3

glob_tecount_dds_prefilt <- glob_tecount_dds[keep, ]

glob_tecount_dds_prefilt_de <- DESeq(glob_tecount_dds_prefilt)


png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/DispEsts_glob_tecount_dds_prefilt_de.png"))

plotDispEsts(glob_tecount_dds_prefilt_de)

dev.off()
```

####### Summarizing DE results

```{r}
library("IHW")
```

```{r}
resultsNames(glob_tecount_ddsClean_de)

# [1] "Intercept"
# [2] "Batch_Bryant_et_al._2017_vs_Ruiz_Perez_et_al._n.d."
# [3] "Batch_Caballero_Perez_et_al._2018_vs_Ruiz_Perez_et_al._n.d."
# [4] "Batch_Nowoshilow_et_al._2018_vs_Ruiz_Perez_et_al._n.d."
# [5] "Batch_Ye_et_al._2022_vs_Ruiz_Perez_et_al._n.d."
# [6] "Age_group_Adult_vs_Sub_adult"
# [7] "Tissue_group_Limb_vs_Non_limb"
# [8] "Tissue_group_Limb_blastema_vs_Non_limb"
# [9] "Age_groupAdult.Tissue_groupLimb"
# [10] "Age_groupAdult.Tissue_groupLimb_blastema"
```

######### Limb vs Non-limb (for subadults)

```{r}
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 1, 0, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1358, 1.5%
# LFC < -0.58 (down) : 1026, 1.1%
# outliers [1]       : 346, 0.37%

## LFC Shrinkage

library(ashr)

glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 920, 0.99%
# LFC < -0.58 (down) : 596, 0.64%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs) <- paste0(colnames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs), "_MLE")
colnames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink) <- paste0(colnames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink), "_MMSE")
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE <- glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink$log2FoldChange_MMSE
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$lfcSE_MMSE <- glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink$lfcSE_MMSE
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$svalue_MMSE <- glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol[is.na(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol)] <- rownames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs)[is.na(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$gene_symbol)]
# add metadata columns
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$quant_tool <- "TEcount"
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$diffexp_contrast <- "subadult_limbs_vs_subadult_nonlimbs"
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$sample_group <- "global"
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) | abs(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$padj_MLE) | glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs <- as_tibble(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs, rownames = "gene_id")

print(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs, width = Inf)

readr::write_tsv(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs.csv")
)
```

######### Limb vs Non-limb (for adults)

```{r}
glob_tecount_res_adult_limbs_vs_adult_nonlimbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 1, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_adult_limbs_vs_adult_nonlimbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 161, 0.17%
# LFC < -0.58 (down) : 677, 0.73%
# outliers [1]       : 346, 0.37%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_adult_limbs_vs_adult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 117, 0.13%
# LFC < -0.58 (down) : 441, 0.48%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs) <- paste0(colnames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs), "_MLE")
colnames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink) <- paste0(colnames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink), "_MMSE")
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE <- glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink$log2FoldChange_MMSE
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$lfcSE_MMSE <- glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink$lfcSE_MMSE
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$svalue_MMSE <- glob_tecount_res_adult_limbs_vs_adult_nonlimbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$gene_symbol[is.na(glob_tecount_res_adult_limbs_vs_adult_nonlimbs$gene_symbol)] <- rownames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs)[is.na(glob_tecount_res_adult_limbs_vs_adult_nonlimbs$gene_symbol)]
# add metadata columns
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$quant_tool <- "TEcount"
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$diffexp_contrast <- "adult_limbs_vs_adult_nonlimbs"
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$sample_group <- "global"
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_adult_limbs_vs_adult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE) | abs(glob_tecount_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_adult_limbs_vs_adult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_adult_limbs_vs_adult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_adult_limbs_vs_adult_nonlimbs$padj_MLE) | glob_tecount_res_adult_limbs_vs_adult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_adult_limbs_vs_adult_nonlimbs <- as_tibble(glob_tecount_res_adult_limbs_vs_adult_nonlimbs, rownames = "gene_id")

print(glob_tecount_res_adult_limbs_vs_adult_nonlimbs, width = Inf)

readr::write_tsv(glob_tecount_res_adult_limbs_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_limbs_vs_adult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_adult_limbs_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_limbs_vs_adult_nonlimbs.csv")
)
```

######### Blastema vs Non-limb (for subadults)

```{r}
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 1, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 244, 0.26%
# LFC < -0.58 (down) : 1884, 2%
# outliers [1]       : 346, 0.37%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 144, 0.16%
# LFC < -0.58 (down) : 1248, 1.3%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs) <- paste0(colnames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs), "_MLE")
colnames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink) <- paste0(colnames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink), "_MMSE")
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE <- glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink$log2FoldChange_MMSE
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$lfcSE_MMSE <- glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink$lfcSE_MMSE
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$svalue_MMSE <- glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol[is.na(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol)] <- rownames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs)[is.na(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$gene_symbol)]
# add metadata columns
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$quant_tool <- "TEcount"
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_nonlimbs"
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$sample_group <- "global"
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE) | abs(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$padj_MLE) | glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs <- as_tibble(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs, rownames = "gene_id")

print(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs, width = Inf)

readr::write_tsv(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs.csv")
)
```

######### Blastema vs Non-limb (for adults)

```{r}
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 1, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1470, 1.6%
# LFC < -0.58 (down) : 8707, 9.4%
# outliers [1]       : 346, 0.37%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_adult_blastemas_vs_adult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 1488, 1.6%
# LFC < -0.58 (down) : 8232, 8.9%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs) <- paste0(colnames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs), "_MLE")
colnames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink) <- paste0(colnames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink), "_MMSE")
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE <- glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink$log2FoldChange_MMSE
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$lfcSE_MMSE <- glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink$lfcSE_MMSE
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$svalue_MMSE <- glob_tecount_res_adult_blastemas_vs_adult_nonlimbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol[is.na(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol)] <- rownames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs)[is.na(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$gene_symbol)]
# add metadata columns
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$quant_tool <- "TEcount"
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$diffexp_contrast <- "adult_blastemas_vs_adult_nonlimbs"
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$sample_group <- "global"
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE) | abs(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$padj_MLE) | glob_tecount_res_adult_blastemas_vs_adult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_adult_blastemas_vs_adult_nonlimbs <- as_tibble(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs, rownames = "gene_id")

print(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs, width = Inf)

readr::write_tsv(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_blastemas_vs_adult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_adult_blastemas_vs_adult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_blastemas_vs_adult_nonlimbs.csv")
)
```

######### Blastema vs Limb (for subadults)

```{r}
glob_tecount_res_subadult_blastemas_vs_subadult_limbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, -1, 1, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_subadult_blastemas_vs_subadult_limbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 210, 0.23%
# LFC < -0.58 (down) : 2868, 3.1%
# outliers [1]       : 346, 0.37%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_subadult_blastemas_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 157, 0.17%
# LFC < -0.58 (down) : 2202, 2.4%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs) <- paste0(colnames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs), "_MLE")
colnames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink) <- paste0(colnames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink), "_MMSE")
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE <- glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink$log2FoldChange_MMSE
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$lfcSE_MMSE <- glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink$lfcSE_MMSE
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$svalue_MMSE <- glob_tecount_res_subadult_blastemas_vs_subadult_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$gene_symbol[is.na(glob_tecount_res_subadult_blastemas_vs_subadult_limbs$gene_symbol)] <- rownames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs)[is.na(glob_tecount_res_subadult_blastemas_vs_subadult_limbs$gene_symbol)]
# add metadata columns
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$quant_tool <- "TEcount"
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$diffexp_contrast <- "subadult_blastemas_vs_subadult_limbs"
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$sample_group <- "global"
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_subadult_blastemas_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE) | abs(glob_tecount_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_subadult_blastemas_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_subadult_blastemas_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_subadult_blastemas_vs_subadult_limbs$padj_MLE) | glob_tecount_res_subadult_blastemas_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_subadult_blastemas_vs_subadult_limbs <- as_tibble(glob_tecount_res_subadult_blastemas_vs_subadult_limbs, rownames = "gene_id")

print(glob_tecount_res_subadult_blastemas_vs_subadult_limbs, width = Inf)

readr::write_tsv(glob_tecount_res_subadult_blastemas_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_subadult_blastemas_vs_subadult_limbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_subadult_blastemas_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_subadult_blastemas_vs_subadult_limbs.csv")
)
```

######### Blastema vs Limb (for adults)

```{r}
glob_tecount_res_adult_blastemas_vs_adult_limbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, -1, 1, -1, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_adult_blastemas_vs_adult_limbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1023, 1.1%
# LFC < -0.58 (down) : 4480, 4.8%
# outliers [1]       : 346, 0.37%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_adult_blastemas_vs_adult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 940, 1%
# LFC < -0.58 (down) : 3472, 3.7%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_adult_blastemas_vs_adult_limbs) <- paste0(colnames(glob_tecount_res_adult_blastemas_vs_adult_limbs), "_MLE")
colnames(glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink) <- paste0(colnames(glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink), "_MMSE")
glob_tecount_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE <- glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink$log2FoldChange_MMSE
glob_tecount_res_adult_blastemas_vs_adult_limbs$lfcSE_MMSE <- glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink$lfcSE_MMSE
glob_tecount_res_adult_blastemas_vs_adult_limbs$svalue_MMSE <- glob_tecount_res_adult_blastemas_vs_adult_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_adult_blastemas_vs_adult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_adult_blastemas_vs_adult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_adult_blastemas_vs_adult_limbs$gene_symbol[is.na(glob_tecount_res_adult_blastemas_vs_adult_limbs$gene_symbol)] <- rownames(glob_tecount_res_adult_blastemas_vs_adult_limbs)[is.na(glob_tecount_res_adult_blastemas_vs_adult_limbs$gene_symbol)]
# add metadata columns
glob_tecount_res_adult_blastemas_vs_adult_limbs$quant_tool <- "TEcount"
glob_tecount_res_adult_blastemas_vs_adult_limbs$diffexp_contrast <- "adult_blastemas_vs_adult_limbs"
glob_tecount_res_adult_blastemas_vs_adult_limbs$sample_group <- "global"
glob_tecount_res_adult_blastemas_vs_adult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_adult_blastemas_vs_adult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_adult_blastemas_vs_adult_limbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE) | abs(glob_tecount_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_adult_blastemas_vs_adult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_adult_blastemas_vs_adult_limbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_adult_blastemas_vs_adult_limbs$padj_MLE) | glob_tecount_res_adult_blastemas_vs_adult_limbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_adult_blastemas_vs_adult_limbs <- as_tibble(glob_tecount_res_adult_blastemas_vs_adult_limbs, rownames = "gene_id")

print(glob_tecount_res_adult_blastemas_vs_adult_limbs, width = Inf)

readr::write_tsv(glob_tecount_res_adult_blastemas_vs_adult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_blastemas_vs_adult_limbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_adult_blastemas_vs_adult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_blastemas_vs_adult_limbs.csv")
)
```

######### Adults vs Sub-adults (for non-limb)

```{r}
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 0, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs)

# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 4339, 4.7%
# LFC < -0.58 (down) : 1279, 1.4%
# outliers [1]       : 346, 0.37%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 3288, 3.5%
# LFC < -0.58 (down) : 1079, 1.2%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs) <- paste0(colnames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs), "_MLE")
colnames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink) <- paste0(colnames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink), "_MMSE")
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE <- glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink$log2FoldChange_MMSE
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$lfcSE_MMSE <- glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink$lfcSE_MMSE
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$svalue_MMSE <- glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol[is.na(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol)] <- rownames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs)[is.na(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_symbol)]
# add metadata columns
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$quant_tool <- "TEcount"
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$diffexp_contrast <- "adult_nonlimbs_vs_subadult_nonlimbs"
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$sample_group <- "global"
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) | abs(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$padj_MLE) | glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs <- as_tibble(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs, rownames = "gene_id")

print(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs, width = Inf)

readr::write_tsv(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs.csv")
)
```

######### Adults vs Sub-adults (for limb)

```{r}
glob_tecount_res_adult_limbs_vs_subadult_limbs <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_adult_limbs_vs_subadult_limbs)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 5406, 5.8%
# LFC < -0.58 (down) : 2661, 2.9%
# outliers [1]       : 346, 0.37%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_adult_limbs_vs_subadult_limbs,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 4712, 5.1%
# LFC < -0.58 (down) : 2423, 2.6%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_adult_limbs_vs_subadult_limbs) <- paste0(colnames(glob_tecount_res_adult_limbs_vs_subadult_limbs), "_MLE")
colnames(glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink) <- paste0(colnames(glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink), "_MMSE")
glob_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE <- glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink$log2FoldChange_MMSE
glob_tecount_res_adult_limbs_vs_subadult_limbs$lfcSE_MMSE <- glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink$lfcSE_MMSE
glob_tecount_res_adult_limbs_vs_subadult_limbs$svalue_MMSE <- glob_tecount_res_adult_limbs_vs_subadult_limbs_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_adult_limbs_vs_subadult_limbs), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol[is.na(glob_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol)] <- rownames(glob_tecount_res_adult_limbs_vs_subadult_limbs)[is.na(glob_tecount_res_adult_limbs_vs_subadult_limbs$gene_symbol)]
# add metadata columns
glob_tecount_res_adult_limbs_vs_subadult_limbs$quant_tool <- "TEcount"
glob_tecount_res_adult_limbs_vs_subadult_limbs$diffexp_contrast <- "adult_limbs_vs_subadult_limbs"
glob_tecount_res_adult_limbs_vs_subadult_limbs$sample_group <- "global"
glob_tecount_res_adult_limbs_vs_subadult_limbs$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_adult_limbs_vs_subadult_limbs)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_adult_limbs_vs_subadult_limbs$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) | abs(glob_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_adult_limbs_vs_subadult_limbs$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_adult_limbs_vs_subadult_limbs$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_adult_limbs_vs_subadult_limbs$padj_MLE) | glob_tecount_res_adult_limbs_vs_subadult_limbs$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_adult_limbs_vs_subadult_limbs <- as_tibble(glob_tecount_res_adult_limbs_vs_subadult_limbs, rownames = "gene_id")

print(glob_tecount_res_adult_limbs_vs_subadult_limbs, width = Inf)

readr::write_tsv(glob_tecount_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_limbs_vs_subadult_limbs.tsv")
)
readr::write_excel_csv(glob_tecount_res_adult_limbs_vs_subadult_limbs,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_limbs_vs_subadult_limbs.csv")
)
```

######### Adults vs Sub-adults (for blastema)

```{r}
glob_tecount_res_adult_vs_subadult_blastema <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 1, 0, 0, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_adult_vs_subadult_blastema)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 623, 0.67%
# LFC < -0.58 (down) : 331, 0.36%
# outliers [1]       : 346, 0.37%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_adult_vs_subadult_blastema_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_adult_vs_subadult_blastema,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_adult_vs_subadult_blastema_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 476, 0.51%
# LFC < -0.58 (down) : 215, 0.23%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_adult_vs_subadult_blastema) <- paste0(colnames(glob_tecount_res_adult_vs_subadult_blastema), "_MLE")
colnames(glob_tecount_res_adult_vs_subadult_blastema_shrink) <- paste0(colnames(glob_tecount_res_adult_vs_subadult_blastema_shrink), "_MMSE")
glob_tecount_res_adult_vs_subadult_blastema$log2FoldChange_MMSE <- glob_tecount_res_adult_vs_subadult_blastema_shrink$log2FoldChange_MMSE
glob_tecount_res_adult_vs_subadult_blastema$lfcSE_MMSE <- glob_tecount_res_adult_vs_subadult_blastema_shrink$lfcSE_MMSE
glob_tecount_res_adult_vs_subadult_blastema$svalue_MMSE <- glob_tecount_res_adult_vs_subadult_blastema_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_adult_vs_subadult_blastema$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_adult_vs_subadult_blastema), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_adult_vs_subadult_blastema$gene_symbol[is.na(glob_tecount_res_adult_vs_subadult_blastema$gene_symbol)] <- rownames(glob_tecount_res_adult_vs_subadult_blastema)[is.na(glob_tecount_res_adult_vs_subadult_blastema$gene_symbol)]
# add metadata columns
glob_tecount_res_adult_vs_subadult_blastema$quant_tool <- "TEcount"
glob_tecount_res_adult_vs_subadult_blastema$diffexp_contrast <- "adults_blastemas_vs_subadults_blastemas"
glob_tecount_res_adult_vs_subadult_blastema$sample_group <- "global"
glob_tecount_res_adult_vs_subadult_blastema$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_adult_vs_subadult_blastema)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_adult_vs_subadult_blastema$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_adult_vs_subadult_blastema$log2FoldChange_MMSE) | abs(glob_tecount_res_adult_vs_subadult_blastema$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_adult_vs_subadult_blastema$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_adult_vs_subadult_blastema$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_adult_vs_subadult_blastema$padj_MLE) | glob_tecount_res_adult_vs_subadult_blastema$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_adult_vs_subadult_blastema <- as_tibble(glob_tecount_res_adult_vs_subadult_blastema, rownames = "gene_id")

print(glob_tecount_res_adult_vs_subadult_blastema, width = Inf)

readr::write_tsv(glob_tecount_res_adult_vs_subadult_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_vs_subadult_blastema.tsv")
)
readr::write_excel_csv(glob_tecount_res_adult_vs_subadult_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_adult_vs_subadult_blastema.csv")
)
```

######### Interaction term: Age_group effect in Limb vs Non-limb (Age_groupAdult.Tissue_groupLimb)

 this tests if the Age_group effect is different in Limb compared to Non-limb

```{r}
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 0, 1, 0),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 380, 0.41%
# LFC < -0.58 (down) : 954, 1%
# outliers [1]       : 346, 0.37%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 211, 0.23%
# LFC < -0.58 (down) : 507, 0.55%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb) <- paste0(colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb), "_MLE")
colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink) <- paste0(colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink), "_MMSE")
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE <- glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink$log2FoldChange_MMSE
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$lfcSE_MMSE <- glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink$lfcSE_MMSE
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$svalue_MMSE <- glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol[is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol)] <- rownames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb)[is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$gene_symbol)]
# add metadata columns
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$quant_tool <- "TEcount"
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_contrast <- "age_effect_limbs_vs_nonlimbs"
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$sample_group <- "global"
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE) | abs(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$padj_MLE) | glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb <- as_tibble(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb, rownames = "gene_id")

print(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb, width = Inf)

readr::write_tsv(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb.tsv")
)
readr::write_excel_csv(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb.csv")
)
```

######### Interaction term: Age_group effect in Blastema vs Non-limb (Age_groupAdult.Tissue_groupLimb_blastema)
 
 this tests if the Age_group effect is different in Blastema compared to Non-limb


```{r}
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 1),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 663, 0.72%
# LFC < -0.58 (down) : 1746, 1.9%
# outliers [1]       : 346, 0.37%


## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink)

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema) <- paste0(colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema), "_MLE")
colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink) <- paste0(colnames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink), "_MMSE")
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE <- glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$log2FoldChange_MMSE
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$lfcSE_MMSE <- glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$lfcSE_MMSE
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$svalue_MMSE <- glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol[is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol)] <- rownames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema)[is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$gene_symbol)]
# add metadata columns
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$quant_tool <- "TEcount"
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_contrast <- "age_effect_blastemas_vs_nonlimbs"
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$sample_group <- "global"
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE) | abs(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$padj_MLE) | glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- as_tibble(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema, rownames = "gene_id")

print(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema, width = Inf)

readr::write_tsv(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema.tsv")
)
readr::write_excel_csv(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema.csv")
)
```

######### Interaction term: Age_group effect in Blastema vs Limb (Age_groupAdult.Tissue_groupLimb_blastema vs Age_groupAdult.Tissue_groupLimb)
 
 this tests if the Age_group effect is different in Blastema compared to Limb

```{r}
glob_tecount_res_int_Age_group_effect_blastema_vs_limb <- results(glob_tecount_ddsClean_de,
  filterFun = ihw,
  contrast = list("Age_groupAdult.Tissue_groupLimb_blastema", "Age_groupAdult.Tissue_groupLimb"),
  lfcThreshold = log2(1.5),
  alpha = 0.05
)

summary(glob_tecount_res_int_Age_group_effect_blastema_vs_limb)
# out of 92642 with nonzero total read count
# adjusted p-value < 0.05
# LFC > 0.58 (up)    : 1036, 1.1%
# LFC < -0.58 (down) : 1386, 1.5%
# outliers [1]       : 346, 0.37%

## LFC Shrinkage

library(ashr)

# Diferencia entre Limb y Non-limb en subadultos
glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink <- lfcShrink(glob_tecount_ddsClean_de,
  res = glob_tecount_res_int_Age_group_effect_blastema_vs_limb,
  lfcThreshold = log2(1.5),
  alpha = 0.05,
  type = "ashr"
)

summary(glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink)
# out of 92642 with nonzero total read count
# s-value < 0.005
# LFC > 0.58 (up)    : 607, 0.66%
# LFC < -0.58 (down) : 744, 0.8%

# Saving all columns (MLE and posterior LFC) in the same dataframe.
colnames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb) <- paste0(colnames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb), "_MLE")
colnames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink) <- paste0(colnames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink), "_MMSE")
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE <- glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink$log2FoldChange_MMSE
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$lfcSE_MMSE <- glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink$lfcSE_MMSE
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$svalue_MMSE <- glob_tecount_res_int_Age_group_effect_blastema_vs_limb_shrink$svalue_MMSE

# adding gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$gene_symbol <- gene_metadata$gene_symbol[match(rownames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb), gene_metadata$gene_id)]
# if gene_symbol is NA, replace with rowname
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$gene_symbol[is.na(glob_tecount_res_int_Age_group_effect_blastema_vs_limb$gene_symbol)] <- rownames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb)[is.na(glob_tecount_res_int_Age_group_effect_blastema_vs_limb$gene_symbol)]
# add metadata columns
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$quant_tool <- "TEcount"
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$diffexp_contrast <- "age_effect_blastemas_vs_limbs"
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$sample_group <- "global"
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$feature_biotype <- ifelse(grepl("^AMEX", rownames(glob_tecount_res_int_Age_group_effect_blastema_vs_limb)), "gene", "repetitive_element")
# NA or abs(lfcThreshold) < log2(1.5) -> not_diffexp, else upregulated or downregulated
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$diffexp_res_lfc <- ifelse(is.na(glob_tecount_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE) | abs(glob_tecount_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE) < log2(1.5), "not_diffexp",
  ifelse(glob_tecount_res_int_Age_group_effect_blastema_vs_limb$log2FoldChange_MMSE > log2(1.5), "upregulated", "downregulated")
)
glob_tecount_res_int_Age_group_effect_blastema_vs_limb$diffexp_res_padj <- ifelse(is.na(glob_tecount_res_int_Age_group_effect_blastema_vs_limb$padj_MLE) | glob_tecount_res_int_Age_group_effect_blastema_vs_limb$padj_MLE > 0.05, "not_significant", "significant")

glob_tecount_res_int_Age_group_effect_blastema_vs_limb <- as_tibble(glob_tecount_res_int_Age_group_effect_blastema_vs_limb, rownames = "gene_id")

print(glob_tecount_res_int_Age_group_effect_blastema_vs_limb, width = Inf)

readr::write_tsv(glob_tecount_res_int_Age_group_effect_blastema_vs_limb,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_int_Age_group_effect_blastema_vs_limb.tsv")
)
readr::write_excel_csv(glob_tecount_res_int_Age_group_effect_blastema_vs_limb,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/results/glob_tecount_res_int_Age_group_effect_blastema_vs_limb.csv")
)
```

####### Volcano plots

```{r}
library(ggh4x)
library(EnhancedVolcano)
library(ggpubr)
library(dplyr)
library(readr)

dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes"))

```

```{r}

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))


```

######## Genes

```{r}

glob_tecount_res_genes_1 <- all_diffexp_results %>%
    filter(quant_tool == "TEcount",
           feature_biotype == "gene",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c( "subadult_blastemas_vs_subadult_limbs",
                                    "adult_blastemas_vs_adult_limbs",
                                    "adult_limbs_vs_subadult_limbs",
                                    "adults_blastemas_vs_subadults_blastemas",
                                    "age_effect_blastemas_vs_limbs")
    )
           

```

```{r}
library(ggh4x)
library(EnhancedVolcano)


# Associating colors to each type of value
keyvals <- ifelse(glob_tecount_res_genes_1$padj_MLE > 0.05 & glob_tecount_res_genes_1$log2FoldChange_MMSE > log2(1.5),
  "#c5c5c5",
  ifelse(glob_tecount_res_genes_1$padj_MLE > 0.05 & glob_tecount_res_genes_1$log2FoldChange_MMSE < -log2(1.5),
    "#c5c5c5",
    ifelse(glob_tecount_res_genes_1$padj_MLE > 0.05 & abs(glob_tecount_res_genes_1$log2FoldChange_MMSE) < log2(1.5),
      "#c5c5c5",
      ifelse(glob_tecount_res_genes_1$padj_MLE < 0.05 & abs(glob_tecount_res_genes_1$log2FoldChange_MMSE) < log2(1.5),
        "#c5c5c5",
        ifelse(glob_tecount_res_genes_1$padj_MLE < 0.05 & glob_tecount_res_genes_1$log2FoldChange_MMSE < -log2(1.5),
          "#0072B2",
          "#D55E00"
        )
      )
    )
  )
)

keyvals[is.na(keyvals)] <- "#c5c5c5"

names(keyvals)[keyvals == "#c5c5c5"] <- "up_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "mid_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "down_NS"
names(keyvals)[keyvals == "#0072B2"] <- "down_S"
names(keyvals)[keyvals == "#D55E00"] <- "up_S"
names(keyvals)[keyvals == "#c5c5c5"] <- "middle_S"

shapevals <- keyvals
names(shapevals) <- keyvals

shapevals <- ifelse(keyvals == "#c5c5c5", 21,
      ifelse(keyvals == "#0072B2", 25,
        ifelse(keyvals == "#D55E00", 24, 21)
      )
  )


# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- glob_tecount_res_genes_1 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE < -log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes1_2 <- glob_tecount_res_genes_1 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE > log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes2 <- glob_tecount_res_genes_1 %>%
                  group_by(diffexp_contrast) %>%
                  slice_min(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes3 <- glob_tecount_res_genes_1 %>%
                  group_by(diffexp_contrast) %>%
                  slice_max(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes <- unique(c(selectGenes1, selectGenes1_2, selectGenes2, selectGenes3))


#if element starts with "AMEX", remove from selectGenes
selectGenes <- selectGenes[!grepl("^AMEX", selectGenes)]
selectGenes <- selectGenes[!grepl("^LOC", selectGenes)]


# Creating volcano plot
volcanoes_glob_tecount_gene_1 <- EnhancedVolcano(

  as.data.frame(glob_tecount_res_genes_1),
  lab = as.data.frame(glob_tecount_res_genes_1)$SYMBOL,
  selectLab = selectGenes,
  drawConnectors = TRUE,
  widthConnectors = 0.3,
  arrowheads = FALSE,
  maxoverlapsConnectors = Inf,
  labSize = 2,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +


  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +

  facetted_pos_scales(
  y = list(
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_1)$padj_MLE)[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_1)$padj_MLE)[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_1)$padj_MLE)[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_1)$padj_MLE)[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_1)$padj_MLE)[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_gene_1.png"),
  plot = volcanoes_glob_tecount_gene_1,
  scale = 1, width = 400, height = 72, units = "mm", bg = "white", dpi = 1200
)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_gene_1.svg"),
  width = 15.748, height = 2.83465
)

volcanoes_glob_tecount_gene_1

dev.off()

```


glob_tecount_res_genes_2

```{r}

glob_tecount_res_genes_2 <- all_diffexp_results %>%
    filter(quant_tool == "TEcount",
           feature_biotype == "gene",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                          "adult_blastemas_vs_adult_nonlimbs",
                          "subadult_blastemas_vs_subadult_nonlimbs",
                          "adult_limbs_vs_adult_nonlimbs",
                          "subadult_limbs_vs_subadult_nonlimbs",
                          "age_effect_blastemas_vs_nonlimbs",
                          "age_effect_limbs_vs_nonlimbs")
    )

library(ggh4x)
library(EnhancedVolcano)


# Associating colors to each type of value
keyvals <- ifelse(glob_tecount_res_genes_2$padj_MLE > 0.05 & glob_tecount_res_genes_2$log2FoldChange_MMSE > log2(1.5),
  "#c5c5c5",
  ifelse(glob_tecount_res_genes_2$padj_MLE > 0.05 & glob_tecount_res_genes_2$log2FoldChange_MMSE < -log2(1.5),
    "#c5c5c5",
    ifelse(glob_tecount_res_genes_2$padj_MLE > 0.05 & abs(glob_tecount_res_genes_2$log2FoldChange_MMSE) < log2(1.5),
      "#c5c5c5",
      ifelse(glob_tecount_res_genes_2$padj_MLE < 0.05 & abs(glob_tecount_res_genes_2$log2FoldChange_MMSE) < log2(1.5),
        "#c5c5c5",
        ifelse(glob_tecount_res_genes_2$padj_MLE < 0.05 & glob_tecount_res_genes_2$log2FoldChange_MMSE < -log2(1.5),
          "#0072B2",
          "#D55E00"
        )
      )
    )
  )
)

keyvals[is.na(keyvals)] <- "#c5c5c5"

names(keyvals)[keyvals == "#c5c5c5"] <- "up_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "mid_NS"
names(keyvals)[keyvals == "#c5c5c5"] <- "down_NS"
names(keyvals)[keyvals == "#0072B2"] <- "down_S"
names(keyvals)[keyvals == "#D55E00"] <- "up_S"
names(keyvals)[keyvals == "#c5c5c5"] <- "middle_S"

shapevals <- keyvals
names(shapevals) <- keyvals

shapevals <- ifelse(keyvals == "#c5c5c5", 21,
      ifelse(keyvals == "#0072B2", 25,
        ifelse(keyvals == "#D55E00", 24, 21)
      )
  )


# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- glob_tecount_res_genes_2 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE < -log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes1_2 <- glob_tecount_res_genes_2 %>%
                  group_by(diffexp_contrast) %>%
                  filter(log2FoldChange_MMSE > log2(1.5)) %>%
                  slice_min(order_by = padj_MLE, n = 3) %>%
                  pull(SYMBOL)

selectGenes2 <- glob_tecount_res_genes_2 %>%
                  group_by(diffexp_contrast) %>%
                  slice_min(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes3 <- glob_tecount_res_genes_2 %>%
                  group_by(diffexp_contrast) %>%
                  slice_max(order_by = log2FoldChange_MMSE, n = 3) %>%
                  pull(SYMBOL)

selectGenes <- unique(c(selectGenes1, selectGenes1_2, selectGenes2, selectGenes3))


#if element starts with "AMEX", remove from selectGenes
selectGenes <- selectGenes[!grepl("^AMEX", selectGenes)]
selectGenes <- selectGenes[!grepl("^LOC", selectGenes)]


# Creating volcano plot
volcanoes_glob_tecount_gene_2 <- EnhancedVolcano(

  as.data.frame(glob_tecount_res_genes_2),
  lab = as.data.frame(glob_tecount_res_genes_2)$SYMBOL,
  selectLab = selectGenes,
  drawConnectors = TRUE,
  widthConnectors = 0.3,
  arrowheads = FALSE,
  maxoverlapsConnectors = Inf,
  labSize = 2,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +


  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +

  facetted_pos_scales(
  y = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_genes_2)$padj_MLE)[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))),
  x = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_genes_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_genes_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))))) +

  force_panelsizes(rows = unit(2, "in"), cols = unit(2, "in"))


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_gene_2.png"),
  plot = volcanoes_glob_tecount_gene_2, width = 50, height = 7.2, units = "cm", bg = "white", dpi = 600)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_gene_2.svg"),
  plot = volcanoes_glob_tecount_gene_2, width = 50, height = 7.2, units = "cm", bg = "white", fix_text_size = FALSE)

```


######## RE

```{r}

library(ggh4x)
library(EnhancedVolcano)
library(tidyr)
library(RColorBrewer)

```

Set 1

```{r}

glob_tecount_res_re_1 <- all_diffexp_results %>%
    filter(quant_tool == "TEcount",
           feature_biotype == "repetitive_element",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c( "subadult_blastemas_vs_subadult_limbs",
                                    "adult_blastemas_vs_adult_limbs",
                                    "adult_limbs_vs_subadult_limbs",
                                    "adults_blastemas_vs_subadults_blastemas",
                                    "age_effect_blastemas_vs_limbs")
)

# split gene_id column into three columns, based on the ":" separator, rename: c("RE_subfam", "RE_superfam", "RE_class")
glob_tecount_res_re_1 <- glob_tecount_res_re_1 %>%
  tidyr::separate(ID, c("RE_subfam", "RE_superfam", "RE_class"), sep = ":", remove = FALSE) %>%
  # if RE_subfam starts with "rnd-", replace with "Unknown"
  mutate(RE_superfam = ifelse(grepl("^rnd-", RE_superfam), "Unknown", RE_superfam)) %>%
  # replace Gypsy with Ty3, not necessarilly the full word
  mutate(RE_superfam = gsub("Gypsy", "Ty3", RE_superfam)) %>%
  # remove "?" in RE_class
  mutate(RE_class = gsub("\\?", "", RE_class))

# add shape, color and fill columns
glob_tecount_res_re_1 <- glob_tecount_res_re_1 %>%
  mutate(shape = ifelse(log2FoldChange_MMSE < -log2(1.5), 25, ifelse(log2FoldChange_MMSE > log2(1.5), 24, 21)),
  # for each unique value in RE_class, assign a color from a palette
        color = factor(RE_class, levels = sort(unique(RE_class)), labels = brewer.pal(n = length(unique(RE_class)), name = "Dark2")))

# Associating colors to each type of value
# convert to named vector
keyvals <- setNames(as.character(glob_tecount_res_re_1$color), as.character(glob_tecount_res_re_1$RE_class))
shapevals <- setNames(glob_tecount_res_re_1$shape, glob_tecount_res_re_1$shape)

# Creating volcano plot
volcanoes_glob_tecount_re_1 <- EnhancedVolcano(

  as.data.frame(glob_tecount_res_re_1),
  lab = NA,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +

  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +


  # change y scale of first plot, keep the second plot as it is
  
  facetted_pos_scales(
  y = list(
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_1)$padj_MLE)[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_1)$padj_MLE)[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_1)$padj_MLE)[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_1)$padj_MLE)[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
    diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_1)$padj_MLE)[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))
  ),
  x = list(
    diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "subadult_blastemas_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adult_blastemas_vs_adult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "adult_blastemas_vs_adult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adult_limbs_vs_subadult_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "adult_limbs_vs_subadult_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "adults_blastemas_vs_subadults_blastemas"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
    diffexp_contrast == "age_effect_blastemas_vs_limbs"  ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_1)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_1)$diffexp_contrast == "age_effect_blastemas_vs_limbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1)))
  )
) +

  force_panelsizes(
    rows = unit(2, "in"),
    cols = unit(2, "in")
  )

# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_re_1.png"),
  plot = volcanoes_glob_tecount_re_1,
  scale = 1, width = 400, height = 90, units = "mm", bg = "white", dpi = 1200
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_re_1.svg"),
  plot = volcanoes_glob_tecount_re_1,
  scale = 1, width = 400, height = 90, units = "mm", bg = "white", fix_text_size = FALSE
)

```

Set 2

```{r}

glob_tecount_res_re_2 <- all_diffexp_results %>%
    filter(quant_tool == "TEcount",
           feature_biotype == "repetitive_element",
           !is.na(padj_MLE),
           !is.na(log2FoldChange_MMSE),
           sample_group == "global",
           diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                          "adult_blastemas_vs_adult_nonlimbs",
                          "subadult_blastemas_vs_subadult_nonlimbs",
                          "adult_limbs_vs_adult_nonlimbs",
                          "subadult_limbs_vs_subadult_nonlimbs",
                          "age_effect_blastemas_vs_nonlimbs",
                          "age_effect_limbs_vs_nonlimbs")
)

# split gene_id column into three columns, based on the ":" separator, rename: c("RE_subfam", "RE_superfam", "RE_class")
glob_tecount_res_re_2 <- glob_tecount_res_re_2 %>%
  tidyr::separate(ID, c("RE_subfam", "RE_superfam", "RE_class"), sep = ":", remove = FALSE) %>%
  # if RE_subfam starts with "rnd-", replace with "Unknown"
  mutate(RE_superfam = ifelse(grepl("^rnd-", RE_superfam), "Unknown", RE_superfam)) %>%
  # replace Gypsy with Ty3, not necessarilly the full word
  mutate(RE_superfam = gsub("Gypsy", "Ty3", RE_superfam)) %>%
  # remove "?" in RE_class
  mutate(RE_class = gsub("\\?", "", RE_class))

# add shape, color and fill columns
glob_tecount_res_re_2 <- glob_tecount_res_re_2 %>%
  mutate(shape = ifelse(log2FoldChange_MMSE < -log2(1.5), 25, ifelse(log2FoldChange_MMSE > log2(1.5), 24, 21)),
  # for each unique value in RE_class, assign a color from a palette
        color = factor(RE_class, levels = sort(unique(RE_class)), labels = brewer.pal(n = length(unique(RE_class)), name = "Dark2")))

# Associating colors to each type of value
# convert to named vector
keyvals <- setNames(as.character(glob_tecount_res_re_2$color), as.character(glob_tecount_res_re_2$RE_class))
shapevals <- setNames(glob_tecount_res_re_2$shape, glob_tecount_res_re_2$shape)

# Creating volcano plot
volcanoes_glob_tecount_re_2 <- EnhancedVolcano(

  as.data.frame(glob_tecount_res_re_2),
  lab = NA,
  x = "log2FoldChange_MMSE",
  y = "padj_MLE",
  pCutoffCol = "padj_MLE",
  pCutoff = 0.05,
  FCcutoff = log2(1.5),
  title = NULL,
  pointSize = 1,
  ylab = expression(-log[10]("IHW-adj" ~ italic("p"))),
  xlab = bquote("ash-log"[2](FC)),
  axisLabSize = 11,
  shapeCustom = shapevals,
  subtitle = NULL,
  colCustom = keyvals,
  colAlpha = 4 / 5,
  caption = NULL,
  legendPosition = "right",
  legendLabSize = 8,
  legendIconSize = 2,
  legendDropLevels = FALSE,
  border = "full",
  borderWidth = 0.8
) +

  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    axis.title.y = element_text(face = "bold", size = 9)
  ) +
  guides(color = guide_legend(title = "DE gene category")) +

  facet_grid2(NULL, vars(diffexp_contrast), scales = "free", independent = "all" ) +
  theme(
    strip.text = element_text(size = 6, face = "bold"),
    strip.background = element_blank(),
    panel.spacing = unit(4, "mm")
  ) +
 
  facetted_pos_scales(
  y = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.7))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_y_continuous(limits = range(-log10(as.data.frame(glob_tecount_res_re_2)$padj_MLE)[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.05, 0.15)))),
  x = list(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "adult_limbs_vs_adult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "adult_limbs_vs_adult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.3))),
           diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_blastemas_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "age_effect_blastemas_vs_nonlimbs"], na.rm = TRUE), expansion(mult = c(0.1, 0.1))),
           diffexp_contrast == "age_effect_limbs_vs_nonlimbs" ~ scale_x_continuous(limits = range(as.data.frame(glob_tecount_res_re_2)$log2FoldChange_MMSE[as.data.frame(glob_tecount_res_re_2)$diffexp_contrast == "age_effect_limbs_vs_nonlimbs"], na.rm = TRUE), expand = expansion(mult = c(0.1, 0.1))))) +

  force_panelsizes(rows = unit(2, "in"), cols = unit(2, "in"))


# Saving plot
ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_re_2.png"),
  plot = volcanoes_glob_tecount_re_2,
  scale = 1, width = 55, height = 9, units = "cm", bg = "white", dpi = 600
)

ggsave(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/volcanoes/volcanoes_glob_tecount_re_2.svg"),
  plot = volcanoes_glob_tecount_re_2,
  scale = 1, width = 55, height = 9, units = "cm", bg = "white", fix_text_size = FALSE
)

```


####### Heatmap of gene/RE clustering

```{r}
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
library(DESeq2)

glob_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")

glob_tecount_ddsClean_de <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/glob_tecount_ddsClean_de.rds"))

# glob_tecount_ddsClean_de_vst <- vst(glob_tecount_ddsClean_de, blind = FALSE)

# saveRDS(glob_tecount_ddsClean_de_vst, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/DEA/glob_tecount_ddsClean_de_vst.rds")

glob_tecount_ddsClean_de_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/glob_tecount_ddsClean_de_vst.rds"))


library(readr)
library(dplyr)

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))
```

glob_full_heatmap1

```{r}
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
library(gridtext)
library(dplyr)
library(readr)
library(DESeq2)

num_keep <- 10

row_names <- all_diffexp_results %>%
                filter(quant_tool == "TEcount",
                sample_group == "global",
                diffexp_res_padj == "significant",
                diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs",
                                          "adult_blastemas_vs_adult_limbs",
                                          "adult_limbs_vs_subadult_limbs",
                                          "adults_blastemas_vs_subadults_blastemas",
                                          "age_effect_blastemas_vs_limbs")
                ) %>%
                group_by(diffexp_contrast, feature_biotype)

top <- row_names %>%
        slice_max(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

bottom <- row_names %>%
        slice_min(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

row_names <- unique(c(top, bottom))

glob_tecount_ddsClean_de_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/glob_tecount_ddsClean_de_vst.rds"))

# Get vst values
mat <- assay(glob_tecount_ddsClean_de_vst)[row_names, ]
mat <- mat - rowMeans(mat)

glob_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")
colnames(mat) <- glob_metadata$new_name_run[match(colnames(mat), glob_metadata$Run)]

# Get lfc values
lfc_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_1 <- as.matrix(lfc_vals_1)
colnames(lfc_vals_1) <- "A-B vs SA-B"
padj_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_1 <- as.matrix(padj_vals_1)
colnames(padj_vals_1) <- "A-B vs SA-B"

lfc_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_2 <- as.matrix(lfc_vals_2)
colnames(lfc_vals_2) <- "A-L vs SA-L"
padj_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_2 <- as.matrix(padj_vals_2)
colnames(padj_vals_2) <- "A-L vs SA-L"

lfc_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_blastemas_vs_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_3 <- as.matrix(lfc_vals_3)
colnames(lfc_vals_3) <- "Age effect (B vs L)"
padj_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_blastemas_vs_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_3 <- as.matrix(padj_vals_3)
colnames(padj_vals_3) <- "Age effect (B vs L)"

lfc_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_4 <- as.matrix(lfc_vals_4)
colnames(lfc_vals_4) <- "A-B vs A-L"
padj_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_4 <- as.matrix(padj_vals_4)
colnames(padj_vals_4) <- "A-B vs A-L"

lfc_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_5 <- as.matrix(lfc_vals_5)
colnames(lfc_vals_5) <- "SA-B vs SA-L"
padj_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_5 <- as.matrix(padj_vals_5)
colnames(padj_vals_5) <- "SA-B vs SA-L"


# baseMean is the same for all contrasts
# match of row_names to glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$ID
basemean <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(baseMean_MLE)
basemean <- as.matrix(basemean)
colnames(basemean) <- "baseMean"

# Prepare heatmap annotations
annotation <- as.data.frame(colData(glob_tecount_ddsClean_de_vst))
annotation <- annotation[c("Age_group", "Tissue_group", "Batch")]
rownames(annotation) <- colnames(mat)

main_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b",
      "Non_limb" = "#57452dff"
    ),
    Batch = c("Ruiz_Perez_et_al._n.d." = "#d36027ff", "Ye_et_al._2022" = "#0773b3ff", "Nowoshilow_et_al._2018" = "#f2e646ff", "Caballero_Perez_et_al._2018" = "#6b6a66ff", "Bryant_et_al._2017" = "#e79f26ff")
  ),
  annotation_label = gt_render(c("Age group", "Tissue", "Batch")),
  annotation_legend_param = list(
    Age_group = list(title = "Age group", at = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Tissue_group = list(title = "Tissue", at = c("Limb", "Limb_blastema", "Non_limb"), labels = c("Limb", "Limb blastema", "Non-limb"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Batch = list(title = "Batch", at = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"), labels = c("This study", "Ye et al. (2022)", "Nowoshilow et al. (2018)", "Caballero-Perez et al. (2018)", "Bryant et al. (2017)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold"))
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

boxplotCol <- HeatmapAnnotation(
  boxplot = anno_boxplot(
    mat,
    border = TRUE,
    gp = gpar(fill = "#CCCCCC"),
    pch = ".",
    size = unit(2, "mm"),
    axis = TRUE,
    axis_param = list(
      gp = gpar(fontsize = 12),
      side = "left"
    )
  ),
  annotation_width = unit(c(2.0), "cm"),
  which = "col"
)

ha <- HeatmapAnnotation(summary = anno_summary(
  gp = gpar(fill = 2),
  height = unit(2, "cm")
))


# maps values between b/w/r for min and max lfc values
lfc_col <- colorRamp2(
  c(
    min(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5)),
    0,
    max(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5))
  ),
  c("#0072B2", "white", "#D55E00")
)


# maps between 0% quantile, and 75% quantile of basemean values
basemean_col <- colorRamp2(c(quantile(basemean)[1], quantile(basemean)[4], quantile(basemean)[5]), c("white", "red", "darkred"))

# Get gene names
symbols <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(SYMBOL) %>%
  gsub("Gypsy_Chromoviridae_Vclade", "Chromoviridae:Ty3", .) %>%
  gsub("Gypsy_Mag", "Mag:Ty3", .) %>%
  gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", .) %>%
  # this replaces instances of "rnd-...:rnd-...:Unknown" with "rnd-...:Unknown"
  gsub("([^:]+):\\1:([^:]+)", "\\1:\\2", .)


# Generate main heatmap
heatmap_main <- Heatmap(mat[row_names, ],
  cluster_rows = T,
  show_row_names = F,
  column_labels = colnames(mat),
  column_names_gp = gpar(fontsize = 5),
  name = "Deviation from the mean (vst)",
  cluster_columns = T,
  top_annotation = main_ha,
  bottom_annotation = boxplotCol,
  width = unit(24, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9),
  # row_km = 3,
  # row_km_repeats = 100,
  split = 4,
  row_title = "Cluster %s",
  row_gap = unit(1.5, "mm"),
  row_title_gp = gpar(fontface = "bold")
)

##### TO SAVE CLUSTERS

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap"))

r.dend <- row_dend(heatmap_main) # If needed, extract row dendrogram
rcl.list <- row_order(heatmap_main) # Extract clusters (output is a list)


lapply(rcl.list, function(x) length(x)) # check/confirm size gene clusters

library(magrittr) # needed to load the pipe function '%%'

clu_df <- lapply(seq_along((rcl.list)), function(i) {
  out <- data.frame(
    ID = rownames(mat[rcl.list[[i]], , drop = FALSE]),
    glob_heatmap1_cluster = paste0("Cluster ", i),
    stringsAsFactors = FALSE
  )
  return(out)
}) %>% # pipe (forward) the output 'out' to the function rbind to create 'clu_df'
  do.call(rbind, .)

write.table(clu_df,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap1_clusters.txt"),
  sep = "\t", quote = F, row.names = FALSE
)

##### TO SAVE CLUSTERS

# Generate lfc_vals_1 column
heatmap_lfc_vals_1 <- Heatmap(lfc_vals_1, # row_labels = row_names,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  # cell_fun that adds a "*" if padj_vals_1 is yes
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_1[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = TRUE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_2 <- Heatmap(lfc_vals_2,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_2[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_3 <- Heatmap(lfc_vals_3,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_3[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_4 <- Heatmap(lfc_vals_4,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_4[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_5 <- Heatmap(lfc_vals_5,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_5[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)


# Generate baseMean column
heatmap_basemean <- Heatmap(basemean, # row_labels = row_names,
  cluster_rows = F, name = "baseMean", col = basemean_col,
  cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
    grid.text(round(basemean[i, j], 2), x, y,
      gp = gpar(fontsize = 9.5)
    )
  },
  width = unit(1.4, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Gene symbol annotation
gene_symb_ann <- rowAnnotation(gene_symbol = anno_text(symbols, gp = gpar(fontsize = 9)))

# Merge all heatmaps
final_heatmap <- heatmap_main + heatmap_lfc_vals_1 + heatmap_lfc_vals_2 + heatmap_lfc_vals_3 + heatmap_lfc_vals_4 + heatmap_lfc_vals_5 + heatmap_basemean + gene_symb_ann

png(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap1.png"),
  width = 50, height = 55, units = "cm", res = 600
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()


# ht_opt(RESET = TRUE)


# ht_opt(RESET = TRUE)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap1.svg"),
  width = 19.685, height = 21.6535
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()
```

glob_summary_heatmap1


```{r}
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
library(gridtext)
library(dplyr)
library(readr)
library(DESeq2)

num_keep <- 6

row_names <- all_diffexp_results %>%
                filter(quant_tool == "TEcount",
                sample_group == "global",
                diffexp_res_padj == "significant",
                feature_biotype == "gene",
                # keep only those whose SYMBOL do not start with "AMEX" or "LOC"
                !grepl("^AMEX|^LOC", SYMBOL),
                diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs",
                                          "adult_blastemas_vs_adult_limbs",
                                          "adult_limbs_vs_subadult_limbs",
                                          "adults_blastemas_vs_subadults_blastemas",
                                          "age_effect_blastemas_vs_limbs")
                ) %>%
                group_by(diffexp_contrast)

top <- row_names %>%
        slice_max(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

bottom <- row_names %>%
        slice_min(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

row_names <- unique(c(top, bottom))

glob_tecount_ddsClean_de_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/glob_tecount_ddsClean_de_vst.rds"))

# remove samples where Tissue_group is "Non_limb"
glob_tecount_ddsClean_de_vst_tmp <- glob_tecount_ddsClean_de_vst[, glob_metadata$Tissue_group != "Non_limb"]

# Get vst values
mat <- assay(glob_tecount_ddsClean_de_vst_tmp)[row_names, ]
mat <- mat - rowMeans(mat)

# to use Z-score instead
#mat <- t(scale(t(mat)))


glob_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")
colnames(mat) <- glob_metadata$new_name_run[match(colnames(mat), glob_metadata$Run)]

# Get lfc values
lfc_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_1 <- as.matrix(lfc_vals_1)
colnames(lfc_vals_1) <- "A-B vs SA-B"
padj_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_1 <- as.matrix(padj_vals_1)
colnames(padj_vals_1) <- "A-B vs SA-B"

lfc_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_2 <- as.matrix(lfc_vals_2)
colnames(lfc_vals_2) <- "A-L vs SA-L"
padj_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_2 <- as.matrix(padj_vals_2)
colnames(padj_vals_2) <- "A-L vs SA-L"

lfc_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_blastemas_vs_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_3 <- as.matrix(lfc_vals_3)
colnames(lfc_vals_3) <- "Age effect (B vs L)"
padj_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_blastemas_vs_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_3 <- as.matrix(padj_vals_3)
colnames(padj_vals_3) <- "Age effect (B vs L)"

lfc_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_4 <- as.matrix(lfc_vals_4)
colnames(lfc_vals_4) <- "A-B vs A-L"
padj_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_4 <- as.matrix(padj_vals_4)
colnames(padj_vals_4) <- "A-B vs A-L"

lfc_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_5 <- as.matrix(lfc_vals_5)
colnames(lfc_vals_5) <- "SA-B vs SA-L"
padj_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_5 <- as.matrix(padj_vals_5)
colnames(padj_vals_5) <- "SA-B vs SA-L"


# baseMean is the same for all contrasts
# match of row_names to glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$ID
basemean <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(baseMean_MLE)
basemean <- as.matrix(basemean)
colnames(basemean) <- "baseMean"

# Prepare heatmap annotations
annotation <- as.data.frame(colData(glob_tecount_ddsClean_de_vst_tmp))
annotation <- annotation[c("Age_group", "Tissue_group", "Batch")]
rownames(annotation) <- colnames(mat)

main_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b"),
    Batch = c("Ruiz_Perez_et_al._n.d." = "#d36027ff", "Ye_et_al._2022" = "#0773b3ff", "Nowoshilow_et_al._2018" = "#f2e646ff", "Caballero_Perez_et_al._2018" = "#6b6a66ff", "Bryant_et_al._2017" = "#e79f26ff")
  ),
  annotation_label = gt_render(c("Age group", "Tissue", "Batch")),
  annotation_legend_param = list(
    Age_group = list(title = "Age group", at = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Tissue_group = list(title = "Tissue", at = c("Limb", "Limb_blastema"), labels = c("Limb", "Limb blastema"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Batch = list(title = "Batch", at = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"), labels = c("This study", "Ye et al. (2022)", "Nowoshilow et al. (2018)", "Caballero-Perez et al. (2018)", "Bryant et al. (2017)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold"))
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

# maps values between b/w/r for min and max lfc values
lfc_col <- colorRamp2(
  c(
    min(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5)),
    0,
    max(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5))
  ),
  c("#0072B2", "white", "#D55E00")
)

# Get gene names
symbols <- all_diffexp_results %>%
  filter(diffexp_contrast == "adults_blastemas_vs_subadults_blastemas") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(SYMBOL) %>%
  gsub("Gypsy_Chromoviridae_Vclade", "Chromoviridae:Ty3", .) %>%
  gsub("Gypsy_Mag", "Mag:Ty3", .) %>%
  gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", .) %>%
  # this replaces instances of "rnd-...:rnd-...:Unknown" with "rnd-...:Unknown"
  gsub("([^:]+):\\1:([^:]+)", "\\1:\\2", .)

main_col <- colorRamp2(
  c(
    min(mat),
    0,
    max(mat)
  ),
  c("blue", "white", "red")
)

#main_col <- colorRamp2(seq(quantile(mat, 0.01), quantile(mat, 0.99), length = 3), c("blue", "white", "red"))

# Generate main heatmap
heatmap_main <- Heatmap(mat[row_names, ],
  cluster_rows = T,
  show_row_names = F,
  col = main_col,
  column_labels = colnames(mat),
  column_names_gp = gpar(fontsize = 5),
  name = "Deviation from the mean (vst)",
  cluster_columns = T,
  top_annotation = main_ha,
  width = unit(24, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9),
  # row_km = 3,
  # row_km_repeats = 100,
  split = 4,
  row_title = "Cluster %s",
  row_gap = unit(1.5, "mm"),
  row_title_gp = gpar(fontface = "bold")
)

##### TO SAVE CLUSTERS

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap"))

r.dend <- row_dend(heatmap_main) # If needed, extract row dendrogram
rcl.list <- row_order(heatmap_main) # Extract clusters (output is a list)


lapply(rcl.list, function(x) length(x)) # check/confirm size gene clusters

library(magrittr) # needed to load the pipe function '%%'

clu_df <- lapply(seq_along((rcl.list)), function(i) {
  out <- data.frame(
    ID = rownames(mat[rcl.list[[i]], , drop = FALSE]),
    glob_heatmap1_cluster = paste0("Cluster ", i),
    stringsAsFactors = FALSE
  )
  return(out)
}) %>% # pipe (forward) the output 'out' to the function rbind to create 'clu_df'
  do.call(rbind, .)

write.table(clu_df,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_summary_heatmap1_clusters.txt"),
  sep = "\t", quote = F, row.names = FALSE
)

##### TO SAVE CLUSTERS

# Generate lfc_vals_1 column
heatmap_lfc_vals_1 <- Heatmap(lfc_vals_1, # row_labels = row_names,
  cluster_rows = F, name = "ash-log2(FC)",
  col = lfc_col,
  # cell_fun that adds a "*" if padj_vals_1 is yes
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_1[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = TRUE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_2 <- Heatmap(lfc_vals_2,
  cluster_rows = F, name = "ash-log2(FC)",
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_2[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_3 <- Heatmap(lfc_vals_3,
  cluster_rows = F, name = "ash-log2(FC)",
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_3[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_4 <- Heatmap(lfc_vals_4,
  cluster_rows = F, name = "ash-log2(FC)",
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_4[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_5 <- Heatmap(lfc_vals_5,
  cluster_rows = F, name = "ash-log2(FC)",
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_5[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Gene symbol annotation
gene_symb_ann <- rowAnnotation(gene_symbol = anno_text(symbols, gp = gpar(fontsize = 9)))

# Merge all heatmaps
final_heatmap <- heatmap_main + heatmap_lfc_vals_1 + heatmap_lfc_vals_2 + heatmap_lfc_vals_3 + heatmap_lfc_vals_4 + heatmap_lfc_vals_5 + gene_symb_ann

png(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_summary_heatmap1.png"),
  width = 50, height = 30, units = "cm", res = 600
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()


# ht_opt(RESET = TRUE)


# ht_opt(RESET = TRUE)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_summary_heatmap1.svg"),
  width = 19.685, height = 21.6535
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()
```

plotCounts summary (Figure 4B)

```{r}	

num_keep <- 2

row_names <- all_diffexp_results %>%
                filter(quant_tool == "TEcount",
                sample_group == "global",
                diffexp_res_padj == "significant",
                feature_biotype == "gene",
                # keep only those whose SYMBOL do not start with "AMEX" or "LOC"
                !grepl("^AMEX|^LOC", SYMBOL),
                diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs",
                                          "adult_blastemas_vs_adult_limbs",
                                          "adult_limbs_vs_subadult_limbs",
                                          "adults_blastemas_vs_subadults_blastemas",
                                          "age_effect_blastemas_vs_limbs")
                ) %>%
                group_by(diffexp_contrast)

top <- row_names %>%
        slice_max(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

bottom <- row_names %>%
        slice_min(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

row_names <- unique(c(top, bottom))


num_keep <- 1

row_names_re <- all_diffexp_results %>%
                filter(quant_tool == "TEcount",
                sample_group == "global",
                diffexp_res_padj == "significant",
                feature_biotype == "repetitive_element",
                diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs",
                                          "adult_blastemas_vs_adult_limbs",
                                          "adult_limbs_vs_subadult_limbs",
                                          "adults_blastemas_vs_subadults_blastemas",
                                          "age_effect_blastemas_vs_limbs")
                ) %>%
                group_by(diffexp_contrast)

top <- row_names_re %>%
        slice_max(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

bottom <- row_names_re %>%
        slice_min(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

row_names <- unique(c(row_names, top, bottom))

# add the following to row_names (primers)
row_names <- unique(c(row_names, "AMEX60DD020567", "AMEX60DD039476", "AMEX60DD054790", "AMEX60DD006293", "AMEX60DD026361", "AMEX60DD026859"))

#remove XELAEV_18033243MG.1
row_names <- row_names[row_names != "AMEX60DD009277"]

names(row_names) <- row_names


library(DESeq2)

glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")

counts <- as.matrix(read.csv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/import/cntTable_all.csv"),
  header = TRUE, sep = ",", row.names = 1
))
colnames(counts) <- sub("^X", "", colnames(counts))
colnames(counts) <- gsub("\\.", "_", colnames(counts))

# filtering to remove other datasets and keep only native
glob_tecount_quant <- counts[, rownames(glob_metadata)]

all(rownames(glob_metadata) %in% colnames(glob_tecount_quant))

all(rownames(glob_metadata) == colnames(glob_tecount_quant))

# remove samples where Tissue_group is "Non_limb"
glob_tecount_quant <- glob_tecount_quant[, glob_metadata$Tissue_group != "Non_limb"]

# remove samples in glob_metadata where Tissue_group is "Non_limb"
glob_metadata <- glob_metadata[glob_metadata$Tissue_group != "Non_limb", ]

glob_tecount_dds <- DESeqDataSetFromMatrix(
  countData = glob_tecount_quant,
  colData = glob_metadata,
  design = ~ Batch + Age_group + Tissue_group + Age_group:Tissue_group
)


# Explicitly setting the factors levels
glob_tecount_dds$Age_group <- relevel(glob_tecount_dds$Age_group, ref = "Sub_adult")
glob_tecount_dds$Batch <- relevel(glob_tecount_dds$Batch, ref = "Ruiz_Perez_et_al._n.d.")
glob_tecount_dds$Tissue_group <- relevel(glob_tecount_dds$Tissue_group, ref = "Limb")

glob_tecount_de <- DESeq(glob_tecount_dds)

# apply plotCounts to every gene in the list


summary_geneCounts <- lapply(row_names, \(x) {
    y <- plotCounts(glob_tecount_de, x, intgroup = c("Tissue_group","Age_group"), returnData=TRUE)
    y$gene <- x
    return(y)
})

summary_geneCounts <- do.call(rbind, summary_geneCounts)

# Replace IDs in gene with gene symbol match in gene_set
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), header = TRUE, sep = "\t")

# only replace when gene starts with "AMEX"
summary_geneCounts$gene <- ifelse(grepl("^AMEX", summary_geneCounts$gene), gene_metadata$gene_symbol[match(summary_geneCounts$gene, gene_metadata$gene_id)], summary_geneCounts$gene)
summary_geneCounts$gene <- gsub("Gypsy_Chromoviridae_Vclade", "Chromoviridae:Ty3", summary_geneCounts$gene)
summary_geneCounts$gene <- gsub("Gypsy_Mag", "Mag:Ty3", summary_geneCounts$gene)
summary_geneCounts$gene <- gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", summary_geneCounts$gene)
# this replaces instances of "rnd-...:rnd-...:Unknown" with "rnd-...:Unknown"
summary_geneCounts$gene <- gsub("([^:]+):\\1:([^:]+)", "\\1:\\2", summary_geneCounts$gene)

summary_geneCounts_plot <- ggplot(summary_geneCounts, aes(x = Age_group, y = count, color = Tissue_group, shape = Age_group)) +
  scale_y_log10() +
  scale_shape_manual(values = c(15, 17)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position=position_jitterdodge(), size = 1) +
  # manual color scale
  theme_bw() +
  theme(
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 9),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 8, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank(),
    legend.position="bottom"
  )+
  scale_color_manual(values = c("#4dac26", "#d01c8b")) +
  stat_summary(fun.y = mean, geom="point", shape=23, size=2, position = position_dodge(width = 0.75)) +
  facet_wrap(~gene, scales="free_y", ncol = 4) +
  stat_summary(fun = mean, geom = "line", aes(group = Tissue_group), size = 0.5, linetype = "dashed", position = position_dodge(width = 0.75)) +
  ggh4x::force_panelsizes(
    rows = unit(2.5, "in"),
    cols = unit(2.5, "in")
  )

ggsave(filename = "../sruizperez/projects/axolotl/rnaseq/STAR/TEcount/global/DEA/plotCounts/summary_geneCounts.png",
plot = summary_geneCounts_plot,
width = 34, height = 54, units = "cm",
bg = "white", dpi = 600)



ggsave(filename = "../sruizperez/projects/axolotl/rnaseq/STAR/TEcount/global/DEA/plotCounts/summary_geneCounts.svg",
plot = summary_geneCounts_plot,
fix_text_size = FALSE,
width = 13.3858, height = 21.2598, units = "in", bg = "white")



```

glob_full_heatmap2

```{r}
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
library(gridtext)
library(dplyr)
library(readr)
library(DESeq2)

num_keep <- 10

row_names <- all_diffexp_results %>%
                filter(quant_tool == "TEcount",
                sample_group == "global",
                diffexp_res_padj == "significant",
                diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                                          "adult_blastemas_vs_adult_nonlimbs",
                                          "subadult_blastemas_vs_subadult_nonlimbs",
                                          "adult_limbs_vs_adult_nonlimbs",
                                          "subadult_limbs_vs_subadult_nonlimbs",
                                          "age_effect_blastemas_vs_nonlimbs",
                                          "age_effect_limbs_vs_nonlimbs")
                ) %>%
                group_by(diffexp_contrast, feature_biotype)

top <- row_names %>%
        slice_max(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

bottom <- row_names %>%
        slice_min(order_by = log2FoldChange_MMSE, n = num_keep) %>%
        pull(ID)

row_names <- unique(c(top, bottom))

glob_tecount_ddsClean_de_vst <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/glob_tecount_ddsClean_de_vst.rds"))

# Get vst values
mat <- assay(glob_tecount_ddsClean_de_vst)[row_names, ]
mat <- mat - rowMeans(mat)

glob_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"), header = TRUE, sep = "\t")
colnames(mat) <- glob_metadata$new_name_run[match(colnames(mat), glob_metadata$Run)]

# Get lfc values
lfc_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_1 <- as.matrix(lfc_vals_1)
colnames(lfc_vals_1) <- "A-NL vs SA-NL"
padj_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_1 <- as.matrix(padj_vals_1)
colnames(padj_vals_1) <- "A-NL vs SA-NL"

lfc_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_2 <- as.matrix(lfc_vals_2)
colnames(lfc_vals_2) <- "A-B vs A-NL"
padj_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_2 <- as.matrix(padj_vals_2)
colnames(padj_vals_2) <- "A-B vs A-NL"

lfc_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_3 <- as.matrix(lfc_vals_3)
colnames(lfc_vals_3) <- "SA-B vs SA-NL"
padj_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_3 <- as.matrix(padj_vals_3)
colnames(padj_vals_3) <- "SA-B vs SA-NL"

lfc_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_4 <- as.matrix(lfc_vals_4)
colnames(lfc_vals_4) <- "A-L vs A-NL"
padj_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_4 <- as.matrix(padj_vals_4)
colnames(padj_vals_4) <- "A-L vs A-NL"

lfc_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_5 <- as.matrix(lfc_vals_5)
colnames(lfc_vals_5) <- "SA-L vs SA-NL"
padj_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_5 <- as.matrix(padj_vals_5)
colnames(padj_vals_5) <- "SA-L vs SA-NL"

lfc_vals_6 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_blastemas_vs_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_6 <- as.matrix(lfc_vals_6)
colnames(lfc_vals_6) <- "Age effect (B vs NL)"
padj_vals_6 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_blastemas_vs_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_6 <- as.matrix(padj_vals_6)
colnames(padj_vals_6) <- "Age effect (B vs NL)"

lfc_vals_7 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_limbs_vs_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_7 <- as.matrix(lfc_vals_7)
colnames(lfc_vals_7) <- "Age effect (L vs NL)"
padj_vals_7 <- all_diffexp_results %>%
  filter(diffexp_contrast == "age_effect_limbs_vs_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_7 <- as.matrix(padj_vals_7)
colnames(padj_vals_7) <- "Age effect (L vs NL)"

# baseMean is the same for all contrasts
# match of row_names to adult_nonlimbs_vs_subadult_nonlimbs$ID
basemean <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(baseMean_MLE)
basemean <- as.matrix(basemean)
colnames(basemean) <- "baseMean"

# Prepare heatmap annotations
annotation <- as.data.frame(colData(glob_tecount_ddsClean_de_vst))
annotation <- annotation[c("Age_group", "Tissue_group", "Batch")]
rownames(annotation) <- colnames(mat)

main_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b",
      "Non_limb" = "#57452dff"
    ),
    Batch = c("Ruiz_Perez_et_al._n.d." = "#d36027ff", "Ye_et_al._2022" = "#0773b3ff", "Nowoshilow_et_al._2018" = "#f2e646ff", "Caballero_Perez_et_al._2018" = "#6b6a66ff", "Bryant_et_al._2017" = "#e79f26ff")
  ),
  annotation_label = gt_render(c("Age group", "Tissue", "Batch")),
  annotation_legend_param = list(
    Age_group = list(title = "Age group", at = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Tissue_group = list(title = "Tissue", at = c("Limb", "Limb_blastema", "Non_limb"), labels = c("Limb", "Limb blastema", "Non-limb"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Batch = list(title = "Batch", at = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"), labels = c("This study", "Ye et al. (2022)", "Nowoshilow et al. (2018)", "Caballero-Perez et al. (2018)", "Bryant et al. (2017)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold"))
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

boxplotCol <- HeatmapAnnotation(
  boxplot = anno_boxplot(
    mat,
    border = TRUE,
    gp = gpar(fill = "#CCCCCC"),
    pch = ".",
    size = unit(2, "mm"),
    axis = TRUE,
    axis_param = list(
      gp = gpar(fontsize = 12),
      side = "left"
    )
  ),
  annotation_width = unit(c(2.0), "cm"),
  which = "col"
)

ha <- HeatmapAnnotation(summary = anno_summary(
  gp = gpar(fill = 2),
  height = unit(2, "cm")
))


# maps values between b/w/r for min and max lfc values
lfc_col <- colorRamp2(
  c(
    min(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5, lfc_vals_6, lfc_vals_7)),
    0,
    max(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5, lfc_vals_6, lfc_vals_7))
  ),
  c("#0072B2", "white", "#D55E00")
)


# maps between 0% quantile, and 75% quantile of basemean values
basemean_col <- colorRamp2(c(quantile(basemean)[1], quantile(basemean)[4], quantile(basemean)[5]), c("white", "red", "darkred"))

# Get gene names
symbols <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_nonlimbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(ID %in% row_names) %>%
  arrange(match(ID, row_names)) %>%
  pull(SYMBOL) %>%
  gsub("Gypsy_Chromoviridae_Vclade", "Chromoviridae:Ty3", .) %>%
  gsub("Gypsy_Mag", "Mag:Ty3", .) %>%
  gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", .) %>%
  # this replaces instances of "rnd-...:rnd-...:Unknown" with "rnd-...:Unknown"
  gsub("([^:]+):\\1:([^:]+)", "\\1:\\2", .)


# Generate main heatmap
heatmap_main <- Heatmap(mat[row_names, ],
  cluster_rows = T,
  show_row_names = F,
  column_labels = colnames(mat),
  column_names_gp = gpar(fontsize = 5),
  name = "Deviation from the mean (vst)",
  cluster_columns = T,
  top_annotation = main_ha,
  bottom_annotation = boxplotCol,
  width = unit(24, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9),
  # row_km = 3,
  # row_km_repeats = 100,
  split = 4,
  row_title = "Cluster %s",
  row_gap = unit(1.5, "mm"),
  row_title_gp = gpar(fontface = "bold")
)

##### TO SAVE CLUSTERS

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap"))

r.dend <- row_dend(heatmap_main) # If needed, extract row dendrogram
rcl.list <- row_order(heatmap_main) # Extract clusters (output is a list)


lapply(rcl.list, function(x) length(x)) # check/confirm size gene clusters

library(magrittr) # needed to load the pipe function '%%'

clu_df <- lapply(seq_along((rcl.list)), function(i) {
  out <- data.frame(
    ID = rownames(mat[rcl.list[[i]], , drop = FALSE]),
    glob_heatmap2_cluster = paste0("Cluster ", i),
    stringsAsFactors = FALSE
  )
  return(out)
}) %>% # pipe (forward) the output 'out' to the function rbind to create 'clu_df'
  do.call(rbind, .)

write.table(clu_df,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap2_clusters.txt"),
  sep = "\t", quote = F, row.names = FALSE
)

##### TO SAVE CLUSTERS

# Generate lfc_vals_1 column
heatmap_lfc_vals_1 <- Heatmap(lfc_vals_1, # row_labels = row_names,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  # cell_fun that adds a "*" if padj_vals_1 is yes
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_1[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = TRUE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_2 <- Heatmap(lfc_vals_2,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_2[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_3 <- Heatmap(lfc_vals_3,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_3[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_4 <- Heatmap(lfc_vals_4,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_4[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_5 <- Heatmap(lfc_vals_5,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_5[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_6 <- Heatmap(lfc_vals_6,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_6[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_7 <- Heatmap(lfc_vals_7,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_7[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(1, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)


# Generate baseMean column
heatmap_basemean <- Heatmap(basemean, # row_labels = row_names,
  cluster_rows = F, name = "baseMean", col = basemean_col,
  cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
    grid.text(round(basemean[i, j], 2), x, y,
      gp = gpar(fontsize = 9.5)
    )
  },
  width = unit(1.4, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Gene symbol annotation
gene_symb_ann <- rowAnnotation(gene_symbol = anno_text(symbols, gp = gpar(fontsize = 9)))

# Merge all heatmaps
final_heatmap <- heatmap_main + heatmap_lfc_vals_1 + heatmap_lfc_vals_2 + heatmap_lfc_vals_3 + heatmap_lfc_vals_4 + heatmap_lfc_vals_5 + heatmap_lfc_vals_6 + heatmap_lfc_vals_7 + heatmap_basemean + gene_symb_ann

png(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap2.png"),
  width = 50, height = 65, units = "cm", res = 600
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()


# ht_opt(RESET = TRUE)


# ht_opt(RESET = TRUE)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap2.svg"),
  width = 19.685, height = 25.5906
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()
```


heatmap 2

```{r}
library(dplyr)
library(readr)

num_keep <- 10
row_names <- c()

for (contrast in c("subadult_limbs_vs_subadult_nonlimbs", "adult_limbs_vs_adult_nonlimbs", "subadult_blastemas_vs_subadult_nonlimbs", "adult_blastemas_vs_adult_nonlimbs", "subadult_blastemas_vs_subadult_limbs", "adult_blastemas_vs_adult_limbs")) {
  x <- read_tsv(paste0(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_"), contrast, ".tsv"))
  x_genes <- x %>%
    filter(feature_biotype == "gene", diffexp_res_padj == "significant") %>%
    arrange(desc(log2FoldChange_MMSE)) %>%
    pull(gene_id)
  x_re <- x %>%
    filter(feature_biotype == "repetitive_element", diffexp_res_padj == "significant") %>%
    arrange(desc(log2FoldChange_MMSE)) %>%
    pull(gene_id)
  row_names <- c(unique(c(row_names, head(x_genes, num_keep), tail(x_genes, num_keep), head(x_re, num_keep), tail(x_re, num_keep))))
}

# Get vst values
mat <- assay(glob_tecount_ddsClean_de_vst)[row_names, ]
mat <- mat - rowMeans(mat)

# Get lfc values
lfc_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_1 <- as.matrix(lfc_vals_1)
colnames(lfc_vals_1) <- "subadult_limbs_vs_subadult_nonlimbs"
padj_vals_1 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_1 <- as.matrix(padj_vals_1)
colnames(padj_vals_1) <- "subadult_limbs_vs_subadult_nonlimbs"

lfc_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_2 <- as.matrix(lfc_vals_2)
colnames(lfc_vals_2) <- "adult_limbs_vs_adult_nonlimbs"
padj_vals_2 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_2 <- as.matrix(padj_vals_2)
colnames(padj_vals_2) <- "adult_limbs_vs_adult_nonlimbs"

lfc_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_3 <- as.matrix(lfc_vals_3)
colnames(lfc_vals_3) <- "subadult_blastemas_vs_subadult_nonlimbs"
padj_vals_3 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_3 <- as.matrix(padj_vals_3)
colnames(padj_vals_3) <- "subadult_blastemas_vs_subadult_nonlimbs"

lfc_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_4 <- as.matrix(lfc_vals_4)
colnames(lfc_vals_4) <- "adult_blastemas_vs_adult_nonlimbs"
padj_vals_4 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_4 <- as.matrix(padj_vals_4)
colnames(padj_vals_4) <- "adult_blastemas_vs_adult_nonlimbs"

lfc_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_5 <- as.matrix(lfc_vals_5)
colnames(lfc_vals_5) <- "subadult_blastemas_vs_subadult_limbs"
padj_vals_5 <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_5 <- as.matrix(padj_vals_5)
colnames(padj_vals_5) <- "subadult_blastemas_vs_subadult_limbs"

lfc_vals_6 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(log2FoldChange_MMSE)
lfc_vals_6 <- as.matrix(lfc_vals_6)
colnames(lfc_vals_6) <- "adult_blastemas_vs_adult_limbs"
padj_vals_6 <- all_diffexp_results %>%
  filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(diffexp_res_padj)
padj_vals_6 <- as.matrix(padj_vals_6)
colnames(padj_vals_6) <- "adult_blastemas_vs_adult_limbs"

# baseMean is the same for all contrasts
# match of row_names to glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs$gene_id
basemean <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(baseMean_MLE)
basemean <- as.matrix(basemean)
colnames(basemean) <- "baseMean"


# To replace sample names
colnames(mat) <- glob_metadata$new_name_run[match(colnames(mat), glob_metadata$Run)]

# Prepare heatmap annotations
annotation <- as.data.frame(colData(glob_tecount_ddsClean_de_vst))
annotation <- annotation[c("Age_group", "Tissue_group", "Batch")]
rownames(annotation) <- colnames(mat)

main_ha <- HeatmapAnnotation(
  df = annotation,
  col = list(
    Age_group = c(
      "Adult" = "#252525",
      "Sub_adult" = "#bcbcbc"
    ),
    Tissue_group = c(
      "Limb" = "#4dac26",
      "Limb_blastema" = "#d01c8b",
      "Non_limb" = "#57452dff"
    ),
    Batch = c("Ruiz_Perez_et_al._n.d." = "#d36027ff", "Ye_et_al._2022" = "#0773b3ff", "Nowoshilow_et_al._2018" = "#f2e646ff", "Caballero_Perez_et_al._2018" = "#6b6a66ff", "Bryant_et_al._2017" = "#e79f26ff")
  ),
  annotation_label = gt_render(c("Age group", "Tissue", "Batch")),
  annotation_legend_param = list(
    Age_group = list(title = "Age group", at = c("Adult", "Sub_adult"), labels = c("Adult (8 years)", "Sub-adult (8 months)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Tissue_group = list(title = "Tissue", at = c("Limb", "Limb_blastema", "Non_limb"), labels = c("Limb", "Limb blastema", "Non-limb"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold")),
    Batch = list(title = "Batch", at = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017"), labels = c("Ruiz-Perez et al. (n.d.)", "Ye et al. (2022)", "Nowoshilow et al. (2018)", "Caballero-Perez et al. (2018)", "Bryant et al. (2017)"), labels_gp = gpar(fontsize = 9), title_gp = gpar(fontsize = 11, fontface = "bold"))
  ),
  gp = gpar(col = "white", lwd = 0.5),
  annotation_name_gp = gpar(fontsize = 11, fontface = "bold"),
  simple_anno_size = unit(0.5, "cm")
)

boxplotCol <- HeatmapAnnotation(
  boxplot = anno_boxplot(
    mat,
    border = TRUE,
    gp = gpar(fill = "#CCCCCC"),
    pch = ".",
    size = unit(2, "mm"),
    axis = TRUE,
    axis_param = list(
      gp = gpar(fontsize = 12),
      side = "left"
    )
  ),
  annotation_width = unit(c(2.0), "cm"),
  which = "col"
)

ha <- HeatmapAnnotation(summary = anno_summary(
  gp = gpar(fill = 2),
  height = unit(2, "cm")
))

# maps values between b/w/r for min and max lfc values
lfc_col <- colorRamp2(
  c(
    min(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5, lfc_vals_6)),
    0,
    max(c(lfc_vals_1, lfc_vals_2, lfc_vals_3, lfc_vals_4, lfc_vals_5, lfc_vals_6))
  ),
  c("#0072B2", "white", "#D55E00")
)


# maps between 0% quantile, and 75% quantile of basemean values
basemean_col <- colorRamp2(c(quantile(basemean)[1], quantile(basemean)[4], quantile(basemean)[5]), c("white", "red", "darkred"))

# Get gene names
symbols <- all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_limbs_vs_subadult_nonlimbs") %>%
  filter(quant_tool == "TEcount") %>%
  filter(gene_id %in% row_names) %>%
  arrange(match(gene_id, row_names)) %>%
  pull(gene_symbol)

# Generate main heatmap
heatmap_main <- Heatmap(mat[row_names, ],
  cluster_rows = T,
  show_row_names = F,
  column_labels = colnames(mat),
  column_names_gp = gpar(fontsize = 5),
  name = "Deviation from the mean (vst)",
  cluster_columns = T,
  top_annotation = main_ha,
  bottom_annotation = boxplotCol,
  width = unit(17, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9),
  # row_km = 3,
  # row_km_repeats = 100,
  split = 4,
  row_title = "Cluster %s",
  row_gap = unit(1.5, "mm"),
  row_title_gp = gpar(fontface = "bold")
)

##### TO SAVE CLUSTERS

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap"))

r.dend <- row_dend(heatmap_main) # If needed, extract row dendrogram
rcl.list <- row_order(heatmap_main) # Extract clusters (output is a list)


lapply(rcl.list, function(x) length(x)) # check/confirm size gene clusters

library(magrittr) # needed to load the pipe function '%%'

clu_df <- lapply(seq_along((rcl.list)), function(i) {
  out <- data.frame(
    GeneID = rownames(mat[rcl.list[[i]], , drop = FALSE]),
    Cluster = paste0("cluster", i),
    stringsAsFactors = FALSE
  )
  return(out)
}) %>% # pipe (forward) the output 'out' to the function rbind to create 'clu_df'
  do.call(rbind, .)

write.table(clu_df,
  file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_heatmap2_clusters.txt"),
  sep = "\t", quote = F, row.names = FALSE
)

##### TO SAVE CLUSTERS

# Generate lfc_vals_1 column
heatmap_lfc_vals_1 <- Heatmap(lfc_vals_1, # row_labels = row_names,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  # cell_fun that adds a "*" if padj_vals_1 is yes
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_1[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(0.5, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_2 <- Heatmap(lfc_vals_2,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_2[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(0.5, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_3 <- Heatmap(lfc_vals_3,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_3[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(0.5, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_4 <- Heatmap(lfc_vals_4,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_4[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(0.5, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_5 <- Heatmap(lfc_vals_5,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_5[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(0.5, "cm"),
  show_heatmap_legend = FALSE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

heatmap_lfc_vals_6 <- Heatmap(lfc_vals_6,
  cluster_rows = F, name = "ash-log2(FC)",
  bottom_annotation = ha,
  col = lfc_col,
  cell_fun = function(j, i, x, y, w, h, col) {
    ifelse(padj_vals_6[i] == "significant",
      grid.text("∗", x, y, gp = gpar(fontsize = 5)),
      grid.text("", x, y, )
    )
  },
  width = unit(0.5, "cm"),
  show_heatmap_legend = TRUE,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Generate baseMean column
heatmap_basemean <- Heatmap(basemean, # row_labels = row_names,
  cluster_rows = F, name = "baseMean", col = basemean_col,
  cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
    grid.text(round(basemean[i, j], 2), x, y,
      gp = gpar(fontsize = 9.5)
    )
  },
  width = unit(1.4, "cm"),
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.9)
)

# Gene symbol annotation
gene_symb_ann <- rowAnnotation(gene_symbol = anno_text(symbols, gp = gpar(fontsize = 9)))

# Merge all heatmaps
final_heatmap <- heatmap_main + heatmap_lfc_vals_1 + heatmap_lfc_vals_2 + heatmap_lfc_vals_3 + heatmap_lfc_vals_4 + heatmap_lfc_vals_5 + heatmap_lfc_vals_6 + heatmap_basemean + gene_symb_ann

png(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/heatmap_2.png"),
  width = 50, height = 55, units = "cm", res = 600
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()


# ht_opt(RESET = TRUE)


# ht_opt(RESET = TRUE)

svg(
  filename = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/heatmap_2.svg"),
  width = 19.685, height = 21.6535
)

draw(final_heatmap, merge_legends = TRUE)

dev.off()
```

###### Gene set enrichment analysis

```{r}
library(clusterProfiler)
library(enrichplot)
library(data.table)
library(DESeq2)
library(ggh4x)
library(stringr)
library(ggrepel)
```

######## Generating gene metadata

######## (GO information/TERM2GENE)
```{r}
# load gene metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# get columns gene_id and GO IDs
go_TERM2GENE <- gene_metadata[c("gene_id", "gene_goall")]
# remove rows where gene_goall is NA
go_TERM2GENE <- go_TERM2GENE[!is.na(go_TERM2GENE$gene_goall), ]

# expand, since some genes/proteins will have multiple assigned GO terms
go_TERM2GENE <- data.table(go_TERM2GENE)
go_TERM2GENE <- go_TERM2GENE[, list(gene_goall = unlist(strsplit(gene_goall, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_goall combination (to remove redundant pairs)
go_TERM2GENE <- distinct(go_TERM2GENE, gene_id, gene_goall, .keep_all = TRUE)

go_TERM2GENE <- go_TERM2GENE[, c(2, 1)]
```

######## KEGG metadata

```{r}
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# get columns gene_id and KEGG IDs
kegg_TERM2GENE <- gene_metadata[c("gene_id", "gene_keggpath")]
# remove rows where gene_keggpath is NA
kegg_TERM2GENE <- kegg_TERM2GENE[!is.na(kegg_TERM2GENE$gene_keggpath), ]

# expand, since some genes/proteins will have multiple assigned GO terms
kegg_TERM2GENE <- data.table(kegg_TERM2GENE)
kegg_TERM2GENE <- kegg_TERM2GENE[, list(gene_keggpath = unlist(strsplit(gene_keggpath, ","))), by = gene_id]

library(dplyr)
# Keep only one instance of gene_id and gene_keggpath combination (to remove redundant pairs)
kegg_TERM2GENE <- distinct(kegg_TERM2GENE, gene_id, gene_keggpath, .keep_all = TRUE)

kegg_TERM2GENE <- kegg_TERM2GENE[, c(2, 1)]
```

Group 1 of contrasts
```{r}

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

df <- all_diffexp_results %>%
  filter(quant_tool == "TEcount",
         sample_group == "global",
         feature_biotype == "gene",
         !is.na(padj_MLE),
         !is.na(log2FoldChange_MMSE),
         diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs",
                                 "age_effect_blastemas_vs_limbs",
                                 "subadult_blastemas_vs_subadult_limbs",
                                 "adults_blastemas_vs_subadults_blastemas",
                                 "adult_blastemas_vs_adult_limbs")) %>%
  group_by(diffexp_contrast) %>%
  # sort each group first by log2FoldChange_MMSE and then by padj_MLE
  arrange(desc(log2FoldChange_MMSE), padj_MLE, .by_group = TRUE) %>%
  # get gene vectors where name is ID and value is log2FoldChange_MMSE
  summarise(geneList = list(ID),
            log2FoldChange_MMSE = list(log2FoldChange_MMSE),
            .groups = "keep")
  
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = diffexp_contrast) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            # lapply so that geneList are names and log2FoldChange_MMSE the values
            lapply(function(x) {
              unlist(x$log2FoldChange_MMSE) %>% setNames(unlist(x$geneList))
            })


gsea_glob_tecount_1 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = go_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_glob_tecount_1

```
  
```{r}
library(GO.db)
library(AnnotationDbi)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_glob_tecount_1@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_glob_tecount_1@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_glob_tecount_1@compareClusterResult$ID[i]))
        gsea_glob_tecount_1@compareClusterResult$Description[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_tecount_1@compareClusterResult$ID[i], column = c("TERM"), keytype = "GOID")
        gsea_glob_tecount_1@compareClusterResult$ONTOLOGY[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_tecount_1@compareClusterResult$ID[i], column = c("ONTOLOGY"), keytype = "GOID")
      }
    },
    error = function(e) {
      cat("ERROR while mapping ", gsea_glob_tecount_1@compareClusterResult$ID[i], ":", conditionMessage(e), "\n")
    }
  )
}

# In gsea_glob_tecount_1@compareClusterResult$ONTOLOGY replace BP with Biological Process, MF with Molecular Function, CC with Cellular Component
gsea_glob_tecount_1@compareClusterResult$ONTOLOGY_full <- gsub("BP", "Biological process", gsea_glob_tecount_1@compareClusterResult$ONTOLOGY)
gsea_glob_tecount_1@compareClusterResult$ONTOLOGY_full <- gsub("MF", "Molecular function", gsea_glob_tecount_1@compareClusterResult$ONTOLOGY_full)
gsea_glob_tecount_1@compareClusterResult$ONTOLOGY_full <- gsub("CC", "Cellular component", gsea_glob_tecount_1@compareClusterResult$ONTOLOGY_full)

# To remove obsolete or removed GO terms
# Remove from gsea_glob_tecount_1@compareClusterResult all rows for which Description starts with GO:
gsea_glob_tecount_1@compareClusterResult <- gsea_glob_tecount_1@compareClusterResult[!grepl("^GO:", gsea_glob_tecount_1@compareClusterResult$Description), ]

# Add type column to gsea_glob_tecount_1@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_glob_tecount_1@compareClusterResult$type <- ifelse(gsea_glob_tecount_1@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_glob_tecount_1@compareClusterResult$type <- factor(gsea_glob_tecount_1@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_glob_tecount_1, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_go_glob_tecount_1.rds"))
# gsea_glob_tecount_1 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_go_glob_tecount_1.rds")

gsea_glob_tecount_1_tb <- as_tibble(gsea_glob_tecount_1@compareClusterResult)
write_tsv(gsea_glob_tecount_1_tb, file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_go_glob_tecount_1.tsv"))
```

Dotplot
```{r}
dp_gsea_glob_tecount_1 <- dotplot(gsea_glob_tecount_1,
  showCategory = 12,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
# switch = "y"
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_1.png"),
  plot = dp_gsea_glob_tecount_1, width = 40, height = 28, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_1.svg"),
  plot = dp_gsea_glob_tecount_1, width = 40, height = 28, units = "cm"
)
```

KEGG


```{r}
gsea_kegg_glob_tecount_1 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = kegg_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_kegg_glob_tecount_1
```
  
```{r}
# To map KEGG IDs to Pathway names (there's a 10-element limitation)
library(KEGGREST)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_kegg_glob_tecount_1@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_kegg_glob_tecount_1@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_kegg_glob_tecount_1@compareClusterResult$ID[i]))
        gsea_kegg_glob_tecount_1@compareClusterResult$Description[i] <- keggGet(gsea_kegg_glob_tecount_1@compareClusterResult$ID[i])[[1]]$NAME
      }
    },
    error = function(e) {
      cat("ERROR: ", gsea_kegg_glob_tecount_1@compareClusterResult$ID[i], conditionMessage(e), "\n")
    }
  )
}


# To remove obsolete or removed KEGG terms
# Remove from gsea_kegg_glob_tecount_1@compareClusterResult all rows for which Description starts with GO:
gsea_kegg_glob_tecount_1@compareClusterResult <- gsea_kegg_glob_tecount_1@compareClusterResult[!grepl("^ko", gsea_kegg_glob_tecount_1@compareClusterResult$Description), ]

# Add type column to gsea_kegg_glob_tecount_1@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_kegg_glob_tecount_1@compareClusterResult$type <- ifelse(gsea_kegg_glob_tecount_1@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_kegg_glob_tecount_1@compareClusterResult$type <- factor(gsea_kegg_glob_tecount_1@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_kegg_glob_tecount_1, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_kegg_glob_tecount_1.rds"))
# gsea_kegg_glob_tecount_1 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_kegg_glob_tecount_1.rds")

gsea_kegg_glob_tecount_1_tb <- as_tibble(gsea_kegg_glob_tecount_1@compareClusterResult)
write_tsv(gsea_kegg_glob_tecount_1_tb, file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_kegg_glob_tecount_1.tsv"))
```

######## Dotplot
```{r}
# add dummy column to gsea_kegg_glob_tecount_1@compareClusterResult
gsea_kegg_glob_tecount_1@compareClusterResult$dummy <- "same"

dp_gsea_kegg_glob_tecount_1 <- dotplot(gsea_kegg_glob_tecount_1,
  showCategory = 11,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  # add text labels to each dot
  # geom_text_repel(aes(label = Count), size = 3, direction = "x", nudge_x = ifelse(Count < 0.5, 0.5, -0.5)) +


  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(dummy ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )


ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_kegg_glob_tecount_1.png"),
  plot = dp_gsea_kegg_glob_tecount_1, width = 40, height = 20, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_kegg_glob_tecount_1.svg"),
  plot = dp_gsea_kegg_glob_tecount_1, width = 40, height = 20, units = "cm", dpi = 600
)


# make ggarrange of dp_gsea_nat_tecount and dp_gsea_kegg_glob_tecount_1
library(ggpubr)

# remove strip titles
dp_gsea_kegg_glob_tecount_1_tmp <- dp_gsea_kegg_glob_tecount_1 +
  theme(strip.text.x = element_blank())

# remove x axis text, title and ticks
dp_gsea_glob_tecount_1_tmp <- dp_gsea_glob_tecount_1 +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

dp_gsea_glob_tecount_arrange_1 <- ggarrange(dp_gsea_glob_tecount_1_tmp, dp_gsea_kegg_glob_tecount_1_tmp,
  ncol = 1, nrow = 2,
  common.legend = TRUE,
  align = "v",
  legend = "bottom",
  heights = c(56, 28)
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_arrange_1.png"),
  plot = dp_gsea_glob_tecount_arrange_1, width = 40, height = 25, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_arrange_1.svg"),
  plot = dp_gsea_glob_tecount_arrange_1, width = 40, height = 25, units = "cm"
)
```

Group 2 of contrasts

```{r}

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

df <- all_diffexp_results %>%
  filter(quant_tool == "TEcount",
         sample_group == "global",
         feature_biotype == "gene",
         !is.na(padj_MLE),
         !is.na(log2FoldChange_MMSE),
         diffexp_contrast %in% c("adult_nonlimbs_vs_subadult_nonlimbs",
                          "adult_blastemas_vs_adult_nonlimbs",
                          "subadult_blastemas_vs_subadult_nonlimbs",
                          "adult_limbs_vs_adult_nonlimbs",
                          "subadult_limbs_vs_subadult_nonlimbs",
                          "age_effect_blastemas_vs_nonlimbs",
                          "age_effect_limbs_vs_nonlimbs")) %>%

  group_by(diffexp_contrast) %>%
  # sort each group first by log2FoldChange_MMSE and then by padj_MLE
  arrange(desc(log2FoldChange_MMSE), padj_MLE, .by_group = TRUE) %>%
  # get gene vectors where name is ID and value is log2FoldChange_MMSE
  summarise(geneList = list(ID),
            log2FoldChange_MMSE = list(log2FoldChange_MMSE),
            .groups = "keep")
  
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = diffexp_contrast) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            # lapply so that geneList are names and log2FoldChange_MMSE the values
            lapply(function(x) {
              unlist(x$log2FoldChange_MMSE) %>% setNames(unlist(x$geneList))
            })


gsea_glob_tecount_2 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = go_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_glob_tecount_2

```
  


```{r}
library(GO.db)
library(AnnotationDbi)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_glob_tecount_2@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_glob_tecount_2@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_glob_tecount_2@compareClusterResult$ID[i]))
        gsea_glob_tecount_2@compareClusterResult$Description[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_tecount_2@compareClusterResult$ID[i], column = c("TERM"), keytype = "GOID")
        gsea_glob_tecount_2@compareClusterResult$ONTOLOGY[i] <-
          AnnotationDbi::mapIds(GO.db, keys = gsea_glob_tecount_2@compareClusterResult$ID[i], column = c("ONTOLOGY"), keytype = "GOID")
      }
    },
    error = function(e) {
      cat("ERROR while mapping ", gsea_glob_tecount_2@compareClusterResult$ID[i], ":", conditionMessage(e), "\n")
    }
  )
}

# In gsea_glob_tecount_2@compareClusterResult$ONTOLOGY replace BP with Biological Process, MF with Molecular Function, CC with Cellular Component
gsea_glob_tecount_2@compareClusterResult$ONTOLOGY_full <- gsub("BP", "Biological process", gsea_glob_tecount_2@compareClusterResult$ONTOLOGY)
gsea_glob_tecount_2@compareClusterResult$ONTOLOGY_full <- gsub("MF", "Molecular function", gsea_glob_tecount_2@compareClusterResult$ONTOLOGY_full)
gsea_glob_tecount_2@compareClusterResult$ONTOLOGY_full <- gsub("CC", "Cellular component", gsea_glob_tecount_2@compareClusterResult$ONTOLOGY_full)

# To remove obsolete or removed GO terms
# Remove from gsea_glob_tecount_2@compareClusterResult all rows for which Description starts with GO:
gsea_glob_tecount_2@compareClusterResult <- gsea_glob_tecount_2@compareClusterResult[!grepl("^GO:", gsea_glob_tecount_2@compareClusterResult$Description), ]

# Add type column to gsea_glob_tecount_2@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_glob_tecount_2@compareClusterResult$type <- ifelse(gsea_glob_tecount_2@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_glob_tecount_2@compareClusterResult$type <- factor(gsea_glob_tecount_2@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_glob_tecount_2, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_go_glob_tecount_2.rds"))
# gsea_glob_tecount_2 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_go_glob_tecount_2.rds")

gsea_glob_tecount_2_tb <- as_tibble(gsea_glob_tecount_2@compareClusterResult)
write_tsv(gsea_glob_tecount_2_tb, file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_go_glob_tecount_2.tsv"))
```

Dotplot
```{r}
dp_gsea_glob_tecount_2 <- dotplot(gsea_glob_tecount_2,
  showCategory = 12,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
# switch = "y"
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 6),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 8, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank(),
    legend.position = "bottom"
  )

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_2.png"),
  plot = dp_gsea_glob_tecount_2, width = 44, height = 38.1, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_2.svg"),
  plot = dp_gsea_glob_tecount_2, width = 44, height = 38.1, units = "cm"
)
```

KEGG

```{r}
gsea_kegg_glob_tecount_2 <- compareCluster(
  geneClusters = df_list,
  fun = "GSEA",
  TERM2GENE = kegg_TERM2GENE,
  exponent = 1,
  minGSSize = 10,
  maxGSSize = 500,
  eps = 0,
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  verbose = TRUE,
  seed = FALSE,
  by = "fgsea",
  nPermSimple = 1000000
)

gsea_kegg_glob_tecount_2
```
  
```{r}
# To map KEGG IDs to Pathway names (there's a 10-element limitation)
library(KEGGREST)

# For error catching see https://stackoverflow.com/a/14749552
for (i in seq_along(gsea_kegg_glob_tecount_2@compareClusterResult$ID)) {
  tryCatch(
    {
      if (gsea_kegg_glob_tecount_2@compareClusterResult$ID[i] != "-") {
        print(paste0("Mapping: ", gsea_kegg_glob_tecount_2@compareClusterResult$ID[i]))
        gsea_kegg_glob_tecount_2@compareClusterResult$Description[i] <- keggGet(gsea_kegg_glob_tecount_2@compareClusterResult$ID[i])[[1]]$NAME
      }
    },
    error = function(e) {
      cat("ERROR: ", gsea_kegg_glob_tecount_2@compareClusterResult$ID[i], conditionMessage(e), "\n")
    }
  )
}


# To remove obsolete or removed KEGG terms
# Remove from gsea_kegg_glob_tecount_2@compareClusterResult all rows for which Description starts with GO:
gsea_kegg_glob_tecount_2@compareClusterResult <- gsea_kegg_glob_tecount_2@compareClusterResult[!grepl("^ko", gsea_kegg_glob_tecount_2@compareClusterResult$Description), ]

# Add type column to gsea_kegg_glob_tecount_2@compareClusterResult, if NES > 0 then upregulated, if NES < 0 then downregulated. Make it a factor
gsea_kegg_glob_tecount_2@compareClusterResult$type <- ifelse(gsea_kegg_glob_tecount_2@compareClusterResult$NES > 0, "Activated", "Suppressed")
gsea_kegg_glob_tecount_2@compareClusterResult$type <- factor(gsea_kegg_glob_tecount_2@compareClusterResult$type, levels = c("Suppressed", "Activated"))

saveRDS(gsea_kegg_glob_tecount_2, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_kegg_glob_tecount_2.rds"))
# gsea_kegg_glob_tecount_2 <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_kegg_glob_tecount_2.rds")

gsea_kegg_glob_tecount_2_tb <- as_tibble(gsea_kegg_glob_tecount_2@compareClusterResult)
write_tsv(gsea_kegg_glob_tecount_2_tb, file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/gsea_kegg_glob_tecount_2.tsv"))
```

######## Dotplot
```{r}
# add dummy column to gsea_kegg_glob_tecount_2@compareClusterResult
gsea_kegg_glob_tecount_2@compareClusterResult$dummy <- "same"

dp_gsea_kegg_glob_tecount_2 <- dotplot(gsea_kegg_glob_tecount_2,
  showCategory = 11,
  size = "Count",
  x = "GeneRatio",
  label_format = 60
) +

  # add text labels to each dot
  # geom_text_repel(aes(label = Count), size = 3, direction = "x", nudge_x = ifelse(Count < 0.5, 0.5, -0.5)) +


  scale_size(
    limits = c(0, 500),
    breaks = c(50, 100, 200, 300),
    range = c(1, 8.5)
  ) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  scale_color_gradient(
    limits = c(0, 0.05),
    breaks = c(0.01, 0.02, 0.03, 0.04, 0.05),
    low = "red", high = "blue"
  ) +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(dummy ~ Cluster + type,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 6),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 8, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank(),
    legend.position = "bottom"
  )


ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_kegg_glob_tecount_2.png"),
  plot = dp_gsea_kegg_glob_tecount_2, width = 44, height = 38.1, units = "cm", dpi = 600
)

# save as svg
ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_kegg_glob_tecount_2.svg"),
  plot = dp_gsea_kegg_glob_tecount_2, width = 44, height = 38.1, units = "cm", dpi = 600
)


# make ggarrange of dp_gsea_nat_tecount and dp_gsea_kegg_glob_tecount_2
library(ggpubr)

# remove strip titles
dp_gsea_kegg_glob_tecount_2_tmp <- dp_gsea_kegg_glob_tecount_2 +
  theme(strip.text.x = element_blank())

# remove x axis text, title and ticks
dp_gsea_glob_tecount_2_tmp <- dp_gsea_glob_tecount_2 +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )

dp_gsea_glob_tecount_arrange_2 <- ggarrange(dp_gsea_glob_tecount_2_tmp, dp_gsea_kegg_glob_tecount_2_tmp,
  ncol = 1, nrow = 2,
  common.legend = TRUE,
  align = "v",
  legend = "bottom",
  heights = c(56, 28)
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_arrange_2.png"),
  plot = dp_gsea_glob_tecount_arrange_2, width = 44, height = 32, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/enrich/dp_gsea_glob_tecount_arrange_2.svg"),
  plot = dp_gsea_glob_tecount_arrange_2, width = 44, height = 32, units = "cm"
)

```




###### Network analysis with WGCNA

```{bash}
mamba activate srp_axolote

mkdir -p $PROJECT_DIR/rnaseq/STAR/TEcount/global/WGCNA/combat_seq

```

######## Batch correction with Combat-seq

```{r}
library(sva)

count_matrix <- as.matrix(read.csv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/import/cntTable_all.csv"),
  header = TRUE,
  sep = ",",
  row.names = 1,
  check.names = FALSE
))

colnames(count_matrix) <- gsub("\\-", "_", colnames(count_matrix))

glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"),
  sep = "\t",
  header = TRUE
)

count_matrix <- count_matrix[, rownames(glob_metadata)]

all(rownames(glob_metadata) %in% colnames(count_matrix))

all(rownames(glob_metadata) == colnames(count_matrix))

# Filtering lowly expressed genes
nrow(count_matrix)
keep <- rowSums(count_matrix >= 10) >= 3
count_matrix <- count_matrix[keep, ]
nrow(count_matrix)

# specify biological covariates of interest (to keep)
# see https://github.com/zhangyuqing/ComBat-seq/issues/18#issuecomment-774312140
# https://github.com/zhangyuqing/ComBat-seq/issues/25#issuecomment-857650933

batch <- as.numeric(factor(glob_metadata$Batch, levels = c("Ruiz_Perez_et_al._n.d.", "Ye_et_al._2022", "Nowoshilow_et_al._2018", "Caballero_Perez_et_al._2018", "Bryant_et_al._2017")))
cov1 <- as.numeric(factor(glob_metadata$Tissue_group, levels = c("Non_limb", "Limb", "Limb_blastema")))
cov2 <- as.numeric(factor(glob_metadata$Age_group, levels = c("Sub_adult", "Adult")))
covar_mat <- cbind(cov1, cov2)

adjusted_counts <- ComBat_seq(count_matrix, batch = batch, group = cov1, covar_mod = covar_mat)

# save as tsv
write.table(adjusted_counts,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/combat_seq/combatseq_tecount_all_samples.tsv"),
  sep = "\t", quote = FALSE
)
```

```{r}
library(DESeq2)

all(rownames(glob_metadata) %in% colnames(adjusted_counts))

all(rownames(glob_metadata) == colnames(adjusted_counts))

pre_dds <- DESeqDataSetFromMatrix(
  countData = adjusted_counts,
  colData = glob_metadata,
  design = ~ Age_group + Tissue_group + Age_group:Tissue_group
)


# Explicitly setting the factors levels
pre_dds$Age_group <- relevel(pre_dds$Age_group, ref = "Sub_adult")
pre_dds$Tissue_group <- relevel(pre_dds$Tissue_group, ref = "Non_limb")

dds_vst <- vst(pre_dds)
expr_normalized <- assay(dds_vst)
input_mat <- t(expr_normalized)

write.table(input_mat,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/vst_combatseq_tecount_all_samples.tsv"),
  sep = "\t", quote = FALSE
)

# input_mat <- read.table(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/vst_combatseq_tecount_all_samples.tsv"),
#             sep = "\t", header = TRUE, row.names = 1)
```

```{r}
library(WGCNA)

allowWGCNAThreads()

# Choose a set of soft-thresholding powers
powers <- c(c(1:10), seq(from = 12, to = 30, by = 2))

# Call the network topology analysis function
sft2 <- pickSoftThreshold(
  input_mat, # <= Input data
  # blockSize = 30,
  powerVector = powers,
  networkType = "signed",
  verbose = 5
)


readr::write_tsv(sft2$fitIndices, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/pickSoftThreshold_fitIndices.tsv"))

readr::write_tsv(data.frame(powerEstimate = sft2$powerEstimate), file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/pickSoftThreshold_powerEstimate.tsv"))


png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/plot_sft2.png"),
  width = 18, height = 12.7, units = "cm", res = 600
)

par(mfrow = c(1, 2))
cex1 <- 0.9
plot(sft2$fitIndices[, 1],
  -sign(sft2$fitIndices[, 3]) * sft2$fitIndices[, 2],
  xlab = "Soft Threshold (power)",
  ylab = "Scale Free Topology Model Fit, signed R^2",
  main = paste("Scale independence")
)
text(sft2$fitIndices[, 1],
  -sign(sft2$fitIndices[, 3]) * sft2$fitIndices[, 2],
  labels = powers, cex = cex1, col = "red"
)
abline(h = 0.90, col = "red")
plot(sft2$fitIndices[, 1],
  sft2$fitIndices[, 5],
  xlab = "Soft Threshold (power)",
  ylab = "Mean Connectivity",
  type = "n",
  main = paste("Mean connectivity")
)
text(sft2$fitIndices[, 1],
  sft2$fitIndices[, 5],
  labels = powers,
  cex = cex1, col = "red"
)


dev.off()
```

```{r}
picked_power <- 8
temp_cor <- cor
cor <- WGCNA::cor # Force it to use WGCNA cor function (fix a namespace conflict issue)

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA"))

netwk <- blockwiseModules(input_mat, # <= input here

  # == Adjacency Function ==
  power = picked_power, # <= power here
  networkType = "signed",

  # == Tree and Block Options ==
  deepSplit = 2,
  pamRespectsDendro = F,
  # detectCutHeight = 0.75,
  minModuleSize = 30,

  # increased maxBlockSize (https://support.bioconductor.org/p/127805/)
  maxBlockSize = 52000,

  # == Module Adjustments ==
  reassignThreshold = 0,
  mergeCutHeight = 0.25,

  # == TOM == Archive the run results in TOM file (saves time)
  saveTOMs = T,
  saveTOMFileBase = "global_blockwiseModules_TOM",

  # == Output Options
  numericLabels = T,
  verbose = 3
)

saveRDS(netwk, file = file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/netwk_blockwiseModules_output.rds"))
# netwk <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/netwk_blockwiseModules_output.rds")

cor <- temp_cor # Return cor function to original namespace
```

```{r}
png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/DendroAndColors_tmp.png"),
  width = 20, height = 8, units = "cm", res = 600
)

# Convert labels to colors for plotting
mergedColors <- labels2colors(netwk$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  dendro = netwk$dendrograms[[1]],
  colors = mergedColors[netwk$blockGenes[[1]]],
  groupLabels = "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/DendroAndColors_tmp2.svg"),
  width = 7.87402, height = 3.14961
)

# Convert labels to colors for plotting
mergedColors <- labels2colors(netwk$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(
  dendro = netwk$dendrograms[[1]],
  colors = mergedColors[netwk$blockGenes[[1]]],
  groupLabels = "Module colors",
  dendroLabels = FALSE,
  hang = 0.03,
  addGuide = TRUE,
  guideHang = 0.05
)

dev.off()
```

Networtk heatmap

```{r}
library(WGCNA)
library(gplots)

# Load the network topology file
load(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/global_blockwiseModules_TOM-block.1.RData"))
# this loads "TOM"

netwk <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/blockwiseModules.rds"))

mergedColors <- labels2colors(netwk$colors)
```

with WGCNA functions
```{r}
# Calculate topological overlap anew: this could be done more efficiently by saving the TOM
# calculated during module detection:
dissTOM <- 1 - TOM

# Transform dissTOM with a power to make moderately strong connections more visible in the heatmap
plotTOM <- dissTOM^7

plotTOM <- as.matrix(plotTOM)

# Set diagonal to NA for a nicer plot
diag(plotTOM) <- NA

# my_palette <- colorRampPalette(c("darkred", "red", "orange", "yellow","lemonchiffon"))(n = 250)

# legend_vals <- seq(from = min(plotTOM, na.rm = TRUE), to = max(plotTOM, na.rm = TRUE), length.out = 5)
# legend_vals_r <- round(legend_vals,2)

png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/tom_plot_2.png"),
  width = 18, height = 18, units = "cm", res = 1200
)

WGCNA::TOMplot(plotTOM, netwk$dendrograms[[1]], mergedColors, main = "Network heatmap plot, all genes") # , col=my_palette)

# Plot a legend in bottom right part of heatmap
# legend(x = "bottomright", legend = legend_vals_r,
#       cex = 0.8, fill = colorRampPalette(c("darkred", "red", "orange", "yellow","lemonchiffon"))(n = 5))

dev.off()
```

```{r}
module_df <- data.frame(
  ID = names(netwk$colors),
  wgcna_module = labels2colors(netwk$colors)
)

module_df[1:5, ]
#          gene_id colors
# 1 AMEX60DD000001 yellow
# 2 AMEX60DD000002   grey
# 3 AMEX60DD000003 yellow
# 4 AMEX60DD000004   grey
# 5 AMEX60DD000005   grey

readr::write_tsv(module_df,
  file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/genes_or_REs_and_wgcna_modules.tsv")
)

# module_df <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/genes_or_REs_and_wgcna_modules.tsv"))
```

Upset plot

```{r}
library(readr)
library(dplyr)
library(tidyr)
library(ComplexHeatmap)


module_df <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/genes_or_REs_and_wgcna_modules.tsv"))

# get list of genes per module
genes_per_module <- module_df %>%
  group_by(wgcna_module) %>%
  summarise(gene_list = list(ID)) %>%
  tibble::deframe()

all_diffexp_results <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

diffexp_results <- all_diffexp_results %>%
  # keep only sample_group == "global"
  filter(sample_group == "global") %>%
  filter(diffexp_res_lfc != "not_diffexp" & diffexp_res_padj != "not_significant")

genes_per_contrast <- diffexp_results %>%
  group_by(diffexp_contrast) %>%
  # get list of genes per contrast
  summarise(gene_list = list(ID)) %>%
  tibble::deframe()

gene_or_re <- diffexp_results %>%
  group_by(feature_biotype) %>%
  # get list of genes per feature_biotype, remove duplicates
  summarise(gene_list = list(unique(ID))) %>%
  tibble::deframe()

# change names to DEGs and DEREs
names(gene_or_re) <- c("DEGs", "DEREs")

# merge lists
genes_list <- c(genes_per_module, gene_or_re) #


genes_comb_mat <- make_comb_mat(genes_list)

dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/upset"))

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/upset/degs_and_wgcna.png"),
  width = 14, height = 20, units = "cm", res = 600
)
UpSet(t(genes_comb_mat),
  # set order
  set_order = c(
    "DEGs", "DEREs", "green", "greenyellow", "tan", "pink", "black", "yellow", "blue", "cyan",
    "magenta", "red", "salmon", "midnightblue", "purple", "brown", "turquoise", "grey"
  ),
  top_annotation = upset_top_annotation(t(genes_comb_mat), add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(t(genes_comb_mat), add_numbers = TRUE)
)
dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/upset/degs_and_wgcna.svg"),
  width = 5.51181, height = 7.87402
)
UpSet(t(genes_comb_mat),
  # set order
  set_order = c(
    "DEGs", "DEREs", "green", "greenyellow", "tan", "pink", "black", "yellow", "blue", "cyan",
    "magenta", "red", "salmon", "midnightblue", "purple", "brown", "turquoise", "grey"
  ),
  top_annotation = upset_top_annotation(t(genes_comb_mat), add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(t(genes_comb_mat), add_numbers = TRUE)
)
dev.off()
```

```{r}
# Get Module Eigengenes per cluster
MEs0 <- moduleEigengenes(input_mat, mergedColors)$eigengenes

# Reorder modules so similar modules are next to each other
MEs0 <- orderMEs(MEs0)

sample_data <- glob_metadata[, c("Tissue_group", "Age_group")]

nGenes <- ncol(input_mat)
nSamples <- nrow(input_mat)


# convert sample_metadata$Tissue_group to binary columns
sample_data <- cbind(sample_data, model.matrix(~ Tissue_group - 1, data = sample_data))

# convert sample_metadata$Age_group to binary columns
sample_data <- cbind(sample_data, model.matrix(~ Age_group - 1, data = sample_data))

b_sample_data <- sample_data[, c("Tissue_groupLimb", "Tissue_groupLimb_blastema", "Tissue_groupNon_limb", "Age_groupAdult", "Age_groupSub_adult")]

moduleTraitCor <- cor(MEs0, b_sample_data, use = "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nSamples)

# print how many Tissue_groupLimb_blastema are 1
print(sum(b_sample_data$Tissue_groupLimb_blastema))

png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/heatmap_module_trait.png"),
  width = 18, height = 12.7, units = "cm", res = 600
)

# Will display correlations and their p-values
textMatrix <- paste(signif(moduleTraitCor, 2), "\n(",
  signif(moduleTraitPvalue, 1), ")",
  sep = ""
)
dim(textMatrix) <- dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3))
# Display the correlation values within a heatmap plot
labeledHeatmap(
  Matrix = moduleTraitCor,
  xLabels = names(b_sample_data),
  yLabels = names(MEs0),
  ySymbols = names(MEs0),
  colorLabels = FALSE,
  colors = greenWhiteRed(50),
  textMatrix = textMatrix,
  setStdMargins = FALSE,
  cex.text = 0.5,
  zlim = c(-1, 1),
  main = paste("Module-trait relationships")
)

dev.off()
```

```{r}
# create now with complexheatmap
library(ComplexHeatmap)
library(viridisLite)
library(RColorBrewer)

# get count per module in module_df$wgcna_module
module_count <- as.data.frame(table(module_df$wgcna_module))


col_list <- as.character(substring(rownames(moduleTraitCor), 3))
names(col_list) <- substring(rownames(moduleTraitCor), 3)

# reorder module_count by Var1 to the same order as col_list
module_count <- module_count[match(col_list, module_count$Var1), ]
module_labels <- paste(module_count$Freq)

# annotation is rowname without ME
row_ha <- rowAnnotation(
  Module = anno_text(module_labels,
    gp = gpar(
      fill = col_list,
      col = "white",
      border = "black",
      fontsize = 9
    ),
    rot = 90,
    location = 0.5,
    just = "center",
    width = unit(0.5, "cm")
  ),
  show_legend = FALSE
)


# remove leading "ME" from row names
moduleTraitCor_new <- moduleTraitCor
rownames(moduleTraitCor_new) <- substring(rownames(moduleTraitCor_new), 3)
# add a counter to row names
rownames(moduleTraitCor_new) <- paste0(
  "Module ",
  seq_len(nrow(moduleTraitCor_new)),
  " (",
  rownames(moduleTraitCor_new),
  ")"
)
# fix colnames
colnames(moduleTraitCor_new) <- c("Limb", "Blastema", "Non-limb", "Adult", "Sub-adult")


moduleTraitPvalue_new <- moduleTraitPvalue
rownames(moduleTraitPvalue_new) <- substring(rownames(moduleTraitPvalue_new), 3)
rownames(moduleTraitPvalue_new) <- paste0(
  "Module ",
  seq_len(nrow(moduleTraitPvalue_new)),
  " (",
  rownames(moduleTraitPvalue_new),
  ")"
)
colnames(moduleTraitPvalue_new) <- c("Limb", "Blastema", "Non-limb", "Adult", "Sub-adult")

moduleTraitPadj <- p.adjust(moduleTraitPvalue_new, method = "BH")
dim(moduleTraitPadj) <- dim(moduleTraitCor_new)
rownames(moduleTraitPadj) <- rownames(moduleTraitPvalue_new)
colnames(moduleTraitPadj) <- colnames(moduleTraitPvalue_new)

# save
write_tsv(as_tibble(moduleTraitCor_new, rownames = "module"), file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/moduleTraitCor.tsv"))
write_tsv(as_tibble(moduleTraitPvalue_new, rownames = "module"), file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/moduleTraitPvalue.tsv"))
write_tsv(as_tibble(moduleTraitPadj, rownames = "module"), file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/moduleTraitPadj.tsv"))
```

with nominal p-value
```{r}
moduleTraitCor_cheatmap <- Heatmap(moduleTraitCor_new,
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_side = "right",
  width = unit(6.7, "cm"),

  # cell fun add both moduleTraitCor_new value and moduleTraitPvalue_new
  cell_fun = function(j, i, x, y, width, height, fill) {
    grid.text(
      sprintf(
        "%.2f\n",
        moduleTraitCor_new[i, j]
      ),
      x, y,
      gp = gpar(fontsize = 8.5, col = "black")
    )

    grid.text(
      sprintf(
        "\n%s",
        ifelse(moduleTraitPvalue_new[i, j] > 0.001,
          format(moduleTraitPvalue_new[i, j], digits = 3, nsmall = 3),
          ifelse(is.finite(moduleTraitPvalue_new[i, j]) && moduleTraitPvalue_new[i, j] < 10^(-200),
            "< 10^-200",
            format(moduleTraitPvalue_new[i, j], digits = 2, nsmall = 2, scientific = TRUE)
          )
        )
      ),
      x, y,
      gp = gpar(fontsize = 6.5, col = "black")
    )
  },
  right_annotation = row_ha,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.5),
  heatmap_legend_param = list(legend_height = unit(5, "cm")),
  column_split = c("Tissue", "Tissue", "Tissue", "Age group", "Age group"),
  column_title = NULL
)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/moduleTraitCor_cheatmap.png"),
  width = 18, height = 18, units = "cm", res = 600
)
draw(moduleTraitCor_cheatmap, merge_legends = TRUE)
dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/moduleTraitCor_cheatmap.svg"),
  width = 7.08661, height = 7.08661
)
draw(moduleTraitCor_cheatmap, merge_legends = TRUE)
dev.off()
```

with adjusted p-value
```{r}
moduleTraitCor_cheatmap <- Heatmap(moduleTraitCor_new,
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  row_names_side = "right",
  width = unit(6.7, "cm"),

  # cell fun add both moduleTraitCor_new value and moduleTraitPadj
  cell_fun = function(j, i, x, y, width, height, fill) {
    grid.text(
      sprintf(
        "%.2f\n",
        moduleTraitCor_new[i, j]
      ),
      x, y,
      gp = gpar(fontsize = 8.5, col = "black")
    )

    grid.text(
      sprintf(
        "\n%s",
        ifelse(moduleTraitPadj[i, j] > 0.001,
          format(moduleTraitPadj[i, j], digits = 3, nsmall = 3),
          ifelse(is.finite(moduleTraitPadj[i, j]) && moduleTraitPadj[i, j] < 10^(-200),
            "< 10^-200",
            format(moduleTraitPadj[i, j], digits = 2, nsmall = 2, scientific = TRUE)
          )
        )
      ),
      x, y,
      gp = gpar(fontsize = 6.5, col = "black")
    )
  },
  right_annotation = row_ha,
  border_gp = gpar(col = "black", lwd = 0.7),
  rect_gp = gpar(col = "white", lwd = 0.5),
  heatmap_legend_param = list(legend_height = unit(5, "cm")),
  column_split = c("Tissue", "Tissue", "Tissue", "Age group", "Age group"),
  column_title = NULL
)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/moduleTraitCor_cheatmap_adj.png"),
  width = 18, height = 18, units = "cm", res = 600
)
draw(moduleTraitCor_cheatmap, merge_legends = TRUE)
dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/moduleTraitCor_cheatmap_adj.svg"),
  width = 7.08661, height = 7.08661
)
draw(moduleTraitCor_cheatmap, merge_legends = TRUE)
dev.off()
```

Enrichment analysis

```{r}
library(data.table)
library(dplyr)
library(readr)
library(tidyr)
library(tibble)
library(ggplot2)
library(ggh4x)
library(stringr)

# load gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# see why not using clusterProfiler: https://github.com/YuLab-SMU/clusterProfiler/issues/636
# create a list of pathways suitable for fgseaORA:
gene_metadata_for_fgseaORA <- dplyr::as_tibble(gene_metadata[, c("gene_id", "gene_goall")]) %>%
  # remove rows where gene_goall is NA or ""
  na.omit() %>%
  dplyr::filter(gene_goall != "") %>%
  # gene_goall is a list of GO terms separated by ",", split them into separate rows
  tidyr::separate_rows(gene_goall, sep = ",") %>%
  # get a list where each gene_goall is a name, and it contains a vector of all gene_id that have this GO term
  dplyr::group_by(gene_goall) %>%
  dplyr::summarise(gene_id = list(gene_id)) %>%
  deframe()
```

```{r}
dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/enrich"))

count_matrix <- as.matrix(read.csv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/import/cntTable_all.csv"),
  header = TRUE,
  sep = ",",
  row.names = 1,
  check.names = FALSE
))

colnames(count_matrix) <- gsub("\\-", "_", colnames(count_matrix))

glob_metadata <- read.delim(file.path(PROJECT_DIR, "/metadata/glob_metadata.txt"),
  sep = "\t",
  header = TRUE
)

count_matrix <- count_matrix[, rownames(glob_metadata)]

all(rownames(glob_metadata) %in% colnames(count_matrix))

all(rownames(glob_metadata) == colnames(count_matrix))


# read module-trait correlation data
moduleTraitCor <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/moduleTraitCor.tsv"))
moduleTraitPadj <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/moduleTraitPadj.tsv"))

module_df <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/genes_and_modcolors.tsv"))

fora_res_tb_all <- tibble(
  pathway = character(),
  pval = numeric(),
  padj = numeric(),
  overlap = numeric(),
  size = numeric(),
  overlapGenes = character(),
  GeneRatio = numeric(),
  Cluster = character(),
  Description = character()
)

module_universe_df <- data.frame(
  module = character(),
  ID = character(),
  stringsAsFactors = FALSE
)

for (i in seq_along(moduleTraitCor$module)) {
  # get the list of colnames(moduleTraitCor) for which there is an abs(correlation) > 0.2 and a pvalue < 0.05
  # traits <- names(moduleTraitCor[,-1])[abs(moduleTraitCor[i, -1]) > 0.2 & moduleTraitPadj[i, -1] < 0.05]
  traits <- names(moduleTraitCor[, -1])[moduleTraitPadj[i, -1] < 0.05]
  # replace "-" with "_" in traits
  traits <- gsub("-", "_", traits)
  # replace "Blastema" with "Limb_blastema" in traits
  traits <- gsub("Blastema", "Limb_blastema", traits)

  module_universe <- c()

  for (j in seq_along(traits)) {
    # keep only rows where at least 3 samples of a trait have a count >= 10
    keep <- rowSums(count_matrix[, glob_metadata$Tissue_group == traits[j]] >= 10) >= 3
    keep2 <- rowSums(count_matrix[, glob_metadata$Age_group == traits[j]] >= 10) >= 3
    # keep gene ids
    gids_universe <- rownames(count_matrix[keep, ])
    gids_universe2 <- rownames(count_matrix[keep2, ])
    gids_universe3 <- unique(c(gids_universe, gids_universe2))

    module_universe <- unique(c(module_universe, gids_universe3))
  }

  # keep only string between the parentheses
  module_new_name <- gsub(".*\\(", "", moduleTraitCor$module[i])
  # remove closing parentheses
  module_new_name <- gsub("\\)", "", module_new_name)
  # write vector to file

  module_universe_tmp <- as_tibble(module_universe)
  module_universe_tmp <- na.omit(module_universe)
  write.table(module_universe_tmp,
    file = paste0(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/enrich/module_"), module_new_name, "_universe.tsv"),
    sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE
  )

  # get all coexpressed genes in module
  # get ID for which wgcna_module is  moduleTraitCor$module[i] in module_df

  module_coexgenes <- module_df$ID[module_df$wgcna_module == module_new_name]

  fora_res <- fgsea::fora(
    pathways = gene_metadata_for_fgseaORA,
    genes = module_coexgenes,
    universe = module_universe,
    minSize = 10,
    maxSize = 500
  )


  fora_res_tb <- as_tibble(fora_res)

  # unlist lists of overlapGenes columns, separated by ","
  fora_res_tb <- fora_res_tb %>%
    mutate(
      overlapGenes = paste(unlist(overlapGenes), collapse = "|"),
      GeneRatio = overlap / size,
      Cluster = moduleTraitCor$module[i],
      Description = pathway
    )

  fora_res_tb_all <- bind_rows(fora_res_tb_all, fora_res_tb)
}

print(head(fora_res_tb_all))
```
  
```{r}
library(GO.db)
library(AnnotationDbi)

# get complete GO annotation table
go_annotation_table <- as_tibble(select(GO.db, keys = keys(GO.db), columns = c("GOID", "TERM", "ONTOLOGY")))

fora_res_tb_all <- fora_res_tb_all %>%
  # match pathway with go_annotation_table$GOID
  mutate(
    Description = go_annotation_table$TERM[match(pathway, go_annotation_table$GOID)],
    ONTOLOGY = go_annotation_table$ONTOLOGY[match(pathway, go_annotation_table$GOID)],
    ONTOLOGY_full = ifelse(ONTOLOGY == "BP", "Biological process",
      ifelse(ONTOLOGY == "MF", "Molecular function",
        "Cellular component"
      )
    )
  ) %>%
  # remove rows where Description is NA
  # To remove obsolete or removed GO terms,
  # remove from all rows for which Description starts with GO:
  filter(!grepl("^GO:", Description)) %>%
  na.omit()


write_tsv(fora_res_tb_all,
  file = paste0(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/enrich/all_modules_fora_res.tsv"))
)
```

Functions

```{r}
library(librarian)
source("https://raw.githubusercontent.com/YuLab-SMU/enrichplot/a16c712b9ea431562f42cffd199b4c84b7a15488/R/utilities.R")

append_intersect <- function(x) {
  # if (!inherits(x, 'compareClusterResult')) stop("x should be a compareClusterResult object")

  d <- as.data.frame(x)

  so <- vapply(split(d$Cluster, d$Description),
    FUN = paste,
    FUN.VALUE = character(1),
    collapse = " & "
  )

  set_info <- data.frame(
    intersect = so,
    Description = names(so)
  )

  d2 <- merge(d, set_info, by = "Description")
  n <- levels(d2$Cluster)
  cc <- yulab.utils::combinations(length(n))
  lv <- vapply(cc, function(i) paste(n[i], collapse = " & "), character(1))
  d2$intersect <- factor(d2$intersect, levels = lv)

  x@compareClusterResult <- d2

  return(x)
}

library(plyr)

fortify.compareClusterResult <- function(model, data, showCategory = 5,
                                         by = "geneRatio", split = NULL,
                                         includeAll = TRUE, ...) {
  clProf.df <- as.data.frame(model)
  .split <- split
  if ("core_enrichment" %in% colnames(clProf.df)) {
    clProf.df$Count <- str_count(clProf.df$core_enrichment, "/")
    clProf.df$.sign <- "activated"
    clProf.df$.sign[clProf.df$NES < 0] <- "suppressed"
    clProf.df$GeneRatio <- clProf.df$Count / clProf.df$setSize
  }
  ## get top 5 (default) categories of each gene cluster.
  if (is.null(showCategory)) {
    result <- clProf.df
  } else if (is.numeric(showCategory)) {
    Cluster <- NULL # to satisfy codetools

    topN <- function(res, showCategory) {
      plyr::ddply(
        .data = res,
        .variables = .(Cluster),
        .fun = function(df, N) {
          if (length(df$Count) > N) {
            if (any(colnames(df) == "pvalue")) {
              idx <- order(df$pvalue, decreasing = FALSE)[1:N]
            } else {
              ## for groupGO
              idx <- order(df$Count, decreasing = T)[1:N]
            }
            return(df[idx, ])
          } else {
            return(df)
          }
        },
        N = showCategory
      )
    }

    if (!is.null(.split) && .split %in% colnames(clProf.df)) {
      lres <- split(clProf.df, as.character(clProf.df[, .split]))
      lres <- lapply(lres, topN, showCategory = showCategory)
      result <- do.call("rbind", lres)
    } else {
      result <- topN(clProf.df, showCategory)
    }
  } else {
    result <- subset(clProf.df, Description %in% showCategory)
  }

  pathway <- NULL
  if (includeAll == TRUE) {
    result <- subset(clProf.df, pathway %in% result$pathway)
  }

  ## remove zero count
  result$Description <- as.character(result$Description) ## un-factor
  GOlevel <- result[, c("pathway", "Description")] ## GO pathway and Term
  GOlevel <- unique(GOlevel)



  result <- result[result$Count != 0, ]
  result$Description <- factor(result$Description,
    levels = rev(GOlevel[, 2])
  )
  if (by == "rowPercentage") {
    Description <- Count <- NULL # to satisfy codetools
    result <- plyr::ddply(result,
      .(Description),
      transform,
      Percentage = Count / sum(Count),
      Total = sum(Count)
    )

    ## label GO Description with gene counts.
    x <- mdply(result[, c("Description", "Total")], paste, sep = " (")
    y <- sapply(x[, 3], paste, ")", sep = "")
    result$Description <- y

    ## restore the original order of GO Description
    xx <- result[, c(2, 3)]
    xx <- unique(xx)
    rownames(xx) <- xx[, 1]
    Termlevel <- xx[as.character(GOlevel[, 1]), 2]

    ## drop the *Total* column
    result <- result[, colnames(result) != "Total"]

    result$Description <- factor(result$Description,
      levels = rev(Termlevel)
    )
  } else if (by == "count") {
    ## nothing
  } else if (by == "geneRatio") {
    ## for result of ORA
    # if (class(result$GeneRatio) == "character" && grep("/", result$GeneRatio[1])) {
    if (inherits(result$GeneRatio, "character") && grep("/", result$GeneRatio[1])) {
      gsize <- as.numeric(sub("/\\d+$", "", as.character(result$GeneRatio)))
      gcsize <- as.numeric(sub("^\\d+/", "", as.character(result$GeneRatio)))
      result$GeneRatio <- gsize / gcsize
      if (("ONTOLOGY" %in% colnames(result)) && (length(unique(result$ONTOLOGY)) > 1)) {
        # do nothing
      } else {
        cluster <- paste(as.character(result$Cluster), "\n", "(", gcsize, ")",
          sep = ""
        )
        lv <- unique(cluster)[order(as.numeric(unique(result$Cluster)))]
        result$Cluster <- factor(cluster, levels = lv)
      }
    }
  } else {
    ## nothing
  }
  return(result)
}

dotplot.compareClusterResult <- function(object, x = "Cluster", colorBy = "p.adjust",
                                         showCategory = 5, by = "geneRatio", size = "geneRatio",
                                         split = NULL, includeAll = TRUE,
                                         font.size = 12, title = "", label_format = 30,
                                         group = FALSE, shape = FALSE, facet = NULL, strip_width = 15) {
  color <- NULL
  if (is.null(size)) size <- by ## by may deprecated in future release

  if (!is.null(facet) && facet == "intersect") {
    object <- append_intersect(object)
  }

  df <- fortify.compareClusterResult(object,
    showCategory = showCategory, by = size,
    includeAll = includeAll, split = split
  )

  if (by != "geneRatio") {
    df$GeneRatio <- parse_ratio(df$GeneRatio)
  }
  label_func <- default_labeller(label_format)
  if (is.function(label_format)) {
    label_func <- label_format
  }

  if (size %in% c("rowPercentage", "count", "geneRatio")) {
    by2 <- switch(size,
      rowPercentage = "Percentage",
      count = "Count",
      geneRatio = "GeneRatio"
    )
  } else {
    by2 <- size
  }

  p <- ggplot(df, aes_string(x = x, y = "Description", size = by2)) +
    scale_y_discrete(labels = label_func)

  if (group) {
    p <- p + geom_line(aes_string(color = "Cluster", group = "Cluster"), size = .3) +
      ggnewscale::new_scale_colour()
  }


  if (shape) {
    check_installed("ggstar", "for `dotplot()` with `shape = TRUE`.")
    ggstar <- "ggstar"
    require(ggstar, character.only = TRUE)
    # p <- p + ggsymbol::geom_symbol(aes_string(symbol = "Cluster", fill = colorBy)) +
    p <- p + ggstar::geom_star(aes_string(starshape = "Cluster", fill = colorBy)) +
      set_enrichplot_color(type = "fill")
  } else {
    p <- p + geom_point(aes_string(fill = colorBy)) +
      aes(shape = I(enrichplot_point_shape))
  }

  p <- p + set_enrichplot_color(type = "fill") +
    ylab(NULL) + ggtitle(title) + DOSE::theme_dose(font.size) +
    scale_size_continuous(range = c(3, 8))


  if (!is.null(facet)) {
    p <- p + facet_grid(.data[[facet]] ~ .,
      scales = "free", space = "free",
      switch = "y",
      labeller = ggplot2::label_wrap_gen(strip_width)
    ) +
      theme(strip.text = element_text(size = 14))
  }

  class(p) <- c("enrichplotDot", class(p))

  return(p)
}
```

```{r}
fora_res_tb_all <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/enrich/all_modules_fora_res.tsv"),
  progress = show_progress()
)

# keep only Module 5 (black), Module 6 (yellow), Module 1 (green), and Module 16 (grey)
fora_res_tb_all_filt <- fora_res_tb_all %>%
  filter(Cluster %in% c("Module 5 (black)", "Module 6 (yellow)", "Module 1 (green)", "Module 16 (grey)")) %>%
  filter(padj < 0.05) %>%
  dplyr::rename("p.adjust" = "padj") %>%
  dplyr::rename("Count" = "overlap") %>%
  dplyr::rename("pvalue" = "pval") %>%
  # sort each cluster by p.adjust
  dplyr::arrange(Cluster, p.adjust)


# print rows where CLuster == Module 16 (grey)
fora_res_tb_all_filt %>%
  filter(Cluster == "Module 16 (grey)") %>%
  print(width = Inf, n = 50)


options(enrichplot.colours = c("red", "blue"))
options(enrichplot.colors = c("red", "blue"))


dp_fora_res_tb_all_filt <- dotplot.compareClusterResult(fora_res_tb_all_filt,
  showCategory = 12,
  size = "size",
  colorBy = "p.adjust",
  x = "GeneRatio",
  label_format = 60
) +


  # scale_size(limits = c(0,500),
  #          breaks = c(50,100,200,300),
  #          range = c(1,8.5)) +
  scale_y_discrete(
    position = "right",
    labels = function(x) str_wrap(x, width = 60)
  ) +

  # scale_color_gradient(limits = c(0,0.05),
  #                    breaks=c(0.01,0.02,0.03,0.04,0.05),
  #                    low="red", high="blue") +

  # x scale from 0 to 0.8
  scale_x_continuous(limits = c(0, 0.8)) +

  facet_nested(ONTOLOGY_full ~ Cluster,
    scales = "free_y",
    space = "free",
    drop = FALSE
  ) +
  theme_bw() +
  theme(
    text = element_text(family = "Arial"),
    strip.placement = "outside",
    legend.title = element_text(face = "bold", size = 9),
    legend.text = element_text(size = 7),
    axis.text.y = element_text(size = 7, lineheight = 0.7),
    axis.text.x = element_text(size = 9),
    axis.title.x = element_text(face = "bold", size = 9),
    strip.text = element_text(size = 10, face = "bold"),
    strip.background = element_blank(),
    strip.text.y = element_blank()
  )
# switch = "y"

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/enrich/dp_fora_res_tb_all_filt.png"),
  plot = dp_fora_res_tb_all_filt, width = 25, height = 20, units = "cm", dpi = 600
)

ggsave(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/enrich/dp_fora_res_tb_all_filt.svg"),
  plot = dp_fora_res_tb_all_filt, width = 25, height = 20, units = "cm"
)
```

Hub genes

```{r}
library(dplyr)
library(readr)

# the grey module is omitted
topHubs <- function(
    datExpr, colorh, omitColors = "grey", power = 2, type = "signed",
    ...) {
  # modified from chooseTopHubInEachModule, but return the table of all genes connectivity
  isIndex <- FALSE
  modules <- names(table(colorh))
  if (!is.na(omitColors)[1]) {
    modules <- modules[!is.element(modules, omitColors)]
  }
  if (is.null(colnames(datExpr))) {
    colnames(datExpr) <- seq_len(dim(datExpr)[2])
    isIndex <- TRUE
  }

  connectivity_table <- data.frame(matrix(ncol = 3)) %>% setNames(c("gene", "connectivity_rowSums_adj", "module"))
  hubs <- rep(NA, length(modules))
  names(hubs) <- modules
  for (m in modules) {
    adj <- adjacency(datExpr[, colorh == m],
      power = power,
      type = type, ...
    )

    hub <- which.max(rowSums(adj))

    hubs[m] <- colnames(adj)[hub]

    sorted_genes <- rowSums(adj) %>%
      sort(decreasing = T) %>%
      as.data.frame() %>%
      tibble::rownames_to_column() %>%
      setNames(c("gene", "connectivity_rowSums_adj")) %>%
      mutate(module = m)
    connectivity_table <- connectivity_table %>% rbind(sorted_genes)
  }
  if (isIndex) {
    hubs <- as.numeric(hubs)
    names(hubs) <- modules
  }
  return(connectivity_table %>% na.omit())
}



connectivity_table <- as_tibble(topHubs(input_mat, omitColors = NA, colorh = mergedColors, power = 8, type = "signed"))

write_tsv(connectivity_table, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/wgcna_hubs_connectivity_table.tsv"))

connectivity_table_tops <- connectivity_table %>%
  group_by(module) %>%
  top_n(4, wt = connectivity_rowSums_adj)

# for genes (REs) that start with "X", remove "X"
connectivity_table_tops$gene <- gsub("^X", "", connectivity_table_tops$gene)

# load gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

# Add symbol with dplyr
connectivity_table_tops <- connectivity_table_tops %>%
  ungroup() %>%
  mutate(gene_symbol = gene_metadata$gene_symbol[match(connectivity_table_tops$gene, gene_metadata$gene_id)]) %>%
  # if gene_symbol is NA, replace with gene_id
  mutate(gene_symbol = ifelse(is.na(gene_symbol), gene, gene_symbol))

write_tsv(connectivity_table_tops, file = file.path(PROJECT_DIR, "/rnaseq/quantification/TEcount/ambMex60DD_ucsc_hub.ALL/global/WGCNA/wgcna_hubs_connectivity_table_tops.tsv"))

# connectivity_table_tops <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/connectivity_table_tops.tsv"))
```

GS and MM plots

```{r}
library(ggplot2)
library(ggh4x)
library(ggpubr)
library(ggrepel)

dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/plots/"))
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)

colors_list <- unique(mergedColors)

for (i in seq_along(colnames(b_sample_data))) {
  # Define variable trait containing the trait column of datTrait
  trait <- as.data.frame(b_sample_data[, i])
  names(trait) <- colnames(b_sample_data)[i]
  # names (colors) of the modules
  modNames <- substring(names(MEs0), 3)
  nGenes <- ncol(input_mat)
  nSamples <- nrow(input_mat)
  geneModuleMembership <- as.data.frame(cor(input_mat, MEs0, use = "p"))
  MMPvalue <- as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples))

  names(geneModuleMembership) <- paste("MM", modNames, sep = "")
  names(MMPvalue) <- paste("p.MM", modNames, sep = "")
  geneTraitSignificance <- as.data.frame(cor(input_mat, trait, use = "p"))
  GSPvalue <- as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples))
  names(geneTraitSignificance) <- paste("GS.", names(trait), sep = "")
  names(GSPvalue) <- paste("p.GS.", names(trait), sep = "")

  for (j in seq_along(colors_list)) {
    module <- toString(colors_list[j])
    column <- match(module, modNames)
    moduleGenes <- mergedColors == module

    # Create a data frame for the ggplot
    plot_data <- data.frame(
      MM = abs(geneModuleMembership[moduleGenes, column]),
      GS = abs(geneTraitSignificance[moduleGenes, 1]),
      # if rowname starts with "AMEX", then it is gene
      feature_biotype = ifelse(grepl("^AMEX", rownames(geneTraitSignificance[moduleGenes, , drop = FALSE])), "gene", "RE"),
      gene_symbol = gene_metadata$gene_symbol[match(rownames(geneTraitSignificance[moduleGenes, , drop = FALSE]), gene_metadata$gene_id)]
    )

    # IF GENES ARE NOT IN gene_metadata$gene_symbol, THEN REPLACE WITH gene_id
    plot_data$gene_symbol <- ifelse(is.na(plot_data$gene_symbol), rownames(geneTraitSignificance[moduleGenes, , drop = FALSE]), plot_data$gene_symbol)

    write_tsv(plot_data, file = paste0(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/plots/"), module, "_mod_", colnames(b_sample_data)[i], "_plot_data.tsv"))

    candp <- corAndPvalue(plot_data$MM, plot_data$GS)

    # Create the ggplot
    gs_mm_mod_plot <- ggplot(plot_data, aes(x = MM, y = GS, color = feature_biotype)) +
      geom_point(size = 0.6, alpha = 0.75) +
      scale_color_manual(values = c("#E69F00", "#56B4E9")) +
      # only label points that are in connectivity_table_tops$module == module
      geom_text_repel(
        data = subset(plot_data, gene_symbol %in% connectivity_table_tops$gene_symbol[connectivity_table_tops$module == module]),
        aes(label = gene_symbol),
        size = 1.5,
        show.legend = FALSE,
        hjust = 0,
        color = "black",
        force = 1.6,
        min.segment.length = 0.1,
        segment.size = 0.5,
        nudge_x = 0.01
      ) +
      # add regression line
      xlim(0.25, 1) +
      ylim(0, 1) +
      geom_smooth(method = "lm", se = FALSE, color = "black", fullrange = TRUE, linewidth = 0.5) +
      labs(
        x = paste("Module Membership in", module, "module"),
        y = paste0("Gene significance for ", colnames(b_sample_data)[i]),
        title = "Module membership vs. gene/RE significance"
      ) +
      # add cor and p value
      geom_text(
        x = 0.25, y = 0.975,
        label = paste("cor = ", format(candp$cor, digits = 2, nsmall = 2), sep = ""),
        size = 3,
        show.legend = FALSE,
        inherit.aes = FALSE,
        hjust = 0
      ) +
      # use scientific notation for p-value
      geom_text(
        x = 0.25, y = 0.925,
        label = paste0("p ", ifelse(candp$p > 0.001, paste0("= ", format(candp$p, digits = 3, nsmall = 3)),
          ifelse(is.finite(candp$p) && candp$p < 10^(-200), "< 10^-200",
            paste0("= ", format(candp$p, digits = 2, nsmall = 2, scientific = TRUE))
          )
        )),
        size = 3,
        show.legend = FALSE,
        inherit.aes = FALSE,
        hjust = 0
      ) +
      # add total number of feature_biotype == "gene"
      geom_text(
        x = 0.25, y = 0.875, color = "#E69F00",
        label = paste0("genes = ", sum(plot_data$feature_biotype == "gene")),
        size = 3,
        show.legend = FALSE,
        inherit.aes = FALSE,
        hjust = 0
      ) +
      # add total number of feature_biotype == "RE"
      geom_text(
        x = 0.25, y = 0.825, color = "#56B4E9",
        label = paste0("REs = ", sum(plot_data$feature_biotype == "RE")),
        size = 3,
        show.legend = FALSE,
        inherit.aes = FALSE,
        hjust = 0
      ) +
      theme_bw() +
      theme(
        axis.text = element_text(size = 6),
        axis.title.x = element_text(face = "bold", size = 8),
        axis.title.y = element_text(face = "bold", size = 8),
        panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.4),
      ) +
      force_panelsizes(
        rows = unit(4, "in"),
        cols = unit(4, "in")
      )

    ggsave(
      filename = paste0(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/plots/"), module, "_mod_", colnames(b_sample_data)[i], "_mm_vs_gs_ggplot.png"),
      plot = gs_mm_mod_plot,
      width = 6, height = 6, units = "in", dpi = 600
    )

    ggsave(
      filename = paste0(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/plots/"), module, "_mod_", colnames(b_sample_data)[i], "_mm_vs_gs_ggplot.svg"),
      plot = gs_mm_mod_plot,
      width = 6, height = 6, units = "in"
    )
  }
}
```

####### Exporting to Cytoscape

```{r}
dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/cytoscape/nodes_and_edges"))
setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/cytoscape/nodes_and_edges"))

# this loads "TOM"
load(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/global_blockwiseModules_TOM-block.1.RData"))

mat_TOM <- as.matrix(TOM)

# for every selected_module do the following:

module_names <- unique(mergedColors)

for (i in seq_along(module_names)) {
  # Select module probes
  probes <- colnames(input_mat)

  inModule <- is.finite(match(mergedColors, module_names[i]))

  modProbes <- probes[inModule]

  # modGenes = annot$gene_symbol[match(modProbes, annot$substanceBXH)]

  # Print if selected modProbes contains strings that do not start with "AMEX"
  if (any(!grepl("^AMEX", modProbes))) {
    print(paste0("The module ", module_names[i], " contains probes that do not start with AMEX:"))
    print(modProbes[!grepl("^AMEX", modProbes)])
  }

  # Select the corresponding Topological Overlap
  modTOM <- mat_TOM[inModule, inModule]

  dimnames(modTOM) <- list(modProbes, modProbes)

  # Export the network into edge and node list files Cytoscape can read
  cyt <- exportNetworkToCytoscape(modTOM,
    edgeFile = paste0(module_names[i], "_module_network_to_cytoscape_edges.txt"),
    nodeFile = paste0(module_names[i], "_module_network_to_cytoscape_nodes.txt"),
    weighted = TRUE,
    threshold = 0.02,
    nodeNames = modProbes,
    # altNodeNames = modGenes,
    nodeAttr = mergedColors[inModule]
  )
}
```

 fix node and edge files

```{r}
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), sep = "\t", header = TRUE)


# for every "*_edges.txt" file do the following:
e_file_list <- list.files(pattern = "*_edges.txt")

for (i in seq_along(e_file_list)) {
  # read the file
  e_file <- read_tsv(e_file_list[i])

  e_file <- e_file %>%
    mutate(
      # if node starts with "X", remove "X"
      fromNode = gsub("^X", "", fromNode),
      toNode = gsub("^X", "", toNode),
      fromAltName = gene_metadata$gene_symbol[match(e_file$fromNode, gene_metadata$gene_id)],
      toAltName = gene_metadata$gene_symbol[match(e_file$toNode, gene_metadata$gene_id)],
      # if altname is NA, replace with node
      fromAltName = ifelse(is.na(fromAltName), fromNode, fromAltName),
      toAltName = ifelse(is.na(toAltName), toNode, toAltName)
    )

  # save the file
  write_tsv(e_file, file = file.path(getwd(), sub(".txt", ".tsv", e_file_list[i])))

  # remove file
  file.remove(e_file_list[i])
}



# for every "*_nodes.txt" file do the following:
n_file_list <- list.files(pattern = "*_nodes.txt")

for (i in seq_along(n_file_list)) {
  # read the file
  n_file <- read_tsv(n_file_list[i])

  n_file <- n_file %>%
    mutate(
      # if node starts with "X", remove "X"
      nodeName = gsub("^X", "", nodeName),
      altName = gene_metadata$gene_symbol[match(n_file$nodeName, gene_metadata$gene_id)],
      # if altname is NA, replace with node
      altName = ifelse(is.na(altName), nodeName, altName)
    )

  # save the file
  write_tsv(n_file, file = file.path(getwd(), sub(".txt", ".tsv", n_file_list[i])))

  # remove file
  file.remove(n_file_list[i])
}
```

 from here, in Cytoscape:

https://www.biostars.org/p/60896/#455176
On Cytoscape 3.8.0, you can import WGCNA network by File -> Import -> Network from File and selecting the module edge file. 
On the import dialogue box, you will typically select the fromNode column as the Source Node and the toNode column as the Target Node. 
The weight column should be left as an Edge Attribute. The direction column should be changed to interaction type. fromAltName is a 
Source Node Attribute while the toAltName is a Target Node Attribute.

## CONSENSUS


```{r}
library(readr)

# bind_rows of every results object
nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_age_tecount_res_adult_limbs_vs_subadult_limbs.tsv"))
nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_reg_tecount_res_sa_blastemas_vs_sa_limbs.tsv"))

glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs.tsv"))
glob_tecount_res_adult_limbs_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_adult_limbs_vs_subadult_limbs.tsv"))
glob_tecount_res_adult_vs_subadult_blastema <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_adult_vs_subadult_blastema.tsv"))
glob_tecount_res_adult_blastemas_vs_adult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_adult_blastemas_vs_adult_limbs.tsv"))
glob_tecount_res_subadult_blastemas_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_subadult_blastemas_vs_subadult_limbs.tsv"))
glob_tecount_res_adult_blastemas_vs_adult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_adult_blastemas_vs_adult_nonlimbs.tsv"))
glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs.tsv"))
glob_tecount_res_adult_limbs_vs_adult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_adult_limbs_vs_adult_nonlimbs.tsv"))
glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs.tsv"))
glob_tecount_res_int_Age_group_effect_blastema_vs_limb <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_int_Age_group_effect_blastema_vs_limb.tsv"))
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema.tsv"))
glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb.tsv"))

nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs.tsv"))
nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs.tsv"))

glob_explorate_fam_res_a_vs_sa_b <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_a_vs_sa_b.tsv"))
glob_explorate_fam_res_a_vs_sa_l <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_a_vs_sa_l.tsv"))
glob_explorate_fam_res_a_vs_sa_nl <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_a_vs_sa_nl.tsv"))
glob_explorate_fam_res_b_vs_l_a <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_l_a.tsv"))
glob_explorate_fam_res_b_vs_l_sa <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_l_sa.tsv"))
glob_explorate_fam_res_b_vs_nl_a <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_nl_a.tsv"))
glob_explorate_fam_res_b_vs_nl_sa <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_b_vs_nl_sa.tsv"))
glob_explorate_fam_res_l_vs_nl_a <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_l_vs_nl_a.tsv"))
glob_explorate_fam_res_l_vs_nl_sa <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_l_vs_nl_sa.tsv"))
glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb.tsv"))
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema.tsv"))
glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/global/DEA/results/glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb.tsv"))

nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs.tsv"))
nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/native/DEA/results/nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs.tsv"))

glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs.tsv"))
glob_salmon_res_adult_limbs_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_limbs_vs_subadult_limbs.tsv"))
glob_salmon_res_adult_vs_subadult_blastema <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_vs_subadult_blastema.tsv"))
glob_salmon_res_adult_blastemas_vs_adult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_blastemas_vs_adult_limbs.tsv"))
glob_salmon_res_subadult_blastemas_vs_subadult_limbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_blastemas_vs_subadult_limbs.tsv"))
glob_salmon_res_adult_blastemas_vs_adult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_blastemas_vs_adult_nonlimbs.tsv"))
glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs.tsv"))
glob_salmon_res_adult_limbs_vs_adult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_adult_limbs_vs_adult_nonlimbs.tsv"))
glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs.tsv"))
glob_salmon_res_int_Age_group_effect_blastema_vs_limb <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_group_effect_blastema_vs_limb.tsv"))
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema.tsv"))
glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb.tsv"))


all_diffexp_results <- dplyr::bind_rows(
  nat_age_tecount_res_adult_limbs_vs_subadult_limbs,
  nat_reg_tecount_res_sa_blastemas_vs_sa_limbs,
  glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs,
  glob_tecount_res_adult_limbs_vs_subadult_limbs,
  glob_tecount_res_adult_vs_subadult_blastema,
  glob_tecount_res_adult_blastemas_vs_adult_limbs,
  glob_tecount_res_subadult_blastemas_vs_subadult_limbs,
  glob_tecount_res_adult_blastemas_vs_adult_nonlimbs,
  glob_tecount_res_subadult_blastemas_vs_subadult_nonlimbs,
  glob_tecount_res_adult_limbs_vs_adult_nonlimbs,
  glob_tecount_res_subadult_limbs_vs_subadult_nonlimbs,
  glob_tecount_res_int_Age_group_effect_blastema_vs_limb,
  glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb,
  nat_age_explorate_fam_res_adult_limbs_vs_subadult_limbs,
  nat_reg_explorate_fam_res_sa_blastemas_vs_sa_limbs,
  glob_explorate_fam_res_a_vs_sa_b,
  glob_explorate_fam_res_a_vs_sa_l,
  glob_explorate_fam_res_a_vs_sa_nl,
  glob_explorate_fam_res_b_vs_l_a,
  glob_explorate_fam_res_b_vs_l_sa,
  glob_explorate_fam_res_b_vs_nl_a,
  glob_explorate_fam_res_b_vs_nl_sa,
  glob_explorate_fam_res_l_vs_nl_a,
  glob_explorate_fam_res_l_vs_nl_sa,
  glob_explorate_fam_res_int_Age_group_effect_blastema_vs_limb,
  glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  glob_explorate_fam_res_int_Age_groupAdult_Tissue_groupLimb,
  nat_age_salmon_gene_res_adult_limbs_vs_subadult_limbs,
  nat_reg_salmon_gene_res_sa_blastemas_vs_sa_limbs,
  glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs,
  glob_salmon_res_adult_limbs_vs_subadult_limbs,
  glob_salmon_res_adult_vs_subadult_blastema,
  glob_salmon_res_adult_blastemas_vs_adult_limbs,
  glob_salmon_res_subadult_blastemas_vs_subadult_limbs,
  glob_salmon_res_adult_blastemas_vs_adult_nonlimbs,
  glob_salmon_res_subadult_blastemas_vs_subadult_nonlimbs,
  glob_salmon_res_adult_limbs_vs_adult_nonlimbs,
  glob_salmon_res_subadult_limbs_vs_subadult_nonlimbs,
  glob_salmon_res_int_Age_group_effect_blastema_vs_limb,
  glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema,
  glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb
)

# rename gene_id column to ID
all_diffexp_results <- all_diffexp_results %>%
  dplyr::rename(
    ID = gene_id,
    SYMBOL = gene_symbol
  )

# new column, re_superfamily is the second element of ID split by :
all_diffexp_results <- all_diffexp_results %>%
  dplyr::mutate(
    re_subfamily = ifelse(feature_biotype == "repetitive_element",
      sapply(strsplit(ID, ":"), function(x) x[1]),
      NA
    ),
    re_superfamily = ifelse(feature_biotype == "repetitive_element",
      sapply(strsplit(ID, ":"), function(x) x[2]),
      NA
    ),
    re_class = ifelse(feature_biotype == "repetitive_element",
      sapply(strsplit(ID, ":"), function(x) x[3]),
      NA
    )
  )

# reorder columns
all_diffexp_results <- all_diffexp_results %>%
  dplyr::select(
    ID, SYMBOL, feature_biotype, re_subfamily, re_superfamily, re_class, quant_tool, sample_group, diffexp_contrast,
    diffexp_res_lfc, diffexp_res_padj, baseMean_MLE, log2FoldChange_MLE, lfcSE_MLE,
    stat_MLE, pvalue_MLE, padj_MLE, weight_MLE, log2FoldChange_MMSE, lfcSE_MMSE, svalue_MMSE
  )

write_tsv(all_diffexp_results, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))
# all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

```


### Native samples

#### DEGs

###### UpSet

```{r}
library(ggplot2)
library(DESeq2)
library(ComplexHeatmap)

```

```{r}
# read gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), header = TRUE, sep = "\t")

# keep only sample_group == native
df <- all_diffexp_results %>%
        filter(sample_group == "native",
              feature_biotype == "gene",
              padj_MLE < 0.05,
              abs(log2FoldChange_MMSE) > log2(1.5)) %>%
        # group by quant_tool, diffexp_contrast and diffexp_res_lfc
        group_by(quant_tool, diffexp_contrast, diffexp_res_lfc) %>%
        summarise(genes = list(ID), .groups = "keep")
        
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(quant_tool, diffexp_contrast, diffexp_res_lfc, sep = "_")) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            lapply(function(x) sort(unlist(x$genes)))

all_genes <- unique(unname(unlist(df_list)))

df_list$annotated_genes <- all_genes[which(!grepl("^AMEX", gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]) & !is.na(gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]))]
df_list$unannotated_genes <- all_genes[which(grepl("^AMEX", gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]) | is.na(gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]))]


genes_comb_mat <- make_comb_mat(df_list)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/nat_genes_upset.png"),
    width = 20, height = 10.5, units = "cm", res = 1200)

UpSet(genes_comb_mat,
  #set_order = rownames(genes_comb_mat),
  set_order = c("TEcount_adult_limbs_vs_subadult_limbs_upregulated",
                "TEcount_adult_limbs_vs_subadult_limbs_downregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_upregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_downregulated",
                "Salmon_adult_limbs_vs_subadult_limbs_upregulated",
                "Salmon_adult_limbs_vs_subadult_limbs_downregulated",
                "Salmon_subadult_blastemas_vs_subadult_limbs_upregulated",
                "Salmon_subadult_blastemas_vs_subadult_limbs_downregulated",
                "annotated_genes",
                "unannotated_genes"),
  top_annotation = upset_top_annotation(genes_comb_mat,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/nat_genes_upset.svg"), width = 7.87402, height = 4.133858)
UpSet(genes_comb_mat,
    #set_order = rownames(genes_comb_mat),
  set_order = c("TEcount_adult_limbs_vs_subadult_limbs_upregulated",
                "TEcount_adult_limbs_vs_subadult_limbs_downregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_upregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_downregulated",
                "Salmon_adult_limbs_vs_subadult_limbs_upregulated",
                "Salmon_adult_limbs_vs_subadult_limbs_downregulated",
                "Salmon_subadult_blastemas_vs_subadult_limbs_upregulated",
                "Salmon_subadult_blastemas_vs_subadult_limbs_downregulated",
                "annotated_genes",
                "unannotated_genes"),
  top_annotation = upset_top_annotation(genes_comb_mat,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)
dev.off()

```

REVISED FIGURE

```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(DESeq2)
library(ComplexHeatmap)

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))


# read gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), header = TRUE, sep = "\t")


# keep only sample_group == native
df <- all_diffexp_results %>%
        filter(sample_group == "native",
              feature_biotype == "gene",
              padj_MLE < 0.05,
              quant_tool == "TEcount",
              abs(log2FoldChange_MMSE) > log2(1.5)) %>%
        # group by quant_tool, diffexp_contrast and diffexp_res_lfc
        group_by(quant_tool, diffexp_contrast, diffexp_res_lfc) %>%
        summarise(genes = list(ID), .groups = "keep")
        
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(quant_tool, diffexp_contrast, diffexp_res_lfc, sep = "_")) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            lapply(function(x) sort(unlist(x$genes)))

all_genes <- unique(unname(unlist(df_list)))

df_list$annotated_genes <- all_genes[which(!grepl("^AMEX", gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]) & !is.na(gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]))]
df_list$unannotated_genes <- all_genes[which(grepl("^AMEX", gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]) | is.na(gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]))]


genes_comb_mat <- make_comb_mat(df_list)

# Save as PNG
png(file.path(NEW_PROJECT_DIR, "/revised_figures/nat_genes_upset_revised.png"),
    width = 15, height = 9, units = "cm", res = 1200)

UpSet(genes_comb_mat,
  #set_order = rownames(genes_comb_mat),
  set_order = c("TEcount_adult_limbs_vs_subadult_limbs_upregulated",
                "TEcount_adult_limbs_vs_subadult_limbs_downregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_upregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_downregulated",
                "annotated_genes",
                "unannotated_genes"),
  top_annotation = upset_top_annotation(genes_comb_mat,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

svg(file.path(NEW_PROJECT_DIR, "/revised_figures/nat_genes_upset_revised.svg"), width = 5.90551, height = 3.54331)
UpSet(genes_comb_mat,
    #set_order = rownames(genes_comb_mat),
  set_order = c("TEcount_adult_limbs_vs_subadult_limbs_upregulated",
                "TEcount_adult_limbs_vs_subadult_limbs_downregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_upregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_downregulated",
                "annotated_genes",
                "unannotated_genes"),
  top_annotation = upset_top_annotation(genes_comb_mat,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)
dev.off()


```

##### RE families

###### UpSet

```{r}

# keep only sample_group == native
df <- all_diffexp_results %>%
        filter(sample_group == "native",
              feature_biotype == "repetitive_element",
              padj_MLE < 0.05,
              abs(log2FoldChange_MMSE) > log2(1.5)) %>%
        # convert columns to factors so that .drop = FALSE works
        mutate(quant_tool = as.factor(quant_tool),
               diffexp_contrast = as.factor(diffexp_contrast),
               diffexp_res_lfc = as.factor(diffexp_res_lfc)) %>%
        # group by quant_tool, diffexp_contrast and diffexp_res_lfc
        group_by(quant_tool, diffexp_contrast, diffexp_res_lfc, .drop = FALSE) %>%
        summarise(re = list(ID), .groups = "keep")

df_keys <- df %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(quant_tool, diffexp_contrast, diffexp_res_lfc, sep = "_")) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            lapply(function(x) sort(unlist(x$re)))

all_re <- unique(unname(unlist(df_list)))

all_re_classification <- all_diffexp_results %>%
        filter(sample_group == "native",
              feature_biotype == "repetitive_element",
              padj_MLE < 0.05,
              abs(log2FoldChange_MMSE) > log2(1.5)) %>%
        # replace re_superfamily that start with "rnd" with "Unknown"
        mutate(re_superfamily = ifelse(grepl("^rnd", re_superfamily), "Unknown", re_superfamily)) %>%
        group_by(re_superfamily, re_class) %>%
        summarise(re = list(ID), .groups = "keep") %>%
        # in re_superfamily, replace "Gypsy" with "Ty3"
        mutate(re_superfamily = gsub("Gypsy_Chromoviridae_Vclade", "Vclade:Chromoviridae:Ty3", re_superfamily),
               re_superfamily = gsub("Gypsy_Mag", "Mag:Ty3", re_superfamily),
               re_superfamily = gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", re_superfamily))

all_re_classification_keys <- all_re_classification %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(re_superfamily, re_class, sep = ":")) %>%
          pull(key) %>%
          #replace "Unknown:Unknown" with "Unknown"
          gsub("Unknown:Unknown", "Unknown", .)

all_re_classification_list <- all_re_classification %>%
            group_split() %>%
            setNames(all_re_classification_keys) %>%
            lapply(function(x) sort(unlist(x$re)))

all_re_list <- c(df_list, all_re_classification_list)

comb_mat_nat_re <- make_comb_mat(all_re_list)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/nat_deres_upset.png"),
    width = 20, height = 10.5, units = "cm", res = 1200)

UpSet(comb_mat_nat_re,
  #set_order = rownames(comb_mat_nat_re),
  set_order = c("TEcount_adult_limbs_vs_subadult_limbs_upregulated",
                "TEcount_adult_limbs_vs_subadult_limbs_downregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_upregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_downregulated",
                "ExplorATE_adult_limbs_vs_subadult_limbs_upregulated",
                "ExplorATE_adult_limbs_vs_subadult_limbs_downregulated",
                "ExplorATE_subadult_blastemas_vs_subadult_limbs_upregulated",
                "ExplorATE_subadult_blastemas_vs_subadult_limbs_downregulated",
                sort(all_re_classification_keys)),
  top_annotation = upset_top_annotation(comb_mat_nat_re,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(comb_mat_nat_re, add_numbers = TRUE),
  # rep #999999B3
  bg_col = c(rep("#999999B3", length(all_re_classification_keys)),
             rep(c("#0072B2B3", "#D55E00B3"), 4)),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/nat_deres_upset.svg"),
    width = 7.87402, height = 4.133858)

UpSet(comb_mat_nat_re,
  #set_order = rownames(comb_mat_nat_re),
  set_order = c("TEcount_adult_limbs_vs_subadult_limbs_upregulated",
                "TEcount_adult_limbs_vs_subadult_limbs_downregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_upregulated",
                "TEcount_subadult_blastemas_vs_subadult_limbs_downregulated",
                "ExplorATE_adult_limbs_vs_subadult_limbs_upregulated",
                "ExplorATE_adult_limbs_vs_subadult_limbs_downregulated",
                "ExplorATE_subadult_blastemas_vs_subadult_limbs_upregulated",
                "ExplorATE_subadult_blastemas_vs_subadult_limbs_downregulated",
                sort(all_re_classification_keys)),
  top_annotation = upset_top_annotation(comb_mat_nat_re,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(comb_mat_nat_re, add_numbers = TRUE),
  # rep #999999B3
  bg_col = c(rep("#999999B3", length(all_re_classification_keys)),
             rep(c("#0072B2B3", "#D55E00B3"), 4)),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

```

Upset Figure 3 - GENE Intersection between native and global TEcount


```{r}
# read gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), header = TRUE, sep = "\t")

# keep only sample_group == native
df <- all_diffexp_results %>%
        filter(quant_tool == "TEcount",
              feature_biotype == "gene",
              padj_MLE < 0.05,
              abs(log2FoldChange_MMSE) > log2(1.5),
              (sample_group == "native" & diffexp_contrast %in% c(
#                                                                  "subadult_blastemas_vs_subadult_limbs" 
                                                                  "adult_limbs_vs_subadult_limbs"
                                                                  ))
              | (sample_group == "global" & diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs",
 #                                                                   "age_effect_blastemas_vs_limbs",
 #                                                                   "subadult_blastemas_vs_subadult_limbs",
 #                                                                   "adults_blastemas_vs_subadults_blastemas",
                                                                     "adult_blastemas_vs_adult_limbs"
                                                                     ))) %>%
        # group by sample_group, diffexp_contrast and diffexp_res_lfc
        group_by(sample_group, diffexp_contrast, diffexp_res_lfc) %>%
        summarise(genes = list(ID), .groups = "keep")
          
        
df_keys <- df %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(sample_group, diffexp_contrast, diffexp_res_lfc, sep = "_")) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            lapply(function(x) sort(unlist(x$genes)))

all_genes <- unique(unname(unlist(df_list)))

df_list$annotated_genes <- all_genes[which(!grepl("^AMEX", gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]) & !is.na(gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]))]
df_list$unannotated_genes <- all_genes[which(grepl("^AMEX", gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]) | is.na(gene_metadata$gene_symbol[match(all_genes, gene_metadata$gene_id)]))]


genes_comb_mat <- make_comb_mat(df_list)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/glob_genes_upset.png"),
    width = 16, height = 10, units = "cm", res = 600)

UpSet(genes_comb_mat,
    #set_order = rownames(genes_comb_mat),
  set_order = c("native_adult_limbs_vs_subadult_limbs_upregulated",
                "native_adult_limbs_vs_subadult_limbs_downregulated",
#                "native_subadult_blastemas_vs_subadult_limbs_upregulated",
#                "native_subadult_blastemas_vs_subadult_limbs_downregulated",
#                "global_adults_blastemas_vs_subadults_blastemas_upregulated",
#                "global_adults_blastemas_vs_subadults_blastemas_downregulated",
                "global_adult_limbs_vs_subadult_limbs_upregulated",
                "global_adult_limbs_vs_subadult_limbs_downregulated",
#                "global_age_effect_blastemas_vs_limbs_upregulated",
#                "global_age_effect_blastemas_vs_limbs_downregulated",
#                "global_subadult_blastemas_vs_subadult_limbs_upregulated",
#                "global_subadult_blastemas_vs_subadult_limbs_downregulated",
                "global_adult_blastemas_vs_adult_limbs_upregulated",
                "global_adult_blastemas_vs_adult_limbs_downregulated",
                "annotated_genes",
                "unannotated_genes"),
  top_annotation = upset_top_annotation(genes_comb_mat,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", c(rep(c("#D55E00B3","#0072B2B3"), 4))),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/glob_genes_upset.svg"), 
  width = 6.29921, height = 3.93701)

UpSet(genes_comb_mat,
    #set_order = rownames(genes_comb_mat),
  set_order = c("native_adult_limbs_vs_subadult_limbs_upregulated",
                "native_adult_limbs_vs_subadult_limbs_downregulated",
#                "native_subadult_blastemas_vs_subadult_limbs_upregulated",
#                "native_subadult_blastemas_vs_subadult_limbs_downregulated",
#                "global_adults_blastemas_vs_subadults_blastemas_upregulated",
#                "global_adults_blastemas_vs_subadults_blastemas_downregulated",
                "global_adult_limbs_vs_subadult_limbs_upregulated",
                "global_adult_limbs_vs_subadult_limbs_downregulated",
#                "global_age_effect_blastemas_vs_limbs_upregulated",
#                "global_age_effect_blastemas_vs_limbs_downregulated",
#                "global_subadult_blastemas_vs_subadult_limbs_upregulated",
#                "global_subadult_blastemas_vs_subadult_limbs_downregulated",
                "global_adult_blastemas_vs_adult_limbs_upregulated",
                "global_adult_blastemas_vs_adult_limbs_downregulated",
                "annotated_genes",
                "unannotated_genes"),
  top_annotation = upset_top_annotation(genes_comb_mat,
    add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", c(rep(c("#D55E00B3","#0072B2B3"), 4))),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

```


Upset Figure 4 - RE Intersection between native and global TEcount

```{r}

# keep only sample_group == native
df <- all_diffexp_results %>%
        filter(quant_tool == "TEcount",
               feature_biotype == "repetitive_element") %>%
        # convert columns to factors so that .drop = FALSE works
        mutate(sample_group = as.factor(sample_group),
               diffexp_contrast = as.factor(diffexp_contrast),
               diffexp_res_lfc = as.factor(diffexp_res_lfc),
               diffexp_res_padj = as.factor(diffexp_res_padj)) %>%
        # group by sample_group, diffexp_contrast and diffexp_res_lfc
        group_by(sample_group, diffexp_contrast, diffexp_res_lfc, diffexp_res_padj, .drop = FALSE) %>%
        summarise(re = list(ID), .groups = "keep") %>%
        filter(diffexp_res_lfc %in% c("upregulated", "downregulated"),
               diffexp_res_padj == "significant") %>%
        filter(
          (sample_group == "native" & diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs", "adult_limbs_vs_subadult_limbs"))
          | (sample_group == "global" & diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs", "age_effect_blastemas_vs_limbs",
                                                                     "subadult_blastemas_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",
                                                                     "adult_blastemas_vs_adult_limbs"))
        ) %>%
        ungroup() %>%
        group_by(sample_group, diffexp_contrast, diffexp_res_lfc)

df_keys <- df %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(sample_group, diffexp_contrast, diffexp_res_lfc, sep = "_")) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            lapply(function(x) sort(unlist(x$re)))

all_re_classification <- all_diffexp_results %>%
                          filter(quant_tool == "TEcount",
                                 feature_biotype == "repetitive_element",
                                 diffexp_res_lfc %in% c("upregulated", "downregulated"),
                                 diffexp_res_padj == "significant",
                                (sample_group == "native" & diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs", "adult_limbs_vs_subadult_limbs"))
                                | (sample_group == "global" & diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs", "age_effect_blastemas_vs_limbs",
                                                                     "subadult_blastemas_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",
                                                                     "adult_blastemas_vs_adult_limbs"))
                          ) %>%
        # replace re_superfamily that start with "rnd" with "Unknown"
        mutate(re_superfamily = ifelse(grepl("^rnd", re_superfamily), "Unknown", re_superfamily)) %>%
        group_by(re_superfamily, re_class) %>%
        summarise(re = list(ID), .groups = "keep") %>%
        # in re_superfamily, replace "Gypsy" with "Ty3"
        mutate(re_superfamily = gsub("Gypsy_Chromoviridae_Vclade", "Vclade:Chromoviridae:Ty3", re_superfamily),
               re_superfamily = gsub("Gypsy_Mag", "Mag:Ty3", re_superfamily),
               re_superfamily = gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", re_superfamily))

all_re_classification_keys <- all_re_classification %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(re_superfamily, re_class, sep = ":")) %>%
          pull(key) %>%
          #replace "Unknown:Unknown" with "Unknown"
          gsub("Unknown:Unknown", "Unknown", .)

all_re_classification_list <- all_re_classification %>%
            group_split() %>%
            setNames(all_re_classification_keys) %>%
            lapply(function(x) sort(unlist(x$re)))

all_re_list <- c(df_list, all_re_classification_list)

comb_mat_nat_re <- make_comb_mat(all_re_list)


# to sort by the inverse of alphabetical
sort_by_last_char <- function(char_vector) {
  # Function to reverse a string
  reverse_string <- function(x) {
    sapply(lapply(strsplit(x, NULL), rev), paste, collapse = "")
  }
  
  # Reverse each string in the character vector
  reversed_vector <- reverse_string(char_vector)
  
  # Get the order of the reversed vector
  sorted_indices <- order(reversed_vector)
  
  # Sort the original vector based on the reversed vector
  sorted_char_vector <- char_vector[sorted_indices]
  
  return(sorted_char_vector)
}

# Example usage
sorted_char_vector <- sort_by_last_char(all_re_classification_keys)

# remove "Unknown" from sorted_char_vector
sorted_char_vector <- sorted_char_vector[sorted_char_vector != "Unknown"]
# add "Unknown" at the end
sorted_char_vector <- c(sorted_char_vector, "Unknown")

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/upset_figure4.png"),
    width = 15.24, height = 24, units = "cm", res = 600)

UpSet(t(comb_mat_nat_re),
  set_order = c("native_adult_limbs_vs_subadult_limbs_upregulated",
                "native_adult_limbs_vs_subadult_limbs_downregulated",
                "native_subadult_blastemas_vs_subadult_limbs_upregulated",
                "native_subadult_blastemas_vs_subadult_limbs_downregulated",
                "global_adults_blastemas_vs_subadults_blastemas_upregulated",
                "global_adults_blastemas_vs_subadults_blastemas_downregulated",
                "global_adult_limbs_vs_subadult_limbs_upregulated",
                "global_adult_limbs_vs_subadult_limbs_downregulated",
                "global_age_effect_blastemas_vs_limbs_upregulated",
                "global_age_effect_blastemas_vs_limbs_downregulated",
                "global_subadult_blastemas_vs_subadult_limbs_upregulated",
                "global_subadult_blastemas_vs_subadult_limbs_downregulated",
                "global_adult_blastemas_vs_adult_limbs_upregulated",
                "global_adult_blastemas_vs_adult_limbs_downregulated",
                sorted_char_vector)
                ,
  top_annotation = upset_top_annotation(t(comb_mat_nat_re), add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(t(comb_mat_nat_re), add_numbers = TRUE),
  bg_col = c(rep(c("#D55E00B3","#0072B2B3"), 7),rep("#999999B3", length(all_re_classification_keys))),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/upset_figure4.svg"),
    width = 6, height = 9.44882)

UpSet(t(comb_mat_nat_re),
  set_order = c("native_adult_limbs_vs_subadult_limbs_upregulated",
                "native_adult_limbs_vs_subadult_limbs_downregulated",
                "native_subadult_blastemas_vs_subadult_limbs_upregulated",
                "native_subadult_blastemas_vs_subadult_limbs_downregulated",
                "global_adults_blastemas_vs_subadults_blastemas_upregulated",
                "global_adults_blastemas_vs_subadults_blastemas_downregulated",
                "global_adult_limbs_vs_subadult_limbs_upregulated",
                "global_adult_limbs_vs_subadult_limbs_downregulated",
                "global_age_effect_blastemas_vs_limbs_upregulated",
                "global_age_effect_blastemas_vs_limbs_downregulated",
                "global_subadult_blastemas_vs_subadult_limbs_upregulated",
                "global_subadult_blastemas_vs_subadult_limbs_downregulated",
                "global_adult_blastemas_vs_adult_limbs_upregulated",
                "global_adult_blastemas_vs_adult_limbs_downregulated",
                sorted_char_vector),
  top_annotation = upset_top_annotation(t(comb_mat_nat_re), add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(t(comb_mat_nat_re), add_numbers = TRUE),
  bg_col = c(rep(c("#D55E00B3","#0072B2B3"), 7),rep("#999999B3", length(all_re_classification_keys))),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)


dev.off()


```


REVISED FIGURE

```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(DESeq2)
library(ComplexHeatmap)

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))


# keep only sample_group == native
df <- all_diffexp_results %>%
        filter(quant_tool == "TEcount",
               feature_biotype == "repetitive_element") %>%
        # convert columns to factors so that .drop = FALSE works
        mutate(sample_group = as.factor(sample_group),
               diffexp_contrast = as.factor(diffexp_contrast),
               diffexp_res_lfc = as.factor(diffexp_res_lfc),
               diffexp_res_padj = as.factor(diffexp_res_padj)) %>%
        # group by sample_group, diffexp_contrast and diffexp_res_lfc
        group_by(sample_group, diffexp_contrast, diffexp_res_lfc, diffexp_res_padj, .drop = FALSE) %>%
        summarise(re = list(ID), .groups = "keep") %>%
        filter(lengths(re) > 0) %>%
        filter(diffexp_res_lfc %in% c("upregulated", "downregulated"),
               diffexp_res_padj == "significant") %>%
        filter(
          (sample_group == "native" & diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs", "adult_limbs_vs_subadult_limbs"))
          | (sample_group == "global" & diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs", "age_effect_blastemas_vs_limbs",
                                                                     "subadult_blastemas_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",
                                                                     "adult_blastemas_vs_adult_limbs"))
        ) %>%
        ungroup() %>%
        group_by(sample_group, diffexp_contrast, diffexp_res_lfc)



df_keys <- df %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(sample_group, diffexp_contrast, diffexp_res_lfc, sep = "_")) %>%
          pull(key)

df_list <- df %>%
            group_split() %>%
            setNames(df_keys) %>%
            lapply(function(x) sort(unlist(x$re)))

all_re_classification <- all_diffexp_results %>%
                          filter(quant_tool == "TEcount",
                                 feature_biotype == "repetitive_element",
                                 diffexp_res_lfc %in% c("upregulated", "downregulated"),
                                 diffexp_res_padj == "significant",
                                (sample_group == "native" & diffexp_contrast %in% c("subadult_blastemas_vs_subadult_limbs", "adult_limbs_vs_subadult_limbs"))
                                | (sample_group == "global" & diffexp_contrast %in% c("adult_limbs_vs_subadult_limbs", "age_effect_blastemas_vs_limbs",
                                                                     "subadult_blastemas_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",
                                                                     "adult_blastemas_vs_adult_limbs"))
                          ) %>%
        # replace re_superfamily that start with "rnd" with "Unknown"
        mutate(re_superfamily = ifelse(grepl("^rnd", re_superfamily), "Unknown", re_superfamily)) %>%
        group_by(re_superfamily, re_class) %>%
        summarise(re = list(ID), .groups = "keep") %>%
        # in re_superfamily, replace "Gypsy" with "Ty3"
        mutate(re_superfamily = gsub("Gypsy_Chromoviridae_Vclade", "Vclade:Chromoviridae:Ty3", re_superfamily),
               re_superfamily = gsub("Gypsy_Mag", "Mag:Ty3", re_superfamily),
               re_superfamily = gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", re_superfamily))

all_re_classification_keys <- all_re_classification %>% 
          group_keys() %>%
          mutate(key = stringr::str_c(re_superfamily, re_class, sep = ":")) %>%
          pull(key) %>%
          #replace "Unknown:Unknown" with "Unknown"
          gsub("Unknown:Unknown", "Unknown", .)

all_re_classification_list <- all_re_classification %>%
            group_split() %>%
            setNames(all_re_classification_keys) %>%
            lapply(function(x) sort(unlist(x$re)))

all_re_list <- c(df_list, all_re_classification_list)

comb_mat_nat_re <- make_comb_mat(all_re_list)


# to sort by the inverse of alphabetical
sort_by_last_char <- function(char_vector) {
  # Function to reverse a string
  reverse_string <- function(x) {
    sapply(lapply(strsplit(x, NULL), rev), paste, collapse = "")
  }
  
  # Reverse each string in the character vector
  reversed_vector <- reverse_string(char_vector)
  
  # Get the order of the reversed vector
  sorted_indices <- order(reversed_vector)
  
  # Sort the original vector based on the reversed vector
  sorted_char_vector <- char_vector[sorted_indices]
  
  return(sorted_char_vector)
}

# Example usage
sorted_char_vector <- sort_by_last_char(all_re_classification_keys)

# remove "Unknown" from sorted_char_vector
sorted_char_vector <- sorted_char_vector[sorted_char_vector != "Unknown"]
# add "Unknown" at the end
sorted_char_vector <- c(sorted_char_vector, "Unknown")

# Save as PNG
png(file.path(NEW_PROJECT_DIR, "/revised_figures/upset_figure4_revised.png"),
    width = 15.24, height = 24, units = "cm", res = 600)

UpSet(t(comb_mat_nat_re),
  set_order = c("global_adult_blastemas_vs_adult_limbs_upregulated",
                "global_adult_blastemas_vs_adult_limbs_downregulated",
                "global_adult_limbs_vs_subadult_limbs_upregulated",
                "global_age_effect_blastemas_vs_limbs_downregulated",
                "global_age_effect_blastemas_vs_limbs_upregulated",
                "global_subadult_blastemas_vs_subadult_limbs_upregulated",
                "global_subadult_blastemas_vs_subadult_limbs_downregulated",
                "native_subadult_blastemas_vs_subadult_limbs_downregulated",
                "native_subadult_blastemas_vs_subadult_limbs_upregulated",
                sorted_char_vector),
  top_annotation = upset_top_annotation(t(comb_mat_nat_re), add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(t(comb_mat_nat_re), add_numbers = TRUE),
  bg_col = c(c("#D55E00B3","#0072B2B3","#D55E00B3","#0072B2B3","#D55E00B3","#0072B2B3","#D55E00B3","#0072B2B3","#D55E00B3"),rep("#999999B3", length(all_re_classification_keys))),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)

dev.off()

svg(file.path(NEW_PROJECT_DIR, "/revised_figures/upset_figure4_revised.svg"),
    width = 6, height = 9.44882)

UpSet(t(comb_mat_nat_re),
  set_order = c("global_adult_blastemas_vs_adult_limbs_upregulated",
                "global_adult_blastemas_vs_adult_limbs_downregulated",
                "global_adult_limbs_vs_subadult_limbs_upregulated",
                "global_age_effect_blastemas_vs_limbs_downregulated",
                "global_age_effect_blastemas_vs_limbs_upregulated",
                "global_subadult_blastemas_vs_subadult_limbs_upregulated",
                "global_subadult_blastemas_vs_subadult_limbs_downregulated",
                "native_subadult_blastemas_vs_subadult_limbs_downregulated",
                "native_subadult_blastemas_vs_subadult_limbs_upregulated",
                sorted_char_vector),
  top_annotation = upset_top_annotation(t(comb_mat_nat_re), add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(t(comb_mat_nat_re), add_numbers = TRUE),
  bg_col = c(c("#D55E00B3","#0072B2B3","#D55E00B3","#0072B2B3","#D55E00B3","#0072B2B3","#D55E00B3","#0072B2B3","#D55E00B3"),rep("#999999B3", length(all_re_classification_keys))),
  bg_pt_col = "white",
  row_names_gp = gpar(fontsize = 5)
)


dev.off()




```

###### Annotation of DEREs (alignment, consensus)

Install T-Coffee

```{bash}

cd $SOFTWARE_DIR
wget https://s3.eu-central-1.amazonaws.com/tcoffee-packages/Stable/Latest/T-COFFEE_distribution_Version_13.46.0.919e8c6b.tar.gz

tar -xvzf T-COFFEE_distribution_Version_13.46.0.919e8c6b.tar.gz
rm T-COFFEE_distribution_Version_13.46.0.919e8c6b.tar.gz

cd T-COFFEE_distribution_Version_13.46.0.919e8c6b

# edit MAX_N_PID as described in https://www.biostars.org/p/463869/#463872
# the ouput of cat /proc/sys/kernel/pid_max

./install t_coffee -update -nobinaries

export PATH=$HOME_DIR/.t_coffee/bin/linux:$PATH

export PLUGINS_4_TCOFFEE=$HOME_DIR/.t_coffee/plugins/linux:

# change TMPDIR 
export TMPDIR=/scratch/tmp/${USER}/


```

Installing InterProScan

```{bash}

cd $SOFTWARE_DIR
mkdir interproscan
cd interproscan

wget https://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.66-98.0/interproscan-5.66-98.0-64-bit.tar.gz
wget https://ftp.ebi.ac.uk/pub/software/unix/iprscan/5/5.66-98.0/interproscan-5.66-98.0-64-bit.tar.gz.md5

# Recommended checksum to confirm the download was successful:
md5sum -c interproscan-5.66-98.0-64-bit.tar.gz.md5
# Must return *interproscan-5.66-98.0-64-bit.tar.gz: OK*
# If not - try downloading the file again as it may be a corrupted copy.

tar -pxvzf interproscan-5.66-98.0-*-bit.tar.gz

rm interproscan-5.66-98.0-64-bit.tar.gz

cd interproscan-5.66-98.0

python3 setup.py -f interproscan.properties

```

Generating consensi for selected DERE families

```{bash}

# extract lines of file all_diffexp_results_re.txt to array 
re_subfamilies=$(cut -f1 $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/all_diffexp_results_re.txt)

NTHREADS=8

chmod +x $PROJECT_DIR/axoloTE/scripts/RE_protein_domain_annotation.sh

mkdir -p $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus

cd $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus
parallel \
  --keep-order --line-buffer --jobs 4 \
  $PROJECT_DIR/axoloTE/scripts/RE_protein_domain_annotation.sh {} \
  $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.fa \
  $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/ambMex60DD_ucsc_hub.ALL.mod.bed \
  $NTHREADS \
  $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus/ \
  /scratch/tmp/${USER}/ \
  ::: ${re_subfamilies[@]}


cat $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus/*.cons.fa > $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus/all_cons.fa

# Protein domain annotation of consensus sequence [ InterProScan ]
$PROJECT_DIR/axoloTE/scripts/interproscan/interproscan-5.66-98.0/interproscan.sh \
    --cpu $NTHREADS \
    --input $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus/all_cons.fa \
    --seqtype n \
    --formats TSV,GFF3,XML \
    --iprlookup \
    --pathways \
    --goterms \
    --tempdir $TEMP_DIR \
    --verbose \
    --output-file-base $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/interproscan/all_cons \
    2> $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/interproscan/all_cons.interproscan.log


# TEtrimmer
mamba create -n srp_tetrimmer bioconda::tetrimmer

mamba activate srp_tetrimmer

mkdir -p $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/TEtrimmer

cd $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/TEtrimmer

TEtrimmer \
  --input_file $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/consensus/all_cons.fa \
  --genome_file $PROJECT_DIR/axolotl/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.fa \
  --output_dir $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/TEtrimmer \
  --num_threads 40 \
  --classify_all \
  --dedup \
  --genome_anno \
  --hmm \
  --debug \
  --max_msa_lines 1100 \
  --top_msa_lines 1000 \
  --max_cluster_num 5 \
  > $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/TEtrimmer/TEtrimmer.log 2>&1



```



ExTRAAAAAAAAAA

Extracting BEDs

```{bash}
mkdir -p $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation

# extract first column without header from $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/summary_table.txt
cut -f1 $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/summary_table.txt | tail -n +2 > $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/DEREs.txt

# for each DERE, extract the lines in nat_explorate_and_tecount_all_sig_re_loci.bed that contain it
while read line; do grep $line $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/nat_explorate_and_tecount_all_sig_re_loci.bed > $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/$line.bed; done < $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/DEREs.txt

```

EXTRACTING FASTA

```{bash}
# To get max length of fasta headers to use in fold
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.bed
do
  cat $i | awk -F'\t' '{print $4 "_" $1 ":" $2 "-" $3}' | awk 'length($0) > max_length { max_length = length($0); max_line = $0 } END { print length(max_line) +1 }'
done

for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.bed
do
  #remove extension
  j=${i%.bed}
  bedtools getfasta -name -fi $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.fa -bed $i \
  | fold -w 120 > $j.fa
done

```

ALIGNING

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.fa
do
  #remove .fa extension
  j=${i%.fa}
  mafft --reorder --thread 22 $i > $j.maf.fa 2> $j.maf.log
done

```

REMOVING GAPS

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.fa
do

  #remove .maf.fa extension
  j=${i%.maf.fa}
  t_coffee -other_pg seq_reformat -in $i -action +rm_gap 80 \
    > $j.maf.tcoffee_rm_gap.fa \
    2> $j.maf.tcoffee_rm_gap.log

done

```

CONSENSUS

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.tcoffee_rm_gap.fa
do

  #remove extension
  j=${i%.maf.tcoffee_rm_gap.fa}
  sname=$(basename $j)
  cons -sequence $i -outseq $j.maf.tcoffee_rm_gap.cons.fa -plurality 1 -n $sname \
    > $j.maf.tcoffee_rm_gap.cons.log \
    2> $j.maf.tcoffee_rm_gap.cons.err

done

```

p-HMMs

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.fa
do
  #remove extension
  j=${i%.maf.fa}
  sname=$(basename $j)
  hmmbuild -n $sname --cpu 8 $j.maf.fa.stk $i \
    > $j.maf.fa.stk.hmmbuild.log \
    2> $j.maf.fa.stk.hmmbuild.err

done

```

GETORF and PFAM_SCAN

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.tcoffee_rm_gap.cons.fa
do

  #remove extension
  j=${i%.maf.tcoffee_rm_gap.cons.fa}
  
  getorf -sequence $i -outseq $j.maf.tcoffee_rm_gap.cons.fa.orfs -minsize 200 \
    > $j.maf.tcoffee_rm_gap.cons.fa.getorf.log \
    2> $j.maf.tcoffee_rm_gap.cons.fa.getorf.err

done


for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.tcoffee_rm_gap.cons.fa.orfs
do

  #remove extension
  j=${i%.maf.tcoffee_rm_gap.cons.fa.orfs}
  
  pfam_scan.pl \
    -fasta $i \
    -dir $PROJECT_DIR/ext_data/Pfam_db \
    -outfile $j.maf.tcoffee_rm_gap.cons.fa.getorf.pfam \
    > $j.maf.tcoffee_rm_gap.cons.fa.getorf.pfam.log \
    2> $j.maf.tcoffee_rm_gap.cons.fa.getorf.pfam.err

done

```

TRANSEQ AND PFAM_SCAN

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.tcoffee_rm_gap.cons.fa
do

  #remove extension
  j=${i%.maf.tcoffee_rm_gap.cons.fa}

  transeq \
    -sequence $i \
    -outseq $j.maf.tcoffee_rm_gap.cons.fa.transeq \
    -frame 6 \
    -clean \
    > $j.maf.tcoffee_rm_gap.cons.fa.transeq.log \
    2> $j.maf.tcoffee_rm_gap.cons.fa.transeq.err

done


for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.tcoffee_rm_gap.cons.fa.transeq
do

  #remove extension
  j=${i%.maf.tcoffee_rm_gap.cons.fa.transeq}
  
  pfam_scan.pl \
    -fasta $i \
    -dir $PROJECT_DIR/ext_data/Pfam_db \
    -outfile $j.maf.tcoffee_rm_gap.cons.fa.transeq.pfam \
    > $j.maf.tcoffee_rm_gap.cons.fa.transeq.pfam.log \
    2> $j.maf.tcoffee_rm_gap.cons.fa.transeq.pfam.err

done


```

BLASTX

```{bash}
for i in $PROJECT_DIR/rnaseq/DEA_consensus/DERE_annotation/*.maf.tcoffee_rm_gap.cons.fa
do

  #remove extension
  j=${i%.maf.tcoffee_rm_gap.cons.fa}
  
  blastx -query $i \
    -db $PROJECT_DIR/repeat_annotation/curated_library/archive/manual_protocol/prot_3/blastx/db/TE_peptide_database.lib \
    -outfmt "6 qseqid qgi qacc qaccver qlen sseqid sallseqid sgi sallgi sacc saccver sallacc slen qstart qend sstart send qseq sseq evalue bitscore score length pident nident mismatch positive gapopen gaps ppos frames qframe sframe btop staxids sscinames scomnames sblastnames sskingdoms stitle salltitles sstrand qcovs qcovhsp" \
    -out $j.maf.tcoffee_rm_gap.cons.fa.blastx.tab \
    > $j.maf.tcoffee_rm_gap.cons.fa.blastx.log \
    2> $j.maf.tcoffee_rm_gap.cons.fa.blastx.err

done

# BLASTX 2.14.0+

# Fields: query id, query gi, query acc., query acc.ver, query length, subject id, subject ids, subject gi, subject gis, subject acc., subject acc.ver, subject accs., subject length, q. start, q. end, s. start, s. end, query seq, subject seq, evalue, bit score, score, alignment length, % identity, identical, mismatches, positives, gap opens, gaps, % positives, query/sbjct frames, query frame, sbjct frame, BTOP, subject tax ids, subject sci names, subject com names, subject blast names, subject super kingdoms, subject title, subject titles, subject strand, % query coverage per subject, % query coverage per hsp
# 1285 hits found

```

### Global samples

#### DEGs

##### UpSet

```{bash}
mamba activate srp_axolote

```

```{r}
library(ggplot2)
library(DESeq2)
library(ComplexHeatmap)
```

```{r}
# read gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"), header = TRUE, sep = "\t")

# Salmon

load(file.path(PROJECT_DIR, "/rnaseq/Salmon/global/DEA/results/glob_salmon_res_all.RData"))

glob_salmon_res_a_vs_sa_b_up <- subset(glob_salmon_res_adult_vs_subadult_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_salmon_res_a_vs_sa_b_down <- subset(glob_salmon_res_adult_vs_subadult_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_salmon_res_a_vs_sa_l_up <- subset(glob_salmon_res_adult_limbs_vs_subadult_limbs, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_salmon_res_a_vs_sa_l_down <- subset(glob_salmon_res_adult_limbs_vs_subadult_limbs, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_salmon_res_a_vs_sa_nl_up <- subset(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_salmon_res_a_vs_sa_nl_down <- subset(glob_salmon_res_adult_nonlimbs_vs_subadult_nonlimbs, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_salmon_res_int_b_vs_l_up <- subset(glob_salmon_res_int_Age_group_effect_blastema_vs_limb, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_salmon_res_int_b_vs_l_down <- subset(glob_salmon_res_int_Age_group_effect_blastema_vs_limb, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_salmon_res_int_l_vs_nl_up <- subset(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_salmon_res_int_l_vs_nl_down <- subset(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_salmon_res_int_b_vs_nl_up <- subset(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_salmon_res_int_b_vs_nl_down <- subset(glob_salmon_res_int_Age_groupAdult_Tissue_groupLimb_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))

# TEcount

load(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/glob_tecount_res_all.RData"))

glob_tecount_res_a_vs_sa_b_up <- subset(glob_tecount_res_adult_vs_subadult_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_tecount_res_a_vs_sa_b_down <- subset(glob_tecount_res_adult_vs_subadult_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_tecount_res_a_vs_sa_l_up <- subset(glob_tecount_res_adult_limbs_vs_subadult_limbs, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_tecount_res_a_vs_sa_l_down <- subset(glob_tecount_res_adult_limbs_vs_subadult_limbs, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_tecount_res_a_vs_sa_nl_up <- subset(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_tecount_res_a_vs_sa_nl_down <- subset(glob_tecount_res_adult_nonlimbs_vs_subadult_nonlimbs, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_tecount_res_int_b_vs_l_up <- subset(glob_tecount_res_int_Age_group_effect_blastema_vs_limb, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_tecount_res_int_b_vs_l_down <- subset(glob_tecount_res_int_Age_group_effect_blastema_vs_limb, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_tecount_res_int_l_vs_nl_up <- subset(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_tecount_res_int_l_vs_nl_down <- subset(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_tecount_res_int_b_vs_nl_up <- subset(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_tecount_res_int_b_vs_nl_down <- subset(glob_tecount_res_int_Age_groupAdult_Tissue_groupLimb_blastema, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))


# keep only rows where rowname starts with "AMEX"
glob_tecount_res_a_vs_sa_b_genes_up <- glob_tecount_res_a_vs_sa_b_up[grep("^AMEX", rownames(glob_tecount_res_a_vs_sa_b_up)), ]
glob_tecount_res_a_vs_sa_b_genes_down <- glob_tecount_res_a_vs_sa_b_down[grep("^AMEX", rownames(glob_tecount_res_a_vs_sa_b_down)), ]
glob_tecount_res_a_vs_sa_l_genes_up <- glob_tecount_res_a_vs_sa_l_up[grep("^AMEX", rownames(glob_tecount_res_a_vs_sa_l_up)), ]
glob_tecount_res_a_vs_sa_l_genes_down <- glob_tecount_res_a_vs_sa_l_down[grep("^AMEX", rownames(glob_tecount_res_a_vs_sa_l_down)), ]
glob_tecount_res_a_vs_sa_nl_genes_up <- glob_tecount_res_a_vs_sa_nl_up[grep("^AMEX", rownames(glob_tecount_res_a_vs_sa_nl_up)), ]
glob_tecount_res_a_vs_sa_nl_genes_down <- glob_tecount_res_a_vs_sa_nl_down[grep("^AMEX", rownames(glob_tecount_res_a_vs_sa_nl_down)), ]
glob_tecount_res_int_b_vs_l_genes_up <- glob_tecount_res_int_b_vs_l_up[grep("^AMEX", rownames(glob_tecount_res_int_b_vs_l_up)), ]
glob_tecount_res_int_b_vs_l_genes_down <- glob_tecount_res_int_b_vs_l_down[grep("^AMEX", rownames(glob_tecount_res_int_b_vs_l_down)), ]
glob_tecount_res_int_l_vs_nl_genes_up <- glob_tecount_res_int_l_vs_nl_up[grep("^AMEX", rownames(glob_tecount_res_int_l_vs_nl_up)), ]
glob_tecount_res_int_l_vs_nl_genes_down <- glob_tecount_res_int_l_vs_nl_down[grep("^AMEX", rownames(glob_tecount_res_int_l_vs_nl_down)), ]
glob_tecount_res_int_b_vs_nl_genes_up <- glob_tecount_res_int_b_vs_nl_up[grep("^AMEX", rownames(glob_tecount_res_int_b_vs_nl_up)), ]
glob_tecount_res_int_b_vs_nl_genes_down <- glob_tecount_res_int_b_vs_nl_down[grep("^AMEX", rownames(glob_tecount_res_int_b_vs_nl_down)), ]

# annotated/unannotated DEGs
all_degs <- unique(c(
  rownames(glob_salmon_res_a_vs_sa_b_up),
  rownames(glob_salmon_res_a_vs_sa_b_down),
  rownames(glob_salmon_res_a_vs_sa_l_up),
  rownames(glob_salmon_res_a_vs_sa_l_down),
  rownames(glob_salmon_res_a_vs_sa_nl_up),
  rownames(glob_salmon_res_a_vs_sa_nl_down),
  rownames(glob_salmon_res_int_b_vs_l_up),
  rownames(glob_salmon_res_int_b_vs_l_down),
  rownames(glob_salmon_res_int_l_vs_nl_up),
  rownames(glob_salmon_res_int_l_vs_nl_down),
  rownames(glob_salmon_res_int_b_vs_nl_up),
  rownames(glob_salmon_res_int_b_vs_nl_down),
  rownames(glob_tecount_res_a_vs_sa_b_genes_up),
  rownames(glob_tecount_res_a_vs_sa_b_genes_down),
  rownames(glob_tecount_res_a_vs_sa_l_genes_up),
  rownames(glob_tecount_res_a_vs_sa_l_genes_down),
  rownames(glob_tecount_res_a_vs_sa_nl_genes_up),
  rownames(glob_tecount_res_a_vs_sa_nl_genes_down),
  rownames(glob_tecount_res_int_b_vs_l_genes_up),
  rownames(glob_tecount_res_int_b_vs_l_genes_down),
  rownames(glob_tecount_res_int_l_vs_nl_genes_up),
  rownames(glob_tecount_res_int_l_vs_nl_genes_down),
  rownames(glob_tecount_res_int_b_vs_nl_genes_up),
  rownames(glob_tecount_res_int_b_vs_nl_genes_down)
))

# ann_degs are DEGs that have a gene_symbol that does not start with AMEX in gene_metadata$gene_symbol
ann_degs <- all_degs[which(!grepl("^AMEX", gene_metadata$gene_symbol[match(all_degs, gene_metadata$gene_id)]))]
unann_degs <- all_degs[which(grepl("^AMEX", gene_metadata$gene_symbol[match(all_degs, gene_metadata$gene_id)]))]

# limited to 31 sets
genes_list <- list(
  "salmon A vs SA B UP" = sort(rownames(glob_tecount_res_a_vs_sa_b_up)),
  "salmon A vs SA B DOWN" = sort(rownames(glob_tecount_res_a_vs_sa_b_down)),
  "salmon A vs SA L UP" = sort(rownames(glob_tecount_res_a_vs_sa_l_up)),
  "salmon A vs SA L DOWN" = sort(rownames(glob_tecount_res_a_vs_sa_l_down)),
  "salmon A vs SA NL UP" = sort(rownames(glob_tecount_res_a_vs_sa_nl_up)),
  "salmon A vs SA NL DOWN" = sort(rownames(glob_tecount_res_a_vs_sa_nl_down)),
  "salmon int B vs L UP" = sort(rownames(glob_tecount_res_int_b_vs_l_up)),
  "salmon int B vs L DOWN" = sort(rownames(glob_tecount_res_int_b_vs_l_down)),
  "salmon int L vs NL UP" = sort(rownames(glob_tecount_res_int_l_vs_nl_up)),
  "salmon int L vs NL DOWN" = sort(rownames(glob_tecount_res_int_l_vs_nl_down)),
  "salmon int B vs NL UP" = sort(rownames(glob_tecount_res_int_b_vs_nl_up)),
  "salmon int B vs NL DOWN" = sort(rownames(glob_tecount_res_int_b_vs_nl_down)),
  "TEcount A vs SA B UP" = sort(rownames(glob_tecount_res_a_vs_sa_b_genes_up)),
  "TEcount A vs SA B DOWN" = sort(rownames(glob_tecount_res_a_vs_sa_b_genes_down)),
  "TEcount A vs SA L UP" = sort(rownames(glob_tecount_res_a_vs_sa_l_genes_up)),
  "TEcount A vs SA L DOWN" = sort(rownames(glob_tecount_res_a_vs_sa_l_genes_down)),
  "TEcount A vs SA NL UP" = sort(rownames(glob_tecount_res_a_vs_sa_nl_genes_up)),
  "TEcount A vs SA NL DOWN" = sort(rownames(glob_tecount_res_a_vs_sa_nl_genes_down)),
  "TEcount int B vs L UP" = sort(rownames(glob_tecount_res_int_b_vs_l_genes_up)),
  "TEcount int B vs L DOWN" = sort(rownames(glob_tecount_res_int_b_vs_l_genes_down)),
  "TEcount int L vs NL UP" = sort(rownames(glob_tecount_res_int_l_vs_nl_genes_up)),
  "TEcount int L vs NL DOWN" = sort(rownames(glob_tecount_res_int_l_vs_nl_genes_down)),
  "TEcount int B vs NL UP" = sort(rownames(glob_tecount_res_int_b_vs_nl_genes_up)),
  "TEcount int B vs NL DOWN" = sort(rownames(glob_tecount_res_int_b_vs_nl_genes_down)),
  "annotated DEGs" = sort(ann_degs),
  "unannotated DEGs" = sort(unann_degs)
)


genes_comb_mat <- make_comb_mat(genes_list)

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/global/upset_glob_genes.png"),
  width = 80,
  height = 20,
  units = "cm",
  res = 600
)

UpSet(genes_comb_mat,
  set_order = c("salmon A vs SA B UP", "salmon A vs SA B DOWN", "salmon A vs SA L UP", "salmon A vs SA L DOWN", "salmon A vs SA NL UP", "salmon A vs SA NL DOWN", "salmon int B vs L UP", "salmon int B vs L DOWN", "salmon int L vs NL UP", "salmon int L vs NL DOWN", "salmon int B vs NL UP", "salmon int B vs NL DOWN", "TEcount A vs SA B UP", "TEcount A vs SA B DOWN", "TEcount A vs SA L UP", "TEcount A vs SA L DOWN", "TEcount A vs SA NL UP", "TEcount A vs SA NL DOWN", "TEcount int B vs L UP", "TEcount int B vs L DOWN", "TEcount int L vs NL UP", "TEcount int L vs NL DOWN", "TEcount int B vs NL UP", "TEcount int B vs NL DOWN", "annotated DEGs", "unannotated DEGs"),
  top_annotation = upset_top_annotation(genes_comb_mat, add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white"
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/nat_genes_upset.svg"), width = 7.87402, height = 4.133858)
UpSet(genes_comb_mat,
  set_order = c("salmon A vs SA UP", "salmon A vs SA DOWN", "salmon B vs L UP", "salmon B vs L DOWN", "TEcount A vs SA UP", "TEcount A vs SA DOWN", "TEcount B vs L UP", "TEcount B vs L DOWN", "annotated DEGs", "unannotated DEGs"),
  top_annotation = upset_top_annotation(genes_comb_mat, add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(genes_comb_mat, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white"
)
dev.off()
```

#### DEREs

##### UpSet

```{r}
glob_exp_f_l_vs_nl_sa_up_S <- subset(glob_explorate_fam_res_l_vs_nl_sa, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_l_vs_nl_sa_mid_NS <- subset(glob_explorate_fam_res_l_vs_nl_sa, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_l_vs_nl_sa_down_S <- subset(glob_explorate_fam_res_l_vs_nl_sa, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_l_vs_nl_sa_down_NS <- subset(glob_explorate_fam_res_l_vs_nl_sa, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_l_vs_nl_a_up_S <- subset(glob_explorate_fam_res_l_vs_nl_a, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_l_vs_nl_a_up_NS <- subset(glob_explorate_fam_res_l_vs_nl_a, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_l_vs_nl_a_mid_S <- subset(glob_explorate_fam_res_l_vs_nl_a, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_l_vs_nl_a_mid_NS <- subset(glob_explorate_fam_res_l_vs_nl_a, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_l_vs_nl_a_down_S <- subset(glob_explorate_fam_res_l_vs_nl_a, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_l_vs_nl_a_down_NS <- subset(glob_explorate_fam_res_l_vs_nl_a, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_b_vs_nl_sa_up_S <- subset(glob_explorate_fam_res_b_vs_nl_sa, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_nl_sa_up_NS <- subset(glob_explorate_fam_res_b_vs_nl_sa, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_nl_sa_mid_S <- subset(glob_explorate_fam_res_b_vs_nl_sa, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_nl_sa_mid_NS <- subset(glob_explorate_fam_res_b_vs_nl_sa, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_nl_sa_down_S <- subset(glob_explorate_fam_res_b_vs_nl_sa, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_b_vs_nl_sa_down_NS <- subset(glob_explorate_fam_res_b_vs_nl_sa, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_b_vs_nl_a_up_S <- subset(glob_explorate_fam_res_b_vs_nl_a, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_nl_a_up_NS <- subset(glob_explorate_fam_res_b_vs_nl_a, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_nl_a_mid_S <- subset(glob_explorate_fam_res_b_vs_nl_a, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_nl_a_mid_NS <- subset(glob_explorate_fam_res_b_vs_nl_a, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_nl_a_down_S <- subset(glob_explorate_fam_res_b_vs_nl_a, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_b_vs_nl_a_down_NS <- subset(glob_explorate_fam_res_b_vs_nl_a, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_b_vs_l_sa_up_S <- subset(glob_explorate_fam_res_b_vs_l_sa, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_l_sa_up_NS <- subset(glob_explorate_fam_res_b_vs_l_sa, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_l_sa_mid_S <- subset(glob_explorate_fam_res_b_vs_l_sa, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_l_sa_mid_NS <- subset(glob_explorate_fam_res_b_vs_l_sa, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_l_sa_down_S <- subset(glob_explorate_fam_res_b_vs_l_sa, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_b_vs_l_sa_down_NS <- subset(glob_explorate_fam_res_b_vs_l_sa, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_b_vs_l_a_up_S <- subset(glob_explorate_fam_res_b_vs_l_a, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_l_a_up_NS <- subset(glob_explorate_fam_res_b_vs_l_a, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_b_vs_l_a_mid_S <- subset(glob_explorate_fam_res_b_vs_l_a, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_l_a_mid_NS <- subset(glob_explorate_fam_res_b_vs_l_a, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_b_vs_l_a_down_S <- subset(glob_explorate_fam_res_b_vs_l_a, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_b_vs_l_a_down_NS <- subset(glob_explorate_fam_res_b_vs_l_a, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_a_vs_sa_nl_up_S <- subset(glob_explorate_fam_res_a_vs_sa_nl, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_a_vs_sa_nl_up_NS <- subset(glob_explorate_fam_res_a_vs_sa_nl, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_a_vs_sa_nl_mid_S <- subset(glob_explorate_fam_res_a_vs_sa_nl, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_a_vs_sa_nl_mid_NS <- subset(glob_explorate_fam_res_a_vs_sa_nl, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_a_vs_sa_nl_down_S <- subset(glob_explorate_fam_res_a_vs_sa_nl, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_a_vs_sa_nl_down_NS <- subset(glob_explorate_fam_res_a_vs_sa_nl, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_a_vs_sa_l_up_S <- subset(glob_explorate_fam_res_a_vs_sa_l, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_a_vs_sa_l_up_NS <- subset(glob_explorate_fam_res_a_vs_sa_l, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_a_vs_sa_l_mid_S <- subset(glob_explorate_fam_res_a_vs_sa_l, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_a_vs_sa_l_mid_NS <- subset(glob_explorate_fam_res_a_vs_sa_l, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_a_vs_sa_l_down_S <- subset(glob_explorate_fam_res_a_vs_sa_l, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_a_vs_sa_l_down_NS <- subset(glob_explorate_fam_res_a_vs_sa_l, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))

glob_exp_f_a_vs_sa_b_up_S <- subset(glob_explorate_fam_res_a_vs_sa_b, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_a_vs_sa_b_up_NS <- subset(glob_explorate_fam_res_a_vs_sa_b, padj_MLE >= 0.05 & log2FoldChange_MMSE > log2(1.5))
glob_exp_f_a_vs_sa_b_mid_S <- subset(glob_explorate_fam_res_a_vs_sa_b, padj_MLE < 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_a_vs_sa_b_mid_NS <- subset(glob_explorate_fam_res_a_vs_sa_b, padj_MLE >= 0.05 & abs(log2FoldChange_MMSE) < log2(1.5))
glob_exp_f_a_vs_sa_b_down_S <- subset(glob_explorate_fam_res_a_vs_sa_b, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))
glob_exp_f_a_vs_sa_b_down_NS <- subset(glob_explorate_fam_res_a_vs_sa_b, padj_MLE >= 0.05 & log2FoldChange_MMSE < -log2(1.5))


# ExplorATE

load(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_explorate_fam_res_adults_vs_subadults.RData"))

load(file.path(PROJECT_DIR, "/rnaseq/ExplorATE/native/DEA/results/nat_explorate_fam_res_blastema_vs_limb.RData"))

nat_exp_f_a_vs_sa_up <- subset(nat_explorate_fam_res_adults_vs_subadults, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
nat_exp_f_a_vs_sa_down <- subset(nat_explorate_fam_res_adults_vs_subadults, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))

nat_exp_f_b_vs_l_up <- subset(nat_explorate_fam_res_blastema_vs_limb, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))
nat_exp_f_b_vs_l_down <- subset(nat_explorate_fam_res_blastema_vs_limb, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))


# TEcount

nat_age_tecount_res_adult_limbs_vs_subadult_limbs <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_age_tecount_res_adult_limbs_vs_subadult_limbs.rds"))

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs <- readRDS(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/native/DEA/results/nat_reg_tecount_res_sa_blastemas_vs_sa_limbs.rds"))

# keep only rows where rowname does not start with "AMEX"
nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re <- nat_age_tecount_res_adult_limbs_vs_subadult_limbs[!grepl("^AMEX", nat_age_tecount_res_adult_limbs_vs_subadult_limbs$gene_id), ]

nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re <- nat_reg_tecount_res_sa_blastemas_vs_sa_limbs[!grepl("^AMEX", nat_reg_tecount_res_sa_blastemas_vs_sa_limbs$gene_id), ]

nat_tecount_a_vs_sa_re_up <- subset(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))

nat_tecount_a_vs_sa_re_down <- subset(nat_age_tecount_res_adult_limbs_vs_subadult_limbs_re, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))

nat_tecount_b_vs_l_re_up <- subset(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re, padj_MLE < 0.05 & log2FoldChange_MMSE > log2(1.5))

nat_tecount_b_vs_l_re_down <- subset(nat_reg_tecount_res_sa_blastemas_vs_sa_limbs_re, padj_MLE < 0.05 & log2FoldChange_MMSE < -log2(1.5))



# annotated/unannotated DEREs

all_deres <- unique(c(
  rownames(nat_exp_f_a_vs_sa_up),
  rownames(nat_exp_f_a_vs_sa_down),
  rownames(nat_exp_f_b_vs_l_up),
  rownames(nat_exp_f_b_vs_l_down),
  rownames(nat_tecount_a_vs_sa_re_up),
  rownames(nat_tecount_a_vs_sa_re_down),
  rownames(nat_tecount_b_vs_l_re_up),
  rownames(nat_tecount_b_vs_l_re_down)
))

# divide by : into a 3 column table where the first is RE_subfam, the second is RE_superfam, the third is RE_class
all_deres2 <- data.frame(do.call("rbind", strsplit(as.character(all_deres), ":", fixed = TRUE)))
colnames(all_deres2) <- c("RE_subfam", "RE_superfam", "RE_class")
rownames(all_deres2) <- all_deres

# replace RE_superfam that start with "rnd" with "Unknown"
all_deres2$RE_superfam <- ifelse(grepl("^rnd", all_deres2$RE_superfam), "Unknown", all_deres2$RE_superfam)
# create list where list element names are RE_superfam and values are rownames(all_deres2) that have this RE_superfam
all_deres_list <- lapply(unique(all_deres2$RE_superfam), function(x) rownames(all_deres2)[which(all_deres2$RE_superfam == x)])
# add names
names(all_deres_list) <- unique(all_deres2$RE_superfam)


# limited to 31 sets
f_list <- list(
  "ExplorATE A vs SA UP" = sort(rownames(nat_exp_f_a_vs_sa_up)),
  "ExplorATE A vs SA DOWN" = sort(rownames(nat_exp_f_a_vs_sa_down)),
  "ExplorATE B vs L UP" = sort(rownames(nat_exp_f_b_vs_l_up)),
  "ExplorATE B vs L DOWN" = sort(rownames(nat_exp_f_b_vs_l_down)),
  "TEcount A vs SA UP" = sort(rownames(nat_tecount_a_vs_sa_re_up)),
  "TEcount A vs SA DOWN" = sort(rownames(nat_tecount_a_vs_sa_re_down)),
  "TEcount B vs L UP" = sort(rownames(nat_tecount_b_vs_l_re_up)),
  "TEcount B vs L DOWN" = sort(rownames(nat_tecount_b_vs_l_re_down), )
)
# concat lists
f_list <- c(f_list, all_deres_list)

comb_mat_nat_re <- make_comb_mat(f_list)

dir.create(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/native/"))

# Save as PNG
png(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/native/upset_nat_re.png"),
  width = 20,
  height = 10.5,
  units = "cm",
  res = 1200
)

UpSet(comb_mat_nat_re,
  set_order = c("TEcount A vs SA UP", "TEcount A vs SA DOWN", "TEcount B vs L UP", "TEcount B vs L DOWN", "ExplorATE A vs SA UP", "ExplorATE A vs SA DOWN", "ExplorATE B vs L UP", "ExplorATE B vs L DOWN", names(all_deres_list)),
  top_annotation = upset_top_annotation(comb_mat_nat_re, add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(comb_mat_nat_re, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white"
)

dev.off()

svg(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/native/upset_nat_re.svg"),
  width = 7.87402,
  height = 4.133858
)

UpSet(comb_mat_nat_re,
  set_order = c("TEcount A vs SA UP", "TEcount A vs SA DOWN", "TEcount B vs L UP", "TEcount B vs L DOWN", "ExplorATE A vs SA UP", "ExplorATE A vs SA DOWN", "ExplorATE B vs L UP", "ExplorATE B vs L DOWN", names(all_deres_list)),
  top_annotation = upset_top_annotation(comb_mat_nat_re, add_numbers = TRUE, numbers_rot = 90),
  right_annotation = upset_right_annotation(comb_mat_nat_re, add_numbers = TRUE),
  bg_col = c("#999999B3", "#999999B3", "#999999B3", "#999999B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3", "#0072B2B3", "#D55E00B3"),
  bg_pt_col = "white"
)

dev.off()
```


temporal 

```{r}

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

# how many upregulated in adult_limbs_vs_subadult_limbs are also down-regulated in adult_blastemas_vs_adult_limbs
all_diffexp_results %>%
  filter(diffexp_contrast == "adult_limbs_vs_subadult_limbs" &
    diffexp_res_lfc == "upregulated" &
    diffexp_res_padj == "significant" &
    quant_tool == "TEcount" &
    feature_biotype == "gene") %>%
  # get gene_id as vector
  pull(gene_id) %>%
  intersect(all_diffexp_results %>%
    filter(diffexp_contrast == "adult_blastemas_vs_adult_limbs" &
      diffexp_res_lfc == "downregulated" &
      diffexp_res_padj == "significant" &
      quant_tool == "TEcount" &
      feature_biotype == "gene") %>%
    # get gene_id as vector
    pull(gene_id)) %>%
  length()

##############
all_diffexp_results %>%
  filter(diffexp_contrast == "subadult_blastemas_vs_subadult_limbs" &
    sample_group == "global" &
    diffexp_res_lfc == "upregulated" &
    diffexp_res_padj == "significant" &
    quant_tool == "TEcount" &
    feature_biotype == "gene")


##############
# print the top repetitive_element for "adult_limbs_vs_subadult_limbs"
all_diffexp_results %>%
  # show where gene_symbol contains string "LOC114595149"
  filter(grepl("^ZNF", gene_symbol) &
    diffexp_res_lfc != "not_diffexp" &
    diffexp_res_padj == "significant") %>%
  print(n = 800, width = Inf)

all_diffexp_results %>%
  # show where gene_symbol contains string "LOC114595149"
  filter(grepl("LOC112547415_dup5", gene_symbol) &
    diffexp_res_lfc != "not_diffexp" &
    diffexp_res_padj == "significant") %>%
  print(n = 50, width = Inf)

# pri

mod_and_colors <- mod_and_colors %>%
  left_join(gene_metadata %>% select(gene_id, gene_symbol), by = "gene_id") %>%
  filter(grepl("TRIM", gene_symbol) &
    colors == "yellow") %>%
  # remove "_dup[0-9]" from gene_symbol
  mutate(gene_symbol = gsub("_dup[0-9]", "", gene_symbol)) %>%
  # get unique values in gene_symbol
  distinct(gene_symbol) %>%
  pull(gene_symbol) %>%
  sort()

# print rows where peak_id is "7750_8250_cluster_73"
deres_ann_and_all_dea_results %>%
  filter(peak_id == "7750_8250_cluster_73" &
    diffexp_contrast.x == "adult_limbs_vs_subadult_limbs" &
    diffexp_contrast.y == "adult_limbs_vs_subadult_limbs" &
    diffexp_res_padj.x == "significant" &
    diffexp_res_padj.y == "significant") %>%
  print(n = 1, width = Inf)




# print rows where gene_symbol contains "ZNF569"
all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element" &
    diffexp_contrast == "adult_blastemas_vs_adult_limbs" &
    diffexp_res_padj == "significant" &
    diffexp_res_lfc == "downregulated") %>%
  # print highest 10 and lowest 10 by log2FoldChange_MMSE
  arrange(desc(log2FoldChange_MMSE)) %>%
  print(n = 26, width = Inf)

all_diffexp_results %>%
  filter(feature_biotype == "gene" &
    diffexp_contrast == "adults_vs_subadults" &
    quant_tool == "TEcount" &
    diffexp_res_padj == "significant" &
    diffexp_res_lfc != "not_diffexp") %>%
  # print highest 10 and lowest 10 by log2FoldChange_MMSE
  arrange(log2FoldChange_MMSE) %>%
  print(n = 50, width = Inf)

####
# print summary of Tecount repetitive elements
all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element" &
    quant_tool == "TEcount" &
    diffexp_res_padj == "significant" &
    diffexp_res_lfc != "not_diffexp" &
    sample_group == "global") %>%
  # summarize per diffexp_contrast and diffexp_res_lfc
  group_by(diffexp_contrast, diffexp_res_lfc) %>%
  summarize(counts = n()) %>%
  print(n = Inf)



###

# with dplyr, filter all_diffexp_results to include only feature_biotypes == "repetitive_element"
all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element" &
    #                 sample_group == "global" &
    diffexp_res_padj == "significant") %>%
  # summary count by quant_tool and diffexp_res_lfc
  group_by(quant_tool, diffexp_res_lfc, diffexp_contrast) %>%
  summarize(counts = n()) %>%
  print(n = Inf)

all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element" &
    diffexp_contrast == "age_effect_blastemas_vs_limbs" &
    diffexp_res_padj == "significant") %>%
  print(n = Inf, width = Inf)
```

##### DEREs x DEGs analysis

Getting chromosome sizes

```{bash}
mkdir -p $SOFTWARE_DIR/UCSC_tools
cd $SOFTWARE_DIR/UCSC_tools
rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/ ./

$SOFTWARE_DIR/UCSC_tools/faSize -detailed -tab $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD.chrom.sizes

chmod -w $PROJECT_DIR/ext_data/axolotl_genome/AmexG_v6.0-DD.chrom.sizes

```

Converting RE GTF to BED

```{bash}
cd $PROJECT_DIR/ext_data/axolotl_genome/repeats

wc -l ambMex60DD_ucsc_hub.ALL.mod.gtf

gtf2bed --do-not-sort < ambMex60DD_ucsc_hub.ALL.mod.gtf > ambMex60DD_ucsc_hub.ALL.mod.bed

wc -l ambMex60DD_ucsc_hub.ALL.mod.bed

chmod -w ambMex60DD_ucsc_hub.ALL.mod.bed

```

reading DEREs

```{r}
library(readr)
library(dplyr)

all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))


# keep only rows where feature_biotype == "repetitive_element", !(diffexp_res_lfc == "not_diffexp"), diffexp_res_padj == "significant"
all_diffexp_results_re <- all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element" &
    !(diffexp_res_lfc == "not_diffexp") &
    diffexp_res_padj == "significant") %>%
  # keep only gene_id as vector
  pull(ID)

all_diffexp_results_re <- unique(all_diffexp_results_re)

# split by ":", keep only first element (the family id)
all_diffexp_results_re <- sapply(strsplit(all_diffexp_results_re, ":"), function(x) x[1])

# save vector to file
write.table(all_diffexp_results_re, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/all_diffexp_results_re.txt"),
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE
)
```

```{bash}
mamba activate srp_axolote

rm -rf $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/tmp
mkdir -p $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/tmp
cd $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/tmp

cat $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/all_diffexp_results_re.txt | \
  parallel \
    --keep-order --line-buffer --jobs 16 \
    $PROJECT_DIR/axoloTE/scripts/all_diffexp_results_re.sh

cat $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/tmp/*.bed > $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/all_diffexp_results_re_loci.bed

rm -rf $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/tmp

# keep only necessary columns in BED format
awk -F'\t' -v OFS='\t' '{print $1,$2,$3,$4,0,$6}' $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/all_diffexp_results_re_loci.bed > $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/all_diffexp_results_re_loci.final.bed

```

```{bash}

rm -r $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/uropa
mkdir -p $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/uropa
cd $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/uropa

uropa \
    --input $PROJECT_DIR/axoloTE/scripts/uropa_config.json \
    --prefix all_diffexp_results_re_loci.uropa \
    --outdir $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/uropa \
    --summary \
    --threads 16 \
    -l $PROJECT_DIR/rnaseq/DEA_consensus/DEREsxDEGs/uropa/uropa.log \
    -d

```

```{r}
library(dplyr)
library(readr)
library(tibble)


# add gene_symbol by matching with gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"),
  sep = "\t",
  header = TRUE,
  stringsAsFactors = FALSE
)
gene_metadata <- as_tibble(gene_metadata)

# load DEA results
all_diffexp_results <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

# add RE info to deres_ann
all_diffexp_results_re <- all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element") %>%
  dplyr::select("re_subfamily", "quant_tool", "sample_group", "diffexp_contrast", "diffexp_res_lfc", "diffexp_res_padj") %>%
  # append ".re" to column names
  rename_with(~ paste0(.x, ".re"))

# remove REs from all_diffexp_results
all_diffexp_results_gene <- all_diffexp_results %>%
  filter(feature_biotype == "gene") %>%
  dplyr::select("ID","quant_tool", "sample_group", "diffexp_contrast", "diffexp_res_lfc", "diffexp_res_padj") %>%
  # append ".re" to column names
  rename_with(~ paste0(.x, ".gene"))


# match deres_ann$gene_id with nat_heatmap_clusters$GeneID, get the corresponding nat_heatmap_clusters$Cluster and check if it is the same when using deres_ann$re_subfamily
nat_heatmap_clusters <- read_tsv(file.path(PROJECT_DIR, "rnaseq/STAR/TEcount/native/DEA/results/heatmap/nat_heatmap_clusters.tsv"))

# from nat_heatmap_clusters$GeneID, remove everything after the first colon
nat_heatmap_clusters$ID <- stringr::str_split_fixed(nat_heatmap_clusters$ID, ":", 2)[, 1]


glob1_heatmap_clusters <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap1_clusters.txt"))
# from glob1_heatmap_clusters$GeneID, remove everything after the first colon
glob1_heatmap_clusters$ID <- stringr::str_split_fixed(glob1_heatmap_clusters$ID, ":", 2)[, 1]


glob2_heatmap_clusters <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/DEA/results/heatmap/glob_full_heatmap2_clusters.txt"))
# from glob_heatmap2_clusters$GeneID, remove everything after the first colon
glob2_heatmap_clusters$ID <- stringr::str_split_fixed(glob2_heatmap_clusters$ID, ":", 2)[, 1]


deres_ann <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/uropa/all_diffexp_results_re_loci.uropa_finalhits.txt"))

```

```{r}

# replace NA with "intergenic_region" in column "feature" and column "name"
# these loci are located in intergenic regions that were not annotated by
# AGAT's agat_sp_add_intergenic_regions.pl (https://agat.readthedocs.io/en/latest/tools/agat_sp_add_intergenic_regions.html)
# because they are leading or trailing regions of a chromosome where there are no genes
# and by default GTF files do not contain these regions

nat_deres_ann <- deres_ann %>%
              # replace NA in deres_ann$feature with "intergenic_region"
              mutate(feature = ifelse(is.na(feature), "intergenic_region", feature),
                     name = ifelse(is.na(feature) & is.na(name), "n_intergenic_region", name)) %>%
              # Adding gene_symbols
              mutate(gene_symbol = gene_metadata$gene_symbol[match(deres_ann$gene_id, gene_metadata$gene_id)]) %>%
              # add re quant_tool, diffexp_contrast, sample_group,  diffexp_res_lfc and diffexp_res_padj to deres_ann by merging
              left_join(all_diffexp_results_re,
                        by = join_by(peak_id == re_subfamily.re)) %>%
              # remove rows where diffexp_res_padj == "not_significant" and diffexp_res_lfc == "not_diffexp"
              filter(diffexp_res_padj.re != "not_significant" & diffexp_res_lfc.re != "not_diffexp") %>%
              # add gene quant_tool, diffexp_contrast, sample_group,  diffexp_res_lfc and diffexp_res_padj to deres_ann by merging
              left_join(all_diffexp_results_gene,
                        by = join_by(gene_id == ID.gene)) %>%
              # keep only rows where quant_tool == quant_tool.y, diffexp_contrast == diffexp_contrast.y, sample_group == sample_group.y
              filter(name == "n_intergenic_region"
                    | (quant_tool.gene == quant_tool.re
                    & diffexp_contrast.gene == diffexp_contrast.re
                    & sample_group.gene == sample_group.re)) %>%
              # add native heatmap clusters for genes
              left_join(nat_heatmap_clusters,
                        by = join_by(gene_id == ID)) %>%
              # rename nat_heatmap_cluster to nat_heatmap_gene_cluster
              dplyr::rename(nat_heatmap_gene_cluster = nat_heatmap_cluster) %>%
              # add native heatmap clusters for REs
              left_join(nat_heatmap_clusters,
                        by = join_by(peak_id == ID)) %>%
              # rename nat_heatmap_cluster to nat_heatmap_re_cluster
              dplyr::rename(nat_heatmap_re_cluster = nat_heatmap_cluster) %>%
              # Rows from global contrasts or ExplorATE quant should not have a cluster
              # because the heatmap clusters were generated with native samples and TEcount quant
              mutate(nat_heatmap_gene_cluster = ifelse(sample_group.re == "global"
                                                        | quant_tool.re != "TEcount"
                                                        | sample_group.gene == "global"
                                                        | quant_tool.gene != "TEcount",
                                                        NA, nat_heatmap_gene_cluster),
                      nat_heatmap_re_cluster = ifelse(sample_group.re == "global"
                                                      | quant_tool.re != "TEcount"
                                                      | sample_group.gene == "global"
                                                      | quant_tool.gene != "TEcount",
                                                      NA, nat_heatmap_re_cluster)) %>%
              # add column "same_cluster", YES if gene_cluster and re_subfamily_cluster are the same, NO otherwise
              mutate(nat_same_cluster = ifelse(nat_heatmap_gene_cluster == nat_heatmap_re_cluster, "yes", "no")) %>%
              # replace NA with NO
              mutate(nat_same_cluster = ifelse(is.na(nat_same_cluster), "no", nat_same_cluster)) %>%
              # keep only rows where sample_group.re is "native"
              filter(sample_group.re == "native") %>%
              # remove rows where every column has NA
              filter(!all(is.na(.))) %>%
              # replace "peak" with "dere" in column names
              rename_with(~ stringr::str_replace(.x, "peak", "dere")) %>%
              # prepend "feat_" to columns: gene_id, ID, Parent, exon_number, gene_name, homolog, oRF_type, transcript_id, transcript_name, gene_symbol
              rename_with(~ paste0("feat_", .x), c("gene_id", "ID", "Parent", "exon_number", "gene_name", "homolog", "oRF_type", "transcript_id", "transcript_name", "gene_symbol")) %>%
              # replace colname "name" with "feat_annotation"
              dplyr::rename(feat_annotation = name) %>%
              # remove "n_" from the beginning of feat_annotation column
              mutate(feat_annotation = stringr::str_replace(feat_annotation, "^n_", ""))

write_tsv(nat_deres_ann, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/nat_deres_ann.tsv"))

# Table 1
nat_tecount_deres_ann_summary_feat_x_dere <- nat_deres_ann %>%
  filter(quant_tool.re == "TEcount") %>%
  # convert grouping variables to factors
  mutate_at(vars(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster), as.factor) %>%
  # set diffexp_res_lfc.gene levels as "upregulated", "not_diffexp", "downregulated"
  mutate(diffexp_res_lfc.gene = factor(diffexp_res_lfc.gene, levels = c("upregulated", "not_diffexp", "downregulated"))) %>%
  # set nat_same_cluster levels as "yes", "no"
  mutate(nat_same_cluster = factor(nat_same_cluster, levels = c("yes", "no"))) %>%
  # set feat_annotation levels
  mutate(feat_annotation = factor(feat_annotation, levels = c("inferred_TSS_promoter", "five_prime_utr", "cds", "intronic", "exonic", "three_prime_utr", "inferred_TTS", "other_genic", "up5kb_gene", "down5kb_gene", "up10kb_gene", "down10kb_gene", "up50kb_gene", "down50kb_gene", "intergenic_region"))) %>%
  # group by feat_annotation, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster
  group_by(dere_id, feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, .drop=FALSE) %>%
  # summarize by adding a new column per unique value in dere_id, with the count
  summarize(counts = n()) %>%
  # pivot_wider to get a column per unique value in dere_id
  tidyr::pivot_wider(names_from = dere_id, values_from = counts, values_fill = 0, id_cols = c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster)) %>%
  # ungroup
  ungroup() %>%
  # remove rows where nat_same_cluster is yes and all count columns are 0
  filter(nat_same_cluster == "no" | rowSums(dplyr::select(., -c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster))) > 0) %>%
  # remove rows where intergenic_region does not apply
  filter(!(feat_annotation == "intergenic_region" & (!is.na(sample_group.gene) | !is.na(diffexp_contrast.gene) | !is.na(diffexp_res_lfc.gene))))

write_tsv(nat_tecount_deres_ann_summary_feat_x_dere, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/nat_tecount_deres_ann_summary_feat_x_dere.tsv"))
# nat_tecount_deres_ann_summary_feat_x_dere <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/nat_tecount_deres_ann_summary_feat_x_dere.tsv"))

```


# temporal replacement table to fix contrast names
```{r}

all_diffexp_results <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

all_diffexp_results <- all_diffexp_results %>%
                        # in SYMBOL, replace "_dup" with "."
                        mutate(SYMBOL = stringr::str_replace(SYMBOL, "_dup", "."))

write_tsv(all_diffexp_results, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

gene_metadata <- readr::read_tsv(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"))

gene_metadata <- gene_metadata %>%
                    # in gene_symbol, replace "_dup" with "."
                    mutate(gene_symbol = stringr::str_replace(gene_symbol, "_dup", "."))

write_tsv(gene_metadata, file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"))
```




```{r}

glob_deres_ann <- deres_ann %>%
              # replace NA in deres_ann$feature with "intergenic_region"
              mutate(feature = ifelse(is.na(feature), "intergenic_region", feature),
                     name = ifelse(is.na(feature) & is.na(name), "n_intergenic_region", name)) %>%
              # Adding gene_symbols
              mutate(gene_symbol = gene_metadata$gene_symbol[match(deres_ann$gene_id, gene_metadata$gene_id)]) %>%
              # add re quant_tool, diffexp_contrast, sample_group,  diffexp_res_lfc and diffexp_res_padj to deres_ann by merging
              left_join(all_diffexp_results_re,
                        by = join_by(peak_id == re_subfamily.re)) %>%
              # remove rows where diffexp_res_padj == "not_significant" and diffexp_res_lfc == "not_diffexp"
              filter(diffexp_res_padj.re != "not_significant" & diffexp_res_lfc.re != "not_diffexp") %>%
              # add gene quant_tool, diffexp_contrast, sample_group,  diffexp_res_lfc and diffexp_res_padj to deres_ann by merging
              left_join(all_diffexp_results_gene,
                        by = join_by(gene_id == ID.gene)) %>%
              # keep only rows where quant_tool == quant_tool.y, diffexp_contrast == diffexp_contrast.y, sample_group == sample_group.y
              filter(name == "n_intergenic_region"
                    | (quant_tool.gene == quant_tool.re
                    & diffexp_contrast.gene == diffexp_contrast.re
                    & sample_group.gene == sample_group.re)) %>%
              # add native heatmap clusters for genes
              left_join(nat_heatmap_clusters,
                        by = join_by(gene_id == ID)) %>%
              # rename nat_heatmap_cluster to nat_heatmap_gene_cluster
              dplyr::rename(nat_heatmap_gene_cluster = nat_heatmap_cluster) %>%
              # add native heatmap clusters for REs
              left_join(nat_heatmap_clusters,
                        by = join_by(peak_id == ID)) %>%
              # rename nat_heatmap_cluster to nat_heatmap_re_cluster
              dplyr::rename(nat_heatmap_re_cluster = nat_heatmap_cluster) %>%
              # Rows from global contrasts or ExplorATE quant should not have a cluster
              # because the heatmap clusters were generated with native samples and TEcount quant
              mutate(nat_heatmap_gene_cluster = ifelse(sample_group.re == "global"
                                                        | quant_tool.re != "TEcount"
                                                        | sample_group.gene == "global"
                                                        | quant_tool.gene != "TEcount",
                                                        NA, nat_heatmap_gene_cluster),
                      nat_heatmap_re_cluster = ifelse(sample_group.re == "global"
                                                      | quant_tool.re != "TEcount"
                                                      | sample_group.gene == "global"
                                                      | quant_tool.gene != "TEcount",
                                                      NA, nat_heatmap_re_cluster)) %>%
              # add column "same_cluster", YES if gene_cluster and re_subfamily_cluster are the same, NO otherwise
              mutate(nat_same_cluster = ifelse(nat_heatmap_gene_cluster == nat_heatmap_re_cluster, "yes", "no")) %>%
              # replace NA with NO
              mutate(nat_same_cluster = ifelse(is.na(nat_same_cluster), "no", nat_same_cluster)) %>%
              # add global heatmap1 clusters for genes
              left_join(glob1_heatmap_clusters,
                        by = join_by(gene_id == ID)) %>%
              # rename glob_heatmap1_cluster to glob_heatmap1_gene_cluster
              dplyr::rename(glob_heatmap1_gene_cluster = glob_heatmap1_cluster) %>%
              # add global heatmap1 clusters for REs
              left_join(glob1_heatmap_clusters,
                        by = join_by(peak_id == ID)) %>%
              # rename glob_heatmap1_cluster to glob_heatmap1_re_cluster
              dplyr::rename(glob_heatmap1_re_cluster = glob_heatmap1_cluster) %>%
              # Rows from native contrasts or ExplorATE quant should not have a cluster
              # because the heatmap clusters were generated with global samples and TEcount quant
              mutate(glob_heatmap1_gene_cluster = ifelse(sample_group.re == "native"
                                                        | quant_tool.re != "TEcount"
                                                        | sample_group.gene == "native"
                                                        | quant_tool.gene != "TEcount"
                                                        | !(diffexp_contrast.gene %in% c("subadult_blastemas_vs_subadult_limbs", "adult_blastemas_vs_adult_limbs", "adult_limbs_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",  "age_effect_blastemas_vs_limbs"))
                                                        | !(diffexp_contrast.re %in% c("subadult_blastemas_vs_subadult_limbs", "adult_blastemas_vs_adult_limbs", "adult_limbs_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",  "age_effect_blastemas_vs_limbs")),
                                                        NA, glob_heatmap1_gene_cluster),
                      glob_heatmap1_re_cluster = ifelse(sample_group.re == "native"
                                                      | quant_tool.re != "TEcount"
                                                      | sample_group.gene == "native"
                                                      | quant_tool.gene != "TEcount"
                                                      | !(diffexp_contrast.gene %in% c("subadult_blastemas_vs_subadult_limbs", "adult_blastemas_vs_adult_limbs", "adult_limbs_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",  "age_effect_blastemas_vs_limbs"))
                                                      | !(diffexp_contrast.re %in% c("subadult_blastemas_vs_subadult_limbs", "adult_blastemas_vs_adult_limbs", "adult_limbs_vs_subadult_limbs", "adults_blastemas_vs_subadults_blastemas",  "age_effect_blastemas_vs_limbs")),
                                                      NA, glob_heatmap1_re_cluster)) %>%
              # add column "same_cluster", YES if gene_cluster and re_subfamily_cluster are the same, NO otherwise
              mutate(glob1_same_cluster = ifelse(glob_heatmap1_gene_cluster == glob_heatmap1_re_cluster, "yes", "no")) %>%
              # replace NA with NO
              mutate(glob1_same_cluster = ifelse(is.na(glob1_same_cluster), "no", glob1_same_cluster)) %>%
              # add global heatmap1 clusters for genes
              left_join(glob2_heatmap_clusters,
                        by = join_by(gene_id == ID)) %>%
              # rename glob_heatmap1_cluster to glob_heatmap1_gene_cluster
              dplyr::rename(glob_heatmap2_gene_cluster = glob_heatmap2_cluster) %>%
              # add global heatmap1 clusters for REs
              left_join(glob2_heatmap_clusters,
                        by = join_by(peak_id == ID)) %>%
              # rename glob_heatmap1_cluster to glob_heatmap1_re_cluster
              dplyr::rename(glob_heatmap2_re_cluster = glob_heatmap2_cluster) %>%
              # Rows from native contrasts or ExplorATE quant should not have a cluster
              # because the heatmap clusters were generated with global samples and TEcount quant
              mutate(glob_heatmap2_gene_cluster = ifelse(sample_group.re == "native"
                                                        | quant_tool.re != "TEcount"
                                                        | sample_group.gene == "native"
                                                        | quant_tool.gene != "TEcount"
                                                        | !(diffexp_contrast.gene %in% c("adult_nonlimbs_vs_subadult_nonlimbs", "adult_blastemas_vs_adult_nonlimbs", "subadult_blastemas_vs_subadult_nonlimbs", "adult_limbs_vs_adult_nonlimbs", "subadult_limbs_vs_subadult_nonlimbs", "age_effect_blastemas_vs_nonlimbs", "age_effect_limbs_vs_nonlimbs"))
                                                        | !(diffexp_contrast.re %in% c("adult_nonlimbs_vs_subadult_nonlimbs", "adult_blastemas_vs_adult_nonlimbs", "subadult_blastemas_vs_subadult_nonlimbs", "adult_limbs_vs_adult_nonlimbs", "subadult_limbs_vs_subadult_nonlimbs", "age_effect_blastemas_vs_nonlimbs", "age_effect_limbs_vs_nonlimbs")),
                                                        NA, glob_heatmap2_gene_cluster),
                      glob_heatmap2_re_cluster = ifelse(sample_group.re == "native"
                                                      | quant_tool.re != "TEcount"
                                                      | sample_group.gene == "native"
                                                      | quant_tool.gene != "TEcount"
                                                      | !(diffexp_contrast.gene %in% c("adult_nonlimbs_vs_subadult_nonlimbs", "adult_blastemas_vs_adult_nonlimbs", "subadult_blastemas_vs_subadult_nonlimbs", "adult_limbs_vs_adult_nonlimbs", "subadult_limbs_vs_subadult_nonlimbs", "age_effect_blastemas_vs_nonlimbs", "age_effect_limbs_vs_nonlimbs"))
                                                      | !(diffexp_contrast.re %in% c("adult_nonlimbs_vs_subadult_nonlimbs", "adult_blastemas_vs_adult_nonlimbs", "subadult_blastemas_vs_subadult_nonlimbs", "adult_limbs_vs_adult_nonlimbs", "subadult_limbs_vs_subadult_nonlimbs", "age_effect_blastemas_vs_nonlimbs", "age_effect_limbs_vs_nonlimbs")),
                                                      NA, glob_heatmap2_re_cluster)) %>%
              # add column "same_cluster", YES if gene_cluster and re_subfamily_cluster are the same, NO otherwise
              mutate(glob2_same_cluster = ifelse(glob_heatmap2_gene_cluster == glob_heatmap2_re_cluster, "yes", "no")) %>%
              # replace NA with NO
              mutate(glob2_same_cluster = ifelse(is.na(glob2_same_cluster), "no", glob2_same_cluster)) %>%
              # remove rows where every column has NA
              filter(!all(is.na(.))) %>%
              # replace "peak" with "dere" in column names
              rename_with(~ stringr::str_replace(.x, "peak", "dere")) %>%
              # prepend "feat_" to columns: gene_id, ID, Parent, exon_number, gene_name, homolog, oRF_type, transcript_id, transcript_name, gene_symbol
              rename_with(~ paste0("feat_", .x), c("gene_id", "ID", "Parent", "exon_number", "gene_name", "homolog", "oRF_type", "transcript_id", "transcript_name", "gene_symbol")) %>%
              # replace colname "name" with "feat_annotation"
              dplyr::rename(feat_annotation = name) %>%
              # remove "n_" from the beginning of feat_annotation column
              mutate(feat_annotation = stringr::str_replace(feat_annotation, "^n_", ""))

write_tsv(glob_deres_ann, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/glob_deres_ann.tsv"))
# glob_deres_ann <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/glob_deres_ann.tsv"))


glob_tecount_deres_ann_summary_feat_x_dere <- glob_deres_ann %>%
  filter(quant_tool.re == "TEcount") %>%
  # convert grouping variables to factors
  mutate_at(vars(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster), as.factor) %>%
  # set diffexp_res_lfc.gene levels as "upregulated", "not_diffexp", "downregulated"
  mutate(diffexp_res_lfc.gene = factor(diffexp_res_lfc.gene, levels = c("upregulated", "not_diffexp", "downregulated"))) %>%
  mutate(sample_group.gene = factor(sample_group.gene, levels = c("native", "global"))) %>%
  # set nat_same_cluster levels as "yes", "no"
  mutate(nat_same_cluster = factor(nat_same_cluster, levels = c("yes", "no")),
         glob1_same_cluster = factor(glob1_same_cluster, levels = c("yes", "no")),
         glob2_same_cluster = factor(glob2_same_cluster, levels = c("yes", "no"))) %>%
  # set feat_annotation levels
  mutate(feat_annotation = factor(feat_annotation, levels = c("inferred_TSS_promoter", "five_prime_utr", "cds", "intronic", "exonic", "three_prime_utr", "inferred_TTS", "other_genic", "up5kb_gene", "down5kb_gene", "up10kb_gene", "down10kb_gene", "up50kb_gene", "down50kb_gene", "intergenic_region"))) %>%
  # group by feat_annotation, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster
  group_by(dere_id, feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster, .drop=FALSE) %>%
  # summarize by adding a new column per unique value in dere_id, with the count
  summarize(counts = n()) %>%
  # pivot_wider to get a column per unique value in dere_id
  tidyr::pivot_wider(names_from = dere_id, values_from = counts, values_fill = 0, id_cols = c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster)) %>%
  # ungroup
  ungroup() %>%
  # remove rows where nat_same_cluster is yes and all count columns are 0
  filter(nat_same_cluster == "no" | rowSums(dplyr::select(., -c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster))) > 0,
         glob1_same_cluster == "no" | rowSums(dplyr::select(., -c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster))) > 0,
         glob2_same_cluster == "no" | rowSums(dplyr::select(., -c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster))) > 0) %>%
  # remove rows where feat_annotation is intergenic_region and all columns starting with column 8 are 0
  filter(!(feat_annotation == "intergenic_region" & rowSums(dplyr::select(., -c(feat_annotation, sample_group.gene, diffexp_contrast.gene, diffexp_res_lfc.gene, nat_same_cluster, glob1_same_cluster, glob2_same_cluster))) == 0)) %>%
  # remove rows where sample_group.gene is "native", but diffexp_contrast.gene is not in c("subadult_blastemas_vs_subadult_limbs", "adult_limbs_vs_subadult_limbs")
  filter(feat_annotation == "intergenic_region" | !(sample_group.gene == "native" & !(diffexp_contrast.gene %in% c("subadult_blastemas_vs_subadult_limbs", "adult_limbs_vs_subadult_limbs"))))
  
write_tsv(glob_tecount_deres_ann_summary_feat_x_dere, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/glob_tecount_deres_ann_summary_feat_x_dere.tsv"))
# glob_tecount_deres_ann_summary_feat_x_dere <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/glob_tecount_deres_ann_summary_feat_x_dere.tsv"))



#glob count overview

sum_deres_ann_glob_totals <- glob_tecount_deres_ann_summary_feat_x_dere %>%
                            # remove rows where sample_group.gene is not "global"
                            filter(sample_group.gene == "global") %>%
                            # get sum of row starting with column 5
                            mutate(TOTAL = rowSums(.[8:ncol(.)], na.rm = TRUE)) %>%
                            # add percentage column
                            mutate(PERCENTAGE = TOTAL / sum(TOTAL) * 100) %>%
                            # remove all columns except the first 7 and the last 2
                            select(1:7, TOTAL, PERCENTAGE)

write_tsv(sum_deres_ann_glob_totals, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/DEREsxDEGs/sum_deres_ann_glob_totals.tsv"))

```


Figure 3b numbers
```{r}
all_diffexp_results <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

# count the number of (diffexp_res_lfc == "upregulated" and diffexp_res_padj == "significant") and (diffexp_res_lfc == "downregulated" and diffexp_res_padj == "significant") for each combination of quant_tool and diffexp_contrast

test <- all_diffexp_results %>%
  filter(diffexp_res_lfc != "not_diffexp",
         diffexp_res_padj != "not_significant") %>%
  # get count of ID per quant_tool, diffexp_contrast, diffexp_res_lfc
  group_by(sample_group, quant_tool, feature_biotype, diffexp_contrast, diffexp_res_lfc) %>%
  summarize(count = n())

print(test, n = Inf)

write_tsv(test, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_results_DEGs.tsv"))
```

Table 2

```{r}

library(readr)
library(dplyr)

all_diffexp_results <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

# dplyr, REs are feature_biotype == "repetitive_element"

all_diffexp_results_re <- all_diffexp_results %>%
  filter(feature_biotype == "repetitive_element",
         diffexp_res_padj != "not_significant" & diffexp_res_lfc != "not_diffexp") %>%
  mutate(re_superfamily = ifelse(grepl("^rnd", re_superfamily), "Unknown", re_superfamily),
         re_superfamily = gsub("Gypsy_Chromoviridae_Vclade", "Vclade:Chromoviridae:Ty3", re_superfamily),
         re_superfamily = gsub("Gypsy_Mag", "Mag:Ty3", re_superfamily),
         re_superfamily = gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", re_superfamily),
         re_superfamily = gsub("Gypsy", "Ty3", re_superfamily),
         re_class = gsub("\\?", "", re_class),
         # if re_superfamily is the same as re_class, prepend "Other " to re_superfamily
         re_superfamily = ifelse(re_superfamily == re_class, paste0("Other ", re_superfamily), re_superfamily)) %>%
  # remove duplicates based on gene_id, diffexp_contrast, sample_group, etc (this removes duplicates e.g. if the same RE was found by TEcount and ExplorATE)
  distinct(ID, sample_group, diffexp_contrast, diffexp_res_lfc, .keep_all = TRUE) %>%
  group_by(re_superfamily, re_class, sample_group, diffexp_contrast, diffexp_res_lfc) %>%
  summarize(counts = n()) %>%
  tidyr::pivot_wider(names_from = c(re_superfamily, re_class),
                     values_from = counts,
                     values_fill = 0) %>%
  # remove columns that contain all 0
  select(where(~ any(. != 0))) %>%
  ungroup() %>%
  mutate(sample_group = factor(sample_group, levels = c("native","global")),
         diffexp_res_lfc = factor(diffexp_res_lfc, levels = c("upregulated", "downregulated"))) %>%
  # sort based on factor levels
  arrange(sample_group, diffexp_contrast, diffexp_res_lfc) %>%
  # starting with the fourth column, arrange columns based on alphabetical order but of with the inverse colname
  dplyr::select(1:3, order(sapply(colnames(.[4:ncol(.)]), function(x) paste0(rev(unlist(strsplit(x, "_"))), collapse = "_"))) + 3) %>%
  # add totals column
  mutate(TOTAL = rowSums(.[4:ncol(.)], na.rm = TRUE))


write_tsv(all_diffexp_results_re, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/deres_and_classification.tsv"))
# all_diffexp_results_re_uniq <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/deres_and_classification.tsv")
```


S3 Table

```{r}

library(dplyr)
library(stringr)
library(readr)

all_diffexp_results <- readr::read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/all_diffexp_results.tsv"))

mod_and_colors <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/global/WGCNA/genes_or_REs_and_wgcna_modules.tsv"))

# rename colors
mod_and_colors <- mod_and_colors %>%
  mutate(wgcna_module = ifelse(wgcna_module == "turquoise", "Module 15 (turquoise)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "blue", "Module 7 (blue)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "brown", "Module 14 (brown)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "green", "Module 1 (green)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "yellow", "Module 6 (yellow)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "red", "Module 10 (red)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "pink", "Module 4 (pink)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "purple", "Module 13 (purple)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "grey", "Module 16 (grey)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "black", "Module 5 (black)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "greenyellow", "Module 2 (greenyellow)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "tan", "Module 3 (tan)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "magenta", "Module 9 (magenta)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "salmon", "Module 11 (salmon)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "cyan", "Module 8 (cyan)", wgcna_module),
          wgcna_module = ifelse(wgcna_module == "midnightblue", "Module 12 (midnightblue)", wgcna_module)
  )

gene_metadata <- read_tsv(file.path(PROJECT_DIR, "/metadata/gene_metadata.txt"))



# create selectLab with the top 3 genes based on padj_MLE, and the top and bottom 3 genes based on log2FoldChange_MMSE, for each diffexp_contrast
selectGenes1 <- all_diffexp_results %>%
                filter(quant_tool == "TEcount", diffexp_res_padj == "significant" & log2FoldChange_MMSE < -log2(1.5)) %>%
                group_by(sample_group, diffexp_contrast, feature_biotype) %>%
                slice_min(order_by = log2FoldChange_MMSE, n = 15) %>%
                pull(ID)

selectGenes2 <- all_diffexp_results %>%
                filter(quant_tool == "TEcount", diffexp_res_padj == "significant" & log2FoldChange_MMSE > log2(1.5)) %>%
                group_by(sample_group, diffexp_contrast, feature_biotype) %>%
                slice_max(order_by = log2FoldChange_MMSE, n = 15) %>%
                pull(ID)

selectGenes <- unique(c(selectGenes1, selectGenes2))


new_table <- all_diffexp_results %>%
  filter(quant_tool == "TEcount",
         diffexp_res_lfc != "not_diffexp",
         diffexp_res_padj == "significant",
         ID %in% selectGenes) %>%
  # merge columns sample_group and diffexp_contrast
  tidyr::unite(sample_group_diffexp_contrast, c("sample_group", "diffexp_contrast")) %>%
  select(ID, SYMBOL, sample_group_diffexp_contrast, log2FoldChange_MMSE) %>%
  # summarize by gene_id, add new columns per diffexp_contrast
  group_by(ID) %>%
  tidyr::pivot_wider(names_from = sample_group_diffexp_contrast, values_from = log2FoldChange_MMSE) %>%
  ungroup() %>%
  left_join(mod_and_colors, by = join_by(ID == gene_or_re_id)) %>%
  mutate(ID = gsub("Gypsy_Chromoviridae_Vclade", "Chromoviridae:Ty3", ID),
         ID = gsub("Gypsy_Mag", "Mag:Ty3", ID),
         ID = gsub("Gypsy_Metaviridae_Gmr1", "Gmr1:Metaviridae:Ty3", ID),
         ID = gsub("Gypsy", "Ty3", ID),
         # this replaces instances of "rnd-...:rnd-...:Unknown" with "rnd-...:Unknown"
         ID = gsub("([^:]+):\\1:([^:]+)", "\\1:\\2", ID))

# Supplementary Table S3
write_tsv(new_table, file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/degs_deres_and_wgcna.tsv"))

#new_table <- read_tsv(file.path(PROJECT_DIR, "/rnaseq/DEA_consensus/degs_deres_and_wgcna.tsv"))
```



# Primers, validation, PCR

```{r}


# read gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR,"/metadata/gene_metadata.txt"),
                            sep = "\t",
                            header = TRUE,
                            stringsAsFactors = FALSE)
#read RE metadata
re_metadata <- read.table(file.path(PROJECT_DIR,"/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/ambMex60DD_ucsc_hub.ALL.mod.tsv"),
                            sep = "\t",
                            header = TRUE,
                            stringsAsFactors = FALSE)


```


# LINE-1 zoom in

gene_id         transcript_id         gene_symbol     gene_goall    gene_keggpath
AMEX60DD020567  AMEX60DD301020567.1   ORF2.6       <NA>          <NA> 

alrededor de chr2p:172,929,292-172,947,978

```{r}

library(GenomicFeatures)
library(karyoploteR)
library(AnnotationDbi)
library(regioneR)
library(dplyr)

PROJECT_DIR <- "../sruizperez/projects/axolotl/"

chrs <- read.table(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.chrom.sizes"), header = FALSE, sep = "\t")

seq.info <- Seqinfo(seqnames = as.vector(chrs$V1), 
                    seqlengths = as.integer(chrs$V2), 
                    isCircular = as.vector(rep(FALSE, length(chrs$V1))), 
                    genome = "Ambystoma_mexicanum")

# TxDb.Amexicanum.ambMex60DD <- makeTxDbFromGFF(file = "$PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD_Repeats_merge.gff3.tmp",
#                                               format = "gff3",
#                                               dataSource = "www.axolotl-omics.org: AmexG_v6.0-DD, AmexT_v47, ambMex60DD_ucsc_hub_563_RM_and_LTR_libs",
#                                               chrominfo = seq.info,
#                                               organism = "Ambystoma mexicanum",
#                                               taxonomyId = 8296
#                 )

# AnnotationDbi::saveDb(TxDb.Amexicanum.ambMex60DD, "$PROJECT_DIR/ext_data/axolotl_genome/TxDb.Amexicanum.ambMex60DD.sqlite")

TxDb.Amexicanum.ambMex60DD <- AnnotationDbi::loadDb(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/TxDb.Amexicanum.ambMex60DD.sqlite"))

setwd(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/"))
chromosomes <- seqlevels(TxDb.Amexicanum.ambMex60DD)
chromosome_lengths <- seqlengths(TxDb.Amexicanum.ambMex60DD)

# Create a GRanges object with chromosome ranges
chromosome_ranges <- GRanges(seqnames = chromosomes, ranges = IRanges(start = 1, end = chromosome_lengths))

region <- toGRanges("chr2p:172,929,292-172,947,978")


kp <- plotKaryotype(genome = chromosome_ranges, zoom = region)

genes.data <- makeGenesDataFromTxDb(TxDb.Amexicanum.ambMex60DD,
                                    karyoplot = kp,
                                    plot.transcripts = TRUE, 
                                    plot.transcripts.structure = TRUE)

dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/DERE_context"))

#genes.data <- addGeneNames(genes.data, orgDb=TxDb.Amexicanum.ambMex60DD)
saveRDS(genes.data, file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.ORF2.6.rds"))
# genes.data <- readRDS(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.ORF2.6.rds"))


genes.data$genes$names <- gene_metadata$gene_symbol[match(genes.data$genes$gene_id, gene_metadata$gene_id)]
# replace names with gene_id if names is NA
genes.data$genes$names[is.na(genes.data$genes$names)] <- genes.data$genes$gene_id[is.na(genes.data$genes$names)]

# keep only genes
genes.data.genes <- genes.data
genes.data.genes$genes <- genes.data.genes$genes[genes.data.genes$genes$gene_id %in% gene_metadata$gene_id, ]
genes.data.genes$transcripts <- genes.data.genes$transcripts[names(genes.data.genes$transcripts) %in% gene_metadata$gene_id]
#genes.data.genes$coding.exons <- genes.data.genes$coding.exons[sub(".merged","",names(genes.data.genes$coding.exons)) %in% gene_metadata$gene_id]
#genes.data.genes$non.coding.exons <- genes.data.genes$non.coding.exons[sub(".merged","",names(genes.data.genes$non.coding.exons)) %in% gene_metadata$gene_id]

saveRDS(genes.data.genes, file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.genes.ORF2.6.rds"))
# genes.data.genes <- readRDS(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.genes.ORF2.6.rds"))

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/DERE_context"))


# replace "LINE?" with "LINE"
re_metadata$class_id <- gsub("LINE\\?", "LINE", re_metadata$class_id)

#extract transcript_id where class_id is "LTR"
ltrs <- re_metadata$transcript_id[re_metadata$class_id == "LTR"]
genes.data.ltr <- genes.data
genes.data.ltr$genes <- genes.data.ltr$genes[genes.data.ltr$genes$gene_id %in% ltrs, ]
genes.data.ltr$transcripts <- genes.data.ltr$transcripts[names(genes.data.ltr$transcripts) %in% ltrs]
#genes.data.ltr$coding.exons <- genes.data.ltr$coding.exons[sub(".merged","",names(genes.data.ltr$coding.exons)) %in% ltrs]
#genes.data.ltr$non.coding.exons <- genes.data.ltr$non.coding.exons[sub(".merged","",names(genes.data.ltr$non.coding.exons)) %in% ltrs]

# DNAs
dnas <- re_metadata$transcript_id[re_metadata$class_id == "DNA"]
genes.data.dna <- genes.data
genes.data.dna$genes <- genes.data.dna$genes[genes.data.dna$genes$gene_id %in% dnas, ]
genes.data.dna$transcripts <- genes.data.dna$transcripts[names(genes.data.dna$transcripts) %in% dnas]
#genes.data.dna$coding.exons <- genes.data.dna$coding.exons[sub(".merged","",names(genes.data.dna$coding.exons)) %in% dnas]
#genes.data.dna$non.coding.exons <- genes.data.dna$non.coding.exons[sub(".merged","",names(genes.data.dna$non.coding.exons)) %in% dnas]

# LINEs
lines <- re_metadata$transcript_id[re_metadata$class_id == "LINE"]
genes.data.line <- genes.data
genes.data.line$genes <- genes.data.line$genes[genes.data.line$genes$gene_id %in% lines, ]
genes.data.line$transcripts <- genes.data.line$transcripts[names(genes.data.line$transcripts) %in% lines]
#genes.data.line$coding.exons <- genes.data.line$coding.exons[sub(".merged","",names(genes.data.line$coding.exons)) %in% lines]
#genes.data.line$non.coding.exons <- genes.data.line$non.coding.exons[sub(".merged","",names(genes.data.line$non.coding.exons)) %in% lines]

#RCs
rcs <- re_metadata$transcript_id[re_metadata$class_id == "RC"]
genes.data.rc <- genes.data
genes.data.rc$genes <- genes.data.rc$genes[genes.data.rc$genes$gene_id %in% rcs, ]
genes.data.rc$transcripts <- genes.data.rc$transcripts[names(genes.data.rc$transcripts) %in% rcs]
#genes.data.rc$coding.exons <- genes.data.rc$coding.exons[sub(".merged","",names(genes.data.rc$coding.exons)) %in% rcs]
#genes.data.rc$non.coding.exons <- genes.data.rc$non.coding.exons[sub(".merged","",names(genes.data.rc$non.coding.exons)) %in% rcs]

#Satellite
satellites <- re_metadata$transcript_id[re_metadata$class_id == "Satellite"]
genes.data.satellite <- genes.data
genes.data.satellite$genes <- genes.data.satellite$genes[genes.data.satellite$genes$gene_id %in% satellites, ]
genes.data.satellite$transcripts <- genes.data.satellite$transcripts[names(genes.data.satellite$transcripts) %in% satellites]
#genes.data.satellite$coding.exons <- genes.data.satellite$coding.exons[sub(".merged","",names(genes.data.satellite$coding.exons)) %in% satellites]
#genes.data.satellite$non.coding.exons <- genes.data.satellite$non.coding.exons[sub(".merged","",names(genes.data.satellite$non.coding.exons)) %in% satellites]

#SINE
sines <- re_metadata$transcript_id[re_metadata$class_id == "SINE"]
genes.data.sine <- genes.data
genes.data.sine$genes <- genes.data.sine$genes[genes.data.sine$genes$gene_id %in% sines, ]
genes.data.sine$transcripts <- genes.data.sine$transcripts[names(genes.data.sine$transcripts) %in% sines]
#genes.data.sine$coding.exons <- genes.data.sine$coding.exons[sub(".merged","",names(genes.data.sine$coding.exons)) %in% sines]
#genes.data.sine$non.coding.exons <- genes.data.sine$non.coding.exons[sub(".merged","",names(genes.data.sine$non.coding.exons)) %in% sines]

#Unknown
unknowns <- re_metadata$transcript_id[re_metadata$class_id == "Unknown"]
genes.data.unknown <- genes.data
genes.data.unknown$genes <- genes.data.unknown$genes[genes.data.unknown$genes$gene_id %in% unknowns, ]
genes.data.unknown$transcripts <- genes.data.unknown$transcripts[names(genes.data.unknown$transcripts) %in% unknowns]
#genes.data.unknown$coding.exons <- genes.data.unknown$coding.exons[sub(".merged","",names(genes.data.unknown$coding.exons)) %in% unknowns]
#genes.data.unknown$non.coding.exons <- genes.data.unknown$non.coding.exons[sub(".merged","",names(genes.data.unknown$non.coding.exons)) %in% unknowns]


# mark.genes is rows of genes.data$genes where genes.data$genes$names contains as part "rnd-6_family-8458"
#mark.genes <- genes.data$genes[grepl("ORF2.6", genes.data$genes$names), ]
#mark.genes <- genes.data$genes[grepl("rnd-6_family-8458", genes.data$genes$names), ]

#rnd-6_family-8458_dup4675
#rnd-6_family-8458_dup4674

base.path <- file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/bigwigs/")

bigwigs <- c(SA_B1="18134R-126-01_S65_L004_scaled.bw",
             SA_B2="18134R-126-02_S66_L004_scaled.bw",
             SA_B3="18134R-126-03_S67_L004_scaled.bw",
             SA_B4="18134R-126-04_S68_L004_scaled.bw",
             SA_B5="18134R-126-05_S69_L004_scaled.bw",
             SA_L1="18134R-126-06_S70_L004_scaled.bw",
             SA_L2="18134R-126-07_S71_L004_scaled.bw",
             SA_L3="18134R-126-08_S72_L004_scaled.bw",
             SA_L4="18134R-126-09_S73_L004_scaled.bw",
             SA_L5="18134R-126-10_S74_L004_scaled.bw",
             A_L1="18134R-126-11_S75_L004_scaled.bw",
             A_L2="18134R-126-12_S76_L004_scaled.bw")

pp <- getDefaultPlotParams(plot.type=1)
pp$leftmargin <- 0.1
pp$data1inmargin <- 2
pp$topmargin <- 2
pp$bottommargin <- 10
pp$ideogramheight <- 5
pp$data1inmargin <- 5

png(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/karyoploter.ORF2.6.2.zi.png"),
    width = 60, height = 45, units = "cm", res = 600)

kp <- plotKaryotype(genome = chromosome_ranges, zoom = region, cex=2, plot.params = pp)

kpAddBaseNumbers(kp, tick.dist = 2000, minor.tick.dist = 500,
                 add.units = TRUE, cex=1.3, digits = 6)

kpPlotGenes(kp, data=genes.data.ltr, r0=0, r1=0.02,
            col = "red",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LTR", r0=0, r1=0.02, cex=1)

kpPlotGenes(kp, data = genes.data.dna, r0=0.02, r1=0.04,
            col = "blue",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "DNA", r0=0.02, r1=0.04, cex=1)

kpPlotGenes(kp, data=genes.data.line, r0=0.04, r1=0.06,
            col = "green",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LINE", r0=0.04, r1=0.06, cex=1)

kpPlotGenes(kp, data=genes.data.rc, r0=0.06, r1=0.08,
            col = "orange",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "RC", r0=0.06, r1=0.08, cex=1)

kpPlotGenes(kp, data=genes.data.satellite, r0=0.08, r1=0.10,
            col = "purple",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Satellite", r0=0.08, r1=0.10, cex=1)

kpPlotGenes(kp, data=genes.data.sine, r0=0.10, r1=0.12,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "SINE", r0=0.10, r1=0.12, cex=1)

kpPlotGenes(kp, data=genes.data.unknown, r0=0.12, r1=0.14,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Unknown", r0=0.12, r1=0.14, cex=1)

#kpPlotMarkers(kp, data=mark.genes, labels=mark.genes$names, text.orientation = "horizontal",
#              r0 = 0.27, r1=0.32, cex=0.8, adjust.label.position = TRUE)

# genes.data.genes$genes$gene_id are names and genes.data.genes$genes$names are values of named vector
gene_names <- genes.data.genes$genes$names
names(gene_names) <- genes.data.genes$genes$gene_id

kpPlotGenes(kp, data=genes.data.genes, r0=0.16, r1=0.24,
            col = "black",
            add.gene.names = TRUE,
            gene.names = gene_names,
            gene.name.position = "center",
            gene.name.cex = 0.5)
kpAddLabels(kp, labels = "Genes", r0=0.16, r1=0.24, cex=1)

for(i in seq_len(length(bigwigs))) {
  bigwig.file <- paste0(base.path, bigwigs[i])
  at <- autotrack(i, length(bigwigs), r0=0.26, r1=1, margin = 0.5)
  manual.ymax <- 50
  kp <- kpPlotBigWig(kp, data=bigwig.file, ymax=manual.ymax,
                     r0=at$r0, r1=at$r1, col = "cadetblue2")
  computed.ymax <- ceiling(kp$latest.plot$computed.values$ymax)
  kpAxis(kp, ymin=0, ymax=manual.ymax, numticks = 2, r0=at$r0, r1=at$r1)
  kpAddLabels(kp, labels = names(bigwigs)[i], r0=at$r0, r1=at$r1, label.margin = 0.03)
}

dev.off()


svg(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/karyoploter.ORF2.6.2.zi.svg"),
    width = 23.622, height = 17.7165)

kp <- plotKaryotype(genome = chromosome_ranges, zoom = region, cex=2, plot.params = pp)

kpAddBaseNumbers(kp, tick.dist = 2000, minor.tick.dist = 500,
                 add.units = TRUE, cex=1.3, digits = 6)

kpPlotGenes(kp, data=genes.data.ltr, r0=0, r1=0.02,
            col = "red",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LTR", r0=0, r1=0.02, cex=1)

kpPlotGenes(kp, data = genes.data.dna, r0=0.02, r1=0.04,
            col = "blue",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "DNA", r0=0.02, r1=0.04, cex=1)

kpPlotGenes(kp, data=genes.data.line, r0=0.04, r1=0.06,
            col = "green",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LINE", r0=0.04, r1=0.06, cex=1)

kpPlotGenes(kp, data=genes.data.rc, r0=0.06, r1=0.08,
            col = "orange",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "RC", r0=0.06, r1=0.08, cex=1)

kpPlotGenes(kp, data=genes.data.satellite, r0=0.08, r1=0.10,
            col = "purple",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Satellite", r0=0.08, r1=0.10, cex=1)

kpPlotGenes(kp, data=genes.data.sine, r0=0.10, r1=0.12,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "SINE", r0=0.10, r1=0.12, cex=1)

kpPlotGenes(kp, data=genes.data.unknown, r0=0.12, r1=0.14,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Unknown", r0=0.12, r1=0.14, cex=1)

#kpPlotMarkers(kp, data=mark.genes, labels=mark.genes$names, text.orientation = "horizontal",
#              r0 = 0.27, r1=0.32, cex=0.8, adjust.label.position = TRUE)

# genes.data.genes$genes$gene_id are names and genes.data.genes$genes$names are values of named vector
gene_names <- genes.data.genes$genes$names
names(gene_names) <- genes.data.genes$genes$gene_id

kpPlotGenes(kp, data=genes.data.genes, r0=0.16, r1=0.24,
            col = "black",
            add.gene.names = TRUE,
            gene.names = gene_names,
            gene.name.position = "center",
            gene.name.cex = 0.5)
kpAddLabels(kp, labels = "Genes", r0=0.16, r1=0.24, cex=1)

for(i in seq_len(length(bigwigs))) {
  bigwig.file <- paste0(base.path, bigwigs[i])
  at <- autotrack(i, length(bigwigs), r0=0.26, r1=1, margin = 0.5)
  manual.ymax <- 50
  kp <- kpPlotBigWig(kp, data=bigwig.file, ymax=manual.ymax,
                     r0=at$r0, r1=at$r1, col = "cadetblue2")
  computed.ymax <- ceiling(kp$latest.plot$computed.values$ymax)
  kpAxis(kp, ymin=0, ymax=manual.ymax, numticks = 2, r0=at$r0, r1=at$r1)
  kpAddLabels(kp, labels = names(bigwigs)[i], r0=at$r0, r1=at$r1, label.margin = 0.03)
}

dev.off()


```


# LINE-1 zoom out

gene_id         transcript_id         gene_symbol     gene_goall    gene_keggpath
AMEX60DD020567  AMEX60DD301020567.1   ORF2.6       <NA>          <NA> 

alrededor de chr2p:172,720,000-173,030,000

```{r}

library(GenomicFeatures)
library(karyoploteR)
library(AnnotationDbi)
library(regioneR)
library(dplyr)

PROJECT_DIR <- "../sruizperez/projects/axolotl/"

chrs <- read.table(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/AmexG_v6.0-DD/AmexG_v6.0-DD.chrom.sizes"), header = FALSE, sep = "\t")

seq.info <- Seqinfo(seqnames = as.vector(chrs$V1), 
                    seqlengths = as.integer(chrs$V2), 
                    isCircular = as.vector(rep(FALSE, length(chrs$V1))), 
                    genome = "Ambystoma_mexicanum")

# TxDb.Amexicanum.ambMex60DD <- makeTxDbFromGFF(file = "$PROJECT_DIR/ext_data/axolotl_genome/AmexT_v47-AmexG_v6.0-DD_Repeats_merge.gff3.tmp",
#                                               format = "gff3",
#                                               dataSource = "www.axolotl-omics.org: AmexG_v6.0-DD, AmexT_v47, ambMex60DD_ucsc_hub_563_RM_and_LTR_libs",
#                                               chrominfo = seq.info,
#                                               organism = "Ambystoma mexicanum",
#                                               taxonomyId = 8296
#                 )

# AnnotationDbi::saveDb(TxDb.Amexicanum.ambMex60DD, "$PROJECT_DIR/ext_data/axolotl_genome/TxDb.Amexicanum.ambMex60DD.sqlite")

TxDb.Amexicanum.ambMex60DD <- AnnotationDbi::loadDb(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/TxDb.Amexicanum.ambMex60DD.sqlite"))

setwd(file.path(PROJECT_DIR, "/ext_data/axolotl_genome/"))
chromosomes <- seqlevels(TxDb.Amexicanum.ambMex60DD)
chromosome_lengths <- seqlengths(TxDb.Amexicanum.ambMex60DD)

# Create a GRanges object with chromosome ranges
chromosome_ranges <- GRanges(seqnames = chromosomes, ranges = IRanges(start = 1, end = chromosome_lengths))

region <- toGRanges("chr2p:172,720,000-173,030,000")


kp <- plotKaryotype(genome = chromosome_ranges, zoom = region)

genes.data <- makeGenesDataFromTxDb(TxDb.Amexicanum.ambMex60DD,
                                    karyoplot = kp,
                                    plot.transcripts = TRUE, 
                                    plot.transcripts.structure = TRUE)

dir.create(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/DERE_context"))

#genes.data <- addGeneNames(genes.data, orgDb=TxDb.Amexicanum.ambMex60DD)
saveRDS(genes.data, file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.ORF2.6.zo.rds"))
# genes.data <- readRDS(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.ORF2.6.rds"))

# read gene_metadata
gene_metadata <- read.table(file.path(PROJECT_DIR,"/metadata/gene_metadata.txt"),
                            sep = "\t",
                            header = TRUE,
                            stringsAsFactors = FALSE)

genes.data$genes$names <- gene_metadata$gene_symbol[match(genes.data$genes$gene_id, gene_metadata$gene_id)]
# replace names with gene_id if names is NA
genes.data$genes$names[is.na(genes.data$genes$names)] <- genes.data$genes$gene_id[is.na(genes.data$genes$names)]

# keep only genes
genes.data.genes <- genes.data
genes.data.genes$genes <- genes.data.genes$genes[genes.data.genes$genes$gene_id %in% gene_metadata$gene_id, ]
genes.data.genes$transcripts <- genes.data.genes$transcripts[names(genes.data.genes$transcripts) %in% gene_metadata$gene_id]
#genes.data.genes$coding.exons <- genes.data.genes$coding.exons[sub(".merged","",names(genes.data.genes$coding.exons)) %in% gene_metadata$gene_id]
#genes.data.genes$non.coding.exons <- genes.data.genes$non.coding.exons[sub(".merged","",names(genes.data.genes$non.coding.exons)) %in% gene_metadata$gene_id]

saveRDS(genes.data.genes, file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/TxDb.Amexicanum.ambMex60DD.genes.data.genes.ORF2.6.zo.rds"))

setwd(file.path(PROJECT_DIR, "/rnaseq/STAR/TEcount/DERE_context"))


#read RE metadata
re_metadata <- read.table(file.path(PROJECT_DIR,"/ext_data/axolotl_genome/AmexG_v6.0-DD_Repeats/ambMex60DD_ucsc_hub.ALL.mod.tsv"),
                            sep = "\t",
                            header = TRUE,
                            stringsAsFactors = FALSE)

# replace "LINE?" with "LINE"
re_metadata$class_id <- gsub("LINE\\?", "LINE", re_metadata$class_id)

#extract transcript_id where class_id is "LTR"
ltrs <- re_metadata$transcript_id[re_metadata$class_id == "LTR"]
genes.data.ltr <- genes.data
genes.data.ltr$genes <- genes.data.ltr$genes[genes.data.ltr$genes$gene_id %in% ltrs, ]
genes.data.ltr$transcripts <- genes.data.ltr$transcripts[names(genes.data.ltr$transcripts) %in% ltrs]
#genes.data.ltr$coding.exons <- genes.data.ltr$coding.exons[sub(".merged","",names(genes.data.ltr$coding.exons)) %in% ltrs]
#genes.data.ltr$non.coding.exons <- genes.data.ltr$non.coding.exons[sub(".merged","",names(genes.data.ltr$non.coding.exons)) %in% ltrs]

# DNAs
dnas <- re_metadata$transcript_id[re_metadata$class_id == "DNA"]
genes.data.dna <- genes.data
genes.data.dna$genes <- genes.data.dna$genes[genes.data.dna$genes$gene_id %in% dnas, ]
genes.data.dna$transcripts <- genes.data.dna$transcripts[names(genes.data.dna$transcripts) %in% dnas]
#genes.data.dna$coding.exons <- genes.data.dna$coding.exons[sub(".merged","",names(genes.data.dna$coding.exons)) %in% dnas]
#genes.data.dna$non.coding.exons <- genes.data.dna$non.coding.exons[sub(".merged","",names(genes.data.dna$non.coding.exons)) %in% dnas]

# LINEs
lines <- re_metadata$transcript_id[re_metadata$class_id == "LINE"]
genes.data.line <- genes.data
genes.data.line$genes <- genes.data.line$genes[genes.data.line$genes$gene_id %in% lines, ]
genes.data.line$transcripts <- genes.data.line$transcripts[names(genes.data.line$transcripts) %in% lines]
#genes.data.line$coding.exons <- genes.data.line$coding.exons[sub(".merged","",names(genes.data.line$coding.exons)) %in% lines]
#genes.data.line$non.coding.exons <- genes.data.line$non.coding.exons[sub(".merged","",names(genes.data.line$non.coding.exons)) %in% lines]

#RCs
rcs <- re_metadata$transcript_id[re_metadata$class_id == "RC"]
genes.data.rc <- genes.data
genes.data.rc$genes <- genes.data.rc$genes[genes.data.rc$genes$gene_id %in% rcs, ]
genes.data.rc$transcripts <- genes.data.rc$transcripts[names(genes.data.rc$transcripts) %in% rcs]
#genes.data.rc$coding.exons <- genes.data.rc$coding.exons[sub(".merged","",names(genes.data.rc$coding.exons)) %in% rcs]
#genes.data.rc$non.coding.exons <- genes.data.rc$non.coding.exons[sub(".merged","",names(genes.data.rc$non.coding.exons)) %in% rcs]

#Satellite
satellites <- re_metadata$transcript_id[re_metadata$class_id == "Satellite"]
genes.data.satellite <- genes.data
genes.data.satellite$genes <- genes.data.satellite$genes[genes.data.satellite$genes$gene_id %in% satellites, ]
genes.data.satellite$transcripts <- genes.data.satellite$transcripts[names(genes.data.satellite$transcripts) %in% satellites]
#genes.data.satellite$coding.exons <- genes.data.satellite$coding.exons[sub(".merged","",names(genes.data.satellite$coding.exons)) %in% satellites]
#genes.data.satellite$non.coding.exons <- genes.data.satellite$non.coding.exons[sub(".merged","",names(genes.data.satellite$non.coding.exons)) %in% satellites]

#SINE
sines <- re_metadata$transcript_id[re_metadata$class_id == "SINE"]
genes.data.sine <- genes.data
genes.data.sine$genes <- genes.data.sine$genes[genes.data.sine$genes$gene_id %in% sines, ]
genes.data.sine$transcripts <- genes.data.sine$transcripts[names(genes.data.sine$transcripts) %in% sines]
#genes.data.sine$coding.exons <- genes.data.sine$coding.exons[sub(".merged","",names(genes.data.sine$coding.exons)) %in% sines]
#genes.data.sine$non.coding.exons <- genes.data.sine$non.coding.exons[sub(".merged","",names(genes.data.sine$non.coding.exons)) %in% sines]

#Unknown
unknowns <- re_metadata$transcript_id[re_metadata$class_id == "Unknown"]
genes.data.unknown <- genes.data
genes.data.unknown$genes <- genes.data.unknown$genes[genes.data.unknown$genes$gene_id %in% unknowns, ]
genes.data.unknown$transcripts <- genes.data.unknown$transcripts[names(genes.data.unknown$transcripts) %in% unknowns]
#genes.data.unknown$coding.exons <- genes.data.unknown$coding.exons[sub(".merged","",names(genes.data.unknown$coding.exons)) %in% unknowns]
#genes.data.unknown$non.coding.exons <- genes.data.unknown$non.coding.exons[sub(".merged","",names(genes.data.unknown$non.coding.exons)) %in% unknowns]


# mark.genes is rows of genes.data$genes where genes.data$genes$names contains as part "rnd-6_family-8458"
#mark.genes <- genes.data$genes[grepl("ORF2.6", genes.data$genes$names), ]
#mark.genes <- genes.data$genes[grepl("rnd-6_family-8458", genes.data$genes$names), ]

#rnd-6_family-8458_dup4675
#rnd-6_family-8458_dup4674

base.path <- file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/bigwigs/")

bigwigs <- c(SA_B1="18134R-126-01_S65_L004_scaled.bw",
             SA_B2="18134R-126-02_S66_L004_scaled.bw",
             SA_B3="18134R-126-03_S67_L004_scaled.bw",
             SA_B4="18134R-126-04_S68_L004_scaled.bw",
             SA_B5="18134R-126-05_S69_L004_scaled.bw",
             SA_L1="18134R-126-06_S70_L004_scaled.bw",
             SA_L2="18134R-126-07_S71_L004_scaled.bw",
             SA_L3="18134R-126-08_S72_L004_scaled.bw",
             SA_L4="18134R-126-09_S73_L004_scaled.bw",
             SA_L5="18134R-126-10_S74_L004_scaled.bw",
             A_L1="18134R-126-11_S75_L004_scaled.bw",
             A_L2="18134R-126-12_S76_L004_scaled.bw")

pp <- getDefaultPlotParams(plot.type=1)
pp$leftmargin <- 0.1
pp$data1inmargin <- 2
pp$topmargin <- 2
pp$bottommargin <- 10
pp$ideogramheight <- 5
pp$data1inmargin <- 5


png(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/karyoploter.ORF2.6.2.zo.png"), width = 50, height = 50, units = "cm", res = 600)

kp <- plotKaryotype(genome = chromosome_ranges, zoom = region, cex=2, plot.params = pp)

kpAddBaseNumbers(kp, tick.dist = 40000, minor.tick.dist = 2000,
                 add.units = TRUE, cex=1.3, digits = 6)

kpPlotGenes(kp, data=genes.data.ltr, r0=0, r1=0.02,
            col = "red",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LTR", r0=0, r1=0.02, cex=1)

kpPlotGenes(kp, data = genes.data.dna, r0=0.02, r1=0.04,
            col = "blue",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "DNA", r0=0.02, r1=0.04, cex=1)

kpPlotGenes(kp, data=genes.data.line, r0=0.04, r1=0.06,
            col = "green",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LINE", r0=0.04, r1=0.06, cex=1)

kpPlotGenes(kp, data=genes.data.rc, r0=0.06, r1=0.08,
            col = "orange",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "RC", r0=0.06, r1=0.08, cex=1)

kpPlotGenes(kp, data=genes.data.satellite, r0=0.08, r1=0.10,
            col = "purple",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Satellite", r0=0.08, r1=0.10, cex=1)

kpPlotGenes(kp, data=genes.data.sine, r0=0.10, r1=0.12,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "SINE", r0=0.10, r1=0.12, cex=1)

kpPlotGenes(kp, data=genes.data.unknown, r0=0.12, r1=0.14,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Unknown", r0=0.12, r1=0.14, cex=1)

kpPlotMarkers(kp, data=mark.genes, labels=mark.genes$names, text.orientation = "horizontal",
              r0 = 0.27, r1=0.32, cex=0.8, adjust.label.position = TRUE)

# genes.data.genes$genes$gene_id are names and genes.data.genes$genes$names are values of named vector
gene_names <- genes.data.genes$genes$names
names(gene_names) <- genes.data.genes$genes$gene_id

kpPlotGenes(kp, data=genes.data.genes, r0=0.16, r1=0.24,
            col = "black",
            add.gene.names = TRUE,
            gene.names = gene_names,
            gene.name.position = "center",
            gene.name.cex = 0.5)
kpAddLabels(kp, labels = "Genes", r0=0.16, r1=0.24, cex=1)



for(i in seq_len(length(bigwigs))) {
  bigwig.file <- paste0(base.path, bigwigs[i])
  at <- autotrack(i, length(bigwigs), r0=0.26, r1=1, margin = 0.5)
  kp <- kpPlotBigWig(kp, data=bigwig.file, ymax="visible.region",
                     r0=at$r0, r1=at$r1, col = "cadetblue2")
  computed.ymax <- ceiling(kp$latest.plot$computed.values$ymax)
  kpAxis(kp, ymin=0, ymax=computed.ymax, numticks = 2, r0=at$r0, r1=at$r1)
  kpAddLabels(kp, labels = names(bigwigs)[i], r0=at$r0, r1=at$r1, label.margin = 0.03)
}

dev.off()


svg(file.path(PROJECT_DIR,"/rnaseq/STAR/TEcount/DERE_context/karyoploter.ORF2.6.2.zo.svg"), width = 19.685, height = 19.685)

kp <- plotKaryotype(genome = chromosome_ranges, zoom = region, cex=2, plot.params = pp)

kpAddBaseNumbers(kp, tick.dist = 40000, minor.tick.dist = 2000,
                 add.units = TRUE, cex=1.3, digits = 6)

kpPlotGenes(kp, data=genes.data.ltr, r0=0, r1=0.02,
            col = "red",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LTR", r0=0, r1=0.02, cex=1)

kpPlotGenes(kp, data = genes.data.dna, r0=0.02, r1=0.04,
            col = "blue",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "DNA", r0=0.02, r1=0.04, cex=1)

kpPlotGenes(kp, data=genes.data.line, r0=0.04, r1=0.06,
            col = "green",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "LINE", r0=0.04, r1=0.06, cex=1)

kpPlotGenes(kp, data=genes.data.rc, r0=0.06, r1=0.08,
            col = "orange",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "RC", r0=0.06, r1=0.08, cex=1)

kpPlotGenes(kp, data=genes.data.satellite, r0=0.08, r1=0.10,
            col = "purple",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Satellite", r0=0.08, r1=0.10, cex=1)

kpPlotGenes(kp, data=genes.data.sine, r0=0.10, r1=0.12,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "SINE", r0=0.10, r1=0.12, cex=1)

kpPlotGenes(kp, data=genes.data.unknown, r0=0.12, r1=0.14,
            col = "brown",
            plot.transcript.structure=FALSE,
            avoid.overlapping = FALSE,
            add.gene.names = FALSE)
kpAddLabels(kp, labels = "Unknown", r0=0.12, r1=0.14, cex=1)

kpPlotMarkers(kp, data=mark.genes, labels=mark.genes$names, text.orientation = "horizontal",
              r0 = 0.27, r1=0.32, cex=0.8, adjust.label.position = TRUE)

# genes.data.genes$genes$gene_id are names and genes.data.genes$genes$names are values of named vector
gene_names <- genes.data.genes$genes$names
names(gene_names) <- genes.data.genes$genes$gene_id

kpPlotGenes(kp, data=genes.data.genes, r0=0.16, r1=0.24,
            col = "black",
            add.gene.names = TRUE,
            gene.names = gene_names,
            gene.name.position = "center",
            gene.name.cex = 0.5)
kpAddLabels(kp, labels = "Genes", r0=0.16, r1=0.24, cex=1)



for(i in seq_len(length(bigwigs))) {
  bigwig.file <- paste0(base.path, bigwigs[i])
  at <- autotrack(i, length(bigwigs), r0=0.26, r1=1, margin = 0.5)
  kp <- kpPlotBigWig(kp, data=bigwig.file, ymax="visible.region",
                     r0=at$r0, r1=at$r1, col = "cadetblue2")
  computed.ymax <- ceiling(kp$latest.plot$computed.values$ymax)
  kpAxis(kp, ymin=0, ymax=computed.ymax, numticks = 2, r0=at$r0, r1=at$r1)
  kpAddLabels(kp, labels = names(bigwigs)[i], r0=at$r0, r1=at$r1, label.margin = 0.03)
}

dev.off()



```